\documentclass{article}
\usepackage{a4wide}
\usepackage{longtable}
\usepackage{tocloft}
\usepackage{blindtext}
 \cftpagenumbersoff{section}
 \cftpagenumbersoff{subsection}
\usepackage{color}   %May be necessary if you want to color links
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\usepackage{graphicx}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{7.0-public}
\lhead{\today}

\title{AACR GENIE {{release}}.0-public Data Guide}
\author{AACR}


% Definition of \maketitle
\makeatletter         
\def\@maketitle{
\begin{center}
\includegraphics[width=1.0\textwidth]{genie_banner.png}\\[8ex]
{\Huge \@title }\\[4ex] 
{\Large  \@author}\\[4ex] 
\@date\\[8ex]
\end{center}}
\makeatother

\begin{document}


\SweaveOpts{concordance=TRUE}


<<echo=F, include=F, results=hide>>=
library(xtable)
library(synapser)
library(ggplot2)
library(ggpubr)

get_list_assay_info_table <- function(allowed_values, assayinfodf, col) {
  na_ind = is.na(assayinfodf[,col])
  seq_assays = assayinfodf$SEQ_ASSAY_ID[!na_ind]
  col_per_panel = matrix(nrow = length(seq_assays),
                         ncol = length(allowed_values),
                         dimnames = list(seq_assays, allowed_values))
  for (panel in seq_assays) {
    str_value = assayinfodf[,col][assayinfodf$SEQ_ASSAY_ID == panel]
    str_value_vector = unlist(strsplit(str_value, ";"))
    col_per_panel[panel, ] = FALSE
    col_per_panel[panel, str_value_vector] = TRUE
  }
  col_per_panel
}

synLogin()
centers = suppressMessages(synTableQuery('select "Center Abbreviation", "Center" from syn16982837',
                                         includeRowIdAndRowVersion = F))
assayinfo = suppressMessages(synTableQuery('SELECT * FROM syn17009222',
                                           includeRowIdAndRowVersion = F))
assayinfodf = assayinfo$asDataFrame()
genomic_profile = suppressMessages(synTableQuery('SELECT * FROM syn21642088',
                                                 includeRowIdAndRowVersion = F))

@

\maketitle

\tableofcontents

\newpage

\addcontentsline{toc}{section}{About this Document}
\section*{About this Document}

This document provides an overview of the {{release}}\textsuperscript{th} public release of American Association for Cancer Research (AACR) GENIE data.

\addcontentsline{toc}{section}{Version of Data}
\section*{Version of Data}

AACR GENIE Project Data: Version {{release}}.0-pubic.
\\*
AACR Project GENIE data versions follow a numbering scheme derived from \href{https://semver.org/}{semantic versioning}, where the digits in the version correspond to: major.patch-release-type.  “Major” releases are public releases of new sample data.  “Patch” releases are corrections to major releases, including data retractions.  “Release-type” refers to whether the release is a public AACR Project GENIE release or a private/consortium-only release.  Public releases will be denoted with the nomenclature “X.X-public” and consortium-only private releases will be denoted with the nomenclature “X.X-consortium".


\addcontentsline{toc}{section}{Data Access}
\section*{Data Access}
AACR GENIE Data is currently available via two mechanisms:
\begin{itemize}
    \item{Synapse Platform (Sage Bionetworks):  http://synapse.org/genie}
    \item{cBioPortal for Cancer Genomics (MSK): http://www.cbioportal.org/genie/}
\end{itemize}


\addcontentsline{toc}{section}{Terms Of Access}
\section*{Terms Of Access}
All users of the AACR Project GENIE data must agree to the following terms of use; failure to abide by any term herein will result in revocation of access.

\begin{itemize}
    \item{Users will not attempt to identify or contact individual participants from whom these data were collected by any means.}
    \item{Users will not redistribute the data without express written permission from the AACR Project GENIE Coordinating Center (send email to:  info@aacrgenie.org).}
\end{itemize}

When publishing or presenting work using or referencing the AACR Project GENIE dataset please include the following attributions:

\begin{itemize}
\item{Please cite: \emph{The AACR Project GENIE Consortium. AACR Project GENIE: Powering Precision Medicine Through An International Consortium, Cancer Discov. 2017 Aug;7(8):818-831} and include the version of the dataset used.}
\item{The authors would like to acknowledge the American Association for Cancer Research and its financial and material support in the development of the AACR Project GENIE registry, as well as members of the consortium for their commitment to data sharing. Interpretations are the responsibility of study authors.}
\end{itemize}

Posters and presentations should include the AACR Project GENIE logo.



\addcontentsline{toc}{section}{Introduction to AACR GENIE}
\section*{Introduction to AACR GENIE}

The AACR Project Genomics, Evidence, Neoplasia, Information, Exchange (GENIE) is a multi-phase, multi-year, international data-sharing project that aims to catalyze precision cancer medicine.  The GENIE platform will integrate and link clinical-grade cancer genomic data with clinical outcome data for tens of thousands of cancer patients treated at multiple international institutions. The project fulfills an unmet need in oncology by providing the statistical power necessary to improve clinical decision-making, to identify novel therapeutic targets, to understand of patient response to therapy, and to design new biomarker-driven clinical trials.  The project will also serve as a prototype for aggregating, harmonizing, and sharing clinical-grade, next-generation sequencing (NGS) data obtained during routine medical practice.

The data within GENIE is being shared with the global research community.  The database currently contains CLIA-/ISO-certified genomic data obtained during the course of routine practice at multiple international institutions (Table 1), and will continue to grow as more patients are treated at additional participating centers.

<<results=tex, echo=F>>=
xtab = xtable(centers$asDataFrame(),
              caption = 'Participating Centers')
#cat("\\makebox[\\textwidth]{")
print(xtab, include.rownames = F, floating = F,
      caption.placement = "top", tabular.environment = "longtable")
#cat("}")
@





\addcontentsline{toc}{section}{Human Subjects Protection and Privacy}
\section*{Human Subjects Protection and Privacy}

Protection of patient privacy is paramount, and the AACR GENIE Project therefore requires that each participating center share data in a manner consistent with patient consent and center-specific Institutional Review Board (IRB) policies. The exact approach varies by center, but largely falls into one of three categories: IRB-approved patient-consent to sharing of de-identified data, captured at time of molecular testing; IRB waivers and; and IRB approvals of GENIE-specific research proposals.  Additionally, all data has been de-identified via the HIPAA Safe Harbor Method.  Full details regarding the HIPAA Safe Harbor Method are available online at:  https://www.hhs.gov/hipaa/for-professionals/privacy/special-topics/de-identification/.

\addcontentsline{toc}{section}{Summary of Data by Center}
\section*{Summary of Data by Center}

This section shows the summary of data


<<results=tex, echo=F>>=
number_of_panels = xtable(table(assayinfodf$CENTER),
                          col.names = c("center", "number of panels"),
                          caption = 'Number of Panels per Center')
names(number_of_panels) = "Number of panels"
print(number_of_panels, floating = F,
      caption.placement = "top",
      tabular.environment = "longtable")

@


\begin{figure}
\begin{center}
<<echo = F, fig=TRUE, fig.width = 4, fig.height = 4>>=
library_selection <- ggplot(assayinfodf, aes(library_selection)) + geom_bar() + theme_bw()
library_strategy <- ggplot(assayinfodf, aes(library_strategy)) + geom_bar() + theme_bw()

platform <- ggplot(assayinfodf, aes(platform)) + geom_bar() + theme_bw()
calling_strategy <- ggplot(assayinfodf, aes(calling_strategy)) + geom_bar() + theme_bw()
specimen_tumor_cellularity <- ggplot(assayinfodf,
                                     aes(specimen_tumor_cellularity)) + geom_bar() + theme_bw()

h = ggarrange(library_selection, library_strategy,
          platform, calling_strategy,
          specimen_tumor_cellularity,
          labels = c("A", "B", "C", "D", "E"),
          ncol = 2, nrow = 3)
print(h)
@
\end{center}
\caption{Assay information}
\end{figure}


<<results=tex, echo=F>>=
allowed_coverage = c('hotspot_regions', 'coding_exons',
                     'introns', 'promoters')

coverage_per_panel = get_list_assay_info_table(allowed_coverage,
                                               assayinfodf,
                                               "coverage")
info_tab = xtable(coverage_per_panel,
                  caption = 'Coverage per Panel')
print(info_tab, floating = F, caption.placement = "top",
      tabular.environment = "longtable")

@

<<results=tex, echo=F>>=
allowed_alterations = c("snv", "small_indels", "gene_level_cna",
                        "intragenic_cna", "structural_variants")

alterations_per_panel = get_list_assay_info_table(allowed_alterations,
                                                  assayinfodf,
                                                  "alteration_types")

info_tab = xtable(alterations_per_panel,
                  caption = 'Alteration Types per Panel')
print(info_tab, floating = F, caption.placement = "top",
      tabular.environment = "longtable")


@



<<results=tex, echo=F>>=
allowed_specimen_types = c("formalin_fixed", "FFPE", 'fresh_frozen')

specimen_per_panel = get_list_assay_info_table(allowed_specimen_types,
                                               assayinfodf,
                                               "specimen_type")

info_tab = xtable(specimen_per_panel,
                  caption = 'Specimen Types per Panel')
print(info_tab, floating = F, caption.placement = "top",
      tabular.environment = "longtable")
@


<<results=tex, echo=F>>=
gene_number_info = assayinfodf[, c("SEQ_ASSAY_ID", "gene_padding", "number_of_genes")]

info_tab = xtable(gene_number_info[order(gene_number_info$SEQ_ASSAY_ID), ],
                  caption = 'Sequence Assay genomic information')

print(info_tab, include.rownames = F, floating = F,
      caption.placement = "top", tabular.environment = "longtable")
@





\addcontentsline{toc}{subsection}{Genomic Profiling at Each Center}
\subsection*{Genomic Profiling at Each Center}

\textbf{\href{https://www.cruk.cam.ac.uk/}{Cancer Research UK Cambridge Centre, University of Cambridge} (CRUK)}\\


Sequencing data (SNVs/Indels):\\\\
DNA was quantified using Qubit HS dsDNA assay (Life Technologies, CA) and libraries were prepared from a total of 50 ng of DNA using Illumina's Nextera Custom Target Enrichment kit (Illumina, CA). In brief, a modified Tn5 transposase was used to simultaneously fragment DNA and attach a transposon sequence to both end of the fragments generated. This was followed by a limited cycle PCR amplification (11 cycles) using barcoded oligonucleotides that have primer sites on the transposon sequence generating 96 uniquely barcoded libraries per run. The libraries were then diluted and quantified using Qubit HS dsDNA assay.\\\\
Five hundred nanograms from each library were pooled into a capture pool of 12 samples. Enrichment probes (80-mer) were designed and synthesized by Illumina; these probes were designed to enrich for all exons of the target genes, as well for 500 bp up- and downstream of the gene. The capture was performed twice to increase the specificity of the enrichment. Enriched libraries were amplified using universal primers in a limited cycle PCR (11 cycles). The quality of the libraries was assessed using Bioanalyser (Agilent Technologies, CA) and quantified using KAPA Library Quantification Kits (Kapa Biosystems, MA).\\\\
Products from four capture reactions (that is, 48 samples) were pooled for sequencing in a lane of Illumina HiSeq 2,000. Sequencing (paired-end, 100 bp) of samples and demultiplexing of libraries was performed by Illumina (Great Chesterford, UK).\\\\
The sequenced reads were aligned with Novoalign, and the resulting BAM files were preprocessed using the GATK Toolkit. Sequencing quality statistics were obtained using the GATK's DepthOfCoverage tool and Picard's CalculateHsMetrics. Coverage metrics are presented in Supplementary Fig. 1. Samples were excluded if <25\% of the targeted bases were covered at a minimum coverage of 50 x.\\\\
The identities of those samples with copy number array data available were confirmed by analyzing the samples' genotypes at loci covered by the Affymetrix SNP6 array. Genotype calls from the sequencing data were compared with those from the SNP6 data that was generated for the original studies. This was to identify possible contamination and sample mix-ups, as this would affect associations with other data sets and clinical parameters.\\
To identify all variants in the samples, we used MuTect (without any filtering) for SNVs and the Haplotype Caller for indels. All reads with a mapping quality <70 were removed prior to calling. Variants were annotated with ANNOVAR using the genes' canonical transcripts as defined by Ensembl. Custom scripts were written to identify variants affecting splice sites using exon coordinates provided by Ensembl. Indels were referenced by the first codon they affected irrespective of length; for example, insertions of two bases and five bases at the same codon were classed together
\\\\ To obtain the final set of mutation calls, we used a two-step approach, first removing any spurious variant calls arising as a consequence of sequencing artefacts (generic filtering) and then making use of our normal samples and the existing data to identify somatic mutations (somatic filtering). For both levels of filtering, we used hard thresholds that were obtained, wherever possible, from the data itself. For example, some of our filtering parameters were derived from considering mutations in technical replicates (15 samples sequenced in triplicate). We compared the distributions of key parameters (including quality scores, depth, VAF) for concordant (present in all three replicates) and discordant (present in only one out of three replicates) variants to obtain thresholds, and used ROC analysis to select the parameters that best identified concordant variants.
\\\\ SNV filtering\\
- Based on our analysis of replicates, SNVs with MuTect quality scores <6.95 were removed.\\
- We removed those variants that overlapped with repetitive regions of MUC16 (chromosome 19: 8,955,441–9,044,530). This segment contains multiple tandem repeats (mucin repeats) that are highly susceptible to misalignment due to sequence similarity.\\
- Variants that failed MuTect's internal filters due to 'nearby\_gap\_events' and 'poor\_mapping\_regional\_alternate\_allele\_mapq' were removed.\\
- Fisher's exact test was used to identify variants exhibiting read direction bias (variants occurring significantly more frequently in one read direction than in the other; FDR=0.0001). These were filtered out from the variant calls.\\
- SNVs present at VAFs smaller than 0.1 or at loci covered by fewer than 10 reads were removed, unless they were also present and confirmed somatic in the Catalogue of Somatic Mutations in Cancer (COSMIC). The presence of well-known PIK3CA mutations present at low VAFs was confirmed by digital PCR (see below), and supported the use of COSMIC when filtering SNVs.\\
- We removed all SNVs that were present in any of the three populations (AMR, ASN, AFR) in the 1,000 Genomes study (Phase 1, release 3) with a population alternate allele frequency of >1\%. \\\\\\
Microarray data (Copy number):\\\\
- We used the normal samples in our data set (normal pool) to control for both sequencing noise and germline variants, and removed any SNV observed in the normal pool (at a VAF of at least 0.1). However, for SNVs present in more than two breast cancer samples in COSMIC, we used more stringent thresholds, removing only those that were observed in >5\% of normal breast tissue or in >1\% of blood samples. The different thresholds were used to avoid the possibility of contamination in the normal pool affecting filtering of known somatic mutations. This is analogous to the optional 'panel of normals' filtering step used by MuTect in paired mode, in which mutations present in normal samples are removed unless present in a list of known mutations61.\\
Indel filtering\\
- As for SNVs, we removed all indels falling within tandem repeats of MUC16 (coordinates given above).\\
- We removed all indels deemed to be of 'LowQual' by the Haplotype Caller with default parameters (Phred-scaled confidence threshold=30).\\
- As for SNVs, we removed indels displaying read direction bias. Indels with strand bias Phred-scaled scores >40 were removed.\\
- We downloaded the Simple Repeats and Microsatellites tracks from the UCSC Table Browser, and removed all indels overlapping these regions. We also removed all indels that overlapped homopolymer stretches of six or more bases.\\
- As for SNVs, indels were removed if present in the 1,000 Genomes database at an allele frequency >1\%, or if they were present in normal samples in our data set. Thresholds were adjusted as for SNVs if the indel was present in COSMIC. The same thresholds for depth and VAF were used.\\\\
DNA was hybridized to Affymetrix SNP 6.0 arrays per the manufacturer's instructions. ASCAT was used to obtain segmented copy number calls and estimates of tumour ploidy and purity. Somatic CNAs were obtained by removing germline CNVs as defined in the original METABRIC study3. We defined regions of LOH as those in which there were no copies present of either the major or minor allele, irrespective of total copy number. Recurrent CNAs were identified with GISTIC2, with log2 ratios obtained by dividing the total number of copies by tumour ploidy for each ASCAT segment. Thresholds for identifying gains and losses were set to 0.4 and (-)0.5, respectively; these values were obtained by examining the distribution of log2 ratios to identify peaks associated with copy number states. A broad length cut-off of 0.98 was used, and peaks were assessed to rule out probe artefacts and CNVs that may have been originally missed.\\

\textbf{\href{https://cancer.columbia.edu/}{Herbert Irving Comprehensive Cancer Center, Columbia University} (COLU)}\\


\textbf{\href{https://www.dana-farber.org/}{Dana-Farber Cancer Institute} (DFCI)}\\

\textbf{\href{http://www.dukecancerinstitute.org/}{Duke Cancer Institute} (DUKE)}\\

\textbf{\href{https://www.gustaveroussy.fr/en}{Institut Gustave Roussy} (GRCC)}\\

\textbf{\href{https://www.mdanderson.org/}{The University of Texas MD Anderson Cancer Center} (MDA)}\\

\textbf{\href{https://www.mskcc.org/}{Memorial Sloan Kettering Cancer Center} (MSK)}\\

\textbf{\href{https://www.jhu.edu/}{Johns Hopkins Sidney Kimmel Comprehensive Cancer Center} (JHU)}\\

\textbf{\href{https://www.nki.nl/}{Netherlands Cancer Center, The Netherlands} (NKI)}\\


\textbf{\href{https://oregon.providence.org/our-services/p/providence-cancer-center/}{Providence Health \& Services Cancer Institute} (PHS)}\\


\textbf{\href{https://www.swedish.org/services/cancer-institute}{Swedish Cancer Institute} (SCI)}\\


\textbf{\href{https://www.uhn.ca/PrincessMargaret}{Princess Margaret Cancer Centre, University Health Network} (UHN)}\\


\textbf{\href{https://www.vhio.net/en/}{Vall d'Hebron Institute of Oncology} (VHIO)}\\

\textbf{\href{https://www.vicc.org/}{Vanderbilt-Ingram Cancer Center} (VICC)}\\


\textbf{\href{https://www.wakehealth.edu/}{Wake Forest University Health Sciences, Wake Forest Baptist Medical Center} (WAKE)}\\




\addcontentsline{toc}{section}{Pipeline for Annotating Mutations and Filtering Putative Germline SNPs}
\section*{Pipeline for Annotating Mutations and Filtering Putative Germline SNPs}

Contributing GENIE centers provided mutation data in \href{http://samtools.github.io/hts-specs/}{Variant Call Format (VCF vcf2maf v1.6.17)} or Mutation Annotation Format (\href{https://docs.gdc.cancer.gov/Data/File\_Formats/MAF\_Format}{GDC MAF v1.0.0}) with additional fields for read counts supporting variant alleles, reference alleles, and total depth. Some “MAF-like” text files with \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/data/minimalist\_test\_maf.tsv}{minimal required columns} were also received from the participating centers. These various input formats were converted into a complete tab-separated MAF format, with a \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/docs/vep\_maf\_readme.txt}{standardized set of additional columns} using either vcf2maf or maf2maf \href{https://github.com/mskcc/vcf2maf/tree/v1.6.17}{v1.6.17} wrappers around the Variant Effect Predictor (\href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/docs/vep\_maf\_readme.txt}{VEPv86}). The vcf2maf “custom-enst” option overrode VEP’s canonical isoform for most genes, with \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/data/isoform\_overrides\_uniprot}{Uniprot’s canonical isoform}.\\\\
While the GENIE data available from Sage contains all mutation data, the following mutation types are automatically filtered upon import into the \href{http://www.cbioportal.org/genie}{cBioPortal}: Silent, Intronic, 3' UTR, 3' Flank, 5' UTR, 5' Flank and Intergenic region (IGR).\\\\
Seventeen of the nineteen GENIE participating centers performed tumor-only sequencing i.e. without also sequencing a patient-matched control sample like blood, to isolate somatic events. These centers minimized artifacts and germline events using pooled controls from unrelated individuals, or using databases of known artifacts, common germline variants, and recurrent somatic mutations. However, there remains a risk that such centers may inadvertently release germline variants that can theoretically be used for patient re-identification. To minimize this risk, the GENIE consortium developed a stringent germline filtering pipeline, and applied it uniformly to all variants across all centers. This pipeline flags sufficiently recurrent artifacts and germline events reported by the Exome Aggregation Consortium (\href{http://exac.broadinstitute.org}{ExAC}). Specifically, the non-TCGA subset VCF of ExAC 0.3.1 was used after excluding known somatic events in this \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/data/known\_somatic\_sites.bed}{bed file}:\\\\

\begin{itemize}
    \item{Hotspots from Chang et al. minus some likely artifacts (dx.doi.org/10.1038/nbt.3391).}
    \item{Somatic mutations associated with clonal hematopoietic expansion from Xie et al.(dx.doi.org/10.1038/nm.3733)}
    \item{Somatically mutable germline sites at MSH6:F1088, TP53:R290, TERT:E280, ASXL1:G645\_G646.}
\end{itemize}

The resulting VCF was used with vcf2maf’s "filter-vcf" option, to match each variant position and allele to per-subpopulation allele counts. If a variant was seen more than 10 times in any of the 7 ExAC subpopulations, it was tagged as a "common\_variant" (vcf2maf’s "max-filter-ac" option), and subsequently removed. This >10 allele count (AC) cutoff was selected because it tagged no more than 1\% of the somatic calls across all MSK-IMPACT samples with patient-matched controls.\\\\


\addcontentsline{toc}{section}{Description of Data Files}
\section*{Description of Data Files}


Description on most of the data files can be found under the "Data files" section for \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats}{cBioPortal file formats}.


\begin{center}
\begin{longtable}{ |p{40mm}|p{40mm}|p{50mm}| }
\caption{GENIE Data Files} \\
  \hline
File Name & Description & Details \\ 
  \hline
data\_mutations\_extended.txt & Mutation data & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#mutation-data}{MAF format} \\
   \hline
data\_CNA.txt & Discritized copy number data.  \newline Note: Not all centers contributed copy number data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#discrete-copy-number-data}{CNA format} \\
   \hline
data\_fusions.txt & Structural variant data. \newline Note: not all centers contributed structural rearrangement data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#fusion-data}{Fusion format} \\
   \hline
genomic\_information.txt &  File describing genomic coordinates covered by all platforms for GENIE data.  \newline Note: Not all centers contributed copy number data to GENIE. & This is not a cBioPortal file format.  Chromosome, Start\_Position, End\_Position: Gene positions.  Hugo\_Symbol: Re-mapped gene symbol based on gene positions. ID: Center submitted gene symbols. SEQ\_ASSAY\_ID: The institutional assay identifier for genomic testing platform. Feature\_Type: "exon", "intron", or "intergenic". includeInPanel: Used to define gene panel files for cBioPortal.  clinicalReported: These are the genes that were clinically Reported. Blank means information not provided.\\
   \hline
genie\_data\_cna\_hg19.seg  & Segmented copy number data.  \newline Note: Not all centers contributed copy number data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#segmented-data}{SEG format} \\
   \hline
data\_clinical.txt & De-identified tier 1 clinical data. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#clinical-data}{clinical format}. See Clinical Data section below for more details. \\
   \hline
\end{longtable}
\end{center}



\addcontentsline{toc}{section}{Clinical Data}
\section*{Clinical Data}

\begin{center}
\begin{longtable}{ |p{45mm}|p{40mm}|p{50mm}| }
\caption{GENIE Clinical Data Fields} \\
  \hline
Data Element & Example Values & Data Description \\ 
  \hline
AGE\_AT\_SEQ\_REPORT & Integer values, <18 or >89.  &  The age of the patient at the time that the sequencing results were reported. Age is masked for patients aged 90 years and greater and for patients under 18 years. \\
   \hline
CENTER & MSK & The center submitting the clinical and genomic data   \\
   \hline
ETHNICITY & Non-Spanish/non-Hispanic \newline Spanish/Hispanic \newline Unknown & Indication of Spanish/Hispanic origin of the patient; this data element maps to the NAACCR v16, Element \#190. Institutions not collecting Spanish/Hispanic origin have set this column to Unknown. \\
   \hline
ONCOTREE\_CODE & LUAD & The primary cancer diagnosis “main type”, based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01} \\
   \hline
PATIENT\_ID & GENIE-JHU-1234 & The unique, anonymized patient identifier for the GENIE project. Conforms to the following the convention: GENIE-CENTER-1234. The first component is the string, "GENIE"; the second component is the Center abbreviation. The third component is an anonymized unique identifier for the patient.\\
   \hline
PRIMARY\_RACE & Asian, Black, Native American Other, Unknown, White & The primary race recorded for the patient; this data element maps to the NAACCR v16, Element \#160. For institutions collecting more than one race category, this race code is the primary race for the patient. Institutions not collecting race have set this field to Unknown. \\
   \hline
SAMPLE\_ID & GENIE-JHU-1234-9876 & The unique, anonymized sample identifier for the GENIE project. Conforms to the following the convention: GENIE-CENTER-1234-9876. The first component is the string, "GENIE"; the second component is the Center abbreviation. The third component is an anonymized, unique patient identifier. The fourth component is a unique identifier for the sample that will distinguish between two or more specimens from a single patient. \\
   \hline
SAMPLE\_TYPE &  Primary, Metastasis, Unspecified & Sample type, e.g. Primary or Metastasis. \\
   \hline
SEQ\_ASSAY\_ID & DFCI-ONCOPANEL-1 & The institutional assay identifier for genomic testing platform. Components are separated by hyphens, with the first component corresponding to the Center's abbreviation. All specimens tested by the same platform should have the same identifier. \\
   \hline
SEX & Female, Male & The patient’s sex code; this data element maps to the NAACCR v16, Element \#220. \\
   \hline
CANCER\_TYPE & Non-Small Cell Lung Cancer & The primary cancer diagnosis "main type", based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. For example, the OncoTree code of LUAD maps to: "Non-Small Cell Lung Cancer".  The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01} \\
   \hline
CANCER\_TYPE\_DETAILED & Lung Adenocarcinoma & The primary cancer diagnosis label, based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. For example, the OncoTree code of LUAD maps to the label: "Lung Adenocarcinoma (LUAD)". The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01}  \\
   \hline
BIRTH\_YEAR & 1965 & The four-digit year associated with the patient's date of birth.  \\
   \hline
YEAR\_CONTACT & 1990  & Record the year that the patient is last known to be alive - data to be derived from information including and EHR, tumor registry, or other system. \\
   \hline
YEAR\_DEATH & 2010 & Year of death  \\
   \hline
INT\_CONTACT & 15000  &  Interval in days from DOB to date of last contact. \\
   \hline
INT\_DOD & 15000 & Interval in days from DOB to DOD.  \\
   \hline
DEAD & Yes, No  & Is this patient known to be deceased? \\
   \hline
\end{longtable}
\end{center}

Cancer types are reported using the \href{http://oncotree.mskcc.org}{OncoTree ontology} originally developed at Memorial Sloan Kettering Cancer Center. The GENIE 8.0 data release series uses the OncoTree specification from June 1, 2018, containing diagnosis codes for 686 tumor types from 32 tissues. The centers participating in GENIE applied the OncoTree cancer types to the tested specimens in a variety of methods depending on center-specific workflows. A brief description of how the cancer type assignment process for each center is specified in Table 6.



\begin{center}
\begin{longtable}{ |p{10mm}|p{100mm}| }
\caption{Center Strategies for OncoTree Assignment} \\
  \hline
Center & OncoTree Cancer Type Assignments \\ 
  \hline
COLU &  Original diagnosis from pathologist was mapped to OncoTree diagnosis by medical oncologist and research manager.\\
   \hline
CRUK & Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
DFCI &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
DUKE & Anatomic and molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
GRCC &  OncoTree cancer types were mapped from ICD-O codes. If no ICD-O code was available, a staff scientist and an oncologist mapped the diagnosis made by the pathologist to Onco Tree cancer type. \\
   \hline
JHU &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
MDA &  OncoTree cancer types were mapped from ICD-O codes. \\
   \hline
MSK &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
NKI &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
PHS &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
SCI &  Original diagnosis from the pathology report was mapped to OncoTree diagnosis by a research coordinator and molecular pathologist. \\
   \hline
UHN & The original diagnosis was mapped to OncoTree by a medical oncologist and research manager.  \\
   \hline
VHIO & Original diagnosis from pathologist or medical oncologist was mapped to OncoTree diagnosis by research data curator.  \\
   \hline
VICC & OncoTree cancer types were mapped from ICD-O codes. If no ICD-O code was available, a research manager mapped the diagnosis to an OncoTree cancer type.  \\
   \hline
WAKE &  Diagnoses from Foundation Medicine and Caris Diagnostics to ICD-O-3, then mapped from ICD-O-3 to Oncotree. \\
   \hline
YALE &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
\end{longtable}
\end{center}



\addcontentsline{toc}{section}{Abbreviations and Acronym Glossary}
\section*{Abbreviations and Acronym Glossary}


\begin{center}
\begin{longtable}{ |p{20mm}|p{100mm}| }
  \hline
Abbreviation & Full Term \\
  \hline
AACR & American Association for Cancer Research, Philadelphia, PA, USA \\
   \hline
CNA & Copy number alterations \\
   \hline
CNV & Copy number variants \\
   \hline
COLU & The Herbert Irving Comprehensive Cancer Center, Columbia University, New York, NY, USA \\
   \hline
CRUK & Cancer Research UK Cambridge Centre, University of Cambridge, Cambridge, England \\
   \hline
DFCI & Dana-Farber Cancer Institute, Boston, MA, USA \\
   \hline
DUKE & Duke Cancer Institute, Duke University Health System, Durham, NC, USA \\
   \hline
FFPE & Formalin-fixed, paraffin-embedded \\
   \hline
GENIE & Genomics, Evidence, Neoplasia, Information, Exchange \\
   \hline
GRCC & Institut Gustave Roussy, Paris, France \\
   \hline
HIPAA & Health Insurance Portability and Accountability Act \\
   \hline
IRB & Institutional Review Board \\
   \hline
JHU & Johns Hopkins Sidney Kimmel Comprehensive Cancer Center, Baltimore, MD, USA \\
   \hline
MAF & Mutation annotation format \\
   \hline
MDA & The University of Texas MD Anderson Cancer Center, Houston, TX, USA \\
   \hline
MSK & Memorial Sloan Kettering Cancer Center, New York, NY, USA \\
   \hline
NAACCR & North American Association of Central Cancer Registries \\
   \hline
NGS & Next-generation sequencing \\
   \hline
NKI & Netherlands Cancer Institute, Amsterdam, Netherlands \\
   \hline
PCR & Polymerase chain reaction \\
   \hline
PHS & Providence Health \& Services Cancer Institute, Portland, OR, USA \\
   \hline
SCI & Swedish Cancer Institute, Seattle, WA, USA \\
   \hline
SNP & Single-nucleotide polymorphism \\
   \hline
SNV & Single-nucleotide variants \\
   \hline
UCHI & University of Chicago Comprehensive Cancer Center, Chicago, IL, USA \\
   \hline
UHN & Princess Margaret Cancer Centre, University Health Network, Toronto, Ontario, Canada \\
   \hline
VCF & Variant Call Format \\
   \hline
VHIO & Vall d' Hebron Institute of Oncology, Barcelona, Spain \\
   \hline
VICC & Vanderbilt-Ingram Cancer Center, Nashville, TN, USA \\
   \hline
WAKE & Wake Forest University Health Sciences (Wake Forest Baptist Medical Center), Winston-Salem, NC, USA \\
   \hline
YALE & Yale University (Yale Cancer Center), New Haven, Connecticut, USA \\
   \hline
\end{longtable}
\end{center}



\end{document}