\documentclass{article}
\usepackage{a4wide}
\usepackage{longtable}
\usepackage{tocloft}
\usepackage{blindtext}
 \cftpagenumbersoff{section}
 \cftpagenumbersoff{subsection}
\usepackage{color}   %May be necessary if you want to color links
\usepackage{hyperref}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\usepackage{graphicx}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{7.0-public}
\lhead{\today}

\title{AACR GENIE {{release}}.0-public Data Guide}
\author{AACR}


% Definition of \maketitle
\makeatletter         
\def\@maketitle{
\begin{center}
\includegraphics[width=1.0\textwidth]{genie_banner.png}\\[8ex]
{\Huge \@title }\\[4ex] 
{\Large  \@author}\\[4ex] 
\@date\\[8ex]
\end{center}}
\makeatother

\begin{document}


\SweaveOpts{concordance=TRUE}


<<echo=F, include=F, results=hide>>=
library(xtable)
library(synapser)
synLogin()
centers = suppressMessages(synTableQuery('select "Center Abbreviation", "Center" from syn16982837',
                                         includeRowIdAndRowVersion = F))
assayinfo = suppressMessages(synTableQuery('SELECT * FROM syn17009222',
                                           includeRowIdAndRowVersion = F))

genomic_profile = suppressMessages(synTableQuery('SELECT * FROM syn21642088',
                                                 includeRowIdAndRowVersion = F))

@

\maketitle

\tableofcontents

\newpage

\addcontentsline{toc}{section}{About this Document}
\section*{About this Document}

This document provides an overview of the {{release}}\textsuperscript{th} public release of American Association for Cancer Research (AACR) GENIE data.

\addcontentsline{toc}{section}{Version of Data}
\section*{Version of Data}

AACR GENIE Project Data: Version {{release}}.0-pubic.
\\*
AACR Project GENIE data versions follow a numbering scheme derived from \href{https://semver.org/}{semantic versioning}, where the digits in the version correspond to: major.patch-release-type.  “Major” releases are public releases of new sample data.  “Patch” releases are corrections to major releases, including data retractions.  “Release-type” refers to whether the release is a public AACR Project GENIE release or a private/consortium-only release.  Public releases will be denoted with the nomenclature “X.X-public” and consortium-only private releases will be denoted with the nomenclature “X.X-consortium".


\addcontentsline{toc}{section}{Data Access}
\section*{Data Access}
AACR GENIE Data is currently available via two mechanisms:
\begin{itemize}
    \item{Synapse Platform (Sage Bionetworks):  http://synapse.org/genie}
    \item{cBioPortal for Cancer Genomics (MSK): http://www.cbioportal.org/genie/}
\end{itemize}


\addcontentsline{toc}{section}{Terms Of Access}
\section*{Terms Of Access}
All users of the AACR Project GENIE data must agree to the following terms of use; failure to abide by any term herein will result in revocation of access.

\begin{itemize}
    \item{Users will not attempt to identify or contact individual participants from whom these data were collected by any means.}
    \item{Users will not redistribute the data without express written permission from the AACR Project GENIE Coordinating Center (send email to:  info@aacrgenie.org).}
\end{itemize}

When publishing or presenting work using or referencing the AACR Project GENIE dataset please include the following attributions:

\begin{itemize}
\item{Please cite: \emph{The AACR Project GENIE Consortium. AACR Project GENIE: Powering Precision Medicine Through An International Consortium, Cancer Discov. 2017 Aug;7(8):818-831} and include the version of the dataset used.}
\item{The authors would like to acknowledge the American Association for Cancer Research and its financial and material support in the development of the AACR Project GENIE registry, as well as members of the consortium for their commitment to data sharing. Interpretations are the responsibility of study authors.}
\end{itemize}

Posters and presentations should include the AACR Project GENIE logo.



\addcontentsline{toc}{section}{Introduction to AACR GENIE}
\section*{Introduction to AACR GENIE}

The AACR Project Genomics, Evidence, Neoplasia, Information, Exchange (GENIE) is a multi-phase, multi-year, international data-sharing project that aims to catalyze precision cancer medicine.  The GENIE platform will integrate and link clinical-grade cancer genomic data with clinical outcome data for tens of thousands of cancer patients treated at multiple international institutions. The project fulfills an unmet need in oncology by providing the statistical power necessary to improve clinical decision-making, to identify novel therapeutic targets, to understand of patient response to therapy, and to design new biomarker-driven clinical trials.  The project will also serve as a prototype for aggregating, harmonizing, and sharing clinical-grade, next-generation sequencing (NGS) data obtained during routine medical practice.

The data within GENIE is being shared with the global research community.  The database currently contains CLIA-/ISO-certified genomic data obtained during the course of routine practice at multiple international institutions (Table 1), and will continue to grow as more patients are treated at additional participating centers.

<<results=tex, echo=F>>=
xtab = xtable(centers$asDataFrame(),
              caption = 'Participating Centers')
#cat("\\makebox[\\textwidth]{")
print(xtab, include.rownames = F, floating = F,
      caption.placement = "top", tabular.environment = "longtable")
#cat("}")
@





\addcontentsline{toc}{section}{Human Subjects Protection and Privacy}
\section*{Human Subjects Protection and Privacy}

Protection of patient privacy is paramount, and the AACR GENIE Project therefore requires that each participating center share data in a manner consistent with patient consent and center-specific Institutional Review Board (IRB) policies. The exact approach varies by center, but largely falls into one of three categories: IRB-approved patient-consent to sharing of de-identified data, captured at time of molecular testing; IRB waivers and; and IRB approvals of GENIE-specific research proposals.  Additionally, all data has been de-identified via the HIPAA Safe Harbor Method.  Full details regarding the HIPAA Safe Harbor Method are available online at:  https://www.hhs.gov/hipaa/for-professionals/privacy/special-topics/de-identification/.

\addcontentsline{toc}{section}{Summary of Data by Center}
\section*{Summary of Data by Center}

<<results=tex, echo=F>>=
info_tab = xtable(assayinfo$asDataFrame()[,c("SEQ_ASSAY_ID", "platform", 
                                             "library_strategy", "library_selection",
                                             "instrument_model",
                                             "specimen_type", "specimen_tumor_cellularity")],
                  caption = 'Genomic Data Characterization')
#cat("\\makebox[\\textwidth]{")
print(info_tab, include.rownames = F, floating = F, size = "\\tiny",
      caption.placement = "top", tabular.environment = "longtable")
# cat("}")
#cat("\\endgroup")

@


\addcontentsline{toc}{subsection}{Genomic Profiling at Each Center}
\subsection*{Genomic Profiling at Each Center}

<<results=tex, echo=F>>=
sanitize <- function(str) {
    # gsub('([#$%&~_\\^\\\\{}])', '\\\\\\\\\\1', str, perl = TRUE);
    str <- gsub("%", "\\\\%", str, fixed = TRUE)
    str <- gsub("_", "\\\\_", str, fixed = TRUE)
    str
}
genomic_profiledf = genomic_profile$asDataFrame()
#genomic_profiledf$genomic_profile = gsub("\\%", "\\%",
#                                         genomic_profiledf$genomic_profile)
#cat("\\makebox[\\textwidth]{")
all_profiles = ''
for (profile in genomic_profiledf$genomic_profile) {
  # new_profile = sanitizeLatexS(profile)
  all_profiles = paste(all_profiles, profile, sep = " ")
  #cat(new_profile)
}

#cat("}")
#cat("\\endgroup")
@

\Sexpr{sanitize(all_profiles)}


\addcontentsline{toc}{section}{Description of Data Files}
\section*{Description of Data Files}


Description on most of the data files can be found under the "Data files" section for \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats}{cBioPortal file formats}.


\begin{center}
\begin{longtable}{ |p{40mm}|p{40mm}|p{50mm}| }
  \hline
File Name & Description & Details \\ 
  \hline
data\_mutations\_extended.txt & Mutation data & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#mutation-data}{MAF format} \\
   \hline
data\_CNA.txt & Discritized copy number data.  \newline Note: Not all centers contributed copy number data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#discrete-copy-number-data}{CNA format} \\
   \hline
data\_fusions.txt & Structural variant data. \newline Note: not all centers contributed structural rearrangement data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#fusion-data}{Fusion format} \\
   \hline
genomic\_information.txt &  File describing genomic coordinates covered by all platforms for GENIE data.  \newline Note: Not all centers contributed copy number data to GENIE. & This is not a cBioPortal file format.  Chromosome, Start\_Position, End\_Position: Gene positions.  Hugo\_Symbol: Re-mapped gene symbol based on gene positions. ID: Center submitted gene symbols. SEQ\_ASSAY\_ID: The institutional assay identifier for genomic testing platform. Feature\_Type: "exon", "intron", or "intergenic". includeInPanel: Used to define gene panel files for cBioPortal.  clinicalReported: These are the genes that were clinically Reported. Blank means information not provided.\\
   \hline
genie\_data\_cna\_hg19.seg  & Segmented copy number data.  \newline Note: Not all centers contributed copy number data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#segmented-data}{SEG format} \\
   \hline
data\_clinical.txt & De-identified tier 1 clinical data. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#clinical-data}{clinical format}. See Clinical Data section below for more details. \\
   \hline
\end{longtable}
\end{center}



\addcontentsline{toc}{section}{Clinical Data}
\section*{Clinical Data}

\begin{center}
\begin{longtable}{ |p{45mm}|p{40mm}|p{50mm}| }
  \hline
Data Element & Example Values & Data Description \\ 
  \hline
AGE\_AT\_SEQ\_REPORT & Integer values, <18 or >89.  &  The age of the patient at the time that the sequencing results were reported. Age is masked for patients aged 90 years and greater and for patients under 18 years. \\
   \hline
CENTER & MSK & The center submitting the clinical and genomic data   \\
   \hline
ETHNICITY & Non-Spanish/non-Hispanic \newline Spanish/Hispanic \newline Unknown & Indication of Spanish/Hispanic origin of the patient; this data element maps to the NAACCR v16, Element \#190. Institutions not collecting Spanish/Hispanic origin have set this column to Unknown. \\
   \hline
ONCOTREE\_CODE & LUAD & The primary cancer diagnosis “main type”, based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01} \\
   \hline
PATIENT\_ID & GENIE-JHU-1234 & The unique, anonymized patient identifier for the GENIE project. Conforms to the following the convention: GENIE-CENTER-1234. The first component is the string, "GENIE"; the second component is the Center abbreviation. The third component is an anonymized unique identifier for the patient.\\
   \hline
PRIMARY\_RACE & Asian, Black, Native American Other, Unknown, White & The primary race recorded for the patient; this data element maps to the NAACCR v16, Element \#160. For institutions collecting more than one race category, this race code is the primary race for the patient. Institutions not collecting race have set this field to Unknown. \\
   \hline
SAMPLE\_ID & GENIE-JHU-1234-9876 & The unique, anonymized sample identifier for the GENIE project. Conforms to the following the convention: GENIE-CENTER-1234-9876. The first component is the string, "GENIE"; the second component is the Center abbreviation. The third component is an anonymized, unique patient identifier. The fourth component is a unique identifier for the sample that will distinguish between two or more specimens from a single patient. \\
   \hline
SAMPLE\_TYPE &  Primary, Metastasis, Unspecified & Sample type, e.g. Primary or Metastasis. \\
   \hline
SEQ\_ASSAY\_ID & DFCI-ONCOPANEL-1 & The institutional assay identifier for genomic testing platform. Components are separated by hyphens, with the first component corresponding to the Center's abbreviation. All specimens tested by the same platform should have the same identifier. \\
   \hline
SEX & Female, Male & The patient’s sex code; this data element maps to the NAACCR v16, Element \#220. \\
   \hline
CANCER\_TYPE & Non-Small Cell Lung Cancer & The primary cancer diagnosis "main type", based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. For example, the OncoTree code of LUAD maps to: "Non-Small Cell Lung Cancer".  The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01} \\
   \hline
CANCER\_TYPE\_DETAILED & Lung Adenocarcinoma & The primary cancer diagnosis label, based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. For example, the OncoTree code of LUAD maps to the label: "Lung Adenocarcinoma (LUAD)". The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01}  \\
   \hline
BIRTH\_YEAR & 1965 & The four-digit year associated with the patient's date of birth.  \\
   \hline
YEAR\_CONTACT & 1990  & Record the year that the patient is last known to be alive - data to be derived from information including and EHR, tumor registry, or other system. \\
   \hline
YEAR\_DEATH & 2010 & Year of death  \\
   \hline
INT\_CONTACT & 15000  &  Interval in days from DOB to date of last contact. \\
   \hline
INT\_DOD & 15000 & Interval in days from DOB to DOD.  \\
   \hline
DEAD & Yes, No  & Is this patient known to be deceased? \\
   \hline
\end{longtable}
\end{center}

Cancer types are reported using the \href{http://oncotree.mskcc.org}{OncoTree ontology} originally developed at Memorial Sloan Kettering Cancer Center. The GENIE 8.0 data release series uses the OncoTree specification from June 1, 2018, containing diagnosis codes for 686 tumor types from 32 tissues. The centers participating in GENIE applied the OncoTree cancer types to the tested specimens in a variety of methods depending on center-specific workflows. A brief description of how the cancer type assignment process for each center is specified in Table 6.



\begin{center}
\begin{longtable}{ |p{10mm}|p{100mm}| }
  \hline
Center & OncoTree Cancer Type Assignments \\ 
  \hline
COLU &  Original diagnosis from pathologist was mapped to OncoTree diagnosis by medical oncologist and research manager.\\
   \hline
CRUK & Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
DFCI &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
DUKE & Anatomic and molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
GRCC &  OncoTree cancer types were mapped from ICD-O codes. If no ICD-O code was available, a staff scientist and an oncologist mapped the diagnosis made by the pathologist to Onco Tree cancer type. \\
   \hline
JHU &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
MDA &  OncoTree cancer types were mapped from ICD-O codes. \\
   \hline
MSK &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
NKI &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
PHS &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
SCI &  Original diagnosis from the pathology report was mapped to OncoTree diagnosis by a research coordinator and molecular pathologist. \\
   \hline
UHN & The original diagnosis was mapped to OncoTree by a medical oncologist and research manager.  \\
   \hline
VHIO & Original diagnosis from pathologist or medical oncologist was mapped to OncoTree diagnosis by research data curator.  \\
   \hline
VICC & OncoTree cancer types were mapped from ICD-O codes. If no ICD-O code was available, a research manager mapped the diagnosis to an OncoTree cancer type.  \\
   \hline
WAKE &  Diagnoses from Foundation Medicine and Caris Diagnostics to ICD-O-3, then mapped from ICD-O-3 to Oncotree. \\
   \hline
YALE &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
\end{longtable}
\end{center}



\addcontentsline{toc}{section}{Abbreviations and Acronym Glossary}
\section*{Abbreviations and Acronym Glossary}







\end{document}