\documentclass{article}
\usepackage{a4wide}
\usepackage{longtable}
\usepackage{tocloft}
\usepackage{blindtext}
 \cftpagenumbersoff{section}
 \cftpagenumbersoff{subsection}
\usepackage{color}   %May be necessary if you want to color links
\usepackage{hyperref}
\usepackage{float}
\hypersetup{
    colorlinks=true, %set true if you want colored links
    linktoc=all,     %set to all if you want both sections and subsections linked
    linkcolor=blue,  %choose some color if you want links to stand out
}
\usepackage{graphicx}
\usepackage{fancyhdr}
\pagestyle{fancy}
\rhead{7.0-public}
\lhead{\today}

\title{AACR GENIE {{release}}.0-public Data Guide}
\author{AACR}


% Definition of \maketitle
\makeatletter         
\def\@maketitle{
\begin{center}
\includegraphics[width=1.0\textwidth]{genie_banner.png}\\[8ex]
{\Huge \@title }\\[4ex] 
{\Large  \@author}\\[4ex] 
\@date\\[8ex]
\end{center}}
\makeatother

\begin{document}


\SweaveOpts{concordance=TRUE}


<<echo=F, include=F, results=hide>>=
library(xtable)
library(synapser)
library(ggplot2)
library(ggpubr)

get_list_assay_info_table <- function(allowed_values, assayinfodf, col) {
  na_ind = is.na(assayinfodf[,col])
  seq_assays = assayinfodf$SEQ_ASSAY_ID[!na_ind]
  col_per_panel = matrix(nrow = length(seq_assays),
                         ncol = length(allowed_values),
                         dimnames = list(seq_assays, allowed_values))
  for (panel in seq_assays) {
    str_value = assayinfodf[,col][assayinfodf$SEQ_ASSAY_ID == panel]
    str_value_vector = unlist(strsplit(str_value, ";"))
    col_per_panel[panel, ] = FALSE
    col_per_panel[panel, str_value_vector] = TRUE
  }
  col_per_panel
}

synLogin()
centers = suppressMessages(synTableQuery('select "Center Abbreviation", "Center" from syn16982837',
                                         includeRowIdAndRowVersion = F))
assayinfo = suppressMessages(synTableQuery('SELECT * FROM syn17009222',
                                           includeRowIdAndRowVersion = F))
assayinfodf = assayinfo$asDataFrame()
genomic_profile = suppressMessages(synTableQuery('SELECT * FROM syn21642088',
                                                 includeRowIdAndRowVersion = F))

@

\maketitle

\tableofcontents

\newpage

\addcontentsline{toc}{section}{About this Document}
\section*{About this Document}

This document provides an overview of the {{release}}\textsuperscript{th} public release of American Association for Cancer Research (AACR) GENIE data.

\addcontentsline{toc}{section}{Version of Data}
\section*{Version of Data}

AACR GENIE Project Data: Version {{release}}.0-pubic.
\\*
AACR Project GENIE data versions follow a numbering scheme derived from \href{https://semver.org/}{semantic versioning}, where the digits in the version correspond to: major.patch-release-type.  “Major” releases are public releases of new sample data.  “Patch” releases are corrections to major releases, including data retractions.  “Release-type” refers to whether the release is a public AACR Project GENIE release or a private/consortium-only release.  Public releases will be denoted with the nomenclature “X.X-public” and consortium-only private releases will be denoted with the nomenclature “X.X-consortium".


\addcontentsline{toc}{section}{Data Access}
\section*{Data Access}
AACR GENIE Data is currently available via two mechanisms:
\begin{itemize}
    \item{Synapse Platform (Sage Bionetworks):  http://synapse.org/genie}
    \item{cBioPortal for Cancer Genomics (MSK): http://www.cbioportal.org/genie/}
\end{itemize}


\addcontentsline{toc}{section}{Terms Of Access}
\section*{Terms Of Access}
All users of the AACR Project GENIE data must agree to the following terms of use; failure to abide by any term herein will result in revocation of access.

\begin{itemize}
    \item{Users will not attempt to identify or contact individual participants from whom these data were collected by any means.}
    \item{Users will not redistribute the data without express written permission from the AACR Project GENIE Coordinating Center (send email to:  info@aacrgenie.org).}
\end{itemize}

When publishing or presenting work using or referencing the AACR Project GENIE dataset please include the following attributions:

\begin{itemize}
\item{Please cite: \emph{The AACR Project GENIE Consortium. AACR Project GENIE: Powering Precision Medicine Through An International Consortium, Cancer Discov. 2017 Aug;7(8):818-831} and include the version of the dataset used.}
\item{The authors would like to acknowledge the American Association for Cancer Research and its financial and material support in the development of the AACR Project GENIE registry, as well as members of the consortium for their commitment to data sharing. Interpretations are the responsibility of study authors.}
\end{itemize}

Posters and presentations should include the AACR Project GENIE logo.



\addcontentsline{toc}{section}{Introduction to AACR GENIE}
\section*{Introduction to AACR GENIE}

The AACR Project Genomics, Evidence, Neoplasia, Information, Exchange (GENIE) is a multi-phase, multi-year, international data-sharing project that aims to catalyze precision cancer medicine.  The GENIE platform will integrate and link clinical-grade cancer genomic data with clinical outcome data for tens of thousands of cancer patients treated at multiple international institutions. The project fulfills an unmet need in oncology by providing the statistical power necessary to improve clinical decision-making, to identify novel therapeutic targets, to understand of patient response to therapy, and to design new biomarker-driven clinical trials.  The project will also serve as a prototype for aggregating, harmonizing, and sharing clinical-grade, next-generation sequencing (NGS) data obtained during routine medical practice.

The data within GENIE is being shared with the global research community.  The database currently contains CLIA-/ISO-certified genomic data obtained during the course of routine practice at multiple international institutions (Table 1), and will continue to grow as more patients are treated at additional participating centers.

<<results=tex, echo=F>>=
xtab = xtable(centers$asDataFrame(),
              caption = 'Participating Centers')
#cat("\\makebox[\\textwidth]{")
print(xtab, include.rownames = F, floating = F,
      caption.placement = "top", tabular.environment = "longtable")
#cat("}")
@





\addcontentsline{toc}{section}{Human Subjects Protection and Privacy}
\section*{Human Subjects Protection and Privacy}

Protection of patient privacy is paramount, and the AACR GENIE Project therefore requires that each participating center share data in a manner consistent with patient consent and center-specific Institutional Review Board (IRB) policies. The exact approach varies by center, but largely falls into one of three categories: IRB-approved patient-consent to sharing of de-identified data, captured at time of molecular testing; IRB waivers and; and IRB approvals of GENIE-specific research proposals.  Additionally, all data has been de-identified via the HIPAA Safe Harbor Method.  Full details regarding the HIPAA Safe Harbor Method are available online at:  https://www.hhs.gov/hipaa/for-professionals/privacy/special-topics/de-identification/.

\addcontentsline{toc}{section}{Summary of Data by Center}
\section*{Summary of Data by Center}

This section shows the summary of data


<<results=tex, echo=F>>=
number_of_panels = xtable(table(assayinfodf$CENTER),
                          col.names = c("center", "number of panels"),
                          caption = 'Number of Panels per Center')
names(number_of_panels) = "Number of panels"
print(number_of_panels, floating = F,
      caption.placement = "top",
      tabular.environment = "longtable")

@


\begin{figure}[H]
\begin{center}
<<echo = F, fig=TRUE, fig.width = 4, fig.height = 4>>=
library_selection <- ggplot(assayinfodf, aes(library_selection)) + geom_bar() + theme_bw()
library_strategy <- ggplot(assayinfodf, aes(library_strategy)) + geom_bar() + theme_bw()

platform <- ggplot(assayinfodf, aes(platform)) + geom_bar() + theme_bw()
calling_strategy <- ggplot(assayinfodf, aes(calling_strategy)) + geom_bar() + theme_bw()
specimen_tumor_cellularity <- ggplot(assayinfodf,
                                     aes(specimen_tumor_cellularity)) + geom_bar() + theme_bw()

h = ggarrange(library_selection, library_strategy,
          platform, calling_strategy,
          specimen_tumor_cellularity,
          labels = c("A", "B", "C", "D", "E"),
          ncol = 2, nrow = 3)
print(h)
@
\end{center}
\caption{Assay information}
\end{figure}


<<results=tex, echo=F>>=
allowed_coverage = c('hotspot_regions', 'coding_exons',
                     'introns', 'promoters')

coverage_per_panel = get_list_assay_info_table(allowed_coverage,
                                               assayinfodf,
                                               "coverage")
info_tab = xtable(coverage_per_panel,
                  caption = 'Coverage per Panel')
print(info_tab, floating = F, caption.placement = "top",
      tabular.environment = "longtable")

@

<<results=tex, echo=F>>=
allowed_alterations = c("snv", "small_indels", "gene_level_cna",
                        "intragenic_cna", "structural_variants")

alterations_per_panel = get_list_assay_info_table(allowed_alterations,
                                                  assayinfodf,
                                                  "alteration_types")

info_tab = xtable(alterations_per_panel,
                  caption = 'Alteration Types per Panel')
print(info_tab, floating = F, caption.placement = "top",
      tabular.environment = "longtable")


@



<<results=tex, echo=F>>=
allowed_specimen_types = c("formalin_fixed", "FFPE", 'fresh_frozen')

specimen_per_panel = get_list_assay_info_table(allowed_specimen_types,
                                               assayinfodf,
                                               "specimen_type")

info_tab = xtable(specimen_per_panel,
                  caption = 'Specimen Types per Panel')
print(info_tab, floating = F, caption.placement = "top",
      tabular.environment = "longtable")
@


<<results=tex, echo=F>>=
gene_number_info = assayinfodf[, c("SEQ_ASSAY_ID", "gene_padding", "number_of_genes")]

info_tab = xtable(gene_number_info[order(gene_number_info$SEQ_ASSAY_ID), ],
                  caption = 'Sequence Assay genomic information')

print(info_tab, include.rownames = F, floating = F,
      caption.placement = "top", tabular.environment = "longtable")
@





\addcontentsline{toc}{subsection}{Genomic Profiling at Each Center}
\subsection*{Genomic Profiling at Each Center}

\textbf{\href{https://www.cruk.cam.ac.uk/}{Cancer Research UK Cambridge Centre, University of Cambridge} (CRUK)}\\\\
Sequencing data (SNVs/Indels):\\\\
DNA was quantified using Qubit HS dsDNA assay (Life Technologies, CA) and libraries were prepared from a total of 50 ng of DNA using Illumina's Nextera Custom Target Enrichment kit (Illumina, CA). In brief, a modified Tn5 transposase was used to simultaneously fragment DNA and attach a transposon sequence to both end of the fragments generated. This was followed by a limited cycle PCR amplification (11 cycles) using barcoded oligonucleotides that have primer sites on the transposon sequence generating 96 uniquely barcoded libraries per run. The libraries were then diluted and quantified using Qubit HS dsDNA assay.\\\\
Five hundred nanograms from each library were pooled into a capture pool of 12 samples. Enrichment probes (80-mer) were designed and synthesized by Illumina; these probes were designed to enrich for all exons of the target genes, as well for 500 bp up- and downstream of the gene. The capture was performed twice to increase the specificity of the enrichment. Enriched libraries were amplified using universal primers in a limited cycle PCR (11 cycles). The quality of the libraries was assessed using Bioanalyser (Agilent Technologies, CA) and quantified using KAPA Library Quantification Kits (Kapa Biosystems, MA).\\\\
Products from four capture reactions (that is, 48 samples) were pooled for sequencing in a lane of Illumina HiSeq 2,000. Sequencing (paired-end, 100 bp) of samples and demultiplexing of libraries was performed by Illumina (Great Chesterford, UK).\\\\
The sequenced reads were aligned with Novoalign, and the resulting BAM files were preprocessed using the GATK Toolkit. Sequencing quality statistics were obtained using the GATK's DepthOfCoverage tool and Picard's CalculateHsMetrics. Coverage metrics are presented in Supplementary Fig. 1. Samples were excluded if <25\% of the targeted bases were covered at a minimum coverage of 50 x.\\\\
The identities of those samples with copy number array data available were confirmed by analyzing the samples' genotypes at loci covered by the Affymetrix SNP6 array. Genotype calls from the sequencing data were compared with those from the SNP6 data that was generated for the original studies. This was to identify possible contamination and sample mix-ups, as this would affect associations with other data sets and clinical parameters.\\
To identify all variants in the samples, we used MuTect (without any filtering) for SNVs and the Haplotype Caller for indels. All reads with a mapping quality <70 were removed prior to calling. Variants were annotated with ANNOVAR using the genes' canonical transcripts as defined by Ensembl. Custom scripts were written to identify variants affecting splice sites using exon coordinates provided by Ensembl. Indels were referenced by the first codon they affected irrespective of length; for example, insertions of two bases and five bases at the same codon were classed together
\\\\ To obtain the final set of mutation calls, we used a two-step approach, first removing any spurious variant calls arising as a consequence of sequencing artefacts (generic filtering) and then making use of our normal samples and the existing data to identify somatic mutations (somatic filtering). For both levels of filtering, we used hard thresholds that were obtained, wherever possible, from the data itself. For example, some of our filtering parameters were derived from considering mutations in technical replicates (15 samples sequenced in triplicate). We compared the distributions of key parameters (including quality scores, depth, VAF) for concordant (present in all three replicates) and discordant (present in only one out of three replicates) variants to obtain thresholds, and used ROC analysis to select the parameters that best identified concordant variants.
\\\\ SNV filtering\\
- Based on our analysis of replicates, SNVs with MuTect quality scores <6.95 were removed.\\
- We removed those variants that overlapped with repetitive regions of MUC16 (chromosome 19: 8,955,441–9,044,530). This segment contains multiple tandem repeats (mucin repeats) that are highly susceptible to misalignment due to sequence similarity.\\
- Variants that failed MuTect's internal filters due to 'nearby\_gap\_events' and 'poor\_mapping\_-regional\_alternate\_allele\_mapq' were removed.\\
- Fisher's exact test was used to identify variants exhibiting read direction bias (variants occurring significantly more frequently in one read direction than in the other; FDR=0.0001). These were filtered out from the variant calls.\\
- SNVs present at VAFs smaller than 0.1 or at loci covered by fewer than 10 reads were removed, unless they were also present and confirmed somatic in the Catalogue of Somatic Mutations in Cancer (COSMIC). The presence of well-known PIK3CA mutations present at low VAFs was confirmed by digital PCR (see below), and supported the use of COSMIC when filtering SNVs.\\
- We removed all SNVs that were present in any of the three populations (AMR, ASN, AFR) in the 1,000 Genomes study (Phase 1, release 3) with a population alternate allele frequency of >1\%. \\\\
Microarray data (Copy number):\\\\
- We used the normal samples in our data set (normal pool) to control for both sequencing noise and germline variants, and removed any SNV observed in the normal pool (at a VAF of at least 0.1). However, for SNVs present in more than two breast cancer samples in COSMIC, we used more stringent thresholds, removing only those that were observed in >5\% of normal breast tissue or in >1\% of blood samples. The different thresholds were used to avoid the possibility of contamination in the normal pool affecting filtering of known somatic mutations. This is analogous to the optional 'panel of normals' filtering step used by MuTect in paired mode, in which mutations present in normal samples are removed unless present in a list of known mutations61.\\
Indel filtering\\
- As for SNVs, we removed all indels falling within tandem repeats of MUC16 (coordinates given above).\\
- We removed all indels deemed to be of 'LowQual' by the Haplotype Caller with default parameters (Phred-scaled confidence threshold=30).\\
- As for SNVs, we removed indels displaying read direction bias. Indels with strand bias Phred-scaled scores >40 were removed.\\
- We downloaded the Simple Repeats and Microsatellites tracks from the UCSC Table Browser, and removed all indels overlapping these regions. We also removed all indels that overlapped homopolymer stretches of six or more bases.\\
- As for SNVs, indels were removed if present in the 1,000 Genomes database at an allele frequency >1\%, or if they were present in normal samples in our data set. Thresholds were adjusted as for SNVs if the indel was present in COSMIC. The same thresholds for depth and VAF were used.\\\\
DNA was hybridized to Affymetrix SNP 6.0 arrays per the manufacturer's instructions. ASCAT was used to obtain segmented copy number calls and estimates of tumour ploidy and purity. Somatic CNAs were obtained by removing germline CNVs as defined in the original METABRIC study3. We defined regions of LOH as those in which there were no copies present of either the major or minor allele, irrespective of total copy number. Recurrent CNAs were identified with GISTIC2, with log2 ratios obtained by dividing the total number of copies by tumour ploidy for each ASCAT segment. Thresholds for identifying gains and losses were set to 0.4 and (-)0.5, respectively; these values were obtained by examining the distribution of log2 ratios to identify peaks associated with copy number states. A broad length cut-off of 0.98 was used, and peaks were assessed to rule out probe artefacts and CNVs that may have been originally missed.\\\\
\textbf{\href{https://cancer.columbia.edu/}{Herbert Irving Comprehensive Cancer Center, Columbia University} (COLU)}\\\\
Columbia University Irving Medical Center usesthe Illumina TruSeq Amplicon –Cancer Panel (TSACP) to detect known cancer hotspots.DNA is extracted from unstained sections of FFPE tissue paired with an H\&E stained section that is used to ensure adequate tumor cellularity (human assessment > 30\%) and marking of the tumor region of interest (macrodissection). Extraction for FFPE tissue is performed on the QiaCube instrument (Qiagen). 50-250ng of genomic DNA is used as input.Tumors are sequenced to an average depth of at least 1000X.  Alignment (to hg19) and variant calling is performed using NextGENe v2.4.2software.  Variants lower than 1\% allele frequency in all three control populations (White, African American, Asian) of the Exome Variant Server database, the 1000 genome project database are retained, and annotation of variants is performed using a custom pipeline.  All cases are reviewed and interpreted by a molecular pathologist.\\\\
\textbf{\href{https://www.dana-farber.org/}{Dana-Farber Cancer Institute} (DFCI)}\\\\
DFCI uses a custom, hybridization-based capture panel (OncoPanel) to detect single nucleotide variants,   small   indels,   copy   number   alterations,   and   structural   variants   from   tumor-only sequencing  data.  Three  (3)  versions  of  the  panel  have  been  submitted  to  GENIE:version  1 containing  275  genes,  version  2  containing  300  genes,  version  3  containing  447  genes.  Specimens are reviewed by a pathologist to ensure tumor cellularity of at least 20\%. Tumors are sequenced to an average unique depth of coverage of approximately 200x for version 1 and 350x for version 2. Reads are aligned using BWA, flagged for duplicate read pairs using Picard Tools, and locally realigned using GATK.  Sequence mutations are called using MuTect for SNVs and GATK SomaticIndelDetector for small indels. Putative germline variants are filtered out using a panel of historical normals or if present in ESP at a frequency $\ge$ .1\%, unless the variant is also present in COSMIC. Copy number alterations are called using a custom pipeline and reported for fold-change >1. Structural rearrangements are called using BreaKmer. Testing is performed for all patients across all solid tumor types.  Version 3 includes the exonicregions of 447 genes and 191 intronic regions across 60 genes targeted for rearrangement detection. 52 genes present in previous versions were retired in the v3 test.\\\\
\textbf{\href{http://www.dukecancerinstitute.org/}{Duke Cancer Institute} (DUKE)}\\\\
Foundation medicine panels: Duke uses Illumina hybridization-based capture panels from Foundation Medicine to detect single nucleotide variants, small indels, copy number alterations and structural variants from FFPE, tumor-only sequencing data. Three gene panels were used: Panel 1 (T5a bait set), covering 326 genes, Panel 2 (T7 bait set), covering 434 genes, and Panel 3 (DX1 bait set), covering 324 genes. The clinical sequencing data were analyzed by Foundation Medicine-developed pipelines.  Briefly:  A pool of 5'-biotinylated DNA 120bp oligonucleotides were designed as baits with 60bp overlap in targeted exon regions and 20bp overlap in targeted introns with a minimum of 3 baits per target and 1 bait per SNP target. The goal was a depth of sequencing between 750x and 1000x. Mapping to the reference genome was accomplished using BWA, local alignment optimizations with GATK, and PCR duplicate read removal and sequence metric collection with Picard and Samtools. A Bayesian methodology incorporating tissue-specific prior expectations allowed for detection of novel somatic mutations at low MAF and increased sensitivity at hotspots. Final single nucleotide variant (SNV) calls were made at MAF$\ge$ 5\% (MAF$\ge$ 1\% at hotspots) with filtering for strand bias, read location bias and presence of two or more normal controls. Indels were detected using the deBrujn approach of de novo local assembly within each targeted exon and through direct read alignment and then filtered as described for SNVs. Copy number alterations were detected utilizing a comparative genomic hybridization-like method to obtain a log-ratio profile of the sample to estimate tumor purity and copy number. Absolute copy number was assigned to segments based on Gibbs sampling. To detect gene fusions, chimeric read pairs were clustered by genomic coordinates and clusters containing at least 10 chimeric pairs were identified as rearrangement candidates.\\\\
\textbf{\href{https://www.gustaveroussy.fr/en}{Institut Gustave Roussy} (GRCC)}\\\\
Gustave  Roussy  Cancer  Centre  submitted  data  includes  somatic  variants  (single  nucleotide variants and small indels) identified with CancerHotspot Panel v2 from tumor-only sequence data. Several  versions of  the  panel  have  been  used:  CHP2  covering  hotspots  in  50  genes,  MOSC3 covering  hotspots  in  74  genes  and  MOSC4  covering  89  genes.  Tumors  are  sequenced  to  an average unique depth of coverage of >500X. The sequencing data were analyzed with the Torrent SuiteTMVariant  Caller  4.2  and  higher  and  reported  somatic  variants  were  compared  with  the reference  genome  GRCh37  (hg19). The  variants  were  called  if  >5 reads  supported  the  variant and/or total base depth >50 and/or variant allele frequency >1\% was observed. All the variants identified  were  visually  controlled  on  .bam  files  using  Alamut  v2.4.2  software  (Interactive Biosoftware).  All  the  germline  variants  found  in  1000  Genomes  Project  or  ESP  (Exome Sequencing Project database) with frequency >0.1\% were removed. All somatic mutations were annotated,  sorted,  and  interpreted  by  an  expert  molecular  biologist  according  to  available databases (COSMIC, TCGA) and medical literature.\\\\
The submitted data set was obtained from selected patients that were included in the MOSCATO trial  (Molecular Screening for CAncer Treatment  Optimization)  (NCT01566019).  This  trial collected  on-purpose  tumour  samples  (from  the  primary  or  from  a  metastatic  site)  that  are immediately fresh-frozen,  and  subsequently  analyzed  for  targeted  gene  panel  sequencing.  Tumour cellularity was assessed by a senior pathologist on a haematoxylin and eosin slide from the same biopsy core to ensure tumor cellularity of at least 10\%.\\\\
\textbf{\href{https://www.mdanderson.org/}{The University of Texas MD Anderson Cancer Center} (MDA)}\\\\
The University of Texas MD Anderson Cancer Centersubmitted data in the current data set includes sequence variants (small indels and point mutations) identified using an amplicon-based targeted hotspot tumor-only assay, and sequence variants/gene level amplifications identified on anamplicon-based exonic gene panel which incorporates germline variant subtraction (MDA-409).  Two different amplicon pools and pipeline versions are included for the hotspot tumor-only assays: a 46-gene assay (MDA-46) corresponding to customized version of AmpliSeq Cancer Hotspot Panel, v1 (Life Technologies), and a 50-gene assay (MDA-50) corresponding to the AmpliSeq Hotspot Panel v2.  The exonic assay with germline variant subtraction and amplification detection corresponds to the AmpliSeq Comprehensive Cancer Panel.  DNA wasextracted from unstained sections of tissue paired with astained section that was used to ensure adequate tumor cellularity (human assessment > 20\%) and marking of the tumor region of interest (macrodissection). Sequencing was performed on an Ion Torrent PGM (hotspot) or Proton (exonic). Tumors were sequenced to a minimum depth of coverage (per amplicon) of approximately 250X. Bioinformatics pipeline for MDA-46 was executed using TorrentSuite 2.0.1 signal processing, basecalling, alignment and variant calling.  For MDA-50, TorrentSuite 3.6 was used. Initial calls were made by Torrent Variant Caller (TVC) using low-stringency somatic parameters. For MDA-50, TorrentSuite 3.6 was used.  For MDA-409, TorrentSuite 4.4 was used. For MDA-409, TorrentSuite 4.4 was used.  Initial calls were made by Torrent Variant Caller (TVC) using low-stringency somatic parameters. All called variants were parsed into a custom annotation \& reporting system, OncoSeek, with a back-end SQL Server database using a convergent data model for all sequencing platforms used by the laboratory.  Calls were reviewed with initial low stringency to help ensure that low effective tumor cellularity samples do not get reported as false negative samples.  Nominal variant filters (5\%  variantallelic  frequency  minimum,  25 variant coverage minimum, variant not present in paired germline DNA for the exonic assay) can then be applied dynamically.  Clinical sequencing reports were generated using OncoSeek to transform genomic representations into HGVS nomenclature. To create VCF files for this project,  unfiltered low stringency VCF files were computationally cross checked against a regular expressions-based variant extract from clinical reports. Only cases where all extracted variants from the clinical report were deterministically mappable to the unfiltered VCF file  and corresponding genomic coordinates were marked for inclusion in this dataset.  This method filters a small number of cases where complex indels may not have originally been called correctly at the VCF level.  Testing is performed for patients with advanced metastatic cancer across all solid tumor types.\\\\
\textbf{\href{https://www.mskcc.org/}{Memorial Sloan Kettering Cancer Center} (MSK)}\\\\
MSK uses a custom, hybridization-based capture panel (MSK-IMPACT) to detect single nucleotide variants, small indels, copy number alterations, and structural variants from matched tumor-normal sequence data.  Three (3) versions of the panel have been submitted to GENIE: version 1 containing 341 genes, version 2 containing 410 genes, version 3 containing 468 genes.  Specimens are reviewed by a pathologist to ensure tumor cellularity of at least 10\%. Tumors are sequenced to an average unique depth of coverage of approximately 750X.  Reads are aligned using BWA, flagged for duplicate read pairs using GATK, and locally realigned using ABRA.  Sequence mutations are called using MuTect, VarDict, and Somatic indel detector, and reported for >5\% allele frequency (novel variants) or >2\% allele frequency (recurrent hotspots). Copy number alterations are called using a custom pipeline and reported for fold-change >2. Structural rearrangements are called using Delly.  All somatic mutations are reported without regard to biological function.  Testing is performed for patients with advanced metastatic cancer across all solid tumor types.\\\\
\textbf{\href{https://www.jhu.edu/}{Johns Hopkins Sidney Kimmel Comprehensive Cancer Center} (JHU)}\\
Johns Hopkins submitted genomic data from the Ion AmpliSeqCancer Hotspot Panel v2, which detects mutations in cancer hotspots from tumor-only analysis.  Data from the  JHU\_50GP\_V2 panel covering frequently mutated regions in 50 genes was submitted to GENIE. Pathologist inspection of an H\&E section ensured adequatetumor cellularity (approximately 10\% or greater). DNA was extracted from the macro-dissected FFPE tumor region of interest.  Tumors are sequenced to an average unique read depth of coverage of greater than 500X. For alignment the TMAP aligner developed by Life Technology for the Ion Torrent sequencing platform is used to align to hg19/GRCh37 using the manufacturer’s suggested settings. Tumor variants are called with a variety of tools. Samtools mpileup is run on the aligned .bam file and then processed withcustom perl  scripts (via a naive variant caller) to identify SNV and INS/DEL.  Specimen variant filters have a total read depth filter of $\ge$ 100, a variant allele coverage of $\ge$ 10, variant allele frequency for substitutions $\ge$ 0.05, variant allele  frequency for small (less than 50 base pair) insertions or deletions $\ge$ 0.05, and "strand bias" of total reads and of variant alleles are both less than 2-fold when comparing forward and reverse reads. Additionally, variants seen in greater than 20\% of a set of non-neoplastic control tissues (>3 of 16 samples) with the same filter criteria are excluded. Finally, variants documented as “common” in dbSNP and not known to COSMIC are excluded. The cohort includes both primary and metastatic lesions and some repeated sampling of the same patient.\\\\
\textbf{\href{https://www.nki.nl/}{Netherlands Cancer Center, The Netherlands} (NKI)}\\\\
NKI uses Illumina TruSeq Amplicon –Cancer Panel (TSACP) to detect known cancer hotspots from tumor-only sequencing data. A single gene panel, NKI-TSACP covering known hotspots in 48 genes with 212 amplicons has been used. Specimens are reviewed by a pathologist to ensure tumor cellularity of at least 10\%. Tumors are sequenced to an average unique depth of coverage of approximately  4000x. The sample plate and sample sheet are made using the Illumina Experiment Manager software before running the sample on the MiSeq Sequencing System (Illumina, SY-410-1003) and MiSeq Reporter (v2.5) is used for data analysis. Reads are aligned using Banded Smith Waterman (v2.5.1.3), and samtools isused to further sort and index the BAM files. Variant calling is performed via the Illumina somatic variant caller (v3.5.2.1). Further detailed variant analysis (e.g. removal of known artifacts, known benign SNPs and variants with read depth < 200 or VAF < 0.05 and manual classification) is performed via Cartagenia BenchLab (https://cartagenia.com/). Testing is performed for all patients across all solid tumor types.\\\\
\textbf{\href{https://oregon.providence.org/our-services/p/providence-cancer-center/}{Providence Health \& Services Cancer Institute} (PHS)}\\\\
PHS has submitted data from two assays for GENIE 6.1: PHS-focus-v1 and PHS-triseq-v1. For the PHS-focus-v1 we have employed the Thermo Fisher Oncomine Focus Assay for amplification of 52 genes from DNA extracted from macro-dissected FFPE samples taken from Pathologist specified tumor regionsof interest (ROIs), with 20\% minimal tumor cellularity. Samples include primary and metastatic tumor ROIs. The assay is a tumor only assay, no paired "normal" DNA is extracted from each case.\\\\
The PHS-focus-v1 BED file describes the positions of the genome assayed by the PHS-focus-v1 panel relative to hg19.\\\\
Amplification products are sequenced on the Life Technology Ion Torrent platform to an average read depth of coverage greater than 500X average per base coverage.\\\\
The TMAP aligner developed by Life Technology for the ION torrent sequencing platform was used to align reads to hg19 using the manufacture suggested settings. Variants are called with the Torrent Suite Variant Caller 4.2 software plug-in.\\\\
Variant filters requiring a total read depth of greater than 100X, variant allele coverage of greater than 10X, and a variant allele frequency for substitutions of greater than or equal to 0.03 are applied. Also, the specimen variant must not be annotated as "COMMON" (a variant allele frequency for substitutions of $\ge$ 0.05) in dnSNP. VCF files were created for upload to GENIE 6.1 by further filtering all detected variants to only those reported after expert review by clinicians.\\\\
For the PHS-triseq-v1 we used DNA extracted from macro-dissected FFPE samples taken from Pathologist specified tumor regions of interest (ROIs), with 20\% minimal tumor cellularity for extraction of tumor DNA, and whole peripheral blood for extraction of normal DNA. Tumor samples include both primary and metastatic tumor ROIs.\\\\
The PHS-triseq-v1 BED file describes the positions of the genome assayed by the PHS-triseq-v1 panel relative to hg19.\\\\
Libraries are prepared using the KAPA for Illumina reagents protocols. Indexed libraries are pooled for exome capture on the xGen V1.0 panel (https://www.idtdna.com/). Sequencing is performed on Illumina 2500, 4000, or Novaseq platforms.\\\\
Raw sequencing data in the form of BCL files are uploaded to the Providence secure computing cloud environment maintained by Amazon Web Services. Following upload, raw files are converted to unaligned reads in FASTQ format using the software program bcl2fastq2, and resultant FASTQ files are aligned to the hg19 human reference genome using the Burrows-Wheeler Aligner (BWA). Aligned reads in the SAM format are subsequently converted to binary BAM format using the samtools software package, and aligned reads are processed for single-nucleotide variants (SNVs) and short insertions and deletions (indels) using our custom variant calling pipeline (see below). FASTQ and aligned BAM files are analyzed with FastQC and Picard metrics for Molecular Genomics Lab staff run-level and sample-level review.\\\\
The Providence variant calling pipeline includes multiple variant calling algorithms including VarScan2, SomaticSniper, Mutect2 and Strelka. Variant filters requiring a total read depth of greater than 100X, variant allele coverage of greater than 10X, and a variant allele frequency for substitutions of greater than or equal to 0.03 are applied. Calls with low-quality variants, silent mutations, and germline variants are also filtered. Annotations from SnpEff, ClinVar, ExAC, 1000 Genomes, ANNOVAR, and COSMIC are incorporated for each call. Finally, all common variants, with non-zero allele frequencies the ExAC or 1000 Genomes databases, are removed.\\\\
GENIE 6.1 VCF files containing annotated calls from Mutect2 were created for upload to GENIE 6.1.\\\\
\textbf{\href{https://www.swedish.org/services/cancer-institute}{Swedish Cancer Institute} (SCI)}\\\\
SCI uses CellNetixPMP gene panel to detect hotspot mutations in known cancer genes from solid tumor DNA (Formalin-fixed, paraffin-embedded tissue). The hotspot gene panel covers 68 genes. Tumor cell content is greater than 10\% verified by pathologist. Tumor DNA is sequenced to >200x on average (Variant that allele frequency is less than 10\% requires more than 400X) on Illumina MiSeq (TruSeq Amplicon) platform, and data is analyzed in MiSeq Reporter 2.5. Reads are aligned to hg19 reference genome by the BWA (v0.6.1-r104-tpx)aligner adapted by the MiSeq Reporter Software (v2.4.1 or v2.5) using the manufacture suggested settings. MiSeq Reporter provided Somatic Variant Caller (v2.1.12) is run on the aligned .bam files to identify variants present in DNA samples. Detailed stepsplease refer toIllumina MiSeq Reporter User Guide. Variants are filtered for allele frequency greater than 3\% except for actionable mutations. Variants that are observed in $\ge$ 75\% samples on the same run, or common variant with population frequency of > 50\%, or average population frequency >5\% reported in the 1000 genome and/or in ExAc. are filtered.\\\\
\textbf{\href{https://www.uhn.ca/PrincessMargaret}{Princess Margaret Cancer Centre, University Health Network} (UHN)}\\\\
Princess Margaret Cancer Centre used four (4) panels to sequence samples for the GENIE 6.1-publicrelease -UHN-48-V1, UHN-50-V2, UHN-2-V,UHN-54-V1, and UHN-555-V1.  Each panel is described below:\\\\
Illumina TruSeq Amplicon panel (UHN-48-V1):  Princess Margaret Cancer Centre used the TruSeq Amplicon Cancer Panel (TSACP, Illumina) to detect single nucleotide variants and small indels from matched tumor-normal sequencing data. Specimens are reviewed by a pathologist to ensure tumor cellularity of at least 20\%. Tumors are sequenced to an average unique depth of coverage of approximately 500x and normal blood samples to 100x. Data was processed using one of two workflows:\\\\
1. Data analysis of tumor-normal pairs processed by UHN\_TSACP\_workflow\_v2: MiSeq fastq were aligned using (MiSeq Reporter v2.4.60 and the corresponding default version of hg19) followed by local realignment and BQSR using GATK v3.3.0.  Somatic sequence mutations were called, using MuTect (v1.1.5) for SNVs and Varscan (v2.3.8) for indels, using both normal and tumor data. Data were filtered to ensure there are no variants included with frequency of 3\% or more in the normal sample. Results were filtered to keep only those with tumor variant allele frequency of at least 10\%.\\\\
2. Data analysis of tumor only processed by UHN\_TSACP\_tumorONLY\_v2\_workflow: MiSeq fastq were aligned using (MiSeqReporter v2.4.60 and the corresponding default version of hg19) followed by local realignment and BQSR using GATK v3.3.0. Sequence mutations (SNV and indel) were called using Varscan (v2.3.8). Results were filtered to keep only those with tumor variant allele frequency of at least 10\%.\\\\
ThermoFisher Ion AmpliSeq Cancer Panel (UHN-50-V2):  Princess Margaret Cancer Centre also used the TruSeq Amplicon Cancer Panel (TSACP, Illumina) to detect single nucleotide variants and small indels from matched tumor-normal sequencing data. Specimens were reviewed by a pathologist to ensure tumor cellularity of at least 20\%. Tumors were sequenced to an average unique depth of coverage of approximately 500x and normal blood samples to 100x.  Ion Torrent data was converted to fastq and sequences were aligned using NextGENe Software v2.3.1. NextGENe Software v2.3.1 provides a version of hg19 (Human\_v37\_3\_dbsnp\_135\_dna). NextGENe was used to call SNV and indels. Results were then filtered to keep all with VAF of at least 10\% and total coverage of at least 100x.\\\\
Sequenom MassArray Panel (UHN-2-V1): Princess Margaret Cancer Centre also used the Sequenom MassArray Solid Tumor Panel v1.0 assay to detect variants from tumor samples. Mutation calling was determined using the TyperAnalyzer software and results were filtered to include only postions listed in the bed file.\\\\
Illumina TruSeq Myeloid Sequencing Panel (UHN-54-V1):  Princess Margaret Cancer Centre also used the TruSeq Myeloid Sequencing Panel (Illumina) to detect single nucleotide variants and small indels in DNA from bone marrow or peripheral blood samples from patients with acute leukemia, myelodysplastic syndrome, or myeloproliferative neoplasms. The diagnosis of each patient was confirmed by hematopathologist using the 2016 revision of the World Health Organization classification system for myeloid neoplasms.  Tumors were sequenced to an average unique depth of coverage of approximately 500x.  MiSeq fastq were aligned using (MiSeq Reporter v2.4.60 and the corresponding default version of hg19). MiSeq Reporter was then used to call variants. In the "Illumina Experiment Manager", "TruSeq Amplicon Workflow –specific settings" were adjusted as follows: “Export to gVCF –MaxIndelSize” from default “25” to “55”. Results were then filtered to keep only those with tumor variant allele frequency of at least 10\% and a depth of coverage greater than 500x.\\\\
Hybrid Capture –Sure Select Custom Panel (Agilent) -UHN-555-V1: Princess Margaret Cancer Centre also used the Sure Select Custom Panel (Agilent) to detect single nucleotide variants and small indels in DNA from tumor (FFPE) for patients with advanced disease from select tumor types. FFPE tissue specimens were reviewed by a pathologist to ensure tumor cellularity of at least 10\%. Tumors were sequenced to an average depth of coverage of approximately 500x FFPE. Data analysis tumour only processed by UHN\_555\_v1\_workflow: Reads are aligned to hg19 using BWA mem version 0.7.12 followed by local realignment and BQSR using GATK v3.3.0 followed by local realignment and BQSR using GATK v3.3.0. Sequence mutations (SNV and indel) are called using Varscan (v2.3.8). SNV results are filtered to keep only those with tumor variant allele frequency of at least 10\%.\\\\
\textbf{\href{https://www.vhio.net/en/}{Vall d'Hebron Institute of Oncology} (VHIO)}\\\\
Vall d’Hebron institute of Oncology (VHIO) submitted data that includes somatic variants (single nucleotide variants and small indels) identified with VHIO Card  Amplicon panels that target frequently mutated regions in oncogenes and tumor suppressors. A total of fifteen panels have been submitted taking different tumor types into consideration. The panels are:\\\\
1-VHIO\_GENERAL\_v01: Panel containing 56 oncogenes andtumor suppressor genes;\\
2-VHIO\_BRAIN\_v01 (General + NF1 v1: 57 genes);\\
3-VHIO\_BILIARY\_v01 (General + *FGFR v1 + **NOTCH v1: 59 genes);\\
4-VHIO\_COLORECTAL\_v01 (General + RingFingers v1 + **NOTCH v1: 60 genes);\\
5-VHIO\_HEAD\_NECK\_v1 (General + MTOR v1 + **NOTCH v1: 61 genes);\\
6-VHIO\_ENDOMETRIUM\_v01 (General + RingFingers v1 + *FGFR v1 + NF1 v1: 60 genes);\\
7-VHIO\_GASTRIC\_v01 (General + RingFingers v1 + MTOR v1 + **NOTCH v1: 63 genes);\\
8-VHIO\_PAROTIDE\_v01 (General + **NOTCH v1: 58 genes);\\
9-VHIO\_BREAST\_v01 (General + *FGFR v1 + **NOTCH v1+ GATA3 v1: 60 genes);\\
10-VHIO\_OVARY\_v01 (General + BRCA v1: 58 genes);\\
11-VHIO\_PANCREAS\_v01 (General + Ring Fingers v1 + BRCA v1: 60 genes);\\
12-VHIO\_SKIN\_v01 (General + NF1 v1 + MTOR v1: 60 genes);\\
13-VHIO\_LUNG\_v01 (General + NF1 v1 + MET v1 + FGFRw7 v1: 58);\\
14-VHIO\_KIDNEY\_v01 (General + MTOR v1: 59 genes);\\
15-VHIO\_URINARY\_BLADDER\_v01 (General + *FGFR v1 + NF1 v1 + MTOR v1: 61 genes).\\\\
*FGFRv1 panel includes extra regions in FGFR1, FGFR2 and FGFR3 genes. **NOTCHv1 panel includes extra regions in FBXW7 and NOTCH1 genes. FGFRw7 v1 panel includes extra regions in FGFR1 gene. ¥MET v1 panel includes intronic regions flanking Exon 14 of MET gene.\\\\
Tumor samples are reviewed by a pathologist to ensure tumor cellularity of at least 20\%. For the sample loading into tumor-specific panels, we use a FREEDOM EVO 150 Platform from TECAN. Tumors  are  sequenced  in  an  Illumina  MiSeq  instrument,  to  an  average  depth  of  coverage  of approximately 1000X. Samples are sequenced, and two independent chemistries are performed and sequenced. Sequencing reads are aligned (BWA v0.7.17, Samtools v1.9), base recalibrated, Indel  realigned  (GATK  v3.7.0),  and  variant  called  (VarScan2  v2.4.3).  A  minimum  of  7  reads supporting  the  variant  allele  is  required  in  order  to  call  a  mutation.  Frequent  SNPs  in  the population  are  filtered  with  the  1000g  database  (MAF>0.005).  The  average  number  of  reads representing a given nucleotide in the panel (Sample Average Coverage) is calculated. Manual curation  of  variants  is  performed  after  manual  search  of  available  literature  and  databases,  in terms of their clinical significance.\\\\
\textbf{\href{https://www.vicc.org/}{Vanderbilt-Ingram Cancer Center} (VICC)}\\\\
Foundation medicine panels: VICC uses Illumina hybridization-based capture panels  from Foundation Medicine to detect single nucleotide variants, small indels, copy number alterations and structural variants from tumor-only sequencing data. Two gene panels were used: Panel 1 (T5a bait set), covering 326 genes and; and Panel 2 (T7 bait set), covering 434 genes. DNA was extracted from unstained FFPE sections, and H\&E stained sections were used to ensure nucleated cellularity $\ge$ 80\% and tumor cellularity $\ge$ 20\%, with use of macro-dissection to enrich samples with $\le$ 20\% tumor content. A pool of 5'-biotinylated DNA 120bp oligonucleotides were designed as baits with 60bp overlap in targeted exon regions and 20bp overlap in targeted introns with a minimum of 3 baits per target and 1 bait per SNP target.  The goal was a depth of sequencing between 750x and 1000x. Mapping to the reference genome was accomplished using BWA, local alignment optimizations with GATK, and PCR duplicate read removal and sequence metric collection with Picard and Samtools. A Bayesian methodology incorporating tissue-specific prior expectations allowed for detection of novel somatic mutations at low MAF and increased sensitivity at hotspots. Final single nucleotide variant (SNV) calls were made at MAF $\ge$ 5\% (MAF $\ge$ 1\%  at hotspots) with filtering for strand bias, read location bias and presence of two or more normal controls.  Indels were detected using the deBrujn approach of de novo local assembly within each targeted exon and through direct read alignment and then filtered as described for SNVs. Copy number alterations were detected utilizing a comparative genomic hybridization-like method to obtain a log-ratio profile of the sample to estimate tumor purity and copy number. Absolute copy number was assigned to segments based on Gibbs sampling.  To detect gene fusions, chimeric read pairs were clustered  by genomic coordinates and clusters containing at least 10 chimeric pairs were identified as rearrangement candidates. Rare tumors and metastatic samples  were  prioritized for sequencing, but ultimately sequencing was at the clinician's discretion.\\\\
VICC also submitted data from 2 smaller hotspot amplicon panels, one used for all myeloid (VICC-01-myeloid) tumors and 1 used for some solid tumors (VICC-01-solidtumor). These panels detect point mutations and small indels from 37 and 31 genes, respectively. Solid tumor H\&E were inspected to ensure adequate tumor cellularity (>10\%). Sections were macrodissected if necessary, and DNA was extracted. Tumors were sequenced to an average depth greater than 1000X.  Reads were aligned to hg19/GRCh37  with novoalign, and single nucleotide variants, insertions and deletions greater than 5\% were called utilizing a customized bioinformatic pipeline. Large (15bp and greater) FLT3 insertions were called using a specialized protocol and were detected to a 0.5\% allelic burden.\\\\
\textbf{\href{https://www.wakehealth.edu/}{Wake Forest University Health Sciences, Wake Forest Baptist Medical Center} (WAKE)}\\\\
We utilized thesequencing analysis pipelines from Foundation Medicine and Caristo analyze clinical samples and support.Enrichment of target sequences was achieved by solution-based hybrid capture with custom biotinylated oligonucleotide bases. Enriched libraries were sequenced to an average median depth of >500× with 99\% of bases covered >100× (IlluminaHiSeq2000 platform using 49 × 49 paired-end reads).The clinical sequencing data were analyzed by Foundation Medicine and Carisdevelopedpipelines.Sequenced readswere mapped to the reference human genome (hg19) using the Burrows-Wheeler Aligner and the publicly available SAM tools, Picard, and Genome Analysis Toolkit.Point mutations were identified by a Bayesian algorithm; short insertions and deletions determined by local assembly; gene copy numberalterations identified by comparison to process-matched normal controls; and gene fusions/rearrangements determined by clustering chimeric reads mapped to targeted introns.Following by computational analysis with tools such asMutSigand CHASM, the driver mutations can be identified which may help the selection of treatment strategy. In addition, the initial report of the analysis of 470 cases has been published and highlightedon the cover of the journalTheranosticsin 2017.\\\\
\textbf{\href{https://www.yale.edu/}{Yale University, Yale Cancer Center} (YALE)}\\\\
GENIE samples submitted by Yale belong to one of three targeted NGS panels (1) YALE-HSM-V1, (2) YALE-OCP-V2, or (3) YALE-OCP-V3.  The first panel corresponds to the Thermofisher Ion AmpliSeq Cancer Hotspot Panel v2, which is designed to assess hotspot variants in 50 of the most frequently mutated genes in cancer, and is performed as a tumor-only analysis.  The latter two panels refer to v2 and v3C of the Thermofisher Oncomine Comprehensive Assay, which provides a more comprehensive assessment of somatic alterations including single nucleotide variants, insertions, deletions, copy number alterations (CNAs), and gene fusions across 143 and 161 genes, respectively.  Target region design (i.e. full exonic, hotspot only, intronic, promoter) varies based on known relevance of each gene.  Pathologist inspection of an H\&E section ensured adequate tumor cellularity (approximately 10\% or greater).  Tumor samples are enriched for malignant cells by manual microdissection of unstained formalin-fixed, paraffin-embedded (FFPE) tissue sections.  If available, germline control DNA from the same patient is obtained either from FFPE non-tumor tissue, from the patient’s blood, or from a buccal swab.  Subsequent libraries are barcoded and sequenced on either an Ion Torrent PGM™ or an Ion S5™ XL next generation sequencer.\\\\
Pre-processing and alignment of reads is performed within Torrent Suite, with TMAP serving as the alignment algorithm.  Resulting BAM files are uploaded to the Ion Reporter software for variant detection, as well as CNA and gene fusion assessment for Oncomine samples.  The bioinformatics pipeline also uses MuTect2 (GATK) and Strelka (Illumina) to assess somatic variants.  Variants are initially filtered based on quality metrics; a minimum read depth of 20x and avariant allelic fraction (VAF) of 0.02 is required.  All variants passing quality filters are passed through the Ensembl Variant Effect Predictor for variant annotation.  Variants that are intronic or synonymous are filtered at this stage; all other variants are manually reviewed for accuracy before submission to the attending pathologist.  Variants below a VAF below 0.05 are not typically reviewed unless tumor cellularity estimates are low.  CNA assessment is performed using the IonReporter CNV algorithm,as well as an internally developed workflow that uses the DNAcopy R package.  Custom visualizations for amplified genes are used to confirm accuracy of CNAs reported by the pipeline.  Only CNAs with a ploidy of 5 or higher are reported.  Gene fusion assessment is handled by a custom workflow in the Ion Reporter software which aligns cDNA reads to known fusion breakpoints.  A fusion read is mapped successfully if there is an overlap of 70\% and exact matches of 66.66\%.



\addcontentsline{toc}{section}{Pipeline for Annotating Mutations and Filtering Putative Germline SNPs}
\section*{Pipeline for Annotating Mutations and Filtering Putative Germline SNPs}

Contributing GENIE centers provided mutation data in \href{http://samtools.github.io/hts-specs/}{Variant Call Format (VCF vcf2maf v1.6.17)} or Mutation Annotation Format (\href{https://docs.gdc.cancer.gov/Data/File\_Formats/MAF\_Format}{GDC MAF v1.0.0}) with additional fields for read counts supporting variant alleles, reference alleles, and total depth. Some “MAF-like” text files with \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/data/minimalist\_test\_maf.tsv}{minimal required columns} were also received from the participating centers. These various input formats were converted into a complete tab-separated MAF format, with a \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/docs/vep\_maf\_readme.txt}{standardized set of additional columns} using either vcf2maf or maf2maf \href{https://github.com/mskcc/vcf2maf/tree/v1.6.17}{v1.6.17} wrappers around the Variant Effect Predictor (\href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/docs/vep\_maf\_readme.txt}{VEPv86}). The vcf2maf “custom-enst” option overrode VEP’s canonical isoform for most genes, with \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/data/isoform\_overrides\_uniprot}{Uniprot’s canonical isoform}.\\\\
While the GENIE data available from Sage contains all mutation data, the following mutation types are automatically filtered upon import into the \href{http://www.cbioportal.org/genie}{cBioPortal}: Silent, Intronic, 3' UTR, 3' Flank, 5' UTR, 5' Flank and Intergenic region (IGR).\\\\
Seventeen of the nineteen GENIE participating centers performed tumor-only sequencing i.e. without also sequencing a patient-matched control sample like blood, to isolate somatic events. These centers minimized artifacts and germline events using pooled controls from unrelated individuals, or using databases of known artifacts, common germline variants, and recurrent somatic mutations. However, there remains a risk that such centers may inadvertently release germline variants that can theoretically be used for patient re-identification. To minimize this risk, the GENIE consortium developed a stringent germline filtering pipeline, and applied it uniformly to all variants across all centers. This pipeline flags sufficiently recurrent artifacts and germline events reported by the Exome Aggregation Consortium (\href{http://exac.broadinstitute.org}{ExAC}). Specifically, the non-TCGA subset VCF of ExAC 0.3.1 was used after excluding known somatic events in this \href{https://github.com/mskcc/vcf2maf/blob/v1.6.17/data/known\_somatic\_sites.bed}{bed file}:\\\\

\begin{itemize}
    \item{Hotspots from Chang et al. minus some likely artifacts.\\(dx.doi.org/10.1038/nbt.3391)}
    \item{Somatic mutations associated with clonal hematopoietic expansion from Xie et al.\\(dx.doi.org/10.1038/nm.3733)}
    \item{Somatically mutable germline sites at MSH6:F1088, TP53:R290, TERT:E280,\\ ASXL1:G645\_G646.}
\end{itemize}

The resulting VCF was used with vcf2maf’s "filter-vcf" option, to match each variant position and allele to per-subpopulation allele counts. If a variant was seen more than 10 times in any of the 7 ExAC subpopulations, it was tagged as a "common\_variant" (vcf2maf’s "max-filter-ac" option), and subsequently removed. This >10 allele count (AC) cutoff was selected because it tagged no more than 1\% of the somatic calls across all MSK-IMPACT samples with patient-matched controls.\\\\


\addcontentsline{toc}{section}{Description of Data Files}
\section*{Description of Data Files}


Description on most of the data files can be found under the "Data files" section for \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats}{cBioPortal file formats}.


\begin{center}
\begin{longtable}{ |p{45mm}|p{40mm}|p{55mm}| }
\caption{GENIE Data Files} \\
  \hline
File Name & Description & Details \\ 
  \hline
data\_mutations\_extended.txt & Mutation data & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#mutation-data}{MAF format} \\
   \hline
data\_CNA.txt & Discritized copy number data.  \newline Note: Not all centers contributed copy number data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#discrete-copy-number-data}{CNA format} \\
   \hline
data\_fusions.txt & Structural variant data. \newline Note: not all centers contributed structural rearrangement data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#fusion-data}{Fusion format} \\
   \hline
genomic\_information.txt &  File describing genomic coordinates covered by all platforms for GENIE data.  \newline Note: Not all centers contributed copy number data to GENIE. & This is not a cBioPortal file format. \newline \textbf{Chromosome}, \textbf{Start\_Position}, \textbf{End\_Position}: Gene positions. \newline \textbf{Hugo\_Symbol}: Re-mapped gene symbol based on gene positions. \textbf{ID}: Center submitted gene symbols. \newline \textbf{SEQ\_ASSAY\_ID}: The institutional assay identifier for genomic testing platform. \newline \textbf{Feature\_Type}: "exon", "intron", or "intergenic". \newline \textbf{includeInPanel}: Used to define gene panel files for cBioPortal. \newline \textbf{clinicalReported}: These are the genes that were clinically Reported. Blank means information not provided.\\
   \hline
assay\_information.txt & Assay information & This is not a cBioPortal file format. \newline \textbf{is\_paried\_end}, \textbf{library\_selection}, , \textbf{library\_strategy}, \textbf{platform}, \textbf{read\_length}, \textbf{target\_capture\_kit}, \textbf{instrument\_model}: defined by \href{https://docs.gdc.cancer.gov/Data\_Dictionary/viewer/#?view=table-definition-view&id=read\_group}{GDC read group} \newline \textbf{number\_of\_genes}: Number of genes from which variants are called \newline \textbf{variant\_classifications}: List of types of variants that are reported for this assay. \newline \textbf{gene\_padding}: Number of basepairs to add to exon endpoints for the inBED filter.  \newline \textbf{alteration\_types}: List of alteration types. \newline \textbf{specimen\_type}: List of specimen types \newline \textbf{specimen\_tumor\_cellularity}: Tumor Cellularity Cutoff \newline \textbf{calling\_strategy}: Yumor only or tumor normal \newline \textbf{coverage}: List of coverage \\
   \hline
genie\_data\_cna\_hg19.seg  & Segmented copy number data.  \newline Note: Not all centers contributed copy number data to GENIE. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#segmented-data}{SEG format} \\
   \hline
data\_clinical.txt & De-identified tier 1 clinical data. & \href{https://docs.cbioportal.org/5.1-data-loading/data-loading/file-formats#clinical-data}{clinical format}. See Clinical Data section below for more details. \\
   \hline
\end{longtable}
\end{center}



\addcontentsline{toc}{section}{Clinical Data}
\section*{Clinical Data}

\begin{center}
\begin{longtable}{ |p{45mm}|p{40mm}|p{50mm}| }
\caption{GENIE Clinical Data Fields} \\
  \hline
Data Element & Example Values & Data Description \\ 
  \hline
AGE\_AT\_SEQ\_REPORT & Integer values, <18 or >89.  &  The age of the patient at the time that the sequencing results were reported. Age is masked for patients aged 90 years and greater and for patients under 18 years. \\
   \hline
CENTER & MSK & The center submitting the clinical and genomic data   \\
   \hline
ETHNICITY & Non-Spanish/non-Hispanic \newline Spanish/Hispanic \newline Unknown & Indication of Spanish/Hispanic origin of the patient; this data element maps to the NAACCR v16, Element \#190. Institutions not collecting Spanish/Hispanic origin have set this column to Unknown. \\
   \hline
ONCOTREE\_CODE & LUAD & The primary cancer diagnosis “main type”, based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01} \\
   \hline
PATIENT\_ID & GENIE-JHU-1234 & The unique, anonymized patient identifier for the GENIE project. Conforms to the following the convention: GENIE-CENTER-1234. The first component is the string, "GENIE"; the second component is the Center abbreviation. The third component is an anonymized unique identifier for the patient.\\
   \hline
PRIMARY\_RACE & Asian, Black, Native American Other, Unknown, White & The primary race recorded for the patient; this data element maps to the NAACCR v16, Element \#160. For institutions collecting more than one race category, this race code is the primary race for the patient. Institutions not collecting race have set this field to Unknown. \\
   \hline
SAMPLE\_ID & GENIE-JHU-1234-9876 & The unique, anonymized sample identifier for the GENIE project. Conforms to the following the convention: GENIE-CENTER-1234-9876. The first component is the string, "GENIE"; the second component is the Center abbreviation. The third component is an anonymized, unique patient identifier. The fourth component is a unique identifier for the sample that will distinguish between two or more specimens from a single patient. \\
   \hline
SAMPLE\_TYPE &  Primary, Metastasis, Unspecified & Sample type, e.g. Primary or Metastasis. \\
   \hline
SEQ\_ASSAY\_ID & DFCI-ONCOPANEL-1 & The institutional assay identifier for genomic testing platform. Components are separated by hyphens, with the first component corresponding to the Center's abbreviation. All specimens tested by the same platform should have the same identifier. \\
   \hline
SEX & Female, Male & The patient’s sex code; this data element maps to the NAACCR v16, Element \#220. \\
   \hline
CANCER\_TYPE & Non-Small Cell Lung Cancer & The primary cancer diagnosis "main type", based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. For example, the OncoTree code of LUAD maps to: "Non-Small Cell Lung Cancer".  The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01} \\
   \hline
CANCER\_TYPE\_DETAILED & Lung Adenocarcinoma & The primary cancer diagnosis label, based on the \href{http://oncotree.mskcc.org}{OncoTree ontology}. For example, the OncoTree code of LUAD maps to the label: "Lung Adenocarcinoma (LUAD)". The version of Oncotree ontology that was used for GENIE 8.0 series is \href{http://oncotree.mskcc.org/#/home?vers
ion=oncotree_2018_06_01}{oncotree 2018\_06\_01}  \\
   \hline
BIRTH\_YEAR & 1965 & The four-digit year associated with the patient's date of birth.  \\
   \hline
YEAR\_CONTACT & 1990  & Record the year that the patient is last known to be alive - data to be derived from information including and EHR, tumor registry, or other system. \\
   \hline
YEAR\_DEATH & 2010 & Year of death  \\
   \hline
INT\_CONTACT & 15000  &  Interval in days from DOB to date of last contact. \\
   \hline
INT\_DOD & 15000 & Interval in days from DOB to DOD.  \\
   \hline
DEAD & Yes, No  & Is this patient known to be deceased? \\
   \hline
\end{longtable}
\end{center}

Cancer types are reported using the \href{http://oncotree.mskcc.org}{OncoTree ontology} originally developed at Memorial Sloan Kettering Cancer Center. The GENIE 8.0 data release series uses the OncoTree specification from June 1, 2018, containing diagnosis codes for 686 tumor types from 32 tissues. The centers participating in GENIE applied the OncoTree cancer types to the tested specimens in a variety of methods depending on center-specific workflows. A brief description of how the cancer type assignment process for each center is specified in Table 6.



\begin{center}
\begin{longtable}{ |p{10mm}|p{100mm}| }
\caption{Center Strategies for OncoTree Assignment} \\
  \hline
Center & OncoTree Cancer Type Assignments \\ 
  \hline
COLU &  Original diagnosis from pathologist was mapped to OncoTree diagnosis by medical oncologist and research manager.\\
   \hline
CRUK & Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
DFCI &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
DUKE & Anatomic and molecular pathologists assigned diagnosis and mapped to OncoTree cancer type.  \\
   \hline
GRCC &  OncoTree cancer types were mapped from ICD-O codes. If no ICD-O code was available, a staff scientist and an oncologist mapped the diagnosis made by the pathologist to Onco Tree cancer type. \\
   \hline
JHU &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
MDA &  OncoTree cancer types were mapped from ICD-O codes. \\
   \hline
MSK &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
NKI &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
PHS &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
SCI &  Original diagnosis from the pathology report was mapped to OncoTree diagnosis by a research coordinator and molecular pathologist. \\
   \hline
UHN & The original diagnosis was mapped to OncoTree by a medical oncologist and research manager.  \\
   \hline
VHIO & Original diagnosis from pathologist or medical oncologist was mapped to OncoTree diagnosis by research data curator.  \\
   \hline
VICC & OncoTree cancer types were mapped from ICD-O codes. If no ICD-O code was available, a research manager mapped the diagnosis to an OncoTree cancer type.  \\
   \hline
WAKE &  Diagnoses from Foundation Medicine and Caris Diagnostics to ICD-O-3, then mapped from ICD-O-3 to Oncotree. \\
   \hline
YALE &  Molecular pathologists assigned diagnosis and mapped to OncoTree cancer type. \\
   \hline
\end{longtable}
\end{center}



\addcontentsline{toc}{section}{Abbreviations and Acronym Glossary}
\section*{Abbreviations and Acronym Glossary}


\begin{center}
\begin{longtable}{ |p{20mm}|p{100mm}| }
  \hline
Abbreviation & Full Term \\
  \hline
AACR & American Association for Cancer Research, Philadelphia, PA, USA \\
   \hline
CNA & Copy number alterations \\
   \hline
CNV & Copy number variants \\
   \hline
COLU & The Herbert Irving Comprehensive Cancer Center, Columbia University, New York, NY, USA \\
   \hline
CRUK & Cancer Research UK Cambridge Centre, University of Cambridge, Cambridge, England \\
   \hline
DFCI & Dana-Farber Cancer Institute, Boston, MA, USA \\
   \hline
DUKE & Duke Cancer Institute, Duke University Health System, Durham, NC, USA \\
   \hline
FFPE & Formalin-fixed, paraffin-embedded \\
   \hline
GENIE & Genomics, Evidence, Neoplasia, Information, Exchange \\
   \hline
GRCC & Institut Gustave Roussy, Paris, France \\
   \hline
HIPAA & Health Insurance Portability and Accountability Act \\
   \hline
IRB & Institutional Review Board \\
   \hline
JHU & Johns Hopkins Sidney Kimmel Comprehensive Cancer Center, Baltimore, MD, USA \\
   \hline
MAF & Mutation annotation format \\
   \hline
MDA & The University of Texas MD Anderson Cancer Center, Houston, TX, USA \\
   \hline
MSK & Memorial Sloan Kettering Cancer Center, New York, NY, USA \\
   \hline
NAACCR & North American Association of Central Cancer Registries \\
   \hline
NGS & Next-generation sequencing \\
   \hline
NKI & Netherlands Cancer Institute, Amsterdam, Netherlands \\
   \hline
PCR & Polymerase chain reaction \\
   \hline
PHS & Providence Health \& Services Cancer Institute, Portland, OR, USA \\
   \hline
SCI & Swedish Cancer Institute, Seattle, WA, USA \\
   \hline
SNP & Single-nucleotide polymorphism \\
   \hline
SNV & Single-nucleotide variants \\
   \hline
UCHI & University of Chicago Comprehensive Cancer Center, Chicago, IL, USA \\
   \hline
UHN & Princess Margaret Cancer Centre, University Health Network, Toronto, Ontario, Canada \\
   \hline
VCF & Variant Call Format \\
   \hline
VHIO & Vall d' Hebron Institute of Oncology, Barcelona, Spain \\
   \hline
VICC & Vanderbilt-Ingram Cancer Center, Nashville, TN, USA \\
   \hline
WAKE & Wake Forest University Health Sciences (Wake Forest Baptist Medical Center), Winston-Salem, NC, USA \\
   \hline
YALE & Yale University (Yale Cancer Center), New Haven, Connecticut, USA \\
   \hline
\end{longtable}
\end{center}



\end{document}