
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="This repository documents code used to gather, QC, standardize, and analyze data uploaded by institutes participating in AACR's Project GENIE (Genomics, Evidence, Neoplasia, Information, Exchange).">
      
      
      
      
        <link rel="prev" href="../cna/">
      
      
        <link rel="next" href="../mutations_in_cis/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Maf - AACR Project GENIE Repository Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="green" data-md-color-accent="green">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genie_registry.maf" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="AACR Project GENIE Repository Documentation" class="md-header__button md-logo" aria-label="AACR Project GENIE Repository Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            AACR Project GENIE Repository Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Maf
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="green" data-md-color-accent="green"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="green" data-md-color-accent="green"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Sage-Bionetworks/Genie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    AACR Project GENIE
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../getting_started/" class="md-tabs__link">
        
  
  
    
  
  Getting Started

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../tutorials/local_file_validation/" class="md-tabs__link">
          
  
  
    
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../main_pipeline_commands/input_to_database/" class="md-tabs__link">
          
  
  
    
  
  Reference

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="AACR Project GENIE Repository Documentation" class="md-nav__button md-logo" aria-label="AACR Project GENIE Repository Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    AACR Project GENIE Repository Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Sage-Bionetworks/Genie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    AACR Project GENIE
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../tutorials/local_file_validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local File Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Main Pipeline Commands
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Main Pipeline Commands
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/input_to_database/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Input to Database
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/database_to_staging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/consortium_to_public/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Consortium to Public
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Main Pipeline Steps
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Main Pipeline Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2_1" >
        
          
          <label class="md-nav__link" for="__nav_4_2_1" id="__nav_4_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Input to Database
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2_1">
            <span class="md-nav__icon md-icon"></span>
            Input to Database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/input_to_database/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    input_to_database
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/validate.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    validate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/write_invalid_reasons/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    write_invalid_reasons
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/process_mutation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process_mutation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/toRetract.py" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    to_retract
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/database_to_staging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/consortium_to_public/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Consortium to Public
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" checked>
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Fileformats
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Fileformats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fileformat/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Fileformat
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assay/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Assay Information
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../bed/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Bed
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../clinical/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Clinical
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cna/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cna
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Maf
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Maf
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#genie_registry.maf" class="md-nav__link">
    <span class="md-ellipsis">
      maf
    </span>
  </a>
  
    <nav class="md-nav" aria-label="maf">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf-attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Attributes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Attributes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.logger" class="md-nav__link">
    <span class="md-ellipsis">
      logger
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf-classes" class="md-nav__link">
    <span class="md-ellipsis">
      Classes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat" class="md-nav__link">
    <span class="md-ellipsis">
      FileTypeFormat
    </span>
  </a>
  
    <nav class="md-nav" aria-label="FileTypeFormat">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Functions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.__init__" class="md-nav__link">
    <span class="md-ellipsis">
      __init__
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.read_file" class="md-nav__link">
    <span class="md-ellipsis">
      read_file
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.validateFilename" class="md-nav__link">
    <span class="md-ellipsis">
      validateFilename
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.process_steps" class="md-nav__link">
    <span class="md-ellipsis">
      process_steps
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.preprocess" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.process" class="md-nav__link">
    <span class="md-ellipsis">
      process
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.FileTypeFormat.validate" class="md-nav__link">
    <span class="md-ellipsis">
      validate
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.maf" class="md-nav__link">
    <span class="md-ellipsis">
      maf
    </span>
  </a>
  
    <nav class="md-nav" aria-label="maf">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.maf-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Functions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf.maf.process_steps" class="md-nav__link">
    <span class="md-ellipsis">
      process_steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Functions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf._check_allele_col_validity" class="md-nav__link">
    <span class="md-ellipsis">
      _check_allele_col_validity
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.maf._check_allele_col" class="md-nav__link">
    <span class="md-ellipsis">
      _check_allele_col
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mutations_in_cis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Mutations In Cis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../patient_retraction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Patient retraction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../sample_retraction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sample retraction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Segmented
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structural_variant/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Structural variant
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../vcf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Vcf
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Helper modules
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            Helper modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/create_case_lists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    create_case_lists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/dashboard_table_updater/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dashboard_table_updater
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/extract/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/load/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/process_functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process_functions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/transform/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    transform
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/Sage-Bionetworks/Genie/edit/master/docs/reference/fileformats/maf.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


  <h1>Maf</h1>

<div class="doc doc-object doc-module">



<h2 id="genie_registry.maf" class="doc doc-heading">
            <code>genie_registry.maf</code>


<a href="#genie_registry.maf" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">










  <div class="doc doc-children">





<h3 id="genie_registry.maf-attributes">Attributes<a href="#genie_registry.maf-attributes" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-attribute">



<h4 id="genie_registry.maf.logger" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie_registry.maf.logger" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<h3 id="genie_registry.maf-classes">Classes<a href="#genie_registry.maf-classes" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-class">



<h4 id="genie_registry.maf.FileTypeFormat" class="doc doc-heading">
            <code>FileTypeFormat</code>


<a href="#genie_registry.maf.FileTypeFormat" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">









              <details class="quote">
                <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">FileTypeFormat</span><span class="p">(</span><span class="n">metaclass</span><span class="o">=</span><span class="n">ABCMeta</span><span class="p">):</span>
    <span class="n">_process_kwargs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;newPath&quot;</span><span class="p">,</span> <span class="s2">&quot;databaseSynId&quot;</span><span class="p">]</span>

    <span class="n">_fileType</span> <span class="o">=</span> <span class="s2">&quot;fileType&quot;</span>

    <span class="n">_validation_kwargs</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">]</span> <span class="o">=</span> <span class="p">[]</span>

    <span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span>
        <span class="n">center</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">genie_config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
        <span class="n">ancillary_files</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">synapseclient</span><span class="o">.</span><span class="n">Entity</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;A validator helper class for a center&#39;s files.</span>

<span class="sd">        Args:</span>
<span class="sd">            syn (synapseclient.Synapse): a synapseclient.Synapse object</span>
<span class="sd">            center (str): The participating center name.</span>
<span class="sd">            genie_config (dict): The configurations needed for the GENIE codebase.</span>
<span class="sd">                GENIE table type/name to Synapse Id. Defaults to None.</span>
<span class="sd">            ancillary_files (List[List[synapseclient.Entity]]): all files downloaded for validation. Defaults to None.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">syn</span> <span class="o">=</span> <span class="n">syn</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">center</span> <span class="o">=</span> <span class="n">center</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genie_config</span> <span class="o">=</span> <span class="n">genie_config</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span> <span class="o">=</span> <span class="n">ancillary_files</span>

        <span class="c1"># self.pool = multiprocessing.Pool(poolSize)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_get_dataframe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePathList</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This function by defaults assumes the filePathList is length of 1</span>
<span class="sd">        and is a tsv file.  Could change depending on file type.</span>

<span class="sd">        Args:</span>
<span class="sd">            filePathList:  A list of file paths (Max is 2 for the two</span>
<span class="sd">                           clinical files)</span>

<span class="sd">        Returns:</span>
<span class="sd">            df: Pandas dataframe of file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">filePath</span> <span class="o">=</span> <span class="n">filePathList</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filePath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">read_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePathList</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Each file is to be read in for validation and processing.</span>
<span class="sd">        This is not to be changed in any functions.</span>

<span class="sd">        Args:</span>
<span class="sd">            filePathList:  A list of file paths (Max is 2 for the two</span>
<span class="sd">                           clinical files)</span>

<span class="sd">        Returns:</span>
<span class="sd">            df: Pandas dataframe of file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_dataframe</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validateFilename</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Function that changes per file type for validating its filename</span>
<span class="sd">        Expects an assertion error.</span>

<span class="sd">        Args:</span>
<span class="sd">            filePath: Path to file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># assert True</span>
        <span class="k">raise</span> <span class="ne">NotImplementedError</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validateFilename</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validation of file name.  The filename is what maps the file</span>
<span class="sd">        to its validation and processing.</span>

<span class="sd">        Args:</span>
<span class="sd">            filePath: Path to file</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: file type defined by self._fileType</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">_validateFilename</span><span class="p">(</span><span class="n">filePath</span><span class="p">)</span>
        <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">process_steps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This function is modified for every single file.</span>
<span class="sd">        It reformats the file and stores the file into database and Synapse.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">pass</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">preprocess</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">newpath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is for any preprocessing that has to occur to the entity name</span>
<span class="sd">        to add to kwargs for processing.  entity name is included in</span>
<span class="sd">        the new path</span>

<span class="sd">        Args:</span>
<span class="sd">            newpath: Path to file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">return</span> <span class="nb">dict</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is the main processing function.</span>

<span class="sd">        Args:</span>
<span class="sd">            filePath: Path to file</span>
<span class="sd">            kwargs: The kwargs are determined by self._process_kwargs</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: file path of processed file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">preprocess_args</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">preprocess</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;newPath&quot;</span><span class="p">))</span>
        <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">preprocess_args</span><span class="p">)</span>
        <span class="n">mykwargs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_kwargs</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="p">(</span>
                <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> not in parameter list&quot;</span> <span class="o">%</span> <span class="n">required_parameter</span>
            <span class="p">)</span>
            <span class="n">mykwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;PROCESSING </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">filePath</span><span class="p">)</span>
        <span class="c1"># This is done because the clinical files are being merged into a list</span>
        <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span> <span class="o">==</span> <span class="s2">&quot;clinical&quot;</span><span class="p">:</span>
            <span class="n">path_or_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="n">filePath</span><span class="p">)</span>
        <span class="c1"># If file type is vcf or maf file, processing requires a filepath</span>
        <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">,</span> <span class="s2">&quot;mafSP&quot;</span><span class="p">,</span> <span class="s2">&quot;md&quot;</span><span class="p">]:</span>
            <span class="n">path_or_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_file</span><span class="p">([</span><span class="n">filePath</span><span class="p">])</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">path_or_df</span> <span class="o">=</span> <span class="n">filePath</span>
        <span class="n">path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_steps</span><span class="p">(</span><span class="n">path_or_df</span><span class="p">,</span> <span class="o">**</span><span class="n">mykwargs</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">path</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is the base validation function.</span>
<span class="sd">        By default, no validation occurs.</span>

<span class="sd">        Args:</span>
<span class="sd">            df (pd.DataFrame): A dataframe of the file</span>
<span class="sd">            kwargs: The kwargs are determined by self._validation_kwargs</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: The errors and warnings as a file from validation.</span>
<span class="sd">                   Defaults to blank strings</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">errors</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">warnings</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;NO VALIDATION for </span><span class="si">%s</span><span class="s2"> files&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_cross_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is the base cross-validation function.</span>
<span class="sd">        By default, no cross-validation occurs.</span>

<span class="sd">        Args:</span>
<span class="sd">            df (pd.DataFrame): A dataframe of the file</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: The errors and warnings as a file from cross-validation.</span>
<span class="sd">                   Defaults to blank strings</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">errors</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">warnings</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;NO CROSS-VALIDATION for </span><span class="si">%s</span><span class="s2"> files&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePathList</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ValidationResults</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This is the main validation function.</span>
<span class="sd">        Every file type calls self._validate, which is different.</span>

<span class="sd">        Args:</span>
<span class="sd">            filePathList: A list of file paths.</span>
<span class="sd">            kwargs: The kwargs are determined by self._validation_kwargs</span>

<span class="sd">        Returns:</span>
<span class="sd">            tuple: The errors and warnings as a file from validation.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">mykwargs</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_kwargs</span><span class="p">:</span>
            <span class="k">assert</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="p">(</span>
                <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> not in parameter list&quot;</span> <span class="o">%</span> <span class="n">required_parameter</span>
            <span class="p">)</span>
            <span class="n">mykwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span>

        <span class="n">errors</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">try</span><span class="p">:</span>
            <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
            <span class="n">errors</span> <span class="o">=</span> <span class="p">(</span>
                <span class="sa">f</span><span class="s2">&quot;The file(s) (</span><span class="si">{</span><span class="n">filePathList</span><span class="si">}</span><span class="s2">) cannot be read. Original error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
            <span class="p">)</span>
            <span class="n">warnings</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">errors</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;VALIDATING </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)))</span>
            <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">mykwargs</span><span class="p">)</span>
            <span class="c1"># only cross-validate if validation passes or ancillary files exist</span>
            <span class="c1"># Assumes that self.ancillary_files won&#39;t be [[]] due to whats returned from</span>
            <span class="c1"># extract.get_center_input_files</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">errors</span> <span class="ow">and</span> <span class="p">(</span>
                <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span>
            <span class="p">):</span>
                <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="s2">&quot;CROSS-VALIDATING </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filePathList</span><span class="p">))</span>
                <span class="p">)</span>
                <span class="n">errors_cross_validate</span><span class="p">,</span> <span class="n">warnings_cross_validate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cross_validate</span><span class="p">(</span>
                    <span class="n">df</span>
                <span class="p">)</span>
                <span class="n">errors</span> <span class="o">+=</span> <span class="n">errors_cross_validate</span>
                <span class="n">warnings</span> <span class="o">+=</span> <span class="n">warnings_cross_validate</span>

        <span class="n">result_cls</span> <span class="o">=</span> <span class="n">ValidationResults</span><span class="p">(</span><span class="n">errors</span><span class="o">=</span><span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span><span class="o">=</span><span class="n">warnings</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">result_cls</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">








<h5 id="genie_registry.maf.FileTypeFormat-functions">Functions<a href="#genie_registry.maf.FileTypeFormat-functions" class="headerlink" title="Permanent link">&para;</a></h5>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.__init__" class="doc doc-heading">
            <code class="highlight language-python"><span class="fm">__init__</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">center</span><span class="p">,</span> <span class="n">genie_config</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">ancillary_files</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.__init__" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>A validator helper class for a center's files.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>a synapseclient.Synapse object</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>The participating center name.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_config</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>The configurations needed for the GENIE codebase.
GENIE table type/name to Synapse Id. Defaults to None.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="dict">dict</span></code>
                  </span>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>None</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ancillary_files</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>all files downloaded for validation. Defaults to None.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="typing.List">List</span>[<span title="typing.List">List</span>[<span title="synapseclient.Entity">Entity</span>]]</code>
                  </span>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>None</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">60</span>
<span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="fm">__init__</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span>
    <span class="n">center</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genie_config</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="nb">dict</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
    <span class="n">ancillary_files</span><span class="p">:</span> <span class="n">Optional</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">List</span><span class="p">[</span><span class="n">synapseclient</span><span class="o">.</span><span class="n">Entity</span><span class="p">]]]</span> <span class="o">=</span> <span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;A validator helper class for a center&#39;s files.</span>

<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): a synapseclient.Synapse object</span>
<span class="sd">        center (str): The participating center name.</span>
<span class="sd">        genie_config (dict): The configurations needed for the GENIE codebase.</span>
<span class="sd">            GENIE table type/name to Synapse Id. Defaults to None.</span>
<span class="sd">        ancillary_files (List[List[synapseclient.Entity]]): all files downloaded for validation. Defaults to None.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">syn</span> <span class="o">=</span> <span class="n">syn</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">center</span> <span class="o">=</span> <span class="n">center</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">genie_config</span> <span class="o">=</span> <span class="n">genie_config</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span> <span class="o">=</span> <span class="n">ancillary_files</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.read_file" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">read_file</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.read_file" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>Each file is to be read in for validation and processing.
This is not to be changed in any functions.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filePathList</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>A list of file paths (Max is 2 for the two
           clinical files)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Pandas dataframe of file</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">read_file</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePathList</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Each file is to be read in for validation and processing.</span>
<span class="sd">    This is not to be changed in any functions.</span>

<span class="sd">    Args:</span>
<span class="sd">        filePathList:  A list of file paths (Max is 2 for the two</span>
<span class="sd">                       clinical files)</span>

<span class="sd">    Returns:</span>
<span class="sd">        df: Pandas dataframe of file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_get_dataframe</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">df</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.validateFilename" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validateFilename</span><span class="p">(</span><span class="n">filePath</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.validateFilename" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>Validation of file name.  The filename is what maps the file
to its validation and processing.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filePath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Path to file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>file type defined by self._fileType</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validateFilename</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Validation of file name.  The filename is what maps the file</span>
<span class="sd">    to its validation and processing.</span>

<span class="sd">    Args:</span>
<span class="sd">        filePath: Path to file</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: file type defined by self._fileType</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="bp">self</span><span class="o">.</span><span class="n">_validateFilename</span><span class="p">(</span><span class="n">filePath</span><span class="p">)</span>
    <span class="k">return</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.process_steps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_steps</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.process_steps" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>This function is modified for every single file.
It reformats the file and stores the file into database and Synapse.</p>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_steps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function is modified for every single file.</span>
<span class="sd">    It reformats the file and stores the file into database and Synapse.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">pass</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.preprocess" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">preprocess</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.preprocess" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>This is for any preprocessing that has to occur to the entity name
to add to kwargs for processing.  entity name is included in
the new path</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>newpath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Path to file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">preprocess</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">newpath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is for any preprocessing that has to occur to the entity name</span>
<span class="sd">    to add to kwargs for processing.  entity name is included in</span>
<span class="sd">    the new path</span>

<span class="sd">    Args:</span>
<span class="sd">        newpath: Path to file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="nb">dict</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.process" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process</span><span class="p">(</span><span class="n">filePath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.process" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>This is the main processing function.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filePath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Path to file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>kwargs</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>The kwargs are determined by self._process_kwargs</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>{}</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>file path of processed file</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePath</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is the main processing function.</span>

<span class="sd">    Args:</span>
<span class="sd">        filePath: Path to file</span>
<span class="sd">        kwargs: The kwargs are determined by self._process_kwargs</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: file path of processed file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">preprocess_args</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">preprocess</span><span class="p">(</span><span class="n">kwargs</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;newPath&quot;</span><span class="p">))</span>
    <span class="n">kwargs</span><span class="o">.</span><span class="n">update</span><span class="p">(</span><span class="n">preprocess_args</span><span class="p">)</span>
    <span class="n">mykwargs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process_kwargs</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> not in parameter list&quot;</span> <span class="o">%</span> <span class="n">required_parameter</span>
        <span class="p">)</span>
        <span class="n">mykwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;PROCESSING </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">filePath</span><span class="p">)</span>
    <span class="c1"># This is done because the clinical files are being merged into a list</span>
    <span class="k">if</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span> <span class="o">==</span> <span class="s2">&quot;clinical&quot;</span><span class="p">:</span>
        <span class="n">path_or_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="n">filePath</span><span class="p">)</span>
    <span class="c1"># If file type is vcf or maf file, processing requires a filepath</span>
    <span class="k">elif</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;vcf&quot;</span><span class="p">,</span> <span class="s2">&quot;maf&quot;</span><span class="p">,</span> <span class="s2">&quot;mafSP&quot;</span><span class="p">,</span> <span class="s2">&quot;md&quot;</span><span class="p">]:</span>
        <span class="n">path_or_df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_file</span><span class="p">([</span><span class="n">filePath</span><span class="p">])</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">path_or_df</span> <span class="o">=</span> <span class="n">filePath</span>
    <span class="n">path</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">process_steps</span><span class="p">(</span><span class="n">path_or_df</span><span class="p">,</span> <span class="o">**</span><span class="n">mykwargs</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">path</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.FileTypeFormat.validate" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">validate</span><span class="p">(</span><span class="n">filePathList</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span></code>

<a href="#genie_registry.maf.FileTypeFormat.validate" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>This is the main validation function.
Every file type calls self._validate, which is different.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filePathList</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>A list of file paths.</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>kwargs</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>The kwargs are determined by self._validation_kwargs</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>{}</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>tuple</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>The errors and warnings as a file from validation.</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><a class="autorefs autorefs-internal" title="ValidationResults


  
      dataclass
   (genie.example_filetype_format.ValidationResults)" href="../fileformat/#genie.example_filetype_format.ValidationResults">ValidationResults</a></code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/example_filetype_format.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePathList</span><span class="p">,</span> <span class="o">**</span><span class="n">kwargs</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">ValidationResults</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This is the main validation function.</span>
<span class="sd">    Every file type calls self._validate, which is different.</span>

<span class="sd">    Args:</span>
<span class="sd">        filePathList: A list of file paths.</span>
<span class="sd">        kwargs: The kwargs are determined by self._validation_kwargs</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: The errors and warnings as a file from validation.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mykwargs</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validation_kwargs</span><span class="p">:</span>
        <span class="k">assert</span> <span class="n">required_parameter</span> <span class="ow">in</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">keys</span><span class="p">(),</span> <span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">%s</span><span class="s2"> not in parameter list&quot;</span> <span class="o">%</span> <span class="n">required_parameter</span>
        <span class="p">)</span>
        <span class="n">mykwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span> <span class="o">=</span> <span class="n">kwargs</span><span class="p">[</span><span class="n">required_parameter</span><span class="p">]</span>

    <span class="n">errors</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">try</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">read_file</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">Exception</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">errors</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;The file(s) (</span><span class="si">{</span><span class="n">filePathList</span><span class="si">}</span><span class="s2">) cannot be read. Original error: </span><span class="si">{</span><span class="nb">str</span><span class="p">(</span><span class="n">e</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">warnings</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">errors</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;VALIDATING </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filePathList</span><span class="p">)))</span>
        <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_validate</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="o">**</span><span class="n">mykwargs</span><span class="p">)</span>
        <span class="c1"># only cross-validate if validation passes or ancillary files exist</span>
        <span class="c1"># Assumes that self.ancillary_files won&#39;t be [[]] due to whats returned from</span>
        <span class="c1"># extract.get_center_input_files</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">errors</span> <span class="ow">and</span> <span class="p">(</span>
            <span class="nb">isinstance</span><span class="p">(</span><span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span><span class="p">,</span> <span class="nb">list</span><span class="p">)</span> <span class="ow">and</span> <span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span>
        <span class="p">):</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                <span class="s2">&quot;CROSS-VALIDATING </span><span class="si">%s</span><span class="s2">&quot;</span> <span class="o">%</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">filePathList</span><span class="p">))</span>
            <span class="p">)</span>
            <span class="n">errors_cross_validate</span><span class="p">,</span> <span class="n">warnings_cross_validate</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_cross_validate</span><span class="p">(</span>
                <span class="n">df</span>
            <span class="p">)</span>
            <span class="n">errors</span> <span class="o">+=</span> <span class="n">errors_cross_validate</span>
            <span class="n">warnings</span> <span class="o">+=</span> <span class="n">warnings_cross_validate</span>

    <span class="n">result_cls</span> <span class="o">=</span> <span class="n">ValidationResults</span><span class="p">(</span><span class="n">errors</span><span class="o">=</span><span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span><span class="o">=</span><span class="n">warnings</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">result_cls</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>

<div class="doc doc-object doc-class">



<h4 id="genie_registry.maf.maf" class="doc doc-heading">
            <code>maf</code>


<a href="#genie_registry.maf.maf" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="FileTypeFormat (genie.example_filetype_format.FileTypeFormat)" href="../fileformat/#genie.example_filetype_format.FileTypeFormat">FileTypeFormat</a></code></p>


        <p>MAF file format validation / processing</p>








              <details class="quote">
                <summary>Source code in <code>genie_registry/maf.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span>
<span class="normal">249</span>
<span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">maf</span><span class="p">(</span><span class="n">FileTypeFormat</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    MAF file format validation / processing</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">_fileType</span> <span class="o">=</span> <span class="s2">&quot;maf&quot;</span>

    <span class="n">_process_kwargs</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">_allele_cols</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SEQ_ALLELE1&quot;</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">]</span>
    <span class="n">_allowed_comb_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;A&quot;</span><span class="p">,</span> <span class="s2">&quot;T&quot;</span><span class="p">,</span> <span class="s2">&quot;C&quot;</span><span class="p">,</span> <span class="s2">&quot;G&quot;</span><span class="p">,</span> <span class="s2">&quot;N&quot;</span><span class="p">]</span>
    <span class="n">_allowed_ind_alleles</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;-&quot;</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validateFilename</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validates filename.  Should be</span>
<span class="sd">        data_mutations_extended_CENTER.txt</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filePath</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span> <span class="o">==</span> <span class="s2">&quot;data_mutations_extended_</span><span class="si">{}</span><span class="s2">.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">center</span>
        <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">process_steps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;The processing of maf files is specific to GENIE, so</span>
<span class="sd">        not included in this function&quot;&quot;&quot;</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
            <span class="s2">&quot;Please run with `--process mutation` parameter &quot;</span>
            <span class="s2">&quot;if you want to reannotate the mutation files&quot;</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutationDF</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        This function validates the mutation file to make sure it</span>
<span class="sd">        adheres to the mutation SOP.</span>

<span class="sd">        t_depth: This column is conditionally optional.</span>
<span class="sd">        1. If this column is missing, the data must include the t_ref_count column. Otherwise, it will cause a validation error.</span>
<span class="sd">        2. If this column is present, it must have one of the following:</span>
<span class="sd">            - A mix of numeric values and NAs</span>
<span class="sd">            - All NAs</span>
<span class="sd">            - All numeric values</span>

<span class="sd">        There are no other checks on the actual values in this column.</span>

<span class="sd">        t_ref_count: This column is conditionally optional.</span>
<span class="sd">        1. If this column is missing, the data must include the t_depth column. Otherwise, it will cause a validation error.</span>
<span class="sd">        2. If this column is present, it must have one of the following:</span>
<span class="sd">            - A mix of numeric values and NAs</span>
<span class="sd">            - All NAs</span>
<span class="sd">            - All numeric values</span>

<span class="sd">        There are no other checks on the actual values in this column.</span>

<span class="sd">        t_alt_count: This column is entirely optional.</span>
<span class="sd">        1. If this column is present, it must have one of the following:</span>
<span class="sd">            - A mix of numeric values and NAs</span>
<span class="sd">            - All NAs</span>
<span class="sd">            - All numeric values</span>

<span class="sd">        There are no other checks on the actual values in this column.</span>

<span class="sd">        Args:</span>
<span class="sd">            mutationDF: mutation dataframe</span>

<span class="sd">        Returns:</span>
<span class="sd">            Text with all the errors in the mutation file</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="n">first_header</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;CHROMOSOME&quot;</span><span class="p">,</span> <span class="s2">&quot;HUGO_SYMBOL&quot;</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">]</span>
        <span class="n">SP</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span> <span class="o">==</span> <span class="s2">&quot;mafSP&quot;</span>
        <span class="k">if</span> <span class="n">SP</span><span class="p">:</span>
            <span class="n">correct_column_headers</span> <span class="o">=</span> <span class="p">[</span>
                <span class="s2">&quot;CHROMOSOME&quot;</span><span class="p">,</span>
                <span class="s2">&quot;START_POSITION&quot;</span><span class="p">,</span>
                <span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">,</span>
            <span class="p">]</span>
            <span class="c1"># T_REF_COUNT + T_ALT_COUNT = T_DEPTH</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">correct_column_headers</span> <span class="o">=</span> <span class="p">[</span>
                <span class="s2">&quot;CHROMOSOME&quot;</span><span class="p">,</span>
                <span class="s2">&quot;START_POSITION&quot;</span><span class="p">,</span>
                <span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;T_ALT_COUNT&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">,</span>
            <span class="p">]</span>
            <span class="c1"># T_REF_COUNT + T_ALT_COUNT = T_DEPTH</span>
        <span class="n">optional_headers</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;T_REF_COUNT&quot;</span><span class="p">,</span> <span class="s2">&quot;N_DEPTH&quot;</span><span class="p">,</span> <span class="s2">&quot;N_REF_COUNT&quot;</span><span class="p">,</span> <span class="s2">&quot;N_ALT_COUNT&quot;</span><span class="p">]</span>

        <span class="n">mutationDF</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">mutationDF</span><span class="o">.</span><span class="n">columns</span><span class="p">]</span>

        <span class="c1"># total_error = &quot;&quot;</span>
        <span class="n">total_error</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>
        <span class="n">warning</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">()</span>

        <span class="c1"># CHECK: Everything in correct_column_headers must be in mutation file</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span>
            <span class="p">[</span>
                <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
                <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">correct_column_headers</span>
            <span class="p">]</span>
        <span class="p">):</span>
            <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;maf: Must at least have these headers: </span><span class="si">{}</span><span class="s2">. &quot;</span>
                <span class="s2">&quot;If you are writing your maf file with R, please make&quot;</span>
                <span class="s2">&quot;sure to specify the &#39;quote=FALSE&#39; parameter.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="p">[</span>
                            <span class="n">i</span>
                            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">correct_column_headers</span>
                            <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mutationDF</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span>
                        <span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># CHECK: First column must be in the first_header list</span>
            <span class="k">if</span> <span class="n">mutationDF</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">first_header</span><span class="p">:</span>
                <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="s2">&quot;maf: First column header must be &quot;</span>
                    <span class="s2">&quot;one of these: </span><span class="si">{}</span><span class="s2">.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">first_header</span><span class="p">))</span>
                <span class="p">)</span>
            <span class="c1"># No duplicated values</span>
            <span class="n">primary_cols</span> <span class="o">=</span> <span class="p">[</span>
                <span class="s2">&quot;CHROMOSOME&quot;</span><span class="p">,</span>
                <span class="s2">&quot;START_POSITION&quot;</span><span class="p">,</span>
                <span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">,</span>
            <span class="p">]</span>
            <span class="c1"># Strip white space if string column</span>
            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">primary_cols</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">mutationDF</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="o">==</span> <span class="nb">object</span><span class="p">:</span>
                    <span class="n">mutationDF</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutationDF</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span>
            <span class="n">duplicated_idx</span> <span class="o">=</span> <span class="n">mutationDF</span><span class="o">.</span><span class="n">duplicated</span><span class="p">(</span><span class="n">primary_cols</span><span class="p">)</span>
            <span class="c1"># Find samples with duplicated variants</span>
            <span class="n">duplicated_variants</span> <span class="o">=</span> <span class="p">(</span>
                <span class="n">mutationDF</span><span class="p">[</span><span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">][</span><span class="n">duplicated_idx</span><span class="p">]</span>
                <span class="o">.</span><span class="n">unique</span><span class="p">()</span>
                <span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
                <span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="p">)</span>

            <span class="k">if</span> <span class="n">duplicated_idx</span><span class="o">.</span><span class="n">any</span><span class="p">():</span>
                <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                    <span class="s2">&quot;maf: Must not have duplicated variants. &quot;</span>
                    <span class="s2">&quot;Samples with duplicated variants: &quot;</span>
                    <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="s1">&#39;, &#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">duplicated_variants</span><span class="p">)</span><span class="si">}</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="p">)</span>

        <span class="n">t_depth_exists</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;T_DEPTH&quot;</span><span class="p">)</span>
        <span class="n">t_ref_exists</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;T_REF_COUNT&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">t_depth_exists</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">t_ref_exists</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">SP</span><span class="p">:</span>
            <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s2">&quot;maf: If missing T_DEPTH, must have T_REF_COUNT!</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">numerical_cols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;T_DEPTH&quot;</span><span class="p">,</span>
            <span class="s2">&quot;T_ALT_COUNT&quot;</span><span class="p">,</span>
            <span class="s2">&quot;T_REF_COUNT&quot;</span><span class="p">,</span>
            <span class="s2">&quot;N_DEPTH&quot;</span><span class="p">,</span>
            <span class="s2">&quot;N_REF_COUNT&quot;</span><span class="p">,</span>
            <span class="s2">&quot;N_ALT_COUNT&quot;</span><span class="p">,</span>
            <span class="s2">&quot;START_POSITION&quot;</span><span class="p">,</span>
            <span class="s2">&quot;END_POSITION&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">actual_numerical_cols</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">numerical_cols</span><span class="p">:</span>
            <span class="n">col_exists</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">col_exists</span><span class="p">:</span>
                <span class="c1"># Attempt to convert column to float</span>
                <span class="k">try</span><span class="p">:</span>
                    <span class="n">mutationDF</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">mutationDF</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span>
                <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
                    <span class="k">pass</span>
                <span class="k">if</span> <span class="n">mutationDF</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">dtype</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="nb">int</span><span class="p">,</span> <span class="nb">float</span><span class="p">]:</span>
                    <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;maf: </span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> must be a numerical column.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">actual_numerical_cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">col</span><span class="p">)</span>

        <span class="c1"># CHECK: Must have TUMOR_SEQ_ALLELE2</span>
        <span class="n">error</span><span class="p">,</span> <span class="n">warn</span> <span class="o">=</span> <span class="n">_check_allele_col</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">)</span>
        <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>
        <span class="n">warning</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">warn</span><span class="p">)</span>

        <span class="c1"># CHECK: Mutation file would benefit from columns in optional_headers</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span>
                <span class="p">[</span>
                    <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="n">i</span><span class="p">)</span>
                    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">optional_headers</span>
                <span class="p">]</span>
            <span class="p">)</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="n">SP</span>
        <span class="p">):</span>
            <span class="n">warning</span><span class="o">.</span><span class="n">write</span><span class="p">(</span>
                <span class="s2">&quot;maf: Does not have the column headers that can give extra &quot;</span>
                <span class="s2">&quot;information to the processed maf: </span><span class="si">{}</span><span class="s2">.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
                        <span class="p">[</span>
                            <span class="n">i</span>
                            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">optional_headers</span>
                            <span class="k">if</span> <span class="n">i</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">mutationDF</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">values</span>
                        <span class="p">]</span>
                    <span class="p">)</span>
                <span class="p">)</span>
            <span class="p">)</span>

        <span class="c1"># CHECK: Must have REFERENCE_ALLELE</span>
        <span class="n">error</span><span class="p">,</span> <span class="n">warn</span> <span class="o">=</span> <span class="n">_check_allele_col</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">)</span>
        <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>
        <span class="n">warning</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">warn</span><span class="p">)</span>

        <span class="n">error</span><span class="p">,</span> <span class="n">warn</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">_validate_chromosome</span><span class="p">(</span>
            <span class="n">df</span><span class="o">=</span><span class="n">mutationDF</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;CHROMOSOME&quot;</span><span class="p">,</span> <span class="n">fileformat</span><span class="o">=</span><span class="s2">&quot;maf&quot;</span><span class="p">,</span> <span class="n">allow_chr</span><span class="o">=</span><span class="kc">False</span>
        <span class="p">)</span>
        <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>
        <span class="n">warning</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">warn</span><span class="p">)</span>
        <span class="c1"># if process_functions.checkColExist(mutationDF, &quot;CHROMOSOME&quot;):</span>
        <span class="c1">#     # CHECK: Chromosome column can&#39;t have any values that start</span>
        <span class="c1">#     # with chr or have any WT values</span>
        <span class="c1">#     invalid_values = [</span>
        <span class="c1">#         str(i).startswith(&quot;chr&quot;) or str(i) == &quot;WT&quot;</span>
        <span class="c1">#         for i in mutationDF[&quot;CHROMOSOME&quot;]</span>
        <span class="c1">#     ]</span>
        <span class="c1">#     if sum(invalid_values) &gt; 0:</span>
        <span class="c1">#         total_error.write(</span>
        <span class="c1">#             &quot;maf: CHROMOSOME column cannot have any values that &quot;</span>
        <span class="c1">#             &quot;start with &#39;chr&#39; or any &#39;WT&#39; values.\n&quot;</span>
        <span class="c1">#         )</span>

        <span class="n">error</span> <span class="o">=</span> <span class="n">_check_allele_col_validity</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">)</span>
        <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>

        <span class="k">if</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">):</span>
            <span class="n">error</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">validate_genie_identifier</span><span class="p">(</span>
                <span class="n">identifiers</span><span class="o">=</span><span class="n">mutationDF</span><span class="p">[</span><span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">],</span>
                <span class="n">center</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">center</span><span class="p">,</span>
                <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;maf&quot;</span><span class="p">,</span>
                <span class="n">col</span><span class="o">=</span><span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">error</span><span class="p">)</span>

        <span class="c1"># only check end position as start position is required col</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;START_POSITION&quot;</span><span class="p">)</span>
            <span class="ow">and</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="s2">&quot;END_POSITION&quot;</span><span class="p">)</span>
            <span class="ow">and</span> <span class="nb">set</span><span class="p">([</span><span class="s2">&quot;START_POSITION&quot;</span><span class="p">,</span> <span class="s2">&quot;END_POSITION&quot;</span><span class="p">])</span> <span class="o">&lt;=</span> <span class="nb">set</span><span class="p">(</span><span class="n">actual_numerical_cols</span><span class="p">)</span>
        <span class="p">):</span>
            <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">check_variant_start_and_end_positions</span><span class="p">(</span>
                <span class="n">input_df</span><span class="o">=</span><span class="n">mutationDF</span><span class="p">,</span>
                <span class="n">start_pos_col</span><span class="o">=</span><span class="s2">&quot;START_POSITION&quot;</span><span class="p">,</span>
                <span class="n">end_pos_col</span><span class="o">=</span><span class="s2">&quot;END_POSITION&quot;</span><span class="p">,</span>
                <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;maf&quot;</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span>
            <span class="n">warning</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">warnings</span><span class="p">)</span>

        <span class="k">for</span> <span class="n">allele_col</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_allele_cols</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">mutationDF</span><span class="p">,</span> <span class="n">allele_col</span><span class="p">):</span>
                <span class="n">invalid_indices</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">get_invalid_allele_rows</span><span class="p">(</span>
                    <span class="n">mutationDF</span><span class="p">,</span>
                    <span class="n">allele_col</span><span class="p">,</span>
                    <span class="n">allowed_comb_alleles</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_allowed_comb_alleles</span><span class="p">,</span>
                    <span class="n">allowed_ind_alleles</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_allowed_ind_alleles</span><span class="p">,</span>
                    <span class="n">ignore_case</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
                    <span class="n">allow_na</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">get_allele_validation_message</span><span class="p">(</span>
                    <span class="n">invalid_indices</span><span class="p">,</span>
                    <span class="n">invalid_col</span><span class="o">=</span><span class="n">allele_col</span><span class="p">,</span>
                    <span class="n">allowed_comb_alleles</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_allowed_comb_alleles</span><span class="p">,</span>
                    <span class="n">allowed_ind_alleles</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_allowed_ind_alleles</span><span class="p">,</span>
                    <span class="n">fileformat</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">_fileType</span><span class="p">,</span>
                <span class="p">)</span>
                <span class="n">total_error</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">errors</span><span class="p">)</span>
                <span class="n">warning</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">warnings</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">total_error</span><span class="o">.</span><span class="n">getvalue</span><span class="p">(),</span> <span class="n">warning</span><span class="o">.</span><span class="n">getvalue</span><span class="p">()</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_cross_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">mutationDF</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">tuple</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;This function cross-validates the mutation file to make sure it</span>
<span class="sd">        adheres to the mutation SOP.</span>

<span class="sd">        Args:</span>
<span class="sd">            mutationDF (pd.DataFrame): mutation dataframe</span>

<span class="sd">        Returns:</span>
<span class="sd">            Text with all the errors in the mutation file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">errors</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">warnings</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

        <span class="c1"># This section can be removed once we remove the list of lists</span>
        <span class="n">clinical_files</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">parse_file_info_in_nested_list</span><span class="p">(</span>
            <span class="n">nested_list</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">ancillary_files</span><span class="p">,</span> <span class="n">search_str</span><span class="o">=</span><span class="s2">&quot;data_clinical_supp&quot;</span>  <span class="c1"># type: ignore[arg-type]</span>
        <span class="p">)</span>
        <span class="n">clinical_file_paths</span> <span class="o">=</span> <span class="n">clinical_files</span><span class="p">[</span><span class="s2">&quot;file_info&quot;</span><span class="p">][</span><span class="s2">&quot;path&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">clinical_files</span><span class="p">[</span><span class="s2">&quot;files&quot;</span><span class="p">]:</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">clinical_sample_df</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">get_clinical_dataframe</span><span class="p">(</span>
                    <span class="n">filePathList</span><span class="o">=</span><span class="n">clinical_file_paths</span>
                <span class="p">)</span>
                <span class="n">has_file_read_error</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
                <span class="n">has_file_read_error</span> <span class="o">=</span> <span class="kc">True</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">has_file_read_error</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">clinical_sample_df</span><span class="p">,</span> <span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">):</span>
                    <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">check_values_between_two_df</span><span class="p">(</span>
                        <span class="n">df1</span><span class="o">=</span><span class="n">mutationDF</span><span class="p">,</span>
                        <span class="n">df1_filename</span><span class="o">=</span><span class="s2">&quot;MAF&quot;</span><span class="p">,</span>
                        <span class="n">df1_id_to_check</span><span class="o">=</span><span class="s2">&quot;TUMOR_SAMPLE_BARCODE&quot;</span><span class="p">,</span>
                        <span class="n">df2</span><span class="o">=</span><span class="n">clinical_sample_df</span><span class="p">,</span>
                        <span class="n">df2_filename</span><span class="o">=</span><span class="s2">&quot;sample clinical&quot;</span><span class="p">,</span>
                        <span class="n">df2_id_to_check</span><span class="o">=</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">,</span>
                    <span class="p">)</span>
        <span class="k">return</span> <span class="n">errors</span><span class="p">,</span> <span class="n">warnings</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_get_dataframe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filePathList</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">])</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Get mutation dataframe</span>

<span class="sd">        1) Starts reading the first line in the file</span>
<span class="sd">        2) Skips lines that starts with #</span>
<span class="sd">        3) Reads in second line</span>
<span class="sd">        4) Checks that first line fields matches second line. Must do this because</span>
<span class="sd">        pandas.read_csv will allow for a file to have more column headers than content.</span>
<span class="sd">        E.g)  A,B,C,D,E</span>
<span class="sd">              1,2</span>
<span class="sd">              2,3</span>

<span class="sd">        5) We keep the &#39;NA&#39;, &#39;nan&#39;, and &#39;NaN&#39; as strings in the data because</span>
<span class="sd">        these are valid allele values</span>
<span class="sd">        then convert the ones in the non-allele columns back to actual NAs</span>

<span class="sd">        NOTE: Because allele columns are case-insensitive in maf data, we must</span>
<span class="sd">        standardize the case of the columns when checking for the non-allele columns</span>
<span class="sd">        to convert the NA strings to NAs</span>

<span class="sd">        NOTE: This code allows empty dataframes to pass through</span>
<span class="sd">        without errors</span>

<span class="sd">        Args:</span>
<span class="sd">            filePathList (List[str]): list of filepath(s)</span>

<span class="sd">        Raises:</span>
<span class="sd">            ValueError: First line fields doesn&#39;t match second line fields in file</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: mutation data</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filePathList</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">maf_f</span><span class="p">:</span>
            <span class="n">firstline</span> <span class="o">=</span> <span class="n">maf_f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">firstline</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;#&quot;</span><span class="p">):</span>
                <span class="n">firstline</span> <span class="o">=</span> <span class="n">maf_f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="n">secondline</span> <span class="o">=</span> <span class="n">maf_f</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>

        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">firstline</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">secondline</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Number of fields in a line do not match the &quot;</span>
                <span class="s2">&quot;expected number of columns&quot;</span>
            <span class="p">)</span>

        <span class="n">read_csv_params</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;filepath_or_buffer&quot;</span><span class="p">:</span> <span class="n">filePathList</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="s2">&quot;sep&quot;</span><span class="p">:</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
            <span class="s2">&quot;comment&quot;</span><span class="p">:</span> <span class="s2">&quot;#&quot;</span><span class="p">,</span>
            <span class="s2">&quot;keep_default_na&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
            <span class="s2">&quot;na_values&quot;</span><span class="p">:</span> <span class="p">[</span>
                <span class="s2">&quot;-1.#IND&quot;</span><span class="p">,</span>
                <span class="s2">&quot;1.#QNAN&quot;</span><span class="p">,</span>
                <span class="s2">&quot;1.#IND&quot;</span><span class="p">,</span>
                <span class="s2">&quot;-1.#QNAN&quot;</span><span class="p">,</span>
                <span class="s2">&quot;#N/A N/A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;#N/A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;N/A&quot;</span><span class="p">,</span>
                <span class="s2">&quot;#NA&quot;</span><span class="p">,</span>
                <span class="s2">&quot;NULL&quot;</span><span class="p">,</span>
                <span class="s2">&quot;-NaN&quot;</span><span class="p">,</span>
                <span class="s2">&quot;-nan&quot;</span><span class="p">,</span>
                <span class="s2">&quot;&quot;</span><span class="p">,</span>
            <span class="p">],</span>
            <span class="c1"># This is to check if people write files</span>
            <span class="c1"># with R, quote=T</span>
            <span class="s2">&quot;quoting&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
            <span class="c1"># Retain completely blank lines so that</span>
            <span class="c1"># validator will cause the file to fail</span>
            <span class="s2">&quot;skip_blank_lines&quot;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>
        <span class="p">}</span>

        <span class="n">mutationdf</span> <span class="o">=</span> <span class="n">transform</span><span class="o">.</span><span class="n">_convert_df_with_mixed_dtypes</span><span class="p">(</span><span class="n">read_csv_params</span><span class="p">)</span>

        <span class="n">mutationdf</span> <span class="o">=</span> <span class="n">transform</span><span class="o">.</span><span class="n">_convert_values_to_na</span><span class="p">(</span>
            <span class="n">input_df</span><span class="o">=</span><span class="n">mutationdf</span><span class="p">,</span>
            <span class="n">values_to_replace</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;nan&quot;</span><span class="p">,</span> <span class="s2">&quot;NaN&quot;</span><span class="p">],</span>
            <span class="n">columns_to_convert</span><span class="o">=</span><span class="p">[</span>
                <span class="n">col</span>
                <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">mutationdf</span><span class="o">.</span><span class="n">columns</span>
                <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">not</span> <span class="ow">in</span> <span class="bp">self</span><span class="o">.</span><span class="n">_allele_cols</span>
            <span class="p">],</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">mutationdf</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">








<h5 id="genie_registry.maf.maf-functions">Functions<a href="#genie_registry.maf.maf-functions" class="headerlink" title="Permanent link">&para;</a></h5>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.maf.maf.process_steps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_steps</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#genie_registry.maf.maf.process_steps" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>The processing of maf files is specific to GENIE, so
not included in this function</p>


            <details class="quote">
              <summary>Source code in <code>genie_registry/maf.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_steps</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">df</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;The processing of maf files is specific to GENIE, so</span>
<span class="sd">    not included in this function&quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Please run with `--process mutation` parameter &quot;</span>
        <span class="s2">&quot;if you want to reannotate the mutation files&quot;</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
<h3 id="genie_registry.maf-functions">Functions<a href="#genie_registry.maf-functions" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.maf._check_allele_col_validity" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_check_allele_col_validity</span><span class="p">(</span><span class="n">df</span><span class="p">)</span></code>

<a href="#genie_registry.maf._check_allele_col_validity" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>This function checks specific columns in a MAF (Mutation Annotation Format)
file for certain conditions.</p>


<details class="the-following-conditions-must-be-met" open>
  <summary>The following conditions must be met</summary>
  <p><strong>If the MAF file has all three of these columns</strong></p>
<pre><code>- TUMOR_SEQ_ALLELE1 (TSA1)
- TUMOR_SEQ_ALLELE2 (TSA2)
- REFERENCE_ALLELE (REF)
</code></pre>
<p><strong>Then, one of the following must be true</strong></p>
<pre><code>- Every value in TSA1 must be the same as the value in REF
- Every value in TSA1 must be the same as the value in TSA2
</code></pre>
<p><strong>Additionally, if the MAF file has at least these two columns</strong></p>
<pre><code>- REFERENCE_ALLELE (REF)
- TUMOR_SEQ_ALLELE2 (TSA2)
</code></pre>
<p><strong>Then</strong></p>
<pre><code>NO values in REF can match TSA2
</code></pre>
<p>These rules are important because Genome Nexus (GN) uses <code>TSA1</code> to annotate data
when it's not clear which variant to use. So, there can't be a mix of rows where
some have <code>TSA1</code> equal to <code>REF</code> and some have <code>TSA1</code> equal to <code>TSA2</code>.</p>
</details>

<details class="example" open>
  <summary>Valid Examples</summary>
  <div class="highlight"><pre><span></span><code>| REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |
| ---------------- | ----------------- | ----------------- |
| C                | C                 | A                 |
| T                | T                 | C                 |
</code></pre></div>
<div class="highlight"><pre><span></span><code>| REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |
| ---------------- | ----------------- | ----------------- |
| C                | A                 | A                 |
| T                | C                 | C                 |
</code></pre></div>
<div class="highlight"><pre><span></span><code>| REFERENCE_ALLELE | TUMOR_SEQ_ALLELE2 |
| ---------------- | ----------------- |
| C                | A                 |
| T                | C                 |
</code></pre></div>
</details>

<details class="example" open>
  <summary>Invalid Examples</summary>
  <div class="highlight"><pre><span></span><code>| REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |
| ---------------- | ----------------- | ----------------- |
| C                | C                 | A                 |
| C                | A                 | A                 |
</code></pre></div>
<div class="highlight"><pre><span></span><code>| REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |
| ---------------- | ----------------- | ----------------- |
| A                | C                 | A                 |
| T                | C                 | T                 |
</code></pre></div>
<div class="highlight"><pre><span></span><code>| REFERENCE_ALLELE | TUMOR_SEQ_ALLELE2 |
| ---------------- | ----------------- |
| C                | C                 |
| T                | C                 |
</code></pre></div>
</details>        <p>See this <a href="https://github.com/genome-nexus/annotation-tools/issues/26">Genome Nexus issue</a> for
more background regarding why this validation rule was implemented.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>input mutation dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>the error message</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/maf.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 14</span>
<span class="normal"> 15</span>
<span class="normal"> 16</span>
<span class="normal"> 17</span>
<span class="normal"> 18</span>
<span class="normal"> 19</span>
<span class="normal"> 20</span>
<span class="normal"> 21</span>
<span class="normal"> 22</span>
<span class="normal"> 23</span>
<span class="normal"> 24</span>
<span class="normal"> 25</span>
<span class="normal"> 26</span>
<span class="normal"> 27</span>
<span class="normal"> 28</span>
<span class="normal"> 29</span>
<span class="normal"> 30</span>
<span class="normal"> 31</span>
<span class="normal"> 32</span>
<span class="normal"> 33</span>
<span class="normal"> 34</span>
<span class="normal"> 35</span>
<span class="normal"> 36</span>
<span class="normal"> 37</span>
<span class="normal"> 38</span>
<span class="normal"> 39</span>
<span class="normal"> 40</span>
<span class="normal"> 41</span>
<span class="normal"> 42</span>
<span class="normal"> 43</span>
<span class="normal"> 44</span>
<span class="normal"> 45</span>
<span class="normal"> 46</span>
<span class="normal"> 47</span>
<span class="normal"> 48</span>
<span class="normal"> 49</span>
<span class="normal"> 50</span>
<span class="normal"> 51</span>
<span class="normal"> 52</span>
<span class="normal"> 53</span>
<span class="normal"> 54</span>
<span class="normal"> 55</span>
<span class="normal"> 56</span>
<span class="normal"> 57</span>
<span class="normal"> 58</span>
<span class="normal"> 59</span>
<span class="normal"> 60</span>
<span class="normal"> 61</span>
<span class="normal"> 62</span>
<span class="normal"> 63</span>
<span class="normal"> 64</span>
<span class="normal"> 65</span>
<span class="normal"> 66</span>
<span class="normal"> 67</span>
<span class="normal"> 68</span>
<span class="normal"> 69</span>
<span class="normal"> 70</span>
<span class="normal"> 71</span>
<span class="normal"> 72</span>
<span class="normal"> 73</span>
<span class="normal"> 74</span>
<span class="normal"> 75</span>
<span class="normal"> 76</span>
<span class="normal"> 77</span>
<span class="normal"> 78</span>
<span class="normal"> 79</span>
<span class="normal"> 80</span>
<span class="normal"> 81</span>
<span class="normal"> 82</span>
<span class="normal"> 83</span>
<span class="normal"> 84</span>
<span class="normal"> 85</span>
<span class="normal"> 86</span>
<span class="normal"> 87</span>
<span class="normal"> 88</span>
<span class="normal"> 89</span>
<span class="normal"> 90</span>
<span class="normal"> 91</span>
<span class="normal"> 92</span>
<span class="normal"> 93</span>
<span class="normal"> 94</span>
<span class="normal"> 95</span>
<span class="normal"> 96</span>
<span class="normal"> 97</span>
<span class="normal"> 98</span>
<span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_check_allele_col_validity</span><span class="p">(</span><span class="n">df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    This function checks specific columns in a MAF (Mutation Annotation Format)</span>
<span class="sd">    file for certain conditions.</span>

<span class="sd">    The following conditions must be met:</span>
<span class="sd">        **If the MAF file has all three of these columns**</span>

<span class="sd">            - TUMOR_SEQ_ALLELE1 (TSA1)</span>
<span class="sd">            - TUMOR_SEQ_ALLELE2 (TSA2)</span>
<span class="sd">            - REFERENCE_ALLELE (REF)</span>

<span class="sd">        **Then, one of the following must be true**</span>

<span class="sd">            - Every value in TSA1 must be the same as the value in REF</span>
<span class="sd">            - Every value in TSA1 must be the same as the value in TSA2</span>

<span class="sd">        **Additionally, if the MAF file has at least these two columns**</span>

<span class="sd">            - REFERENCE_ALLELE (REF)</span>
<span class="sd">            - TUMOR_SEQ_ALLELE2 (TSA2)</span>

<span class="sd">        **Then**</span>

<span class="sd">            NO values in REF can match TSA2</span>

<span class="sd">        These rules are important because Genome Nexus (GN) uses `TSA1` to annotate data</span>
<span class="sd">        when it&#39;s not clear which variant to use. So, there can&#39;t be a mix of rows where</span>
<span class="sd">        some have `TSA1` equal to `REF` and some have `TSA1` equal to `TSA2`.</span>

<span class="sd">    Example: Valid Examples</span>
<span class="sd">        ```</span>
<span class="sd">        | REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |</span>
<span class="sd">        | ---------------- | ----------------- | ----------------- |</span>
<span class="sd">        | C                | C                 | A                 |</span>
<span class="sd">        | T                | T                 | C                 |</span>
<span class="sd">        ```</span>

<span class="sd">        ```</span>
<span class="sd">        | REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |</span>
<span class="sd">        | ---------------- | ----------------- | ----------------- |</span>
<span class="sd">        | C                | A                 | A                 |</span>
<span class="sd">        | T                | C                 | C                 |</span>
<span class="sd">        ```</span>

<span class="sd">        ```</span>
<span class="sd">        | REFERENCE_ALLELE | TUMOR_SEQ_ALLELE2 |</span>
<span class="sd">        | ---------------- | ----------------- |</span>
<span class="sd">        | C                | A                 |</span>
<span class="sd">        | T                | C                 |</span>
<span class="sd">        ```</span>


<span class="sd">    Example: Invalid Examples</span>
<span class="sd">        ```</span>
<span class="sd">        | REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |</span>
<span class="sd">        | ---------------- | ----------------- | ----------------- |</span>
<span class="sd">        | C                | C                 | A                 |</span>
<span class="sd">        | C                | A                 | A                 |</span>
<span class="sd">        ```</span>

<span class="sd">        ```</span>
<span class="sd">        | REFERENCE_ALLELE | TUMOR_SEQ_ALLELE1 | TUMOR_SEQ_ALLELE2 |</span>
<span class="sd">        | ---------------- | ----------------- | ----------------- |</span>
<span class="sd">        | A                | C                 | A                 |</span>
<span class="sd">        | T                | C                 | T                 |</span>
<span class="sd">        ```</span>

<span class="sd">        ```</span>
<span class="sd">        | REFERENCE_ALLELE | TUMOR_SEQ_ALLELE2 |</span>
<span class="sd">        | ---------------- | ----------------- |</span>
<span class="sd">        | C                | C                 |</span>
<span class="sd">        | T                | C                 |</span>
<span class="sd">        ```</span>


<span class="sd">    See this [Genome Nexus issue](https://github.com/genome-nexus/annotation-tools/issues/26) for</span>
<span class="sd">    more background regarding why this validation rule was implemented.</span>

<span class="sd">    Args:</span>
<span class="sd">        df: input mutation dataframe</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: the error message</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">tsa2_col_exist</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">)</span>
    <span class="n">tsa1_col_exist</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;TUMOR_SEQ_ALLELE1&quot;</span><span class="p">)</span>
    <span class="n">ref_col_exist</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">)</span>
    <span class="n">error</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">tsa2_col_exist</span> <span class="ow">and</span> <span class="n">tsa1_col_exist</span> <span class="ow">and</span> <span class="n">ref_col_exist</span><span class="p">:</span>
        <span class="n">tsa1_eq_ref</span> <span class="o">=</span> <span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;TUMOR_SEQ_ALLELE1&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">])</span>
        <span class="n">tsa1_eq_tsa2</span> <span class="o">=</span> <span class="nb">all</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;TUMOR_SEQ_ALLELE1&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">tsa1_eq_ref</span> <span class="ow">or</span> <span class="n">tsa1_eq_tsa2</span><span class="p">):</span>
            <span class="n">error</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;maf: Contains both &quot;</span>
                <span class="s2">&quot;TUMOR_SEQ_ALLELE1 and TUMOR_SEQ_ALLELE2 columns. &quot;</span>
                <span class="s2">&quot;All values in TUMOR_SEQ_ALLELE1 must match all values in &quot;</span>
                <span class="s2">&quot;REFERENCE_ALLELE or all values in TUMOR_SEQ_ALLELE2.</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="p">)</span>
    <span class="k">if</span> <span class="p">(</span>
        <span class="n">tsa2_col_exist</span>
        <span class="ow">and</span> <span class="n">ref_col_exist</span>
        <span class="ow">and</span> <span class="nb">any</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">])</span>
    <span class="p">):</span>
        <span class="n">error</span> <span class="o">=</span> <span class="p">(</span>
            <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">error</span><span class="si">}</span><span class="s2">maf: Contains instances where values in REFERENCE_ALLELE match values in TUMOR_SEQ_ALLELE2. &quot;</span>
            <span class="s2">&quot;This is invalid. Please correct.</span><span class="se">\n</span><span class="s2">&quot;</span>
        <span class="p">)</span>
        <span class="n">row_index</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s2">&quot;REFERENCE_ALLELE&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">df</span><span class="p">[</span><span class="s2">&quot;TUMOR_SEQ_ALLELE2&quot;</span><span class="p">]]</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">values</span>
    <span class="k">return</span> <span class="n">error</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.maf._check_allele_col" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_check_allele_col</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span></code>

<a href="#genie_registry.maf._check_allele_col" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Check the Allele column is correctly formatted.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>mutation dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>col</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Column header name</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>error, warning</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/maf.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span>
<span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_check_allele_col</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check the Allele column is correctly formatted.</span>

<span class="sd">    Args:</span>
<span class="sd">        df: mutation dataframe</span>
<span class="sd">        col: Column header name</span>

<span class="sd">    Returns:</span>
<span class="sd">        error, warning</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">col_exist</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkColExist</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">col</span><span class="p">)</span>
    <span class="n">error</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">warning</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="k">if</span> <span class="n">col_exist</span><span class="p">:</span>
        <span class="c1"># CHECK: There can&#39;t be any null values</span>
        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">error</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;maf: </span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s2"> can&#39;t have any blank or null values.</span><span class="se">\n</span><span class="s2">&quot;</span>

    <span class="k">return</span> <span class="n">error</span><span class="p">,</span> <span class="n">warning</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tabs", "search.highlight", "search.share", "toc.integrate", "content.tabs.link", "navigation.indexes", "navigation.sections", "content.code.copy", "navigation.sections", "navigation.tabs", "navigation.top", "navigation.tracking", "navigation.instant", "header.autohide", "content.action.edit", "content.tabs.link", "content.tabs", "search.highlight", "search.share", "search.suggest", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>