---
title: "%s Summary"
output: html_document
#  html_document:
#    toc: true
#    toc_depth: 3
params:
  genieUser:
    value: x
  geniePass:
    value: x
---
  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      warning = FALSE,
                      fig.height = 8,
                      fig.width = 10)
```

```{r intro}
suppressMessages(library(synapser))
tryCatch({
  foo = capture.output(synLogin())
}, error = function(err) {
  genieUser = params$genieUser
  geniePass = params$geniePass
  foo = capture.output(synLogin(genieUser, geniePass))
})

suppressMessages(library(ggplot2))
suppressMessages(library(UpSetR))
suppressMessages(library(RColorBrewer))
suppressMessages(library(jsonlite))
suppressMessages(library(knitr))

createCenterColumn <- function(clinicalDf) {        
  if (is.null(clinicalDf$CENTER)) {
    centers = unlist(
      lapply(strsplit(clinicalDf$PATIENT_ID, "-"), function(x) return(x[2])))
  } else {
    centers = clinicalDf$CENTER
  }
  return(centers)
}

getFileDf <- function(fileName, releaseFiles) {
  synId = unlist(lapply(releaseFiles, function(x) {
    if (x$name == fileName) {
      return(x$id)
    }
  }))
  if (length(synId) > 0) {
    ent = synGet(synId,followLink = T)
    df = read.csv(ent$path,sep = "\t", comment.char = "#", stringsAsFactors = F)
    return(df)
  } else {
    return(NULL)
  }
}
getPrimary <- function(code, oncotreeDict, primary) {
  if (code != "") {
    for (x in oncotreeDict) {
      if (length(primary[x == code]) > 0) {
        return(unique(primary[x == code])[1]) 
      }
    }
    return(code)
  } else {
    return("NOT_ANNOTATED")
  }
}

extract <- function(oncotree_json, primary, secondary) {
  oncotree_code_to_info = list()
  data = oncotree_json$children
  for (node in names(data)) {
    if (data[[node]]$level == 1) {
      primary = node
      secondary = ""
    } else if (data[[node]]$level == 2) {
      secondary = node
    }
    cancer_type = data[[node]]$mainType
    cancer_type_detailed = data[[node]]$name
    if (is.null(cancer_type_detailed)) {
      cancer_type_detailed = ""
    }
    oncotree_code_to_info[[node]] = c("CANCER_TYPE" = cancer_type,
                                      "CANCER_TYPE_DETAILED" = cancer_type_detailed , 
                                      "ONCOTREE_PRIMARY_NODE" = primary, 
                                      "ONCOTREE_SECONDARY_NODE" = secondary)
    if (length(data[[node]]$children) > 0) {
      recurseDict = extract(data[[node]], primary, secondary)
      oncotree_code_to_info <- append(oncotree_code_to_info,recurseDict)
    }
  }
  return(oncotree_code_to_info)
}
plotPrimarySites <- function(clinical, oncotreeLink, release) { 
  tryCatch({
    ONCOTREE_MAP <- read.csv(oncotreeLink, sep = "\t", stringsAsFactors = F)
    levels = colnames(ONCOTREE_MAP)[grepl(colnames(ONCOTREE_MAP), pattern = "level")]
    oncotreeDict = list()
    for (level in levels) {
      oncotreeDict[[level]] = sub(".+[(](.+)[)]","\\1",ONCOTREE_MAP[[level]])
    }
    primary = oncotreeDict$level_1
    clinical$PRIMARY_CODES <- unlist(sapply(clinical$ONCOTREE_CODE, function(code) {
      getPrimary(code, oncotreeDict, primary)
    }))
  }, error = function(err) {
    data <- fromJSON(oncotreeLink)
    oncotree_json = data$TISSUE
    oncotreeDict = extract(oncotree_json, "", "")
    clinical$PRIMARY_CODES <- unlist(sapply(clinical$ONCOTREE_CODE, function(code) {
      if (toupper(code) %in% names(oncotreeDict)) {
        oncotreeDict[[toupper(code)]]["ONCOTREE_PRIMARY_NODE"]
      } else {
        "DEPRECATED_CODE"
      }
    }))
  })
  clinical$CENTER = createCenterColumn(clinical)
  primarysites <- matrix(nrow = length(unique(clinical$PRIMARY_CODES)),
                         ncol = length(unique(clinical$CENTER)),
                         dimnames = list(unique(clinical$PRIMARY_CODES),
                                         unique(clinical$CENTER)))
  
  for (center in unique(clinical$CENTER)) {
    mapping <- clinical[clinical$CENTER == center,]
    primary = table(mapping$PRIMARY_CODES)
    primarysites[,center] <- primary[match(row.names(primarysites),names(primary))]
  }
  primarysites[is.na(primarysites)] <- 0
  transposed <- t(primarysites)
  
  maxValue = max(apply(transposed, 2,sum))
  par(mar = c(11, 4.1, 4.1, 7), xpd = TRUE)
  if (nrow(transposed) > 1) {
    toPlot = transposed[,order(apply(transposed, 2,sum), decreasing = T)][,0:50]
  } else {
    toPlot = transposed[,order(apply(transposed, 2,sum), decreasing = T)][0:50]
  }
  barplot(toPlot,
          col = c("red","blue","light green","pink",
                  "purple","black","lavender","yellow"),
          width = 2,
          beside = F,
          las = 2,
          ylab = "Number of Samples",
          main = "Primary Site Distribution (Top 50 Counts)",
          ylim = c(0,ceiling(maxValue * 1.0 / 1000) * 1000))
  legend("topright",
         fill = c("red","blue","light green","pink",
                  "purple","black","lavender","yellow"), 
         legend = rownames(transposed))
}

plotCancerType <- function(this_samples) {
  typeCounts = table(this_samples$CANCER_TYPE)
  op = par(mar = c(13, 4, 4, 2))
  barplot(typeCounts[order(typeCounts,decreasing = TRUE)],
          las = 2,
          cex.names = 0.7,
          ylab = "Number of samples",
          main = "Samples, by cancer type and institution")
  typeCountsByCen = table(this_samples$CENTER,this_samples$CANCER_TYPE)
  
  fractionalCounts = apply(typeCountsByCen,MARGIN = 2,function(x){ x/sum(x)})
  op = par(mar = c(13, 4, 4, 2))
  if (nrow(typeCountsByCen) > 1) {
    toPlot = fractionalCounts[,order(typeCounts,decreasing = TRUE)]
  } else {
    toPlot = fractionalCounts[order(typeCounts,decreasing = TRUE)]
  }
  barplot(toPlot,
          las = 2,
          cex.names = 0.7,
          ylab = "Fraction of samples",
          col = brewer.pal(8,"Spectral"),
          legend.text = rownames(typeCountsByCen),
          main = "Fractional sample counts, per center and type")
}

plotPanelOverlap <- function(samples, bed, assays) {
  #source("../analyses/genomicData/panelOverlap_gene.R")
  #panelOverlap = getPanelOverlap(genieClinData = this_samples,genieBedData = this_bed)
  #kable(panelOverlap)
  makePanelList = function(assay,bed){
    return(unique(as.character(bed$Hugo_Symbol[bed$SEQ_ASSAY_ID == assay])))
  }
  listInput = lapply(as.list(assays), function(x){makePanelList(x,bed)})
  names(listInput) = assays
  upset(fromList(listInput),
        order.by = "freq",
        nsets = length(assays),
        nintersects = 30)
}

plotCenterXRace <- function(genieClinData) {
  t = as.data.frame.matrix(table(genieClinData$CENTER,genieClinData$PRIMARY_RACE))
  t = data.frame(n = rowSums(t),t)
  # kable(t)
  
  # kable(data.frame("Percent.Unknown" = (t[,"Unknown"]/rowSums(t[,colnames(t) != "n"]))*100), digits = 2)
  
  # reduce table to none "Unknown" data
  # t$n = t$n - t$Unknown
  # t = t[,!(colnames(t) == "Unknown")]
  # find sites with greater than 100 nono "Unknown" patients wrt race
  # k = which(t$n > 100)
  t = t[,c("n", "White", "Black", "Asian", "Native.American",
           "Other", "Unknown")]
  t[,colnames(t) != "n"] = t[,colnames(t) != "n"]/rowSums(t[,colnames(t) != "n"])
  # kable(t[k,],digits = 3)
  
  # Center vs Race BARCHARTS
  t$Center = rownames(t)
  l = c("White", "Black", "Asian", "Native.American",
        "Other", "Unknown")
  t = reshape(t,
              direction = "long",
              idvar = ("Center"),
              varying = l,
              v.names = "Percentage",
              times = l,
              drop = "n")
  colnames(t)[colnames(t) == "time"] = "Race"
  t$Race = factor(t$Race,l)
  plot(ggplot(t,aes(Race, Percentage, group = Center)) + 
         geom_bar(stat = "identity") +
         facet_wrap(~Center, ncol = 4) +
         ylim(0,1) +
         coord_flip() +
         ggtitle("Distribution of race across centers"))
}

plotCenterXEthnicity <- function(genieClinData) {
  t = as.data.frame.matrix(table(genieClinData$CENTER,genieClinData$ETHNICITY))
  # t = data.frame(n = rowSums(t),t)
  # kable(t)
  
  # kable(data.frame("Percent.Unknown" = (t[,"Unknown"]/rowSums(t[,colnames(t) != "n"]))*100), digits = 2)
  
  # t$n = t$n - t$Unknown
  # t = t[,!(colnames(t) == "Unknown")]
  # k = which(t$n > 100)
  t[,colnames(t) != "n"] = t[,colnames(t) != "n"]/rowSums(t[,colnames(t) != "n"])
  # kable(t[k,],digits = 3)
  
  # Center vs Ethnicity BARCHARTS
  t$Center = rownames(t)
  l = c("Non-Spanish/non-Hispanic", "Spanish/Hispanic", "Unknown")
  t = reshape(t,
              direction = "long",
              idvar = ("Center"),
              varying = l,
              v.names = "Percentage",
              times = l,
              drop = "n")
  colnames(t)[colnames(t) == "time"] = "Ethnicity"
  t$Ethnicity = factor(t$Ethnicity,l)
  plot(ggplot(t,aes(Ethnicity, Percentage, group = Center)) +
         geom_bar(stat = "identity") +
         facet_wrap(~Center,ncol = 4) +
         ylim(0,1) +
         coord_flip() +
         ggtitle("Distribution of ethnicity across centers"))
}

plotCenterXSex <- function(genieClinData) {
  t = as.data.frame.matrix(table(genieClinData$CENTER,genieClinData$SEX))
  t = data.frame(n = rowSums(t),t)
  kable(t)
  
  t[,colnames(t) != "n"] = t[,colnames(t) != "n"]/rowSums(t[,colnames(t) != "n"])
  kable(t,digits = 3)
  
  # Center vs Sex BARCHARTS
  t$Center = rownames(t)
  l = c("Female","Male")
  t = reshape(t,
              direction = "long",
              idvar = ("Center"),
              varying = l,
              v.names = "Percentage",
              times = l,
              drop = "n")
  colnames(t)[colnames(t) == "time"] = "Sex"
  t$Sex = factor(t$Sex, l)
  plot(ggplot(t,aes(Sex,Percentage,group = Center)) +
         geom_bar(stat = "identity") +
         facet_wrap(~Center,ncol = 4) +
         ylim(0,1) +
         coord_flip() +
         ggtitle("Distribution of sex across centers"))
}

plotCenterXAge <- function(genieClinData) {
  if (mean(as.numeric(genieClinData$AGE_AT_SEQ_REPORT),na.rm = T) < 100) {
    multiply = T
  } else {
    multiply = F
  } 
  genieClinData$age_days <- genieClinData$AGE_AT_SEQ_REPORT
  k = grep("^>",genieClinData$AGE_AT_SEQ_REPORT) # these are the >=90 group that get mapped to 90
  genieClinData$age_days[k] = 32872
  k = grep("^<",genieClinData$AGE_AT_SEQ_REPORT) # these are the <=18 group that get mapped to 18
  genieClinData$age_days[k] = 6573
  genieClinData$age_days = as.numeric(genieClinData$AGE_AT_SEQ_REPORT)
  if (multiply) {
    genieClinData$age_days <- genieClinData$age_days * 365
  }
  
  k = is.na(genieClinData$age_days)
  t = as.data.frame.matrix(table(genieClinData$CENTER,k))
  kable(data.frame("Percent.Unknown"=(t[,"TRUE"]/rowSums(t))*100),digits=2)
  
  k = !is.na(genieClinData$age_days)
  plot(ggplot(genieClinData[k,], aes(x = (age_days/365), group = CENTER)) +
         geom_histogram(aes(y = (..count..)/width/tapply(..count..,..group..,sum)[..group..]),
                        breaks = c(0,seq(19,89,2),90,108),
                        closed = "left") +
         facet_wrap(~CENTER,ncol = 4) +
         ylab("Density") +
         xlab("Age") +
         ggtitle("Distribution of age across centers"))
}

```

```{r downloads, echo=FALSE, message=FALSE, results = 'hide'}
releaseFiles = as.list(synGetChildren('%s'))
this_samples <- getFileDf("data_clinical.txt", releaseFiles)
if (is.null(this_samples)) {
  this_samples <- getFileDf("data_clinical_sample.txt", releaseFiles)
  this_patient <- getFileDf("data_clinical_patient.txt", releaseFiles)
  this_patient$CENTER <- createCenterColumn(this_patient)
  this_samples$CENTER <- createCenterColumn(this_samples)
} else {
  this_patient <- this_samples
  this_patient <- this_patient[!duplicated(this_patient$PATIENT_ID),]
}
this_bed <- getFileDf("genomic_information.txt", releaseFiles)
if (is.null(this_bed)) {
  this_bed <- getFileDf("genie_combined.bed", releaseFiles)
}
this_assays = as.character(unique(this_samples$SEQ_ASSAY_ID))
this_mut <- getFileDf("data_mutations_extended.txt", releaseFiles)
assay_infodf = getFileDf("assay_information.txt", releaseFiles)
black_list_variants <- synTableQuery("select * from syn18459663",
                                     includeRowIdAndRowVersion = F)
black_list_variantsdf = black_list_variants$asDataFrame()
# this_cna <- getFileDf("data_CNA.txt", releaseFiles)
#this_fus <- getFileDf("data_fusions.txt", releaseFiles)
```

## Sample and Variant Count per center
Please confirm that the the number of samples and variants in this release is what you expect.

```{r sampleTable}
sampleCounts = table(this_samples$CENTER)
samplesPerReleaseDf = as.data.frame(sampleCounts)
colnames(samplesPerReleaseDf) = c("Center", "Samples")
variant_counts = table(this_mut$Center)
variant_countsdf = as.data.frame(variant_counts)
colnames(variant_countsdf) = c("Center", "Variants")
release_infodf = merge.data.frame(samplesPerReleaseDf, variant_countsdf, by="Center", all=T)
release_infodf[is.na(release_infodf)] = 0
samplesPerReleaseDf = rbind(release_infodf,
                            data.frame("Center" = "Total",
                                       "Samples" = sum(sampleCounts),
                                       "Variants" = sum(variant_counts)))
kable(samplesPerReleaseDf, row.names = F,
      caption = "Number of samples and variants per center")
```

## Blacklisted Variants
This is a count of how many blacklisted variants a center has. For instance, it is likely that the BRAF p.V600M mutations are likely false positives. If your center shows up in this table, please investigate your variants. If you have suggestions for variants that should be part of this list, please inform Sage Bionetworks.

```{r blacklist}
blacklist_variants = paste(black_list_variantsdf$Hugo_Symbol,
                           black_list_variantsdf$HGVSp_Short)
subset_mut = this_mut[this_mut$Hugo_Symbol %in% black_list_variantsdf$Hugo_Symbol,
                      c("Hugo_Symbol","Center","HGVSp_Short")]
subset_mut$blacklist = paste(subset_mut$Hugo_Symbol,
                             subset_mut$HGVSp_Short)
subset_mut = subset_mut[subset_mut$blacklist %in% blacklist_variants,]
kable(table(subset_mut$Center, subset_mut$blacklist),
      caption = "Blacklist variant count")
```


<!-- ```{r} -->
<!-- #primary site distribution -->
<!-- par(mar = c(10,3,3,1)) -->
<!-- # barplot(sort(log(table(this_mut$FILTER)),decreasing = T), -->
<!-- #         main = "Log Distribution of Mutation FILTERs", las = 2) -->
<!-- # plotPrimarySites(this_samples, -->
<!-- #                  "http://oncotree.mskcc.org/api/tumorTypes/tree?version=oncotree_2018_06_01", -->
<!-- #                  "%s") -->
<!-- ``` -->

## Distribution of Clinical Attributes
Please use the below distributions to access the validity of your clinical data.  There will be some tips for what to look for in each section.

### Race
Some european sites like GRCC, CRUK, and NKI do not collect race information from their patients.  If your institution does not either, you should see empty barplots.

```{r}
#Center X Race
par(mar = c(10,3,3,1))
plotCenterXRace(this_patient)
```

### Ethnicity
Some european sites like GRCC, CRUK, and NKI do not collect ethnicity information from their patients.  If your institution does not either, you should see empty barplots.

```{r}
#Center X Ethnicity
par(mar = c(10,3,3,1))
plotCenterXEthnicity(this_patient)
```

### Sex
CRUK submits samples from breast cancer patients, so their samples are all female.

```{r}
#Center X Sex
par(mar = c(10,3,3,1))
plotCenterXSex(this_patient)
```

### Age
CHOP submits mainly pediatric cases, so their distribution is different from the rest.


```{r}
#Center X Age
par(mar = c(10,3,3,1))
plotCenterXAge(this_samples)
```


<!-- ```{r cancer_type_plot, fig.width=14} -->
<!-- if (!is.null(this_samples$CANCER_TYPE)) { -->
<!--   #Samples, by cancer type and institution -->
<!--   plotCancerType(this_samples) -->
<!-- } -->
<!-- ``` -->

## Gene Panel Overlaps
This section will show the number of genes that overlap between the myeloid, small and large GENIE panels.

```{r bed_process, include=F}
if (!is.null(this_bed)) {
  this_bed <- this_bed[this_bed$Feature_Type == "exon",]
  #Make it so that I use include in panel
  if (!is.null(this_bed$includeInPanel)) {
    this_bed <- this_bed[this_bed$includeInPanel == "True",]
  }
  noneExistentAssays = this_assays[!this_assays %in% this_bed$SEQ_ASSAY_ID]
  if (length(noneExistentAssays) > 0) {
    print(paste("These assays do not have bed files associated with them: ", 
                paste(noneExistentAssays, collapse = ", ")))
  }
  this_assays = this_assays[this_assays %in% this_bed$SEQ_ASSAY_ID]
  myeloid_panels = c("VICC-01-MYELOID","UHN-54-V1","UCHI-ONCOHEME55-V1","CHOP-HEMEP","MSK-IMPACT-HEME-400")
  myeloid = this_assays[this_assays %in% myeloid_panels]

  normal = this_assays[!this_assays %in% myeloid_panels]
  smallPanels = c()
  largePanels = c()
  for (panel in normal) {
    panelDf = this_bed[this_bed$SEQ_ASSAY_ID == panel,]
    if (length(table(panelDf$Hugo_Symbol)) < 100) {
      smallPanels = c(smallPanels, panel)
      # Don't add to panel if more than 1500 genes
    } else if (length(table(panelDf$Hugo_Symbol)) < 1500) {
      largePanels = c(largePanels, panel)
    }
  }
} else {
  largePanels = c()
  smallPanels = c()
  myeloid = c()
}
```
  
### Meyloid Gene Panels

```{r myeloid_panel_upset}
if (length(myeloid) > 1) {
  plotPanelOverlap(this_samples, this_bed, myeloid)
}

```

### Small (<100) Gene panels

```{r small_panel_upset, fig.height=12}
if (length(smallPanels) > 1) {
  plotPanelOverlap(this_samples, this_bed, smallPanels)
}
```

### Large ($\geq$ 100) Gene panels

```{r large_panel_upset, fig.height=15}
if (length(largePanels) > 1) {
  plotPanelOverlap(this_samples, this_bed, largePanels)
}
```


## Top 5 most frequently mutated genes per pipeline

To assist with discovering artifacts in a center's data, this will summarize the top 5 most frequently mutated genes per pipline for all non-synonymous mutations.  All mutations that have `Variant_Classifications` of "Silent", "Intron", "3'UTR", "3'Flank", "5'UTR", "IGR" and "5'Flank" are filtered out with the exception of TERT promoters. One of the easiest questions to answer is:

**Has the frequently mutated gene list changed significantly? Is TP53 still #1?**

### All pipelines

Top 5 most frequently mutated genes per pipeline for all pipelines.

```{r, include=F}
# Filter maf file to only contain non synonymous variants.
# Follow same strategy here:
# https://github.com/cBioPortal/cbioportal/blob/v3.3.4/core/src/main/java/org/mskcc/cbio/portal/scripts/MutationFilter.java#L86
filter_maf <- function(mafdf) {
  filter_classifications = c("Silent", "Intron", "3'UTR",
                             "3'Flank", "5'UTR", "IGR")
  non_coding_mutations = mafdf$Variant_Classification %in% filter_classifications
  # Remove non coding
  mafdf <- mafdf[!non_coding_mutations, ]
  flank_classification = mafdf$Variant_Classification == "5'Flank"
  tert_mutations = mafdf$Hugo_Symbol == "TERT"
  # Remove 5'Flank unless it is TERT
  mafdf[!flank_classification | tert_mutations, ]
}
this_mut <- filter_maf(this_mut)

```


```{r top_5_mutated, echo=FALSE}
mergeddf = merge.data.frame(this_mut[,c("Tumor_Sample_Barcode", "Hugo_Symbol")],
                            this_samples[,c("SAMPLE_ID", "SEQ_ASSAY_ID")],
                            by.x = "Tumor_Sample_Barcode",
                            by.y = "SAMPLE_ID")
# TODO: can remove NA is fixed
na_pipeline = assay_infodf$SEQ_PIPELINE_ID == ""
assay_infodf$SEQ_PIPELINE_ID[na_pipeline] = assay_infodf$SEQ_ASSAY_ID[na_pipeline]

# Merge in assay information to group by seq pipeline id
mergeddf = merge.data.frame(mergeddf,
                            assay_infodf[,c("SEQ_ASSAY_ID", "SEQ_PIPELINE_ID")],
                            by = "SEQ_ASSAY_ID")

mutated_genes = table(mergeddf$Hugo_Symbol, mergeddf$SEQ_PIPELINE_ID)
# Top 5 mutated genes per panel
top_5_mutated_genes = apply(mutated_genes, 2, function(x) {
  list(sort(x, decreasing = T)[1:5])
})
top_5_mutated_genes_string = apply(mutated_genes, 2, function(x) {
  top_5 = sort(x, decreasing = T)[1:5]
  paste(names(top_5), top_5, sep = ": ")
})
transposed_matrix = t(top_5_mutated_genes_string)
final_matrix = cbind(transposed_matrix, table(mergeddf$SEQ_PIPELINE_ID)[row.names(transposed_matrix)])
colnames(final_matrix) = c(1, 2, 3, 4, 5, "Total Variants")
kable(final_matrix,
      caption = "Distribution of top 5 most frequently mutated genes per SEQ_PIPELINE_ID")

```


### Panels Covering TP53

These are all the pipelines that cover TP53 that don't have TP53 as the top mutated genes.  If your pipeline is shown in the table below, it is _possible_ that you have artifacts in your data.  **Please review and reupload if your data is incorrect**


```{r highlight_panels, echo=FALSE}
this_bed = merge.data.frame(this_bed,
                            assay_infodf[,c("SEQ_ASSAY_ID", "SEQ_PIPELINE_ID")],
                            by = "SEQ_ASSAY_ID")
if (!is.null(this_bed)) {
  panels_covering_tp53 = unique(this_bed$SEQ_PIPELINE_ID[this_bed$Hugo_Symbol == "TP53"])
  # Exclude panels if TP53 is top mutated
  tp53_not_top = sapply(names(top_5_mutated_genes), function(seq_assay) {
    if (names(top_5_mutated_genes[[seq_assay]][[1]][1]) != "TP53") {
      return(seq_assay)
    }
  })
  relevant_panels = mutated_genes[, colnames(mutated_genes) %in% panels_covering_tp53]
  relevant_panels = relevant_panels[, colnames(relevant_panels) %in% unlist(tp53_not_top)]
  top_5_mutated_genes_string = apply(relevant_panels, 2, function(x) {
    top_5 = sort(x, decreasing = T)[1:5]
    paste(names(top_5), top_5, sep = ": ")
  })
  transposed_matrix = t(top_5_mutated_genes_string)
  final_matrix = cbind(transposed_matrix, table(mergeddf$SEQ_PIPELINE_ID)[row.names(transposed_matrix)])
  colnames(final_matrix) = c(1, 2, 3, 4, 5, "Total Variants")
  kable(final_matrix,
        caption = "Distribution of top 5 most frequently mutated genes per SEQ_PIPELINE_ID")
}
```

## Assay information vs genomic information discrepancies

During processing, each panel's bed file undergoes gene symbol harmonization and is used to generate a gene panel file used in cBioPortal.
A user of the GENIE data pointed out that the expected gene count numbers in the data guide did not match up to the number of genes in the gene panel files made from each center's bed files.  The table below was created to assist with resolving these issues.

**Here are some action items required if your panel appears in the table**. The assay_information.yaml file is a file that I assisted all the centers to create, and can be found in each center’s input directory.

* `num_genes_remapped`:  number of unique genes a panel has after submitted genes are remapped
* `num_genes_submitted`: number of unique genes found in site's submitted BED file prior to remapping
* `num_genes_expected`: size of panel as defined in site's submitted assay_information.yaml file

1. If the `num_genes_expected` is NA, this means your panel is either missing from your assay_information.yaml, or the file is in an invalid format and cannot be processed.  Please provide a fix.
2. If the `num_genes_remapped` does not match `num_genes_expected`, please make sure one of the two values is correct.
   a. If `num_genes_expected` is incorrect, please update the value in your `assay_information.yaml` file.
   b. If `num_genes_remapped` is incorrect, please update your coordinates or gene symbols in the bed file.  Please view the bed submission guidelines to learn more about the remapping.
3. If there is a huge difference between your `num_genes_submitted` and `num_genes_remapped`, there is a high chance that you are submitting "invalid" gene symbols.  (ie.  PDGFRA1 is not equivalent to PDGFRA and PDGFRA1 is an invalid symbol name)


```{r genepanel_diff, echo=F}
if (!is.null(assay_infodf) & !is.null(this_bed)) {
  gene_panel_bed = this_bed[this_bed$Feature_Type == "exon" & this_bed$includeInPanel == "True" & this_bed$Hugo_Symbol != "", ]
  # Get count of symbols per panel
  symbol_count_per_panel = table(gene_panel_bed$Hugo_Symbol, gene_panel_bed$SEQ_ASSAY_ID)
  # Get whether or not a panel contains a certain symbol
  symbol_bool_per_panel = symbol_count_per_panel > 0
  # Sum of the columns gets number of unique genes per panel
  number_of_genes = data.frame(colSums(symbol_bool_per_panel))

  submitted_count = table(gene_panel_bed$ID, gene_panel_bed$SEQ_ASSAY_ID)
  submitted_bool_per_panel = submitted_count > 0
  num_submitted_genes = colSums(submitted_bool_per_panel)
  colnames(number_of_genes) = "num_genes_remapped"
  number_of_genes$num_genes_submitted = NA
  number_of_genes$num_expected_genes = NA
  number_of_genes[names(num_submitted_genes), "num_genes_submitted"] = num_submitted_genes
  number_of_genes[assay_infodf$SEQ_ASSAY_ID, "num_expected_genes"] = assay_infodf$number_of_genes

  different_from_expected = number_of_genes[c(which(number_of_genes$num_genes_remapped != number_of_genes$num_expected_genes),
                                              which(is.na(number_of_genes$num_expected_genes))),]
  kable(different_from_expected,
        caption = "Number of submitted vs expected genes in a bed file")
}
```
