
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="This repository documents code used to gather, QC, standardize, and analyze data uploaded by institutes participating in AACR's Project GENIE (Genomics, Evidence, Neoplasia, Information, Exchange).">
      
      
      
      
        <link rel="prev" href="../input_to_database/process_mutation/">
      
      
        <link rel="next" href="../consortium_to_public/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Database to Staging - AACR Project GENIE Repository Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="green" data-md-color-accent="green">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genie.database_to_staging" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="AACR Project GENIE Repository Documentation" class="md-header__button md-logo" aria-label="AACR Project GENIE Repository Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            AACR Project GENIE Repository Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Database to Staging
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="green" data-md-color-accent="green"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="green" data-md-color-accent="green"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Sage-Bionetworks/Genie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    AACR Project GENIE
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../getting_started/" class="md-tabs__link">
        
  
  
    
  
  Getting Started

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../tutorials/local_file_validation/" class="md-tabs__link">
          
  
  
    
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../main_pipeline_commands/input_to_database/" class="md-tabs__link">
          
  
  
    
  
  Reference

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="AACR Project GENIE Repository Documentation" class="md-nav__button md-logo" aria-label="AACR Project GENIE Repository Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    AACR Project GENIE Repository Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Sage-Bionetworks/Genie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    AACR Project GENIE
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../tutorials/local_file_validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local File Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Main Pipeline Commands
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Main Pipeline Commands
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/input_to_database/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Input to Database
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/database_to_staging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/consortium_to_public/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Consortium to Public
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Main Pipeline Steps
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Main Pipeline Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2_1" >
        
          
          <label class="md-nav__link" for="__nav_4_2_1" id="__nav_4_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Input to Database
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2_1">
            <span class="md-nav__icon md-icon"></span>
            Input to Database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input_to_database/input_to_database/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    input_to_database
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input_to_database/validate.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    validate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input_to_database/write_invalid_reasons/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    write_invalid_reasons
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input_to_database/process_mutation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process_mutation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input_to_database/toRetract.py" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    to_retract
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#genie.database_to_staging" class="md-nav__link">
    <span class="md-ellipsis">
      database_to_staging
    </span>
  </a>
  
    <nav class="md-nav" aria-label="database_to_staging">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging-attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Attributes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Attributes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.logger" class="md-nav__link">
    <span class="md-ellipsis">
      logger
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.GENIE_RELEASE_DIR" class="md-nav__link">
    <span class="md-ellipsis">
      GENIE_RELEASE_DIR
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.CASE_LIST_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      CASE_LIST_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.CNA_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      CNA_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.SAMPLE_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      SAMPLE_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.PATIENT_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      PATIENT_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.MUTATIONS_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      MUTATIONS_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.FUSIONS_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      FUSIONS_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.SEG_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      SEG_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.SV_CENTER_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      SV_CENTER_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.BED_DIFFS_SEQASSAY_PATH" class="md-nav__link">
    <span class="md-ellipsis">
      BED_DIFFS_SEQASSAY_PATH
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.FULL_MAF_RELEASE_COLUMNS" class="md-nav__link">
    <span class="md-ellipsis">
      FULL_MAF_RELEASE_COLUMNS
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Functions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging._to_redact_interval" class="md-nav__link">
    <span class="md-ellipsis">
      _to_redact_interval
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging._redact_year" class="md-nav__link">
    <span class="md-ellipsis">
      _redact_year
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging._redact_ped_year" class="md-nav__link">
    <span class="md-ellipsis">
      _redact_ped_year
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging._to_redact_difference" class="md-nav__link">
    <span class="md-ellipsis">
      _to_redact_difference
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.redact_phi" class="md-nav__link">
    <span class="md-ellipsis">
      redact_phi
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.remove_maf_samples" class="md-nav__link">
    <span class="md-ellipsis">
      remove_maf_samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.get_whitelist_variants_idx" class="md-nav__link">
    <span class="md-ellipsis">
      get_whitelist_variants_idx
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.configure_maf" class="md-nav__link">
    <span class="md-ellipsis">
      configure_maf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.calculate_missing_variant_counts" class="md-nav__link">
    <span class="md-ellipsis">
      calculate_missing_variant_counts
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.runMAFinBED" class="md-nav__link">
    <span class="md-ellipsis">
      runMAFinBED
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.get_run_maf_in_bed_script_cmd" class="md-nav__link">
    <span class="md-ellipsis">
      get_run_maf_in_bed_script_cmd
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_maf_in_bed_filtered_variants" class="md-nav__link">
    <span class="md-ellipsis">
      store_maf_in_bed_filtered_variants
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.seq_date_filter" class="md-nav__link">
    <span class="md-ellipsis">
      seq_date_filter
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.mutation_in_cis_filter" class="md-nav__link">
    <span class="md-ellipsis">
      mutation_in_cis_filter
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.get_mutation_in_cis_filter_script_cmd" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_in_cis_filter_script_cmd
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.get_mutation_in_cis_filtered_samples" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_in_cis_filtered_samples
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.get_mutation_in_cis_flagged_variants" class="md-nav__link">
    <span class="md-ellipsis">
      get_mutation_in_cis_flagged_variants
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_mutation_in_cis_files_to_staging" class="md-nav__link">
    <span class="md-ellipsis">
      store_mutation_in_cis_files_to_staging
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.seq_assay_id_filter" class="md-nav__link">
    <span class="md-ellipsis">
      seq_assay_id_filter
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.no_genepanel_filter" class="md-nav__link">
    <span class="md-ellipsis">
      no_genepanel_filter
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_gene_panel_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_gene_panel_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.filter_out_germline_variants" class="md-nav__link">
    <span class="md-ellipsis">
      filter_out_germline_variants
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_sv_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_sv_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.append_or_create_release_maf" class="md-nav__link">
    <span class="md-ellipsis">
      append_or_create_release_maf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_maf_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_maf_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.run_genie_filters" class="md-nav__link">
    <span class="md-ellipsis">
      run_genie_filters
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_assay_info_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_assay_info_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_clinical_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_clinical_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_cna_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_cna_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_seg_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_seg_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_data_gene_matrix" class="md-nav__link">
    <span class="md-ellipsis">
      store_data_gene_matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.store_bed_files" class="md-nav__link">
    <span class="md-ellipsis">
      store_bed_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.stagingToCbio" class="md-nav__link">
    <span class="md-ellipsis">
      stagingToCbio
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.revise_metadata_files" class="md-nav__link">
    <span class="md-ellipsis">
      revise_metadata_files
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.search_or_create_folder" class="md-nav__link">
    <span class="md-ellipsis">
      search_or_create_folder
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie.database_to_staging.create_link_version" class="md-nav__link">
    <span class="md-ellipsis">
      create_link_version
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../consortium_to_public/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Consortium to Public
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Fileformats
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Fileformats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/fileformat/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Fileformat
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/assay/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Assay Information
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/bed/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Bed
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/clinical/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Clinical
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/cna/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cna
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/maf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Maf
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/mutations_in_cis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Mutations In Cis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/patient_retraction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Patient retraction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/sample_retraction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sample retraction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/seg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Segmented
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/structural_variant/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Structural variant
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/vcf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Vcf
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../fileformats/workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Helper modules
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            Helper modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/create_case_lists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    create_case_lists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/dashboard_table_updater/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dashboard_table_updater
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/extract/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/load/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/process_functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process_functions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/transform/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    transform
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/Sage-Bionetworks/Genie/edit/master/docs/reference/main_pipeline_steps/database_to_staging.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


  <h1>Database to Staging</h1>

<div class="doc doc-object doc-module">



<h2 id="genie.database_to_staging" class="doc doc-heading">
            <code>genie.database_to_staging</code>


<a href="#genie.database_to_staging" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>Functions for releasing GENIE consortium releases</p>










  <div class="doc doc-children">





<h3 id="genie.database_to_staging-attributes">Attributes<a href="#genie.database_to_staging-attributes" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.logger" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.logger" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.GENIE_RELEASE_DIR" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">GENIE_RELEASE_DIR</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">expanduser</span><span class="p">(</span><span class="s1">&#39;~/.synapseCache&#39;</span><span class="p">),</span> <span class="s1">&#39;GENIE_release&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.GENIE_RELEASE_DIR" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.CASE_LIST_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">CASE_LIST_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;case_lists&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.CASE_LIST_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.CNA_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">CNA_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_CNA_</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.CNA_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.SAMPLE_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">SAMPLE_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_clinical_supp_sample_</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.SAMPLE_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.PATIENT_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">PATIENT_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_clinical_supp_patient_</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.PATIENT_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.MUTATIONS_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">MUTATIONS_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_mutations_extended_</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.MUTATIONS_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.FUSIONS_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">FUSIONS_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_fusions_</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.FUSIONS_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.SEG_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">SEG_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_cna_hg19_</span><span class="si">%s</span><span class="s1">.seg&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.SEG_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.SV_CENTER_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">SV_CENTER_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;data_sv_</span><span class="si">%s</span><span class="s1">.txt&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.SV_CENTER_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.BED_DIFFS_SEQASSAY_PATH" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">BED_DIFFS_SEQASSAY_PATH</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s1">&#39;diff_</span><span class="si">%s</span><span class="s1">.csv&#39;</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.BED_DIFFS_SEQASSAY_PATH" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<div class="doc doc-object doc-attribute">



<h4 id="genie.database_to_staging.FULL_MAF_RELEASE_COLUMNS" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">FULL_MAF_RELEASE_COLUMNS</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;Hugo_Symbol&#39;</span><span class="p">,</span> <span class="s1">&#39;Entrez_Gene_Id&#39;</span><span class="p">,</span> <span class="s1">&#39;Center&#39;</span><span class="p">,</span> <span class="s1">&#39;NCBI_Build&#39;</span><span class="p">,</span> <span class="s1">&#39;Chromosome&#39;</span><span class="p">,</span> <span class="s1">&#39;Start_Position&#39;</span><span class="p">,</span> <span class="s1">&#39;End_Position&#39;</span><span class="p">,</span> <span class="s1">&#39;Strand&#39;</span><span class="p">,</span> <span class="s1">&#39;Consequence&#39;</span><span class="p">,</span> <span class="s1">&#39;Variant_Classification&#39;</span><span class="p">,</span> <span class="s1">&#39;Variant_Type&#39;</span><span class="p">,</span> <span class="s1">&#39;Reference_Allele&#39;</span><span class="p">,</span> <span class="s1">&#39;Tumor_Seq_Allele1&#39;</span><span class="p">,</span> <span class="s1">&#39;Tumor_Seq_Allele2&#39;</span><span class="p">,</span> <span class="s1">&#39;dbSNP_RS&#39;</span><span class="p">,</span> <span class="s1">&#39;dbSNP_Val_Status&#39;</span><span class="p">,</span> <span class="s1">&#39;Tumor_Sample_Barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;Matched_Norm_Sample_Barcode&#39;</span><span class="p">,</span> <span class="s1">&#39;Match_Norm_Seq_Allele1&#39;</span><span class="p">,</span> <span class="s1">&#39;Match_Norm_Seq_Allele2&#39;</span><span class="p">,</span> <span class="s1">&#39;Tumor_Validation_Allele1&#39;</span><span class="p">,</span> <span class="s1">&#39;Tumor_Validation_Allele2&#39;</span><span class="p">,</span> <span class="s1">&#39;Match_Norm_Validation_Allele1&#39;</span><span class="p">,</span> <span class="s1">&#39;Match_Norm_Validation_Allele2&#39;</span><span class="p">,</span> <span class="s1">&#39;Verification_Status&#39;</span><span class="p">,</span> <span class="s1">&#39;Validation_Status&#39;</span><span class="p">,</span> <span class="s1">&#39;Mutation_Status&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequencing_Phase&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequence_Source&#39;</span><span class="p">,</span> <span class="s1">&#39;Validation_Method&#39;</span><span class="p">,</span> <span class="s1">&#39;Score&#39;</span><span class="p">,</span> <span class="s1">&#39;BAM_File&#39;</span><span class="p">,</span> <span class="s1">&#39;Sequencer&#39;</span><span class="p">,</span> <span class="s1">&#39;t_ref_count&#39;</span><span class="p">,</span> <span class="s1">&#39;t_alt_count&#39;</span><span class="p">,</span> <span class="s1">&#39;n_ref_count&#39;</span><span class="p">,</span> <span class="s1">&#39;n_alt_count&#39;</span><span class="p">,</span> <span class="s1">&#39;HGVSc&#39;</span><span class="p">,</span> <span class="s1">&#39;HGVSp&#39;</span><span class="p">,</span> <span class="s1">&#39;HGVSp_Short&#39;</span><span class="p">,</span> <span class="s1">&#39;Transcript_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;RefSeq&#39;</span><span class="p">,</span> <span class="s1">&#39;Protein_position&#39;</span><span class="p">,</span> <span class="s1">&#39;Codons&#39;</span><span class="p">,</span> <span class="s1">&#39;Exon_Number&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_AFR_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_AMR_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_ASJ_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_EAS_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_FIN_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_NFE_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_OTH_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;gnomAD_SAS_AF&#39;</span><span class="p">,</span> <span class="s1">&#39;FILTER&#39;</span><span class="p">,</span> <span class="s1">&#39;Polyphen_Prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;Polyphen_Score&#39;</span><span class="p">,</span> <span class="s1">&#39;SIFT_Prediction&#39;</span><span class="p">,</span> <span class="s1">&#39;SIFT_Score&#39;</span><span class="p">,</span> <span class="s1">&#39;SWISSPROT&#39;</span><span class="p">,</span> <span class="s1">&#39;n_depth&#39;</span><span class="p">,</span> <span class="s1">&#39;t_depth&#39;</span><span class="p">,</span> <span class="s1">&#39;Annotation_Status&#39;</span><span class="p">,</span> <span class="s1">&#39;mutationInCis_Flag&#39;</span><span class="p">]</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie.database_to_staging.FULL_MAF_RELEASE_COLUMNS" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>


<h3 id="genie.database_to_staging-functions">Functions<a href="#genie.database_to_staging-functions" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging._to_redact_interval" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_to_redact_interval</span><span class="p">(</span><span class="n">df_col</span><span class="p">)</span></code>

<a href="#genie.database_to_staging._to_redact_interval" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Determines year values that are "&lt;18" and interval values &gt;89 that need to be redacted
Returns bool because BIRTH_YEAR needs to be redacted as well based
on the results</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df_col</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Dataframe column/pandas.Series of an interval column</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>tuple</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: to redact boolean vector
   pandas.Series: to redact pediatric boolean vector</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><span title="typing.Tuple">Tuple</span>[<span title="pandas.Series">Series</span>, <span title="pandas.Series">Series</span>]</code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span>
<span class="normal">129</span>
<span class="normal">130</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_to_redact_interval</span><span class="p">(</span><span class="n">df_col</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">Tuple</span><span class="p">[</span><span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">]:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Determines year values that are &quot;&lt;18&quot; and interval values &gt;89 that need to be redacted</span>
<span class="sd">    Returns bool because BIRTH_YEAR needs to be redacted as well based</span>
<span class="sd">    on the results</span>

<span class="sd">    Args:</span>
<span class="sd">        df_col: Dataframe column/pandas.Series of an interval column</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: pandas.Series: to redact boolean vector</span>
<span class="sd">               pandas.Series: to redact pediatric boolean vector</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">phi_cutoff</span> <span class="o">=</span> <span class="mi">365</span> <span class="o">*</span> <span class="mi">89</span>
    <span class="c1"># Some centers pre-redact their values by adding &lt; or &gt;. These</span>
    <span class="c1"># must be redacted</span>
    <span class="n">contain_greaterthan</span> <span class="o">=</span> <span class="n">df_col</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">contain_lessthan</span> <span class="o">=</span> <span class="n">df_col</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;&lt;&quot;</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Add in errors=&#39;coerce&#39; to turn strings into NaN</span>
    <span class="n">col_int</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">df_col</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s2">&quot;coerce&quot;</span><span class="p">)</span>
    <span class="n">to_redact</span> <span class="o">=</span> <span class="p">(</span><span class="n">col_int</span> <span class="o">&gt;</span> <span class="n">phi_cutoff</span><span class="p">)</span> <span class="o">|</span> <span class="n">contain_greaterthan</span>
    <span class="n">to_redact_pediatric</span> <span class="o">=</span> <span class="n">contain_lessthan</span>
    <span class="k">return</span> <span class="n">to_redact</span><span class="p">,</span> <span class="n">to_redact_pediatric</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging._redact_year" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_redact_year</span><span class="p">(</span><span class="n">df_col</span><span class="p">)</span></code>

<a href="#genie.database_to_staging._redact_year" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Redacts year values that have &lt; or &gt;</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df_col</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Dataframe column/pandas.Series of a year column</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.Series">Series</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: Redacted series</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_redact_year</span><span class="p">(</span><span class="n">df_col</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Redacts year values that have &lt; or &gt;</span>

<span class="sd">    Args:</span>
<span class="sd">        df_col: Dataframe column/pandas.Series of a year column</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.Series: Redacted series</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">year</span> <span class="o">=</span> <span class="n">df_col</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">contain_greaterthan</span> <span class="o">=</span> <span class="n">year</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">contain_lessthan</span> <span class="o">=</span> <span class="n">year</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;&lt;&quot;</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_col</span><span class="p">[</span><span class="n">contain_greaterthan</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cannotReleaseHIPAA&quot;</span>
    <span class="n">df_col</span><span class="p">[</span><span class="n">contain_lessthan</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;withheld&quot;</span>
    <span class="k">return</span> <span class="n">df_col</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging._redact_ped_year" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_redact_ped_year</span><span class="p">(</span><span class="n">df_col</span><span class="p">)</span></code>

<a href="#genie.database_to_staging._redact_ped_year" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Redacts year values that have &lt;</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df_col</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Dataframe column/pandas.Series of a year column</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.Series">Series</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: Redacted series</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_redact_ped_year</span><span class="p">(</span><span class="n">df_col</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Redacts year values that have &lt;</span>

<span class="sd">    Args:</span>
<span class="sd">        df_col: Dataframe column/pandas.Series of a year column</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.Series: Redacted series</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">year</span> <span class="o">=</span> <span class="n">df_col</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="n">contain_lessthan</span> <span class="o">=</span> <span class="n">year</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s2">&quot;&lt;&quot;</span><span class="p">,</span> <span class="n">na</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">df_col</span><span class="p">[</span><span class="n">contain_lessthan</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;withheld&quot;</span>
    <span class="k">return</span> <span class="n">df_col</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging._to_redact_difference" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_to_redact_difference</span><span class="p">(</span><span class="n">df_col_year1</span><span class="p">,</span> <span class="n">df_col_year2</span><span class="p">)</span></code>

<a href="#genie.database_to_staging._to_redact_difference" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Determine if difference between year2 and year1 is &gt; 89</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df_col_year1</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Dataframe column/pandas.Series of a year column</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>df_col_year2</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Dataframe column/pandas.Series of a year column</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.Series">Series</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: to redact boolean vector</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_to_redact_difference</span><span class="p">(</span>
    <span class="n">df_col_year1</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">df_col_year2</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Determine if difference between year2 and year1 is &gt; 89</span>

<span class="sd">    Args:</span>
<span class="sd">        df_col_year1: Dataframe column/pandas.Series of a year column</span>
<span class="sd">        df_col_year2: Dataframe column/pandas.Series of a year column</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.Series: to redact boolean vector</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Add in errors=&#39;coerce&#39; to turn strings into NaN</span>
    <span class="n">year1</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">df_col_year1</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s2">&quot;coerce&quot;</span><span class="p">)</span>
    <span class="n">year2</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">to_numeric</span><span class="p">(</span><span class="n">df_col_year2</span><span class="p">,</span> <span class="n">errors</span><span class="o">=</span><span class="s2">&quot;coerce&quot;</span><span class="p">)</span>
    <span class="n">to_redact</span> <span class="o">=</span> <span class="n">year2</span> <span class="o">-</span> <span class="n">year1</span> <span class="o">&gt;</span> <span class="mi">89</span>
    <span class="k">return</span> <span class="n">to_redact</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.redact_phi" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">redact_phi</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">,</span> <span class="n">interval_cols_to_redact</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;AGE_AT_SEQ_REPORT&#39;</span><span class="p">,</span> <span class="s1">&#39;INT_CONTACT&#39;</span><span class="p">,</span> <span class="s1">&#39;INT_DOD&#39;</span><span class="p">])</span></code>

<a href="#genie.database_to_staging.redact_phi" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Redacts the PHI by re-annotating the clinical file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>merged clinical dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>interval_cols_to_redact</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>List of interval columns to redact.
                     Defaults to ['AGE_AT_SEQ_REPORT',
                                  'INT_CONTACT',
                                  'INT_DOD']</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="list">list</span></code>
                  </span>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>[&#39;AGE_AT_SEQ_REPORT&#39;, &#39;INT_CONTACT&#39;, &#39;INT_DOD&#39;]</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.DataFrame">DataFrame</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.DataFrame: Redacted clinical dataframe</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">redact_phi</span><span class="p">(</span>
    <span class="n">clinicaldf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">interval_cols_to_redact</span><span class="p">:</span> <span class="nb">list</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">,</span> <span class="s2">&quot;INT_CONTACT&quot;</span><span class="p">,</span> <span class="s2">&quot;INT_DOD&quot;</span><span class="p">],</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Redacts the PHI by re-annotating the clinical file</span>

<span class="sd">    Args:</span>
<span class="sd">        clinicaldf: merged clinical dataframe</span>
<span class="sd">        interval_cols_to_redact: List of interval columns to redact.</span>
<span class="sd">                                 Defaults to [&#39;AGE_AT_SEQ_REPORT&#39;,</span>
<span class="sd">                                              &#39;INT_CONTACT&#39;,</span>
<span class="sd">                                              &#39;INT_DOD&#39;]</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: Redacted clinical dataframe</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Moved to cannotReleaseHIPAA and withheld because the HIPAA</span>
    <span class="c1"># years would change every single year. (e.g. &lt;1926, &gt;1998 would be</span>
    <span class="c1"># inaccurate every year)</span>
    <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">interval_cols_to_redact</span><span class="p">:</span>
        <span class="n">to_redact</span><span class="p">,</span> <span class="n">to_redactpeds</span> <span class="o">=</span> <span class="n">_to_redact_interval</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">[</span><span class="n">col</span><span class="p">])</span>
        <span class="n">clinicaldf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">to_redact</span><span class="p">,</span> <span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cannotReleaseHIPAA&quot;</span>
        <span class="n">clinicaldf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">to_redact</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&gt;32485&quot;</span>
        <span class="n">clinicaldf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">to_redactpeds</span><span class="p">,</span> <span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;withheld&quot;</span>
    <span class="c1"># Redact BIRTH_YEAR values that have &lt; or &gt;</span>
    <span class="c1"># Birth year has to be done separately because it is not an interval</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_redact_year</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">])</span>
    <span class="n">to_redact</span> <span class="o">=</span> <span class="n">_to_redact_difference</span><span class="p">(</span>
        <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">],</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;YEAR_CONTACT&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">clinicaldf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">to_redact</span><span class="p">,</span> <span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cannotReleaseHIPAA&quot;</span>
    <span class="n">to_redact</span> <span class="o">=</span> <span class="n">_to_redact_difference</span><span class="p">(</span>
        <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">],</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;YEAR_DEATH&quot;</span><span class="p">]</span>
    <span class="p">)</span>
    <span class="n">clinicaldf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">to_redact</span><span class="p">,</span> <span class="s2">&quot;BIRTH_YEAR&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;cannotReleaseHIPAA&quot;</span>

    <span class="c1"># redact range year for pediatric data</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;YEAR_CONTACT&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_redact_ped_year</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;YEAR_CONTACT&quot;</span><span class="p">])</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;YEAR_DEATH&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">_redact_ped_year</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;YEAR_DEATH&quot;</span><span class="p">])</span>

    <span class="k">return</span> <span class="n">clinicaldf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.remove_maf_samples" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">remove_maf_samples</span><span class="p">(</span><span class="n">mafdf</span><span class="p">,</span> <span class="n">keep_samples</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.remove_maf_samples" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Remove samples from maf file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mafdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Maf dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="list">list</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.DataFrame">DataFrame</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Filtered maf dataframe</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span>
<span class="normal">248</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">remove_maf_samples</span><span class="p">(</span><span class="n">mafdf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">keep_samples</span><span class="p">:</span> <span class="nb">list</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Remove samples from maf file</span>

<span class="sd">    Args:</span>
<span class="sd">        mafdf: Maf dataframe</span>
<span class="sd">        keep_samples: Samples to keep</span>

<span class="sd">    Returns:</span>
<span class="sd">        Filtered maf dataframe</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">keep_maf</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_samples</span><span class="p">)</span>
    <span class="n">mafdf</span> <span class="o">=</span> <span class="n">mafdf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">keep_maf</span><span class="p">,]</span>
    <span class="k">return</span> <span class="n">mafdf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.get_whitelist_variants_idx" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_whitelist_variants_idx</span><span class="p">(</span><span class="n">mafdf</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.get_whitelist_variants_idx" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Get boolean vector for variants that are known somatic sites
This is to override the germline filter</p>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span>
<span class="normal">281</span>
<span class="normal">282</span>
<span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_whitelist_variants_idx</span><span class="p">(</span><span class="n">mafdf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Get boolean vector for variants that are known somatic sites</span>
<span class="sd">    This is to override the germline filter</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;End&quot;</span><span class="p">,</span> <span class="s2">&quot;Symbol&quot;</span><span class="p">]</span>
    <span class="n">whitelist</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
        <span class="s2">&quot;https://raw.githubusercontent.com/mskcc/vcf2maf/v1.6.19/data/known_somatic_sites.bed&quot;</span><span class="p">,</span>
        <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span>
        <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span>
        <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
        <span class="n">names</span><span class="o">=</span><span class="n">columns</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">rangesdf</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span>
        <span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span> <span class="s2">&quot;End_Position&quot;</span><span class="p">,</span> <span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span> <span class="s2">&quot;HGVSp_Short&quot;</span><span class="p">]</span>
    <span class="p">]</span>
    <span class="n">rangesdf</span> <span class="o">=</span> <span class="n">rangesdf</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">:</span> <span class="s2">&quot;Start&quot;</span><span class="p">,</span> <span class="s2">&quot;End_Position&quot;</span><span class="p">:</span> <span class="s2">&quot;End&quot;</span><span class="p">}</span>
    <span class="p">)</span>
    <span class="n">maf_ranges</span> <span class="o">=</span> <span class="n">pyranges</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">rangesdf</span><span class="p">)</span>
    <span class="n">whitelist_ranges</span> <span class="o">=</span> <span class="n">pyranges</span><span class="o">.</span><span class="n">PyRanges</span><span class="p">(</span><span class="n">whitelist</span><span class="p">)</span>
    <span class="n">whitelisted_variants</span> <span class="o">=</span> <span class="n">maf_ranges</span><span class="o">.</span><span class="n">intersect</span><span class="p">(</span><span class="n">whitelist_ranges</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;containment&quot;</span><span class="p">)</span>
    <span class="n">whitelist_variantsdf</span> <span class="o">=</span> <span class="n">whitelisted_variants</span><span class="o">.</span><span class="n">as_df</span><span class="p">()</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">whitelist_variantsdf</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">whitelist_variantsdf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>
            <span class="o">+</span> <span class="s2">&quot; &quot;</span>
            <span class="o">+</span> <span class="n">whitelist_variantsdf</span><span class="p">[</span><span class="s2">&quot;HGVSp_Short&quot;</span><span class="p">]</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">variants</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">maf_variants</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; &quot;</span> <span class="o">+</span> <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;HGVSp_Short&quot;</span><span class="p">]</span>
    <span class="c1"># For some reason intersect and overlap doesn&#39;t work when</span>
    <span class="c1"># Start and End are the same. Here is an example that won&#39;t be</span>
    <span class="c1"># matched by the intersect function</span>
    <span class="c1"># variant: chr9-10-10</span>
    <span class="c1"># Bed: chr9-9-10</span>
    <span class="n">match_start_end</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">whitelist</span><span class="p">[</span><span class="s2">&quot;Start&quot;</span><span class="p">])</span> <span class="o">|</span> <span class="n">mafdf</span><span class="p">[</span>
        <span class="s2">&quot;End_Position&quot;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">whitelist</span><span class="p">[</span><span class="s2">&quot;End&quot;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">maf_variants</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">variants</span><span class="p">)</span> <span class="o">|</span> <span class="n">match_start_end</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.configure_maf" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">configure_maf</span><span class="p">(</span><span class="n">mafdf</span><span class="p">,</span> <span class="n">remove_variants</span><span class="p">,</span> <span class="n">flagged_variants</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.configure_maf" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Configures each maf dataframe, does germline filtering</p>
<p>Germline filtering for MAF files uses the gnomAD columns that refer to the
allele frequencies (AF) of variants in different population groups
from the gnomAD (Genome Aggregation Database). This filter will filter out
variants with a maximum AF &gt; 0.05% across all populations which are typically
common germline variants.</p>
<p>Germline filtering for MAF files occurs during release instead of during processing
because the MAF file gets re-annotated during processing via genome nexus annotation.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>mafdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Maf dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>remove_variants</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Variants to remove</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>flagged_variants</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Variants to flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>configured maf row</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span>
<span class="normal">303</span>
<span class="normal">304</span>
<span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span>
<span class="normal">351</span>
<span class="normal">352</span>
<span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span>
<span class="normal">394</span>
<span class="normal">395</span>
<span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">configure_maf</span><span class="p">(</span><span class="n">mafdf</span><span class="p">,</span> <span class="n">remove_variants</span><span class="p">,</span> <span class="n">flagged_variants</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Configures each maf dataframe, does germline filtering</span>

<span class="sd">    Germline filtering for MAF files uses the gnomAD columns that refer to the</span>
<span class="sd">    allele frequencies (AF) of variants in different population groups</span>
<span class="sd">    from the gnomAD (Genome Aggregation Database). This filter will filter out</span>
<span class="sd">    variants with a maximum AF &gt; 0.05% across all populations which are typically</span>
<span class="sd">    common germline variants.</span>

<span class="sd">    Germline filtering for MAF files occurs during release instead of during processing</span>
<span class="sd">    because the MAF file gets re-annotated during processing via genome nexus annotation.</span>

<span class="sd">    Args:</span>
<span class="sd">        mafdf: Maf dataframe</span>
<span class="sd">        remove_variants: Variants to remove</span>
<span class="sd">        flagged_variants: Variants to flag</span>

<span class="sd">    Returns:</span>
<span class="sd">        configured maf row</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">variant</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;End_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Reference_Allele&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Tumor_Seq_Allele2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">str</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
    <span class="n">mergecheck_variant</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span>
        <span class="p">[</span>
            <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;HGVSp_Short&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Reference_Allele&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Tumor_Seq_Allele2&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">,</span>
        <span class="p">]</span>
    <span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">x</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="nb">str</span><span class="p">)),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

    <span class="c1"># Flag mutation in cis variants</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;mutationInCis_Flag&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">mergecheck_variant</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">flagged_variants</span><span class="p">)</span>
    <span class="c1"># Remove common variants</span>
    <span class="c1"># na=False to resolve this linked error</span>
    <span class="c1"># https://stackoverflow.com/questions/52297740</span>
    <span class="c1"># common_variants = mafdf[&#39;FILTER&#39;].astype(str).str.contains(</span>
    <span class="c1">#     &quot;common_variant&quot;, na=False</span>
    <span class="c1"># )</span>
    <span class="c1"># Germline Filter</span>
    <span class="n">gnomad_cols</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;gnomAD_AFR_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_AMR_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_ASJ_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_EAS_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_FIN_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_NFE_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_OTH_AF&quot;</span><span class="p">,</span>
        <span class="s2">&quot;gnomAD_SAS_AF&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="c1"># location of germline variants</span>
    <span class="n">common_variants_idx</span> <span class="o">=</span> <span class="n">mafdf</span><span class="o">.</span><span class="n">loc</span><span class="p">[:,</span> <span class="n">gnomad_cols</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">(</span><span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">skipna</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mf">0.0005</span>

    <span class="c1"># Remove specific variants</span>
    <span class="n">to_remove_variants</span> <span class="o">=</span> <span class="n">variant</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">remove_variants</span><span class="p">)</span>
    <span class="c1"># Genome Nexus successfully annotated (vcf2maf does not have this column)</span>
    <span class="k">if</span> <span class="n">mafdf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;Annotation_Status&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;Annotation_Status&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;SUCCESS&quot;</span>
    <span class="c1"># Make sure to only get variants that were successfully annotated</span>
    <span class="n">success</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;Annotation_Status&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;SUCCESS&quot;</span>
    <span class="c1"># Get whitelisted variants</span>
    <span class="n">whitelist_variants_idx</span> <span class="o">=</span> <span class="n">get_whitelist_variants_idx</span><span class="p">(</span><span class="n">mafdf</span><span class="p">)</span>
    <span class="n">mafdf</span> <span class="o">=</span> <span class="n">mafdf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span>
        <span class="p">(</span>
            <span class="p">(</span><span class="o">~</span><span class="n">common_variants_idx</span> <span class="o">|</span> <span class="n">whitelist_variants_idx</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="o">~</span><span class="n">to_remove_variants</span>
            <span class="o">&amp;</span> <span class="n">success</span>
        <span class="p">),</span>
    <span class="p">]</span>
    <span class="c1"># May not need to do this because these columns are always</span>
    <span class="c1"># returned as numerical values now</span>
    <span class="c1"># fillnas = [&#39;t_depth&#39;, &#39;t_ref_count&#39;, &#39;t_alt_count&#39;,</span>
    <span class="c1">#            &#39;n_depth&#39;, &#39;n_ref_count&#39;, &#39;n_alt_count&#39;]</span>
    <span class="c1"># for col in fillnas:</span>
    <span class="c1">#     mafdf[col][mafdf[col].astype(str) == &quot;.&quot;] = float(&#39;nan&#39;)</span>
    <span class="n">n_depth_ind</span> <span class="o">=</span> <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_depth&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="s2">&quot;NA&quot;</span><span class="p">,</span> <span class="s2">&quot;0.0&quot;</span><span class="p">,</span> <span class="s2">&quot;0&quot;</span><span class="p">,</span> <span class="s2">&quot;nan&quot;</span><span class="p">])</span>
    <span class="n">mafdf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">n_depth_ind</span><span class="p">,</span> <span class="s2">&quot;Match_Norm_Seq_Allele2&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">mafdf</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">n_depth_ind</span><span class="p">,</span> <span class="s2">&quot;Match_Norm_Seq_Allele1&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="c1"># Calculate missing t_depth, t_ref_count, t_alt_count</span>
    <span class="n">t_counts</span> <span class="o">=</span> <span class="n">calculate_missing_variant_counts</span><span class="p">(</span>
        <span class="n">depth</span><span class="o">=</span><span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;t_depth&quot;</span><span class="p">],</span>
        <span class="n">alt_count</span><span class="o">=</span><span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;t_alt_count&quot;</span><span class="p">],</span>
        <span class="n">ref_count</span><span class="o">=</span><span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;t_ref_count&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;t_depth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">t_counts</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;t_ref_count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">t_counts</span><span class="p">[</span><span class="s2">&quot;ref_count&quot;</span><span class="p">]</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;t_alt_count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">t_counts</span><span class="p">[</span><span class="s2">&quot;alt_count&quot;</span><span class="p">]</span>
    <span class="c1"># Calculate missing n_depth, n_ref_count, n_alt_count</span>
    <span class="n">n_counts</span> <span class="o">=</span> <span class="n">calculate_missing_variant_counts</span><span class="p">(</span>
        <span class="n">depth</span><span class="o">=</span><span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_depth&quot;</span><span class="p">],</span>
        <span class="n">alt_count</span><span class="o">=</span><span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_alt_count&quot;</span><span class="p">],</span>
        <span class="n">ref_count</span><span class="o">=</span><span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_ref_count&quot;</span><span class="p">],</span>
    <span class="p">)</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_depth&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">n_counts</span><span class="p">[</span><span class="s2">&quot;depth&quot;</span><span class="p">]</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_ref_count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">n_counts</span><span class="p">[</span><span class="s2">&quot;ref_count&quot;</span><span class="p">]</span>
    <span class="n">mafdf</span><span class="p">[</span><span class="s2">&quot;n_alt_count&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">n_counts</span><span class="p">[</span><span class="s2">&quot;alt_count&quot;</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">mafdf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.calculate_missing_variant_counts" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">calculate_missing_variant_counts</span><span class="p">(</span><span class="n">depth</span><span class="p">,</span> <span class="n">alt_count</span><span class="p">,</span> <span class="n">ref_count</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.calculate_missing_variant_counts" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Calculate missing counts. t_depth = t_alt_count + t_ref_count</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>depth</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Allele Depth</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>alt_count</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Allele alt counts</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>ref_count</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Allele ref counts</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.Series">Series</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="dict">dict</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>filled in depth, alt_count and ref_count values</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">calculate_missing_variant_counts</span><span class="p">(</span>
    <span class="n">depth</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">alt_count</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">,</span> <span class="n">ref_count</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Calculate missing counts. t_depth = t_alt_count + t_ref_count</span>

<span class="sd">    Args:</span>
<span class="sd">        depth: Allele Depth</span>
<span class="sd">        alt_count: Allele alt counts</span>
<span class="sd">        ref_count: Allele ref counts</span>

<span class="sd">    Returns:</span>
<span class="sd">        filled in depth, alt_count and ref_count values</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Avoid SettingWithCopyWarning</span>
    <span class="n">depth</span> <span class="o">=</span> <span class="n">depth</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">alt_count</span> <span class="o">=</span> <span class="n">alt_count</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="n">ref_count</span> <span class="o">=</span> <span class="n">ref_count</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
    <span class="c1"># t_depth = t_ref_count + t_alt_count</span>
    <span class="n">null_depth</span> <span class="o">=</span> <span class="n">depth</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span>
    <span class="c1"># The notation null_depth_ref means all the reference values for which</span>
    <span class="c1"># depth is NA</span>
    <span class="n">null_depth_ref</span> <span class="o">=</span> <span class="n">ref_count</span><span class="p">[</span><span class="n">null_depth</span><span class="p">]</span>
    <span class="n">null_depth_alt</span> <span class="o">=</span> <span class="n">alt_count</span><span class="p">[</span><span class="n">null_depth</span><span class="p">]</span>
    <span class="n">depth</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">null_depth</span><span class="p">]</span> <span class="o">=</span> <span class="n">null_depth_ref</span> <span class="o">+</span> <span class="n">null_depth_alt</span>
    <span class="c1"># t_ref_count = t_depth - t_alt_count</span>
    <span class="n">null_ref</span> <span class="o">=</span> <span class="n">ref_count</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span>
    <span class="n">null_ref_depth</span> <span class="o">=</span> <span class="n">depth</span><span class="p">[</span><span class="n">null_ref</span><span class="p">]</span>
    <span class="n">null_ref_alt</span> <span class="o">=</span> <span class="n">alt_count</span><span class="p">[</span><span class="n">null_ref</span><span class="p">]</span>
    <span class="n">ref_count</span><span class="p">[</span><span class="n">null_ref</span><span class="p">]</span> <span class="o">=</span> <span class="n">null_ref_depth</span> <span class="o">-</span> <span class="n">null_ref_alt</span>
    <span class="c1"># t_alt_count = t_depth - t_ref_count</span>
    <span class="n">null_alt</span> <span class="o">=</span> <span class="n">alt_count</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span>
    <span class="n">null_alt_depth</span> <span class="o">=</span> <span class="n">depth</span><span class="p">[</span><span class="n">null_alt</span><span class="p">]</span>
    <span class="n">null_alt_ref</span> <span class="o">=</span> <span class="n">ref_count</span><span class="p">[</span><span class="n">null_alt</span><span class="p">]</span>
    <span class="n">alt_count</span><span class="p">[</span><span class="n">null_alt</span><span class="p">]</span> <span class="o">=</span> <span class="n">null_alt_depth</span> <span class="o">-</span> <span class="n">null_alt_ref</span>
    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;depth&quot;</span><span class="p">:</span> <span class="n">depth</span><span class="p">,</span> <span class="s2">&quot;ref_count&quot;</span><span class="p">:</span> <span class="n">ref_count</span><span class="p">,</span> <span class="s2">&quot;alt_count&quot;</span><span class="p">:</span> <span class="n">alt_count</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.runMAFinBED" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">runMAFinBED</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">staging</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">genieVersion</span><span class="o">=</span><span class="s1">&#39;test&#39;</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.runMAFinBED" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Run MAF in BED script, filter data and update MAFinBED files</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse client connection</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>center mapping dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>test</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Testing parameter. Defaults to False.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging parameter. Defaults to False.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genieVersion</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version. Defaults to "test".</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>&#39;test&#39;</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.Series">Series</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.Series: Variants to remove</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">runMAFinBED</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">test</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">staging</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="n">genieVersion</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;test&quot;</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run MAF in BED script, filter data and update MAFinBED files</span>

<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): Synapse client connection</span>
<span class="sd">        center_mappingdf (pd.DataFrame): center mapping dataframe</span>
<span class="sd">        test (bool, optional):Testing parameter. Defaults to False.</span>
<span class="sd">        staging (bool, optional): Staging parameter. Defaults to False.</span>
<span class="sd">        genieVersion (str, optional): GENIE version. Defaults to &quot;test&quot;.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: Variants to remove</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">script_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span>
    <span class="c1"># TODO: Use tempfile</span>
    <span class="n">notinbed_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">script_dir</span><span class="p">,</span> <span class="s2">&quot;../R/notinbed.csv&quot;</span><span class="p">)</span>
    <span class="n">command</span> <span class="o">=</span> <span class="n">get_run_maf_in_bed_script_cmd</span><span class="p">(</span>
        <span class="n">notinbed_file</span><span class="o">=</span><span class="n">notinbed_file</span><span class="p">,</span> <span class="n">script_dir</span><span class="o">=</span><span class="n">script_dir</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">staging</span><span class="o">=</span><span class="n">staging</span>
    <span class="p">)</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>

    <span class="c1"># mutationSynId = databaseSynIdMappingDf[&#39;Id&#39;][</span>
    <span class="c1">#     databaseSynIdMappingDf[&#39;Database&#39;] == &quot;vcf2maf&quot;][0]</span>
    <span class="c1"># removedVariants = syn.tableQuery(</span>
    <span class="c1">#     &quot;select Chromosome, Start_Position, End_Position, Reference_Allele, &quot;</span>
    <span class="c1">#     &quot;Tumor_Seq_Allele2, Tumor_Sample_Barcode, Center from {} where inBED&quot;</span>
    <span class="c1">#     &quot; is False and Center in (&#39;{}&#39;)&quot;.format(</span>
    <span class="c1">#         mutationSynId, &quot;&#39;,&#39;&quot;.join(center_mappingdf.center)))</span>
    <span class="c1"># removedVariantsDf = removedVariants.asDataFrame()</span>
    <span class="n">removed_variantsdf</span> <span class="o">=</span> <span class="n">store_maf_in_bed_filtered_variants</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">notinbed_file</span><span class="o">=</span><span class="n">notinbed_file</span><span class="p">,</span>
        <span class="n">center_mapping_df</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">,</span>
        <span class="n">genie_version</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;removeVariants&quot;</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.get_run_maf_in_bed_script_cmd" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_run_maf_in_bed_script_cmd</span><span class="p">(</span><span class="n">notinbed_file</span><span class="p">,</span> <span class="n">script_dir</span><span class="p">,</span> <span class="n">test</span><span class="p">,</span> <span class="n">staging</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.get_run_maf_in_bed_script_cmd" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>This function gets the MAFinBED R script command call based on
   whether we are running in test, staging or production mode</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>notinbed_file</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Full path to the notinbed csv</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>script_dir</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>directory of the MAFinBED script</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>test</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Testing parameter.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging parameter</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Full command call for the MAFinBED script</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_run_maf_in_bed_script_cmd</span><span class="p">(</span>
    <span class="n">notinbed_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">script_dir</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">test</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">staging</span><span class="p">:</span> <span class="nb">bool</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;This function gets the MAFinBED R script command call based on</span>
<span class="sd">       whether we are running in test, staging or production mode</span>

<span class="sd">    Args:</span>
<span class="sd">        notinbed_file (str): Full path to the notinbed csv</span>
<span class="sd">        script_dir (str): directory of the MAFinBED script</span>
<span class="sd">        test (bool): Testing parameter.</span>
<span class="sd">        staging (bool): Staging parameter</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Full command call for the MAFinBED script</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mafinbed_script</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">script_dir</span><span class="p">,</span> <span class="s2">&quot;../R/MAFinBED.R&quot;</span><span class="p">)</span>
    <span class="c1"># The MAFinBED script filters out the centers that aren&#39;t being processed</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Rscript&quot;</span><span class="p">,</span> <span class="n">mafinbed_script</span><span class="p">,</span> <span class="n">notinbed_file</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">test</span><span class="p">:</span>
        <span class="n">command</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--testing&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">staging</span><span class="p">:</span>
        <span class="n">command</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--staging&quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">return</span> <span class="n">command</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_maf_in_bed_filtered_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_maf_in_bed_filtered_variants</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">notinbed_file</span><span class="p">,</span> <span class="n">center_mapping_df</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_maf_in_bed_filtered_variants" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>This script retrieves and stores the filtered variants generated
    from running the MAFinBed script as files on Synapse. This script
    then returns the not in bed file with the removed variants column
    added.
    TODO: Add handling for empty file?</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse client connection</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>notinbed_file</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>input file</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mapping_df</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>the center mapping dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>version of this genie run</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.DataFrame">DataFrame</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.DataFrame: filtered variants dataframe with
removed variants column added</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_maf_in_bed_filtered_variants</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span>
    <span class="n">notinbed_file</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">center_mapping_df</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;This script retrieves and stores the filtered variants generated</span>
<span class="sd">        from running the MAFinBed script as files on Synapse. This script</span>
<span class="sd">        then returns the not in bed file with the removed variants column</span>
<span class="sd">        added.</span>
<span class="sd">        TODO: Add handling for empty file?</span>

<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): Synapse client connection</span>
<span class="sd">        notinbed_file (str): input file</span>
<span class="sd">        center_mapping_df (pd.DataFrame): the center mapping dataframe</span>
<span class="sd">        genie_version (str): version of this genie run</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: filtered variants dataframe with</span>
<span class="sd">            removed variants column added</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">removed_variantsdf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">notinbed_file</span><span class="p">)</span>
    <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;removeVariants&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Reference_Allele&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Tumor_Seq_Allele2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Store filtered variants</span>
    <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Center&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="n">center_mutation</span> <span class="o">=</span> <span class="n">removed_variantsdf</span><span class="p">[</span><span class="n">removed_variantsdf</span><span class="p">[</span><span class="s2">&quot;Center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span><span class="p">]</span>
        <span class="c1"># mafText = process.removePandasDfFloat(center_mutation)</span>
        <span class="n">center_mutation</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;mafinbed_filtered_variants.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
            <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
            <span class="n">filepath</span><span class="o">=</span><span class="s2">&quot;mafinbed_filtered_variants.csv&quot;</span><span class="p">,</span>
            <span class="n">parentid</span><span class="o">=</span><span class="n">center_mapping_df</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                <span class="n">center_mapping_df</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
            <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
            <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">os</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="s2">&quot;mafinbed_filtered_variants.csv&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">removed_variantsdf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.seq_date_filter" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">seq_date_filter</span><span class="p">(</span><span class="n">clinicalDf</span><span class="p">,</span> <span class="n">processingDate</span><span class="p">,</span> <span class="n">consortiumReleaseCutOff</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.seq_date_filter" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Filter samples by seq date</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clinicalDf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>processingDate</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Processing date in form of Apr-XXXX</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>consortiumReleaseCutOff</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Release cut off days</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>list</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Samples to remove</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">seq_date_filter</span><span class="p">(</span><span class="n">clinicalDf</span><span class="p">,</span> <span class="n">processingDate</span><span class="p">,</span> <span class="n">consortiumReleaseCutOff</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Filter samples by seq date</span>

<span class="sd">    Args:</span>
<span class="sd">        clinicalDf: Clinical dataframe</span>
<span class="sd">        processingDate: Processing date in form of Apr-XXXX</span>
<span class="sd">        consortiumReleaseCutOff: Release cut off days</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: Samples to remove</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">removeSeqDateSamples</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">seqDateFilter</span><span class="p">(</span>
        <span class="n">clinicalDf</span><span class="p">,</span> <span class="n">processingDate</span><span class="p">,</span> <span class="n">consortiumReleaseCutOff</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">removeSeqDateSamples</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.mutation_in_cis_filter" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">mutation_in_cis_filter</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">skipMutationsInCis</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">genieVersion</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">staging</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.mutation_in_cis_filter" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Run mutation in cis filter, look up samples to remove.</p>
<p>The mutation in cis script ONLY runs
WHEN the skipMutationsInCis parameter is FALSE</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>skipMutationsInCis</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Skip this filter</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_filtering_synId</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>mergeCheck database dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genieVersion</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version. Default is test.</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>test</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Testing parameter. Default is False.</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging parameter. Default is False.</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.Series: Samples to remove</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">mutation_in_cis_filter</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">skipMutationsInCis</span><span class="p">,</span>
    <span class="n">variant_filtering_synId</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">genieVersion</span><span class="p">,</span>
    <span class="n">test</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">staging</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run mutation in cis filter, look up samples to remove.</span>

<span class="sd">    The mutation in cis script ONLY runs</span>
<span class="sd">    WHEN the skipMutationsInCis parameter is FALSE</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        skipMutationsInCis: Skip this filter</span>
<span class="sd">        variant_filtering_synId: mergeCheck database dataframe</span>
<span class="sd">        center_mappingdf: center mapping dataframe</span>
<span class="sd">        genieVersion: GENIE version. Default is test.</span>
<span class="sd">        test: Testing parameter. Default is False.</span>
<span class="sd">        staging: Staging parameter. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: Samples to remove</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">skipMutationsInCis</span><span class="p">:</span>
        <span class="n">command</span> <span class="o">=</span> <span class="n">get_mutation_in_cis_filter_script_cmd</span><span class="p">(</span><span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span> <span class="n">staging</span><span class="o">=</span><span class="n">staging</span><span class="p">)</span>
        <span class="c1"># TODO: use subprocess.run instead</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="n">command</span><span class="p">)</span>
        <span class="n">store_mutation_in_cis_files_to_staging</span><span class="p">(</span>
            <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
            <span class="n">center_mappingdf</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">,</span>
            <span class="n">variant_filtering_synId</span><span class="o">=</span><span class="n">variant_filtering_synId</span><span class="p">,</span>
            <span class="n">genieVersion</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">,</span>
        <span class="p">)</span>
    <span class="n">remove_samples</span> <span class="o">=</span> <span class="n">get_mutation_in_cis_filtered_samples</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="o">=</span><span class="n">variant_filtering_synId</span>
    <span class="p">)</span>
    <span class="n">flagged_variants</span> <span class="o">=</span> <span class="n">get_mutation_in_cis_flagged_variants</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="o">=</span><span class="n">variant_filtering_synId</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">remove_samples</span><span class="p">,</span> <span class="n">flagged_variants</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.get_mutation_in_cis_filter_script_cmd" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_mutation_in_cis_filter_script_cmd</span><span class="p">(</span><span class="n">test</span><span class="p">,</span> <span class="n">staging</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.get_mutation_in_cis_filter_script_cmd" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>This function gets the mutation_in_cis_filter R script
    command call based on whether we are running in test, staging
    or production mode</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>test</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Testing parameter.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging parameter. Default is False.</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Full command call for the mergeCheck script</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_mutation_in_cis_filter_script_cmd</span><span class="p">(</span><span class="n">test</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span> <span class="n">staging</span><span class="p">:</span> <span class="nb">bool</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;This function gets the mutation_in_cis_filter R script</span>
<span class="sd">        command call based on whether we are running in test, staging</span>
<span class="sd">        or production mode</span>

<span class="sd">    Args:</span>
<span class="sd">        test (bool): Testing parameter.</span>
<span class="sd">        staging (bool): Staging parameter. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Full command call for the mergeCheck script</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mergeCheck_script</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">)),</span> <span class="s2">&quot;../R/mergeCheck.R&quot;</span>
    <span class="p">)</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Rscript&quot;</span><span class="p">,</span> <span class="n">mergeCheck_script</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">test</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">staging</span><span class="p">:</span>
        <span class="n">command</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--testing&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">staging</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">test</span><span class="p">:</span>
        <span class="n">command</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="s2">&quot;--staging&quot;</span><span class="p">)</span>
    <span class="k">elif</span> <span class="n">test</span> <span class="ow">and</span> <span class="n">staging</span><span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
            <span class="s2">&quot;Mutation in cis only available in staging or testing mode not both&quot;</span>
        <span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">return</span> <span class="n">command</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.get_mutation_in_cis_filtered_samples" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_mutation_in_cis_filtered_samples</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.get_mutation_in_cis_filtered_samples" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Gets the samples to remove in our variant filtering table
   TODO: Add handling for when we have 0 row query results
Args:
    syn (synapseclient.Synapse): synapse client connection
    variant_filtering_synId (str): variant filtering synapse id</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="list">list</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.Series: removed samples</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_mutation_in_cis_filtered_samples</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">list</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gets the samples to remove in our variant filtering table</span>
<span class="sd">       TODO: Add handling for when we have 0 row query results</span>
<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): synapse client connection</span>
<span class="sd">        variant_filtering_synId (str): variant filtering synapse id</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: removed samples</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">query_str</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;SELECT Tumor_Sample_Barcode FROM </span><span class="si">{</span><span class="n">variant_filtering_synId</span><span class="si">}</span><span class="s2"> where &quot;</span>
        <span class="s2">&quot;Flag = &#39;TOSS&#39; and Tumor_Sample_Barcode is not null&quot;</span>
    <span class="p">)</span>
    <span class="n">filtered_samplesdf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span><span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span> <span class="n">query_string</span><span class="o">=</span><span class="n">query_str</span><span class="p">)</span>
    <span class="c1"># #Alex script #1 removed patients</span>
    <span class="n">remove_samples</span> <span class="o">=</span> <span class="n">filtered_samplesdf</span><span class="p">[</span><span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">remove_samples</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.get_mutation_in_cis_flagged_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">get_mutation_in_cis_flagged_variants</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.get_mutation_in_cis_flagged_variants" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Gets the flagged variants in our variant filtering table which is
   a unique string concatenation of the Chromosome, Start_Position,
   HGVSp_Short, Reference_Allele, Tumor_Seq_Allele2 and Tumor_Sample_Barcode
   columns
   TODO: Add handling for when we have 0 row query results</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>synapse client connection</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_filtering_synId</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>variant filtering synapse id</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.Series">Series</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.Series: flagged variants</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span>
<span class="normal">719</span>
<span class="normal">720</span>
<span class="normal">721</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">get_mutation_in_cis_flagged_variants</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Gets the flagged variants in our variant filtering table which is</span>
<span class="sd">       a unique string concatenation of the Chromosome, Start_Position,</span>
<span class="sd">       HGVSp_Short, Reference_Allele, Tumor_Seq_Allele2 and Tumor_Sample_Barcode</span>
<span class="sd">       columns</span>
<span class="sd">       TODO: Add handling for when we have 0 row query results</span>

<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): synapse client connection</span>
<span class="sd">        variant_filtering_synId (str): variant filtering synapse id</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: flagged variants</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">query_str</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;SELECT * FROM </span><span class="si">{</span><span class="n">variant_filtering_synId</span><span class="si">}</span><span class="s2"> where &quot;</span>
        <span class="s2">&quot;Flag = &#39;Flag&#39; and Tumor_Sample_Barcode is not null&quot;</span>
    <span class="p">)</span>
    <span class="n">flag_variantsdf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span><span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span> <span class="n">query_string</span><span class="o">=</span><span class="n">query_str</span><span class="p">)</span>

    <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;flaggedVariants&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">(</span>
        <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;HGVSp_Short&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;Reference_Allele&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;Tumor_Seq_Allele2&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="o">+</span> <span class="s2">&quot; &quot;</span>
        <span class="o">+</span> <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;Tumor_Sample_Barcode&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">flag_variantsdf</span><span class="p">[</span><span class="s2">&quot;flaggedVariants&quot;</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_mutation_in_cis_files_to_staging" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_mutation_in_cis_files_to_staging</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">,</span> <span class="n">genieVersion</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_mutation_in_cis_files_to_staging" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Stores the mutation in cis files to synapse per
    center</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>synapse client connection</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>center mapping dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_filtering_synId</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>variant filtering synapse id</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genieVersion</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>version of the genie pipeline run</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">724</span>
<span class="normal">725</span>
<span class="normal">726</span>
<span class="normal">727</span>
<span class="normal">728</span>
<span class="normal">729</span>
<span class="normal">730</span>
<span class="normal">731</span>
<span class="normal">732</span>
<span class="normal">733</span>
<span class="normal">734</span>
<span class="normal">735</span>
<span class="normal">736</span>
<span class="normal">737</span>
<span class="normal">738</span>
<span class="normal">739</span>
<span class="normal">740</span>
<span class="normal">741</span>
<span class="normal">742</span>
<span class="normal">743</span>
<span class="normal">744</span>
<span class="normal">745</span>
<span class="normal">746</span>
<span class="normal">747</span>
<span class="normal">748</span>
<span class="normal">749</span>
<span class="normal">750</span>
<span class="normal">751</span>
<span class="normal">752</span>
<span class="normal">753</span>
<span class="normal">754</span>
<span class="normal">755</span>
<span class="normal">756</span>
<span class="normal">757</span>
<span class="normal">758</span>
<span class="normal">759</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_mutation_in_cis_files_to_staging</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">variant_filtering_synId</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genieVersion</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="kc">None</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Stores the mutation in cis files to synapse per</span>
<span class="sd">        center</span>

<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): synapse client connection</span>
<span class="sd">        center_mappingdf (pd.DataFrame): center mapping dataframe</span>
<span class="sd">        variant_filtering_synId (str): variant filtering synapse id</span>
<span class="sd">        genieVersion (str): version of the genie pipeline run</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Store each centers mutations in cis to their staging folder</span>
    <span class="n">center_str</span> <span class="o">=</span> <span class="s2">&quot;&#39;,&#39;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">center_mappingdf</span><span class="o">.</span><span class="n">center</span><span class="p">)</span>
    <span class="n">query_str</span> <span class="o">=</span> <span class="p">(</span>
        <span class="sa">f</span><span class="s2">&quot;select * from </span><span class="si">{</span><span class="n">variant_filtering_synId</span><span class="si">}</span><span class="s2"> where Center in (&#39;</span><span class="si">{</span><span class="n">center_str</span><span class="si">}</span><span class="s2">&#39;)&quot;</span>
    <span class="p">)</span>
    <span class="n">mergeCheckDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span><span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span> <span class="n">query_string</span><span class="o">=</span><span class="n">query_str</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">mergeCheckDf</span><span class="o">.</span><span class="n">Center</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pd</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">center</span><span class="p">):</span>
            <span class="n">stagingSynId</span> <span class="o">=</span> <span class="n">center_mappingdf</span><span class="o">.</span><span class="n">stagingSynId</span><span class="p">[</span>
                <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
            <span class="p">]</span>
            <span class="n">mergeCheckDf</span><span class="p">[</span><span class="n">mergeCheckDf</span><span class="p">[</span><span class="s2">&quot;Center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span><span class="p">]</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span>
                <span class="s2">&quot;mutationsInCis_filtered_samples.csv&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span>
            <span class="p">)</span>
            <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                <span class="n">filepath</span><span class="o">=</span><span class="s2">&quot;mutationsInCis_filtered_samples.csv&quot;</span><span class="p">,</span>
                <span class="n">parentid</span><span class="o">=</span><span class="n">stagingSynId</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="n">version_comment</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">os</span><span class="o">.</span><span class="n">unlink</span><span class="p">(</span><span class="s2">&quot;mutationsInCis_filtered_samples.csv&quot;</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.seq_assay_id_filter" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">seq_assay_id_filter</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.seq_assay_id_filter" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>(Deprecated)
Remove samples that are part of SEQ_ASSAY_IDs with less
than 50 samples</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clinicalDf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Sample clinical dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.Series: samples to remove</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">763</span>
<span class="normal">764</span>
<span class="normal">765</span>
<span class="normal">766</span>
<span class="normal">767</span>
<span class="normal">768</span>
<span class="normal">769</span>
<span class="normal">770</span>
<span class="normal">771</span>
<span class="normal">772</span>
<span class="normal">773</span>
<span class="normal">774</span>
<span class="normal">775</span>
<span class="normal">776</span>
<span class="normal">777</span>
<span class="normal">778</span>
<span class="normal">779</span>
<span class="normal">780</span>
<span class="normal">781</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">seq_assay_id_filter</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    (Deprecated)</span>
<span class="sd">    Remove samples that are part of SEQ_ASSAY_IDs with less</span>
<span class="sd">    than 50 samples</span>

<span class="sd">    Args:</span>
<span class="sd">        clinicalDf: Sample clinical dataframe</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: samples to remove</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">remove_seqassayid</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()[</span>
        <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span> <span class="o">&lt;</span> <span class="mi">50</span>
    <span class="p">]</span>
    <span class="n">clinicaldf</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span>
        <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">remove_seqassayid</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="k">return</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.no_genepanel_filter" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">no_genepanel_filter</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">,</span> <span class="n">beddf</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.no_genepanel_filter" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Remove samples that don't have bed files associated with
them</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical data</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beddf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>bed data</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.Series: samples to remove</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">784</span>
<span class="normal">785</span>
<span class="normal">786</span>
<span class="normal">787</span>
<span class="normal">788</span>
<span class="normal">789</span>
<span class="normal">790</span>
<span class="normal">791</span>
<span class="normal">792</span>
<span class="normal">793</span>
<span class="normal">794</span>
<span class="normal">795</span>
<span class="normal">796</span>
<span class="normal">797</span>
<span class="normal">798</span>
<span class="normal">799</span>
<span class="normal">800</span>
<span class="normal">801</span>
<span class="normal">802</span>
<span class="normal">803</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">no_genepanel_filter</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">,</span> <span class="n">beddf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remove samples that don&#39;t have bed files associated with</span>
<span class="sd">    them</span>

<span class="sd">    Args:</span>
<span class="sd">        clinicaldf:  Clinical data</span>
<span class="sd">        beddf: bed data</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.Series: samples to remove</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;NO GENE PANEL FILTER&quot;</span><span class="p">)</span>
    <span class="n">has_seq_assay</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">])</span>
    <span class="n">remove_samples</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">][</span><span class="o">~</span><span class="n">has_seq_assay</span><span class="p">]</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
        <span class="s2">&quot;Removing samples with no bed file: </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;,&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">remove_samples</span><span class="p">))</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">remove_samples</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_gene_panel_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_gene_panel_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">fileviewSynId</span><span class="p">,</span> <span class="n">genieVersion</span><span class="p">,</span> <span class="n">data_gene_panel</span><span class="p">,</span> <span class="n">consortiumReleaseSynId</span><span class="p">,</span> <span class="n">wes_seqassayids</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_gene_panel_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">807</span>
<span class="normal">808</span>
<span class="normal">809</span>
<span class="normal">810</span>
<span class="normal">811</span>
<span class="normal">812</span>
<span class="normal">813</span>
<span class="normal">814</span>
<span class="normal">815</span>
<span class="normal">816</span>
<span class="normal">817</span>
<span class="normal">818</span>
<span class="normal">819</span>
<span class="normal">820</span>
<span class="normal">821</span>
<span class="normal">822</span>
<span class="normal">823</span>
<span class="normal">824</span>
<span class="normal">825</span>
<span class="normal">826</span>
<span class="normal">827</span>
<span class="normal">828</span>
<span class="normal">829</span>
<span class="normal">830</span>
<span class="normal">831</span>
<span class="normal">832</span>
<span class="normal">833</span>
<span class="normal">834</span>
<span class="normal">835</span>
<span class="normal">836</span>
<span class="normal">837</span>
<span class="normal">838</span>
<span class="normal">839</span>
<span class="normal">840</span>
<span class="normal">841</span>
<span class="normal">842</span>
<span class="normal">843</span>
<span class="normal">844</span>
<span class="normal">845</span>
<span class="normal">846</span>
<span class="normal">847</span>
<span class="normal">848</span>
<span class="normal">849</span>
<span class="normal">850</span>
<span class="normal">851</span>
<span class="normal">852</span>
<span class="normal">853</span>
<span class="normal">854</span>
<span class="normal">855</span>
<span class="normal">856</span>
<span class="normal">857</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_gene_panel_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">fileviewSynId</span><span class="p">,</span>
    <span class="n">genieVersion</span><span class="p">,</span>
    <span class="n">data_gene_panel</span><span class="p">,</span>
    <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
    <span class="n">wes_seqassayids</span><span class="p">,</span>
<span class="p">):</span>
    <span class="c1"># Only need to upload these files once</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STORING GENE PANELS FILES&quot;</span><span class="p">)</span>
    <span class="n">wes_genepanel_filenames</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;data_gene_panel_</span><span class="si">{}</span><span class="s2">.txt&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seqassayid</span><span class="p">)</span> <span class="k">for</span> <span class="n">seqassayid</span> <span class="ow">in</span> <span class="n">wes_seqassayids</span>
    <span class="p">]</span>
    <span class="c1"># Format string for tableQuery statement</span>
    <span class="n">wes_genepanel_str</span> <span class="o">=</span> <span class="s2">&quot;&#39;,&#39;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">wes_genepanel_filenames</span><span class="p">)</span>
    <span class="c1"># Only need to upload these files once</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STORING GENE PANELS FILES&quot;</span><span class="p">)</span>
    <span class="c1"># This line of code is required to make sure any new files are</span>
    <span class="c1"># pulled into the file view.</span>
    <span class="n">syn</span><span class="o">.</span><span class="n">tableQuery</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;select * from </span><span class="si">{</span><span class="n">fileviewSynId</span><span class="si">}</span><span class="s2"> limit 1&quot;</span><span class="p">)</span>
    <span class="n">genePanelDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;select id from </span><span class="si">{</span><span class="n">fileviewSynId</span><span class="si">}</span><span class="s2"> where &quot;</span>
        <span class="s2">&quot;cBioFileFormat = &#39;genePanel&#39; and &quot;</span>
        <span class="s2">&quot;fileStage = &#39;staging&#39; and &quot;</span>
        <span class="sa">f</span><span class="s2">&quot;name not in (&#39;</span><span class="si">{</span><span class="n">wes_genepanel_str</span><span class="si">}</span><span class="s2">&#39;)&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">genePanelEntities</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">panelNames</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">data_gene_panel</span><span class="p">[</span><span class="s2">&quot;mutations&quot;</span><span class="p">])</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;EXISTING GENE PANELS: </span><span class="si">{</span><span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">panelNames</span><span class="p">)</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="k">for</span> <span class="n">synId</span> <span class="ow">in</span> <span class="n">genePanelDf</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span>
        <span class="n">genePanel</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">synId</span><span class="p">)</span>
        <span class="n">genePanelName</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">genePanel</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
        <span class="n">newGenePanelPath</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="n">genePanelName</span><span class="p">)</span>
        <span class="n">gene_panel</span> <span class="o">=</span> <span class="n">genePanelName</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;data_gene_panel_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">gene_panel</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">gene_panel</span> <span class="ow">in</span> <span class="n">panelNames</span><span class="p">:</span>
            <span class="n">os</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">genePanel</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">newGenePanelPath</span><span class="p">)</span>
            <span class="n">annotations</span> <span class="o">=</span> <span class="p">{</span><span class="s2">&quot;cBioFileFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;genePanel&quot;</span><span class="p">}</span>
            <span class="n">genePanelEntities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span>
                <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                    <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                    <span class="n">filepath</span><span class="o">=</span><span class="n">newGenePanelPath</span><span class="p">,</span>
                    <span class="n">parentid</span><span class="o">=</span><span class="n">consortiumReleaseSynId</span><span class="p">,</span>
                    <span class="n">version_comment</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">,</span>
                    <span class="n">name</span><span class="o">=</span><span class="n">genePanelName</span><span class="p">,</span>
                    <span class="n">annotations</span><span class="o">=</span><span class="n">annotations</span><span class="p">,</span>
                    <span class="n">used</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">synId</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">genePanel</span><span class="o">.</span><span class="n">versionNumber</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
                <span class="p">)</span>
            <span class="p">)</span>
    <span class="k">return</span> <span class="n">genePanelEntities</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.filter_out_germline_variants" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">filter_out_germline_variants</span><span class="p">(</span><span class="n">input_data</span><span class="p">,</span> <span class="n">status_col_str</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.filter_out_germline_variants" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Filters out germline variants given a status col str. Genie pipeline
    cannot have any of these variants. NOTE: We have to search for the
    status column because there's no column name validation in the release
    steps so the status column may have different casing.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>input_data</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>input data with germline variants to filter out</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>status_col_str</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>search string for the status column for the data</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <span class="doc-returns-annotation">
                    <code><span title="pandas.DataFrame">DataFrame</span></code>
                </span>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.DataFrame: filtered out germline variant data</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">860</span>
<span class="normal">861</span>
<span class="normal">862</span>
<span class="normal">863</span>
<span class="normal">864</span>
<span class="normal">865</span>
<span class="normal">866</span>
<span class="normal">867</span>
<span class="normal">868</span>
<span class="normal">869</span>
<span class="normal">870</span>
<span class="normal">871</span>
<span class="normal">872</span>
<span class="normal">873</span>
<span class="normal">874</span>
<span class="normal">875</span>
<span class="normal">876</span>
<span class="normal">877</span>
<span class="normal">878</span>
<span class="normal">879</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">filter_out_germline_variants</span><span class="p">(</span>
    <span class="n">input_data</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">status_col_str</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Filters out germline variants given a status col str. Genie pipeline</span>
<span class="sd">        cannot have any of these variants. NOTE: We have to search for the</span>
<span class="sd">        status column because there&#39;s no column name validation in the release</span>
<span class="sd">        steps so the status column may have different casing.</span>

<span class="sd">    Args:</span>
<span class="sd">        input_data (pd.DataFrame): input data with germline variants to filter out</span>
<span class="sd">        status_col_str (str): search string for the status column for the data</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: filtered out germline variant data</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># find status col SV_Status</span>
    <span class="n">status_col</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">input_data</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span> <span class="o">==</span> <span class="n">status_col_str</span><span class="o">.</span><span class="n">lower</span><span class="p">()</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">input_data</span><span class="p">[</span><span class="n">input_data</span><span class="p">[</span><span class="n">status_col</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;GERMLINE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">reset_index</span><span class="p">(</span><span class="n">drop</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_sv_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_sv_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">synid</span><span class="p">,</span> <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span> <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_sv_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create, filter, configure, and store structural variant file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>SV database synid</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_center_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for center files</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_merged_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for merged file</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="typing.List">List</span>[<span title="str">str</span>]</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging flag</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="bool">bool</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>List of SV Samples</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">883</span>
<span class="normal">884</span>
<span class="normal">885</span>
<span class="normal">886</span>
<span class="normal">887</span>
<span class="normal">888</span>
<span class="normal">889</span>
<span class="normal">890</span>
<span class="normal">891</span>
<span class="normal">892</span>
<span class="normal">893</span>
<span class="normal">894</span>
<span class="normal">895</span>
<span class="normal">896</span>
<span class="normal">897</span>
<span class="normal">898</span>
<span class="normal">899</span>
<span class="normal">900</span>
<span class="normal">901</span>
<span class="normal">902</span>
<span class="normal">903</span>
<span class="normal">904</span>
<span class="normal">905</span>
<span class="normal">906</span>
<span class="normal">907</span>
<span class="normal">908</span>
<span class="normal">909</span>
<span class="normal">910</span>
<span class="normal">911</span>
<span class="normal">912</span>
<span class="normal">913</span>
<span class="normal">914</span>
<span class="normal">915</span>
<span class="normal">916</span>
<span class="normal">917</span>
<span class="normal">918</span>
<span class="normal">919</span>
<span class="normal">920</span>
<span class="normal">921</span>
<span class="normal">922</span>
<span class="normal">923</span>
<span class="normal">924</span>
<span class="normal">925</span>
<span class="normal">926</span>
<span class="normal">927</span>
<span class="normal">928</span>
<span class="normal">929</span>
<span class="normal">930</span>
<span class="normal">931</span>
<span class="normal">932</span>
<span class="normal">933</span>
<span class="normal">934</span>
<span class="normal">935</span>
<span class="normal">936</span>
<span class="normal">937</span>
<span class="normal">938</span>
<span class="normal">939</span>
<span class="normal">940</span>
<span class="normal">941</span>
<span class="normal">942</span>
<span class="normal">943</span>
<span class="normal">944</span>
<span class="normal">945</span>
<span class="normal">946</span>
<span class="normal">947</span>
<span class="normal">948</span>
<span class="normal">949</span>
<span class="normal">950</span>
<span class="normal">951</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_sv_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">synid</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">keep_for_center_consortium_samples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">keep_for_merged_consortium_samples</span><span class="p">:</span> <span class="n">List</span><span class="p">[</span><span class="nb">str</span><span class="p">],</span>
    <span class="n">current_release_staging</span><span class="p">:</span> <span class="nb">bool</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create, filter, configure, and store structural variant file</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        synid: SV database synid</span>
<span class="sd">        keep_for_center_consortium_samples: Samples to keep for center files</span>
<span class="sd">        keep_for_merged_consortium_samples: Samples to keep for merged file</span>
<span class="sd">        current_release_staging: Staging flag</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of SV Samples</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;MERING, FILTERING, STORING SV FILES&quot;</span><span class="p">)</span>
    <span class="n">sv_df</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;select * from </span><span class="si">{</span><span class="n">synid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">create_snapshot_version</span><span class="p">(</span><span class="n">synid</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">)</span>

    <span class="c1"># sv_df[&quot;ENTREZ_GENE_ID&quot;].mask(</span>
    <span class="c1">#     sv_df[&quot;ENTREZ_GENE_ID&quot;] == 0, float(&quot;nan&quot;), inplace=True</span>
    <span class="c1"># )</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">current_release_staging</span><span class="p">:</span>
        <span class="n">sv_staging_df</span> <span class="o">=</span> <span class="n">sv_df</span><span class="p">[</span>
            <span class="n">sv_df</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_for_center_consortium_samples</span><span class="p">)</span>
        <span class="p">]</span>
        <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">center_mappingdf</span><span class="o">.</span><span class="n">center</span><span class="p">:</span>
            <span class="n">center_sv</span> <span class="o">=</span> <span class="n">sv_staging_df</span><span class="p">[</span><span class="n">sv_staging_df</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">center_sv</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">center_sv</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">SV_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                    <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                    <span class="n">filepath</span><span class="o">=</span><span class="n">SV_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span>
                    <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                    <span class="n">parentid</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                        <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
                    <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                <span class="p">)</span>

    <span class="n">sv_df</span> <span class="o">=</span> <span class="n">sv_df</span><span class="p">[</span><span class="n">sv_df</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_for_merged_consortium_samples</span><span class="p">)]</span>
    <span class="n">sv_df</span> <span class="o">=</span> <span class="n">filter_out_germline_variants</span><span class="p">(</span><span class="n">input_data</span><span class="o">=</span><span class="n">sv_df</span><span class="p">,</span> <span class="n">status_col_str</span><span class="o">=</span><span class="s2">&quot;SV_STATUS&quot;</span><span class="p">)</span>
    <span class="n">sv_df</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">transform</span><span class="o">.</span><span class="n">_col_name_to_titlecase</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">sv_text</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">sv_df</span><span class="p">)</span>
    <span class="n">sv_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_sv.txt&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">sv_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">sv_file</span><span class="p">:</span>
        <span class="n">sv_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">sv_text</span><span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">sv_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_sv.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">synid</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">sv_df</span><span class="p">[</span><span class="s2">&quot;Sample_Id&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.append_or_create_release_maf" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">append_or_create_release_maf</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">filepath</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.append_or_create_release_maf" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Creates a file with the dataframe or appends to a existing file.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>pandas.dataframe to write out</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>filepath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Filepath to append or create</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">955</span>
<span class="normal">956</span>
<span class="normal">957</span>
<span class="normal">958</span>
<span class="normal">959</span>
<span class="normal">960</span>
<span class="normal">961</span>
<span class="normal">962</span>
<span class="normal">963</span>
<span class="normal">964</span>
<span class="normal">965</span>
<span class="normal">966</span>
<span class="normal">967</span>
<span class="normal">968</span>
<span class="normal">969</span>
<span class="normal">970</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">append_or_create_release_maf</span><span class="p">(</span><span class="n">dataframe</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span> <span class="n">filepath</span><span class="p">:</span> <span class="nb">str</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates a file with the dataframe or appends to a existing file.</span>

<span class="sd">    Args:</span>
<span class="sd">        df: pandas.dataframe to write out</span>
<span class="sd">        filepath: Filepath to append or create</span>

<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span> <span class="ow">or</span> <span class="n">os</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="o">.</span><span class="n">st_size</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">dataframe</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
            <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">dataframe</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="s2">&quot;a&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f_out</span><span class="p">:</span>
            <span class="n">f_out</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">data</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_maf_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_maf_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">flatfiles_view_synid</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span> <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span> <span class="n">remove_mafinbed_variants</span><span class="p">,</span> <span class="n">flagged_mutationInCis_variants</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_maf_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create, filter, configure, and store maf file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>flatfiles_view_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id of fileview with all the flat files</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical dataframe with SAMPLE_ID and CENTER</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_merged_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for merged file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_center_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for center files</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>remove_mafinbed_variants</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Variants to remove</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>flagged_mutationInCis_variants</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Variants to flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 974</span>
<span class="normal"> 975</span>
<span class="normal"> 976</span>
<span class="normal"> 977</span>
<span class="normal"> 978</span>
<span class="normal"> 979</span>
<span class="normal"> 980</span>
<span class="normal"> 981</span>
<span class="normal"> 982</span>
<span class="normal"> 983</span>
<span class="normal"> 984</span>
<span class="normal"> 985</span>
<span class="normal"> 986</span>
<span class="normal"> 987</span>
<span class="normal"> 988</span>
<span class="normal"> 989</span>
<span class="normal"> 990</span>
<span class="normal"> 991</span>
<span class="normal"> 992</span>
<span class="normal"> 993</span>
<span class="normal"> 994</span>
<span class="normal"> 995</span>
<span class="normal"> 996</span>
<span class="normal"> 997</span>
<span class="normal"> 998</span>
<span class="normal"> 999</span>
<span class="normal">1000</span>
<span class="normal">1001</span>
<span class="normal">1002</span>
<span class="normal">1003</span>
<span class="normal">1004</span>
<span class="normal">1005</span>
<span class="normal">1006</span>
<span class="normal">1007</span>
<span class="normal">1008</span>
<span class="normal">1009</span>
<span class="normal">1010</span>
<span class="normal">1011</span>
<span class="normal">1012</span>
<span class="normal">1013</span>
<span class="normal">1014</span>
<span class="normal">1015</span>
<span class="normal">1016</span>
<span class="normal">1017</span>
<span class="normal">1018</span>
<span class="normal">1019</span>
<span class="normal">1020</span>
<span class="normal">1021</span>
<span class="normal">1022</span>
<span class="normal">1023</span>
<span class="normal">1024</span>
<span class="normal">1025</span>
<span class="normal">1026</span>
<span class="normal">1027</span>
<span class="normal">1028</span>
<span class="normal">1029</span>
<span class="normal">1030</span>
<span class="normal">1031</span>
<span class="normal">1032</span>
<span class="normal">1033</span>
<span class="normal">1034</span>
<span class="normal">1035</span>
<span class="normal">1036</span>
<span class="normal">1037</span>
<span class="normal">1038</span>
<span class="normal">1039</span>
<span class="normal">1040</span>
<span class="normal">1041</span>
<span class="normal">1042</span>
<span class="normal">1043</span>
<span class="normal">1044</span>
<span class="normal">1045</span>
<span class="normal">1046</span>
<span class="normal">1047</span>
<span class="normal">1048</span>
<span class="normal">1049</span>
<span class="normal">1050</span>
<span class="normal">1051</span>
<span class="normal">1052</span>
<span class="normal">1053</span>
<span class="normal">1054</span>
<span class="normal">1055</span>
<span class="normal">1056</span>
<span class="normal">1057</span>
<span class="normal">1058</span>
<span class="normal">1059</span>
<span class="normal">1060</span>
<span class="normal">1061</span>
<span class="normal">1062</span>
<span class="normal">1063</span>
<span class="normal">1064</span>
<span class="normal">1065</span>
<span class="normal">1066</span>
<span class="normal">1067</span>
<span class="normal">1068</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_maf_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">flatfiles_view_synid</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">,</span>
    <span class="n">clinicaldf</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span>
    <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span>
    <span class="n">remove_mafinbed_variants</span><span class="p">,</span>
    <span class="n">flagged_mutationInCis_variants</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create, filter, configure, and store maf file</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        flatfiles_view_synid: Synapse id of fileview with all the flat files</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">        clinicaldf: Clinical dataframe with SAMPLE_ID and CENTER</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>
<span class="sd">        keep_for_merged_consortium_samples: Samples to keep for merged file</span>
<span class="sd">        keep_for_center_consortium_samples: Samples to keep for center files</span>
<span class="sd">        remove_mafinbed_variants: Variants to remove</span>
<span class="sd">        flagged_mutationInCis_variants: Variants to flag</span>
<span class="sd">        current_release_staging: Staging flag</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;FILTERING, STORING MUTATION FILES&quot;</span><span class="p">)</span>
    <span class="n">centerMafSynIdsDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">query_string</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;select id from </span><span class="si">{</span><span class="n">flatfiles_view_synid</span><span class="si">}</span><span class="s2"> where name like &#39;%mutation%&#39;&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">mutations_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_mutations_extended.txt&quot;</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">mutations_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">):</span>
        <span class="k">pass</span>
    <span class="c1"># Create maf file per center for their staging directory</span>
    <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">MUTATIONS_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">):</span>
            <span class="k">pass</span>
    <span class="n">used_entities</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">maf_ent</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">centerMafSynIdsDf</span><span class="o">.</span><span class="n">id</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
    <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">mafSynId</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">centerMafSynIdsDf</span><span class="o">.</span><span class="n">id</span><span class="p">):</span>
        <span class="n">maf_ent</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mafSynId</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">maf_ent</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
        <span class="c1"># Extract center name</span>
        <span class="n">center</span> <span class="o">=</span> <span class="n">maf_ent</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">3</span><span class="p">]</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">center_mappingdf</span><span class="o">.</span><span class="n">center</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
            <span class="n">used_entities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">maf_ent</span><span class="o">.</span><span class="n">id</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">maf_ent</span><span class="o">.</span><span class="n">versionNumber</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">mafchunks</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span>
                <span class="n">maf_ent</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">chunksize</span><span class="o">=</span><span class="mi">100000</span>
            <span class="p">)</span>

            <span class="k">for</span> <span class="n">mafchunk</span> <span class="ow">in</span> <span class="n">mafchunks</span><span class="p">:</span>
                <span class="c1"># Get center for center staging maf</span>
                <span class="c1"># Configure maf</span>
                <span class="n">configured_mafdf</span> <span class="o">=</span> <span class="n">configure_maf</span><span class="p">(</span>
                    <span class="n">mafchunk</span><span class="p">,</span> <span class="n">remove_mafinbed_variants</span><span class="p">,</span> <span class="n">flagged_mutationInCis_variants</span>
                <span class="p">)</span>
                <span class="n">configured_mafdf</span> <span class="o">=</span> <span class="n">configured_mafdf</span><span class="p">[</span><span class="n">FULL_MAF_RELEASE_COLUMNS</span><span class="p">]</span>
                <span class="c1"># Create maf for release</span>
                <span class="n">merged_mafdf</span> <span class="o">=</span> <span class="n">remove_maf_samples</span><span class="p">(</span>
                    <span class="n">configured_mafdf</span><span class="p">,</span> <span class="n">keep_for_merged_consortium_samples</span>
                <span class="p">)</span>
                <span class="n">append_or_create_release_maf</span><span class="p">(</span><span class="n">merged_mafdf</span><span class="p">,</span> <span class="n">mutations_path</span><span class="p">)</span>
                <span class="c1"># Create maf for center staging</span>
                <span class="n">center_mafdf</span> <span class="o">=</span> <span class="n">remove_maf_samples</span><span class="p">(</span>
                    <span class="n">configured_mafdf</span><span class="p">,</span> <span class="n">keep_for_center_consortium_samples</span>
                <span class="p">)</span>
                <span class="n">append_or_create_release_maf</span><span class="p">(</span>
                    <span class="n">center_mafdf</span><span class="p">,</span> <span class="n">MUTATIONS_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span>
                <span class="p">)</span>

    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">mutations_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_mutations_extended.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">used_entities</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">current_release_staging</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="n">staging_synid</span> <span class="o">=</span> <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
            <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                <span class="n">filepath</span><span class="o">=</span><span class="n">MUTATIONS_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span>
                <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                <span class="n">parentid</span><span class="o">=</span><span class="n">staging_synid</span><span class="p">,</span>
            <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.run_genie_filters" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">run_genie_filters</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">variant_filtering_synId</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">beddf</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">processing_date</span><span class="p">,</span> <span class="n">skip_mutationsincis</span><span class="p">,</span> <span class="n">consortium_release_cutoff</span><span class="p">,</span> <span class="n">test</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.run_genie_filters" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Run GENIE filters and returns variants and samples to remove</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>variant_filtering_synId</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id of mutationInCis table</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beddf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Bed dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>processing_date</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Processing date</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>skip_mutationsincis</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Skip mutation in cis filter</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>consortium_release_cutoff</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Release cutoff in days</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>test</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Test flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: Variants to remove</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>set</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>samples to remove for release files</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>set</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>samples to remove for center files</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: Variants to flag</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1072</span>
<span class="normal">1073</span>
<span class="normal">1074</span>
<span class="normal">1075</span>
<span class="normal">1076</span>
<span class="normal">1077</span>
<span class="normal">1078</span>
<span class="normal">1079</span>
<span class="normal">1080</span>
<span class="normal">1081</span>
<span class="normal">1082</span>
<span class="normal">1083</span>
<span class="normal">1084</span>
<span class="normal">1085</span>
<span class="normal">1086</span>
<span class="normal">1087</span>
<span class="normal">1088</span>
<span class="normal">1089</span>
<span class="normal">1090</span>
<span class="normal">1091</span>
<span class="normal">1092</span>
<span class="normal">1093</span>
<span class="normal">1094</span>
<span class="normal">1095</span>
<span class="normal">1096</span>
<span class="normal">1097</span>
<span class="normal">1098</span>
<span class="normal">1099</span>
<span class="normal">1100</span>
<span class="normal">1101</span>
<span class="normal">1102</span>
<span class="normal">1103</span>
<span class="normal">1104</span>
<span class="normal">1105</span>
<span class="normal">1106</span>
<span class="normal">1107</span>
<span class="normal">1108</span>
<span class="normal">1109</span>
<span class="normal">1110</span>
<span class="normal">1111</span>
<span class="normal">1112</span>
<span class="normal">1113</span>
<span class="normal">1114</span>
<span class="normal">1115</span>
<span class="normal">1116</span>
<span class="normal">1117</span>
<span class="normal">1118</span>
<span class="normal">1119</span>
<span class="normal">1120</span>
<span class="normal">1121</span>
<span class="normal">1122</span>
<span class="normal">1123</span>
<span class="normal">1124</span>
<span class="normal">1125</span>
<span class="normal">1126</span>
<span class="normal">1127</span>
<span class="normal">1128</span>
<span class="normal">1129</span>
<span class="normal">1130</span>
<span class="normal">1131</span>
<span class="normal">1132</span>
<span class="normal">1133</span>
<span class="normal">1134</span>
<span class="normal">1135</span>
<span class="normal">1136</span>
<span class="normal">1137</span>
<span class="normal">1138</span>
<span class="normal">1139</span>
<span class="normal">1140</span>
<span class="normal">1141</span>
<span class="normal">1142</span>
<span class="normal">1143</span>
<span class="normal">1144</span>
<span class="normal">1145</span>
<span class="normal">1146</span>
<span class="normal">1147</span>
<span class="normal">1148</span>
<span class="normal">1149</span>
<span class="normal">1150</span>
<span class="normal">1151</span>
<span class="normal">1152</span>
<span class="normal">1153</span>
<span class="normal">1154</span>
<span class="normal">1155</span>
<span class="normal">1156</span>
<span class="normal">1157</span>
<span class="normal">1158</span>
<span class="normal">1159</span>
<span class="normal">1160</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">run_genie_filters</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">variant_filtering_synId</span><span class="p">,</span>
    <span class="n">clinicaldf</span><span class="p">,</span>
    <span class="n">beddf</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">processing_date</span><span class="p">,</span>
    <span class="n">skip_mutationsincis</span><span class="p">,</span>
    <span class="n">consortium_release_cutoff</span><span class="p">,</span>
    <span class="n">test</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Run GENIE filters and returns variants and samples to remove</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        variant_filtering_synId: Synapse id of mutationInCis table</span>
<span class="sd">        clinicaldf: Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</span>
<span class="sd">        beddf: Bed dataframe</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>
<span class="sd">        processing_date: Processing date</span>
<span class="sd">        skip_mutationsincis: Skip mutation in cis filter</span>
<span class="sd">        consortium_release_cutoff: Release cutoff in days</span>
<span class="sd">        test: Test flag</span>
<span class="sd">        current_release_staging: Staging flag</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.Series: Variants to remove</span>
<span class="sd">        set: samples to remove for release files</span>
<span class="sd">        set: samples to remove for center files</span>
<span class="sd">        pandas.Series: Variants to flag</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># ADD CHECKS TO CODE BEFORE UPLOAD.</span>
    <span class="c1"># Throw error if things don&#39;t go through</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;RUN GENIE FILTERS&quot;</span><span class="p">)</span>
    <span class="c1"># STORING CLINICAL FILES INTO CBIOPORTAL</span>
    <span class="c1"># FILTERING</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;MAF IN BED FILTER&quot;</span><span class="p">)</span>
    <span class="n">remove_mafinbed_variants</span> <span class="o">=</span> <span class="n">runMAFinBED</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">center_mappingdf</span><span class="p">,</span>
        <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
        <span class="n">staging</span><span class="o">=</span><span class="n">current_release_staging</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;MUTATION IN CIS FILTER&quot;</span><span class="p">)</span>
    <span class="p">(</span>
        <span class="n">remove_mutationincis_samples</span><span class="p">,</span>
        <span class="n">flagged_mutationincis_variants</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">=</span> <span class="n">mutation_in_cis_filter</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">skip_mutationsincis</span><span class="p">,</span>
        <span class="n">variant_filtering_synId</span><span class="p">,</span>
        <span class="n">center_mappingdf</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">test</span><span class="o">=</span><span class="n">test</span><span class="p">,</span>
        <span class="n">staging</span><span class="o">=</span><span class="n">current_release_staging</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">remove_no_genepanel_samples</span> <span class="o">=</span> <span class="n">no_genepanel_filter</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">,</span> <span class="n">beddf</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;SEQ DATE FILTER&quot;</span><span class="p">)</span>
    <span class="n">remove_seqdate_samples</span> <span class="o">=</span> <span class="n">seq_date_filter</span><span class="p">(</span>
        <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">processing_date</span><span class="p">,</span> <span class="n">consortium_release_cutoff</span>
    <span class="p">)</span>

    <span class="c1"># Only certain samples are removed for the files that go into</span>
    <span class="c1"># staging center folder</span>
    <span class="n">remove_center_consortium_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">remove_mutationincis_samples</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
        <span class="nb">set</span><span class="p">(</span><span class="n">remove_no_genepanel_samples</span><span class="p">)</span>
    <span class="p">)</span>
    <span class="c1"># Most filteres are applied for the files that go into the merged</span>
    <span class="c1"># consortium release</span>
    <span class="n">remove_merged_consortium_samples</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">remove_seqdate_samples</span><span class="p">)</span>

    <span class="n">remove_merged_consortium_samples</span> <span class="o">=</span> <span class="n">remove_merged_consortium_samples</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
        <span class="n">remove_center_consortium_samples</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span>
        <span class="n">remove_mafinbed_variants</span><span class="p">,</span>
        <span class="n">remove_merged_consortium_samples</span><span class="p">,</span>
        <span class="n">remove_center_consortium_samples</span><span class="p">,</span>
        <span class="n">flagged_mutationincis_variants</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_assay_info_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_assay_info_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">assay_info_synid</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_assay_info_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Creates, stores assay information and gets WES panel list</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>assay_info_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Assay information database synid</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>List of whole exome sequencing SEQ_ASSAY_IDs</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1164</span>
<span class="normal">1165</span>
<span class="normal">1166</span>
<span class="normal">1167</span>
<span class="normal">1168</span>
<span class="normal">1169</span>
<span class="normal">1170</span>
<span class="normal">1171</span>
<span class="normal">1172</span>
<span class="normal">1173</span>
<span class="normal">1174</span>
<span class="normal">1175</span>
<span class="normal">1176</span>
<span class="normal">1177</span>
<span class="normal">1178</span>
<span class="normal">1179</span>
<span class="normal">1180</span>
<span class="normal">1181</span>
<span class="normal">1182</span>
<span class="normal">1183</span>
<span class="normal">1184</span>
<span class="normal">1185</span>
<span class="normal">1186</span>
<span class="normal">1187</span>
<span class="normal">1188</span>
<span class="normal">1189</span>
<span class="normal">1190</span>
<span class="normal">1191</span>
<span class="normal">1192</span>
<span class="normal">1193</span>
<span class="normal">1194</span>
<span class="normal">1195</span>
<span class="normal">1196</span>
<span class="normal">1197</span>
<span class="normal">1198</span>
<span class="normal">1199</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_assay_info_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">assay_info_synid</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">release_synid</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Creates, stores assay information and gets WES panel list</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        assay_info_synid: Assay information database synid</span>
<span class="sd">        clinicaldf: Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</span>
<span class="sd">        release_synid: Synapse id to store release file</span>

<span class="sd">    Returns:</span>
<span class="sd">        List of whole exome sequencing SEQ_ASSAY_IDs</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Creates assay information file&quot;</span><span class="p">)</span>
    <span class="n">assay_info_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;assay_information.txt&quot;</span><span class="p">)</span>
    <span class="n">seq_assay_str</span> <span class="o">=</span> <span class="s2">&quot;&#39;,&#39;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">create_snapshot_version</span><span class="p">(</span><span class="n">assay_info_synid</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">)</span>
    <span class="n">assay_infodf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;select * from </span><span class="si">{</span><span class="n">assay_info_synid</span><span class="si">}</span><span class="s2"> where SEQ_ASSAY_ID &quot;</span>
        <span class="sa">f</span><span class="s2">&quot;in (&#39;</span><span class="si">{</span><span class="n">seq_assay_str</span><span class="si">}</span><span class="s2">&#39;)&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">assay_infodf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">assay_info_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">assay_info_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;assay_information.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">assay_info_synid</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">wes_index</span> <span class="o">=</span> <span class="n">assay_infodf</span><span class="p">[</span><span class="s2">&quot;library_strategy&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;WXS&quot;</span>
    <span class="n">wes_panels</span> <span class="o">=</span> <span class="n">assay_infodf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">][</span><span class="n">wes_index</span><span class="p">]</span>
    <span class="k">return</span> <span class="n">wes_panels</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_clinical_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_clinical_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">oncotree_url</span><span class="p">,</span> <span class="n">sample_cols</span><span class="p">,</span> <span class="n">patient_cols</span><span class="p">,</span> <span class="n">remove_center_consortium_samples</span><span class="p">,</span> <span class="n">remove_merged_consortium_samples</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">databaseSynIdMappingDf</span><span class="p">,</span> <span class="n">used</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_clinical_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create, filter, configure, and store clinical file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>oncotree_url</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Oncotree URL</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sample_cols</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical sample columns</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>patient_cols</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical patient columns</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>remove_center_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to remove for center files</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>remove_merged_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to remove for merged file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>databaseSynIdMappingDf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Database to Synapse Id mapping</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.DataFrame: configured clinical dataframe</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: samples to keep for center files</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.Series: samples to keep for release files</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1203</span>
<span class="normal">1204</span>
<span class="normal">1205</span>
<span class="normal">1206</span>
<span class="normal">1207</span>
<span class="normal">1208</span>
<span class="normal">1209</span>
<span class="normal">1210</span>
<span class="normal">1211</span>
<span class="normal">1212</span>
<span class="normal">1213</span>
<span class="normal">1214</span>
<span class="normal">1215</span>
<span class="normal">1216</span>
<span class="normal">1217</span>
<span class="normal">1218</span>
<span class="normal">1219</span>
<span class="normal">1220</span>
<span class="normal">1221</span>
<span class="normal">1222</span>
<span class="normal">1223</span>
<span class="normal">1224</span>
<span class="normal">1225</span>
<span class="normal">1226</span>
<span class="normal">1227</span>
<span class="normal">1228</span>
<span class="normal">1229</span>
<span class="normal">1230</span>
<span class="normal">1231</span>
<span class="normal">1232</span>
<span class="normal">1233</span>
<span class="normal">1234</span>
<span class="normal">1235</span>
<span class="normal">1236</span>
<span class="normal">1237</span>
<span class="normal">1238</span>
<span class="normal">1239</span>
<span class="normal">1240</span>
<span class="normal">1241</span>
<span class="normal">1242</span>
<span class="normal">1243</span>
<span class="normal">1244</span>
<span class="normal">1245</span>
<span class="normal">1246</span>
<span class="normal">1247</span>
<span class="normal">1248</span>
<span class="normal">1249</span>
<span class="normal">1250</span>
<span class="normal">1251</span>
<span class="normal">1252</span>
<span class="normal">1253</span>
<span class="normal">1254</span>
<span class="normal">1255</span>
<span class="normal">1256</span>
<span class="normal">1257</span>
<span class="normal">1258</span>
<span class="normal">1259</span>
<span class="normal">1260</span>
<span class="normal">1261</span>
<span class="normal">1262</span>
<span class="normal">1263</span>
<span class="normal">1264</span>
<span class="normal">1265</span>
<span class="normal">1266</span>
<span class="normal">1267</span>
<span class="normal">1268</span>
<span class="normal">1269</span>
<span class="normal">1270</span>
<span class="normal">1271</span>
<span class="normal">1272</span>
<span class="normal">1273</span>
<span class="normal">1274</span>
<span class="normal">1275</span>
<span class="normal">1276</span>
<span class="normal">1277</span>
<span class="normal">1278</span>
<span class="normal">1279</span>
<span class="normal">1280</span>
<span class="normal">1281</span>
<span class="normal">1282</span>
<span class="normal">1283</span>
<span class="normal">1284</span>
<span class="normal">1285</span>
<span class="normal">1286</span>
<span class="normal">1287</span>
<span class="normal">1288</span>
<span class="normal">1289</span>
<span class="normal">1290</span>
<span class="normal">1291</span>
<span class="normal">1292</span>
<span class="normal">1293</span>
<span class="normal">1294</span>
<span class="normal">1295</span>
<span class="normal">1296</span>
<span class="normal">1297</span>
<span class="normal">1298</span>
<span class="normal">1299</span>
<span class="normal">1300</span>
<span class="normal">1301</span>
<span class="normal">1302</span>
<span class="normal">1303</span>
<span class="normal">1304</span>
<span class="normal">1305</span>
<span class="normal">1306</span>
<span class="normal">1307</span>
<span class="normal">1308</span>
<span class="normal">1309</span>
<span class="normal">1310</span>
<span class="normal">1311</span>
<span class="normal">1312</span>
<span class="normal">1313</span>
<span class="normal">1314</span>
<span class="normal">1315</span>
<span class="normal">1316</span>
<span class="normal">1317</span>
<span class="normal">1318</span>
<span class="normal">1319</span>
<span class="normal">1320</span>
<span class="normal">1321</span>
<span class="normal">1322</span>
<span class="normal">1323</span>
<span class="normal">1324</span>
<span class="normal">1325</span>
<span class="normal">1326</span>
<span class="normal">1327</span>
<span class="normal">1328</span>
<span class="normal">1329</span>
<span class="normal">1330</span>
<span class="normal">1331</span>
<span class="normal">1332</span>
<span class="normal">1333</span>
<span class="normal">1334</span>
<span class="normal">1335</span>
<span class="normal">1336</span>
<span class="normal">1337</span>
<span class="normal">1338</span>
<span class="normal">1339</span>
<span class="normal">1340</span>
<span class="normal">1341</span>
<span class="normal">1342</span>
<span class="normal">1343</span>
<span class="normal">1344</span>
<span class="normal">1345</span>
<span class="normal">1346</span>
<span class="normal">1347</span>
<span class="normal">1348</span>
<span class="normal">1349</span>
<span class="normal">1350</span>
<span class="normal">1351</span>
<span class="normal">1352</span>
<span class="normal">1353</span>
<span class="normal">1354</span>
<span class="normal">1355</span>
<span class="normal">1356</span>
<span class="normal">1357</span>
<span class="normal">1358</span>
<span class="normal">1359</span>
<span class="normal">1360</span>
<span class="normal">1361</span>
<span class="normal">1362</span>
<span class="normal">1363</span>
<span class="normal">1364</span>
<span class="normal">1365</span>
<span class="normal">1366</span>
<span class="normal">1367</span>
<span class="normal">1368</span>
<span class="normal">1369</span>
<span class="normal">1370</span>
<span class="normal">1371</span>
<span class="normal">1372</span>
<span class="normal">1373</span>
<span class="normal">1374</span>
<span class="normal">1375</span>
<span class="normal">1376</span>
<span class="normal">1377</span>
<span class="normal">1378</span>
<span class="normal">1379</span>
<span class="normal">1380</span>
<span class="normal">1381</span>
<span class="normal">1382</span>
<span class="normal">1383</span>
<span class="normal">1384</span>
<span class="normal">1385</span>
<span class="normal">1386</span>
<span class="normal">1387</span>
<span class="normal">1388</span>
<span class="normal">1389</span>
<span class="normal">1390</span>
<span class="normal">1391</span>
<span class="normal">1392</span>
<span class="normal">1393</span>
<span class="normal">1394</span>
<span class="normal">1395</span>
<span class="normal">1396</span>
<span class="normal">1397</span>
<span class="normal">1398</span>
<span class="normal">1399</span>
<span class="normal">1400</span>
<span class="normal">1401</span>
<span class="normal">1402</span>
<span class="normal">1403</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_clinical_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">clinicaldf</span><span class="p">,</span>
    <span class="n">oncotree_url</span><span class="p">,</span>
    <span class="n">sample_cols</span><span class="p">,</span>
    <span class="n">patient_cols</span><span class="p">,</span>
    <span class="n">remove_center_consortium_samples</span><span class="p">,</span>
    <span class="n">remove_merged_consortium_samples</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">databaseSynIdMappingDf</span><span class="p">,</span>
    <span class="n">used</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create, filter, configure, and store clinical file</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        clinicaldf: Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</span>
<span class="sd">        oncotree_url: Oncotree URL</span>
<span class="sd">        sample_cols: Clinical sample columns</span>
<span class="sd">        patient_cols: Clinical patient columns</span>
<span class="sd">        remove_center_consortium_samples: Samples to remove for center files</span>
<span class="sd">        remove_merged_consortium_samples: Samples to remove for merged file</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">        current_release_staging: Staging flag</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>
<span class="sd">        databaseSynIdMappingDf: Database to Synapse Id mapping</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: configured clinical dataframe</span>
<span class="sd">        pandas.Series: samples to keep for center files</span>
<span class="sd">        pandas.Series: samples to keep for release files</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;CONFIGURING CLINICAL FILES&quot;</span><span class="p">)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;REMOVING PHI&quot;</span><span class="p">)</span>
    <span class="c1"># clinicaldf = redact_phi(clinicaldf)</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;ADD CANCER TYPES&quot;</span><span class="p">)</span>
    <span class="c1"># This removes support for both oncotree urls (only support json)</span>
    <span class="n">oncotree_dict</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">get_oncotree_code_mappings</span><span class="p">(</span><span class="n">oncotree_url</span><span class="p">)</span>
    <span class="c1"># Add in unknown key which maps to UNKNOWN everything</span>
    <span class="n">oncotree_dict</span><span class="p">[</span><span class="s2">&quot;UNKNOWN&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s2">&quot;CANCER_TYPE&quot;</span><span class="p">:</span> <span class="s2">&quot;UNKNOWN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;CANCER_TYPE_DETAILED&quot;</span><span class="p">:</span> <span class="s2">&quot;UNKNOWN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ONCOTREE_PRIMARY_NODE&quot;</span><span class="p">:</span> <span class="s2">&quot;UNKNOWN&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ONCOTREE_SECONDARY_NODE&quot;</span><span class="p">:</span> <span class="s2">&quot;UNKNOWN&quot;</span><span class="p">,</span>
    <span class="p">}</span>

    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;CANCER_TYPE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="n">oncotree_dict</span><span class="p">[</span><span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()][</span><span class="s2">&quot;CANCER_TYPE&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">oncotree_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">code</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;ONCOTREE_CODE&quot;</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;CANCER_TYPE_DETAILED&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="n">oncotree_dict</span><span class="p">[</span><span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()][</span><span class="s2">&quot;CANCER_TYPE_DETAILED&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">oncotree_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">code</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;ONCOTREE_CODE&quot;</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;ONCOTREE_PRIMARY_NODE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="n">oncotree_dict</span><span class="p">[</span><span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()][</span><span class="s2">&quot;ONCOTREE_PRIMARY_NODE&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">oncotree_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">code</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;ONCOTREE_CODE&quot;</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;ONCOTREE_SECONDARY_NODE&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="n">oncotree_dict</span><span class="p">[</span><span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()][</span><span class="s2">&quot;ONCOTREE_SECONDARY_NODE&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">code</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span> <span class="ow">in</span> <span class="n">oncotree_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()</span>
            <span class="k">else</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">code</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;ONCOTREE_CODE&quot;</span><span class="p">]</span>
    <span class="p">]</span>

    <span class="c1"># All cancer types that are null contain deprecated oncotree codes</span>
    <span class="c1"># And should be removed</span>
    <span class="n">clinicaldf</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="o">~</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;CANCER_TYPE&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
    <span class="c1"># Suggest using AGE_AT_SEQ_REPORT_DAYS instead so that the</span>
    <span class="c1"># descriptions can match</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT_DAYS&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">]</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
        <span class="p">(</span>
            <span class="nb">int</span><span class="p">(</span><span class="n">math</span><span class="o">.</span><span class="n">floor</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">age</span><span class="p">))</span> <span class="o">/</span> <span class="mf">365.25</span><span class="p">))</span>
            <span class="k">if</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">checkInt</span><span class="p">(</span><span class="n">age</span><span class="p">)</span>
            <span class="k">else</span> <span class="n">age</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">age</span> <span class="ow">in</span> <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">]</span>
    <span class="p">]</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">][</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&gt;32485&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&gt;89&quot;</span>
    <span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">][</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;&lt;6570&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="s2">&quot;&lt;18&quot;</span>

    <span class="c1">############################################################</span>
    <span class="c1"># CENTER SPECIFIC CODE FOR RIGHT NOW (REMOVE UHN-555-V1)</span>
    <span class="c1">############################################################</span>
    <span class="c1"># clinicalDf = clinicalDf[clinicalDf[&#39;SEQ_ASSAY_ID&#39;] != &quot;UHN-555-V1&quot;]</span>
    <span class="c1"># clinicalDf = clinicalDf[clinicalDf[&#39;SEQ_ASSAY_ID&#39;] != &quot;PHS-TRISEQ-V1&quot;]</span>

    <span class="c1"># clinicalDf = clinicalDf[clinicalDf[&#39;CENTER&#39;] != &quot;WAKE&quot;]</span>
    <span class="c1"># clinicalDf = clinicalDf[clinicalDf[&#39;CENTER&#39;] != &quot;CRUK&quot;]</span>
    <span class="c1">############################################################</span>
    <span class="c1">############################################################</span>

    <span class="n">clinicaldf</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STORING CLINICAL FILES&quot;</span><span class="p">)</span>
    <span class="c1"># samples must be removed after reading in the clinical file again</span>
    <span class="n">staging_clinicaldf</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span>
        <span class="o">~</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">remove_center_consortium_samples</span><span class="p">)</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">current_release_staging</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">center_mappingdf</span><span class="o">.</span><span class="n">center</span><span class="p">:</span>
            <span class="n">center_clinical</span> <span class="o">=</span> <span class="n">staging_clinicaldf</span><span class="p">[</span><span class="n">staging_clinicaldf</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span><span class="p">]</span>
            <span class="n">center_sample</span> <span class="o">=</span> <span class="n">center_clinical</span><span class="p">[</span><span class="n">sample_cols</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">)</span>
            <span class="n">center_patient</span> <span class="o">=</span> <span class="n">center_clinical</span><span class="p">[</span><span class="n">patient_cols</span><span class="p">]</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s2">&quot;PATIENT_ID&quot;</span><span class="p">)</span>
            <span class="n">center_sample</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">SAMPLE_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">center_patient</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">PATIENT_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
            <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                <span class="n">filepath</span><span class="o">=</span><span class="n">SAMPLE_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span>
                <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                <span class="n">parentid</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                    <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
                <span class="p">][</span><span class="mi">0</span><span class="p">],</span>
            <span class="p">)</span>
            <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                <span class="n">filepath</span><span class="o">=</span><span class="n">PATIENT_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span>
                <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                <span class="n">parentid</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                    <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
                <span class="p">][</span><span class="mi">0</span><span class="p">],</span>
            <span class="p">)</span>

    <span class="n">clinicaldf</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="p">[</span>
        <span class="o">~</span><span class="n">clinicaldf</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">remove_merged_consortium_samples</span><span class="p">)</span>
    <span class="p">]</span>

    <span class="n">keep_center_consortium_samples</span> <span class="o">=</span> <span class="n">staging_clinicaldf</span><span class="o">.</span><span class="n">SAMPLE_ID</span>
    <span class="n">keep_merged_consortium_samples</span> <span class="o">=</span> <span class="n">clinicaldf</span><span class="o">.</span><span class="n">SAMPLE_ID</span>
    <span class="c1"># This mapping table is the GENIE clinical code to description</span>
    <span class="c1"># mapping to generate the headers of the clinical file</span>
    <span class="n">clinical_code_to_desc_map_synid</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;clinical_code_to_desc_map&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">mapping</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span> <span class="n">query_string</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;SELECT * FROM </span><span class="si">{</span><span class="n">clinical_code_to_desc_map_synid</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="n">clinical_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_clinical.txt&quot;</span><span class="p">)</span>
    <span class="n">clinical_sample_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_clinical_sample.txt&quot;</span><span class="p">)</span>
    <span class="n">clinical_patient_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_clinical_patient.txt&quot;</span><span class="p">)</span>
    <span class="n">process_functions</span><span class="o">.</span><span class="n">addClinicalHeaders</span><span class="p">(</span>
        <span class="n">clinicaldf</span><span class="p">,</span>
        <span class="n">mapping</span><span class="p">,</span>
        <span class="n">patient_cols</span><span class="p">,</span>
        <span class="n">sample_cols</span><span class="p">,</span>
        <span class="n">clinical_sample_path</span><span class="p">,</span>
        <span class="n">clinical_patient_path</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">clinical_sample_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_clinical_sample.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">used</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">clinical_patient_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_clinical_patient.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">used</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">clinicaldf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">clinical_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">clinical_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_clinical.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">used</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="s2">&quot;database&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="p">(</span><span class="n">clinicaldf</span><span class="p">,</span> <span class="n">keep_center_consortium_samples</span><span class="p">,</span> <span class="n">keep_merged_consortium_samples</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_cna_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_cna_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">flatfiles_view_synid</span><span class="p">,</span> <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span> <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_cna_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create, filter and store cna file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>flatfiles_view_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id of fileview with all the flat files</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_center_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for center files</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_merged_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for merged file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>
        <p>Returns:
    list: CNA samples</p>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1407</span>
<span class="normal">1408</span>
<span class="normal">1409</span>
<span class="normal">1410</span>
<span class="normal">1411</span>
<span class="normal">1412</span>
<span class="normal">1413</span>
<span class="normal">1414</span>
<span class="normal">1415</span>
<span class="normal">1416</span>
<span class="normal">1417</span>
<span class="normal">1418</span>
<span class="normal">1419</span>
<span class="normal">1420</span>
<span class="normal">1421</span>
<span class="normal">1422</span>
<span class="normal">1423</span>
<span class="normal">1424</span>
<span class="normal">1425</span>
<span class="normal">1426</span>
<span class="normal">1427</span>
<span class="normal">1428</span>
<span class="normal">1429</span>
<span class="normal">1430</span>
<span class="normal">1431</span>
<span class="normal">1432</span>
<span class="normal">1433</span>
<span class="normal">1434</span>
<span class="normal">1435</span>
<span class="normal">1436</span>
<span class="normal">1437</span>
<span class="normal">1438</span>
<span class="normal">1439</span>
<span class="normal">1440</span>
<span class="normal">1441</span>
<span class="normal">1442</span>
<span class="normal">1443</span>
<span class="normal">1444</span>
<span class="normal">1445</span>
<span class="normal">1446</span>
<span class="normal">1447</span>
<span class="normal">1448</span>
<span class="normal">1449</span>
<span class="normal">1450</span>
<span class="normal">1451</span>
<span class="normal">1452</span>
<span class="normal">1453</span>
<span class="normal">1454</span>
<span class="normal">1455</span>
<span class="normal">1456</span>
<span class="normal">1457</span>
<span class="normal">1458</span>
<span class="normal">1459</span>
<span class="normal">1460</span>
<span class="normal">1461</span>
<span class="normal">1462</span>
<span class="normal">1463</span>
<span class="normal">1464</span>
<span class="normal">1465</span>
<span class="normal">1466</span>
<span class="normal">1467</span>
<span class="normal">1468</span>
<span class="normal">1469</span>
<span class="normal">1470</span>
<span class="normal">1471</span>
<span class="normal">1472</span>
<span class="normal">1473</span>
<span class="normal">1474</span>
<span class="normal">1475</span>
<span class="normal">1476</span>
<span class="normal">1477</span>
<span class="normal">1478</span>
<span class="normal">1479</span>
<span class="normal">1480</span>
<span class="normal">1481</span>
<span class="normal">1482</span>
<span class="normal">1483</span>
<span class="normal">1484</span>
<span class="normal">1485</span>
<span class="normal">1486</span>
<span class="normal">1487</span>
<span class="normal">1488</span>
<span class="normal">1489</span>
<span class="normal">1490</span>
<span class="normal">1491</span>
<span class="normal">1492</span>
<span class="normal">1493</span>
<span class="normal">1494</span>
<span class="normal">1495</span>
<span class="normal">1496</span>
<span class="normal">1497</span>
<span class="normal">1498</span>
<span class="normal">1499</span>
<span class="normal">1500</span>
<span class="normal">1501</span>
<span class="normal">1502</span>
<span class="normal">1503</span>
<span class="normal">1504</span>
<span class="normal">1505</span>
<span class="normal">1506</span>
<span class="normal">1507</span>
<span class="normal">1508</span>
<span class="normal">1509</span>
<span class="normal">1510</span>
<span class="normal">1511</span>
<span class="normal">1512</span>
<span class="normal">1513</span>
<span class="normal">1514</span>
<span class="normal">1515</span>
<span class="normal">1516</span>
<span class="normal">1517</span>
<span class="normal">1518</span>
<span class="normal">1519</span>
<span class="normal">1520</span>
<span class="normal">1521</span>
<span class="normal">1522</span>
<span class="normal">1523</span>
<span class="normal">1524</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_cna_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">flatfiles_view_synid</span><span class="p">,</span>
    <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span>
    <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create, filter and store cna file</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        flatfiles_view_synid: Synapse id of fileview with all the flat files</span>
<span class="sd">        keep_for_center_consortium_samples: Samples to keep for center files</span>
<span class="sd">        keep_for_merged_consortium_samples: Samples to keep for merged file</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">    Returns:</span>
<span class="sd">        list: CNA samples</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;MERING, FILTERING, STORING CNA FILES&quot;</span><span class="p">)</span>
    <span class="n">cna_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_CNA.txt&quot;</span><span class="p">)</span>
    <span class="n">query_str</span> <span class="o">=</span> <span class="p">(</span><span class="s2">&quot;select id from </span><span class="si">{}</span><span class="s2"> &quot;</span> <span class="s2">&quot;where name like &#39;data_CNA%&#39;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
        <span class="n">flatfiles_view_synid</span>
    <span class="p">)</span>
    <span class="n">center_cna_synidsdf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">query_str</span><span class="p">)</span>
    <span class="c1"># Grab all unique symbols and form cna_template</span>
    <span class="n">all_symbols</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">cna_synid</span> <span class="ow">in</span> <span class="n">center_cna_synidsdf</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span>
        <span class="n">cna_ent</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">cna_synid</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">cna_ent</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;r&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">cna_file</span><span class="p">:</span>
            <span class="c1"># Read first line first</span>
            <span class="n">cna_file</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>
            <span class="c1"># Get all hugo symbols</span>
            <span class="n">all_symbols</span> <span class="o">=</span> <span class="n">all_symbols</span><span class="o">.</span><span class="n">union</span><span class="p">(</span>
                <span class="nb">set</span><span class="p">(</span><span class="n">line</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">cna_file</span><span class="p">)</span>
            <span class="p">)</span>
    <span class="n">cna_template</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">all_symbols</span><span class="p">)})</span>
    <span class="n">cna_template</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="n">cna_template</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">cna_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="c1"># Loop through to create finalized CNA file</span>
    <span class="n">with_center_hugo_symbol</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">)</span>
    <span class="n">with_center_hugo_symbol</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span><span class="n">with_center_hugo_symbol</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">keep_for_center_consortium_samples</span><span class="p">)]</span>
    <span class="p">)</span>

    <span class="n">with_merged_hugo_symbol</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">)</span>
    <span class="n">with_merged_hugo_symbol</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span><span class="n">with_merged_hugo_symbol</span><span class="p">,</span> <span class="n">pd</span><span class="o">.</span><span class="n">Series</span><span class="p">(</span><span class="n">keep_for_merged_consortium_samples</span><span class="p">)]</span>
    <span class="p">)</span>

    <span class="n">cna_samples</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="n">used_entities</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">cna_synId</span> <span class="ow">in</span> <span class="n">center_cna_synidsdf</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]:</span>
        <span class="n">cna_ent</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">cna_synId</span><span class="p">)</span>
        <span class="n">center</span> <span class="o">=</span> <span class="n">cna_ent</span><span class="o">.</span><span class="n">name</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;data_CNA_&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.txt&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">cna_ent</span><span class="o">.</span><span class="n">path</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">center_mappingdf</span><span class="o">.</span><span class="n">center</span><span class="o">.</span><span class="n">tolist</span><span class="p">():</span>
            <span class="n">used_entities</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cna_synId</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">cna_ent</span><span class="o">.</span><span class="n">versionNumber</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">center_cna</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">cna_ent</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">merged_cna</span> <span class="o">=</span> <span class="n">cna_template</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">center_cna</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">)</span>
            <span class="n">merged_cna</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="k">if</span> <span class="ow">not</span> <span class="n">current_release_staging</span><span class="p">:</span>
                <span class="n">merged_cna</span> <span class="o">=</span> <span class="n">merged_cna</span><span class="p">[</span>
                    <span class="n">merged_cna</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">merged_cna</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">with_center_hugo_symbol</span><span class="p">)]</span>
                <span class="p">]</span>

                <span class="n">cna_text</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">merged_cna</span><span class="p">)</span>
                <span class="c1"># Replace blank with NA&#39;s</span>
                <span class="n">cna_text</span> <span class="o">=</span> <span class="n">cna_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">NA</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">cna_text</span> <span class="o">=</span> <span class="n">cna_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">NA</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
                <span class="n">cna_text</span> <span class="o">=</span> <span class="n">cna_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">NA</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

                <span class="c1"># Store center CNA file in staging dir</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">CNA_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">cna_file</span><span class="p">:</span>
                    <span class="n">cna_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">cna_text</span><span class="p">)</span>
                <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                    <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                    <span class="n">filepath</span><span class="o">=</span><span class="n">CNA_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span>
                    <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                    <span class="n">parentid</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                        <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="c1"># This is to remove more samples for the final cna file</span>
            <span class="n">merged_cna</span> <span class="o">=</span> <span class="n">merged_cna</span><span class="p">[</span>
                <span class="n">merged_cna</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="n">merged_cna</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">with_merged_hugo_symbol</span><span class="p">)]</span>
            <span class="p">]</span>

            <span class="n">cna_text</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">merged_cna</span><span class="p">)</span>
            <span class="n">cna_text</span> <span class="o">=</span> <span class="n">cna_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">NA</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">cna_text</span> <span class="o">=</span> <span class="n">cna_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">NA</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">cna_text</span> <span class="o">=</span> <span class="n">cna_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;</span><span class="se">\t\n</span><span class="s2">&quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">NA</span><span class="se">\n</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">CNA_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">cna_file</span><span class="p">:</span>
                <span class="n">cna_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">cna_text</span><span class="p">)</span>
            <span class="c1"># Join CNA file</span>
            <span class="n">cna_samples</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">merged_cna</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span><span class="o">.</span><span class="n">tolist</span><span class="p">())</span>
            <span class="n">linux_join_command</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;join&quot;</span><span class="p">,</span> <span class="n">cna_path</span><span class="p">,</span> <span class="n">CNA_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">]</span>
            <span class="n">output</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">linux_join_command</span><span class="p">)</span>
            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">cna_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">cna_file</span><span class="p">:</span>
                <span class="n">cna_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">output</span><span class="o">.</span><span class="n">decode</span><span class="p">(</span><span class="s2">&quot;utf-8&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="p">,</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">))</span>

    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">cna_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_CNA.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">used_entities</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">cna_samples</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_seg_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_seg_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">seg_synid</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span> <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_seg_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create, filter and store seg file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seg_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Seg database synid</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_center_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for center files</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>keep_for_merged_consortium_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples to keep for merged file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1528</span>
<span class="normal">1529</span>
<span class="normal">1530</span>
<span class="normal">1531</span>
<span class="normal">1532</span>
<span class="normal">1533</span>
<span class="normal">1534</span>
<span class="normal">1535</span>
<span class="normal">1536</span>
<span class="normal">1537</span>
<span class="normal">1538</span>
<span class="normal">1539</span>
<span class="normal">1540</span>
<span class="normal">1541</span>
<span class="normal">1542</span>
<span class="normal">1543</span>
<span class="normal">1544</span>
<span class="normal">1545</span>
<span class="normal">1546</span>
<span class="normal">1547</span>
<span class="normal">1548</span>
<span class="normal">1549</span>
<span class="normal">1550</span>
<span class="normal">1551</span>
<span class="normal">1552</span>
<span class="normal">1553</span>
<span class="normal">1554</span>
<span class="normal">1555</span>
<span class="normal">1556</span>
<span class="normal">1557</span>
<span class="normal">1558</span>
<span class="normal">1559</span>
<span class="normal">1560</span>
<span class="normal">1561</span>
<span class="normal">1562</span>
<span class="normal">1563</span>
<span class="normal">1564</span>
<span class="normal">1565</span>
<span class="normal">1566</span>
<span class="normal">1567</span>
<span class="normal">1568</span>
<span class="normal">1569</span>
<span class="normal">1570</span>
<span class="normal">1571</span>
<span class="normal">1572</span>
<span class="normal">1573</span>
<span class="normal">1574</span>
<span class="normal">1575</span>
<span class="normal">1576</span>
<span class="normal">1577</span>
<span class="normal">1578</span>
<span class="normal">1579</span>
<span class="normal">1580</span>
<span class="normal">1581</span>
<span class="normal">1582</span>
<span class="normal">1583</span>
<span class="normal">1584</span>
<span class="normal">1585</span>
<span class="normal">1586</span>
<span class="normal">1587</span>
<span class="normal">1588</span>
<span class="normal">1589</span>
<span class="normal">1590</span>
<span class="normal">1591</span>
<span class="normal">1592</span>
<span class="normal">1593</span>
<span class="normal">1594</span>
<span class="normal">1595</span>
<span class="normal">1596</span>
<span class="normal">1597</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_seg_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">seg_synid</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">,</span>
    <span class="n">keep_for_center_consortium_samples</span><span class="p">,</span>
    <span class="n">keep_for_merged_consortium_samples</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create, filter and store seg file</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        seg_synid: Seg database synid</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">        keep_for_center_consortium_samples: Samples to keep for center files</span>
<span class="sd">        keep_for_merged_consortium_samples: Samples to keep for merged file</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>
<span class="sd">        current_release_staging: Staging flag</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;MERING, FILTERING, STORING SEG FILES&quot;</span><span class="p">)</span>
    <span class="n">seg_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_cna_hg19.seg&quot;</span><span class="p">)</span>
    <span class="n">version</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">create_snapshot_version</span><span class="p">(</span><span class="n">seg_synid</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">)</span>

    <span class="n">segdf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">query_string</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;SELECT ID,CHROM,LOCSTART,LOCEND,NUMMARK,SEGMEAN,CENTER FROM </span><span class="si">{</span><span class="n">seg_synid</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">segdf</span> <span class="o">=</span> <span class="n">segdf</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span>
        <span class="n">columns</span><span class="o">=</span><span class="p">{</span>
            <span class="s2">&quot;CHROM&quot;</span><span class="p">:</span> <span class="s2">&quot;chrom&quot;</span><span class="p">,</span>
            <span class="s2">&quot;LOCSTART&quot;</span><span class="p">:</span> <span class="s2">&quot;loc.start&quot;</span><span class="p">,</span>
            <span class="s2">&quot;LOCEND&quot;</span><span class="p">:</span> <span class="s2">&quot;loc.end&quot;</span><span class="p">,</span>
            <span class="s2">&quot;SEGMEAN&quot;</span><span class="p">:</span> <span class="s2">&quot;seg.mean&quot;</span><span class="p">,</span>
            <span class="s2">&quot;NUMMARK&quot;</span><span class="p">:</span> <span class="s2">&quot;num.mark&quot;</span><span class="p">,</span>
        <span class="p">}</span>
    <span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">current_release_staging</span><span class="p">:</span>
        <span class="n">staging_segdf</span> <span class="o">=</span> <span class="n">segdf</span><span class="p">[</span><span class="n">segdf</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_for_center_consortium_samples</span><span class="p">)]</span>
        <span class="k">for</span> <span class="n">center</span> <span class="ow">in</span> <span class="n">center_mappingdf</span><span class="o">.</span><span class="n">center</span><span class="p">:</span>
            <span class="n">center_seg</span> <span class="o">=</span> <span class="n">staging_segdf</span><span class="p">[</span><span class="n">staging_segdf</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span><span class="p">]</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">center_seg</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="k">del</span> <span class="n">center_seg</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span>
                <span class="n">segtext</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">center_seg</span><span class="p">)</span>
                <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">SEG_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">seg_file</span><span class="p">:</span>
                    <span class="n">seg_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">segtext</span><span class="p">)</span>
                <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                    <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                    <span class="n">filepath</span><span class="o">=</span><span class="n">SEG_CENTER_PATH</span> <span class="o">%</span> <span class="n">center</span><span class="p">,</span>
                    <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                    <span class="n">parentid</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span>
                        <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
                    <span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                <span class="p">)</span>
    <span class="k">del</span> <span class="n">segdf</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span>
    <span class="n">segdf</span> <span class="o">=</span> <span class="n">segdf</span><span class="p">[</span><span class="n">segdf</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">keep_for_merged_consortium_samples</span><span class="p">)]</span>
    <span class="n">segtext</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">removePandasDfFloat</span><span class="p">(</span><span class="n">segdf</span><span class="p">)</span>
    <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">seg_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">seg_file</span><span class="p">:</span>
        <span class="n">seg_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">segtext</span><span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">seg_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_cna_hg19.seg&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">seg_synid</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">version</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_data_gene_matrix" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_data_gene_matrix</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">,</span> <span class="n">cna_samples</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">wes_seqassayids</span><span class="p">,</span> <span class="n">sv_samples</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_data_gene_matrix" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create and store data gene matrix file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>clinicaldf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>cna_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples with CNA</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>wes_seqassayids</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Whole exome sequencing SEQ_ASSAY_IDs</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>sv_samples</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Samples with SV</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pandas.DataFrame: data gene matrix dataframe</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1601</span>
<span class="normal">1602</span>
<span class="normal">1603</span>
<span class="normal">1604</span>
<span class="normal">1605</span>
<span class="normal">1606</span>
<span class="normal">1607</span>
<span class="normal">1608</span>
<span class="normal">1609</span>
<span class="normal">1610</span>
<span class="normal">1611</span>
<span class="normal">1612</span>
<span class="normal">1613</span>
<span class="normal">1614</span>
<span class="normal">1615</span>
<span class="normal">1616</span>
<span class="normal">1617</span>
<span class="normal">1618</span>
<span class="normal">1619</span>
<span class="normal">1620</span>
<span class="normal">1621</span>
<span class="normal">1622</span>
<span class="normal">1623</span>
<span class="normal">1624</span>
<span class="normal">1625</span>
<span class="normal">1626</span>
<span class="normal">1627</span>
<span class="normal">1628</span>
<span class="normal">1629</span>
<span class="normal">1630</span>
<span class="normal">1631</span>
<span class="normal">1632</span>
<span class="normal">1633</span>
<span class="normal">1634</span>
<span class="normal">1635</span>
<span class="normal">1636</span>
<span class="normal">1637</span>
<span class="normal">1638</span>
<span class="normal">1639</span>
<span class="normal">1640</span>
<span class="normal">1641</span>
<span class="normal">1642</span>
<span class="normal">1643</span>
<span class="normal">1644</span>
<span class="normal">1645</span>
<span class="normal">1646</span>
<span class="normal">1647</span>
<span class="normal">1648</span>
<span class="normal">1649</span>
<span class="normal">1650</span>
<span class="normal">1651</span>
<span class="normal">1652</span>
<span class="normal">1653</span>
<span class="normal">1654</span>
<span class="normal">1655</span>
<span class="normal">1656</span>
<span class="normal">1657</span>
<span class="normal">1658</span>
<span class="normal">1659</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_data_gene_matrix</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">clinicaldf</span><span class="p">,</span>
    <span class="n">cna_samples</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">,</span>
    <span class="n">wes_seqassayids</span><span class="p">,</span>
    <span class="n">sv_samples</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create and store data gene matrix file</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        clinicaldf: Clinical dataframe with SAMPLE_ID and SEQ_ASSAY_ID</span>
<span class="sd">        cna_samples: Samples with CNA</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">        wes_seqassayids: Whole exome sequencing SEQ_ASSAY_IDs</span>
<span class="sd">        sv_samples: Samples with SV</span>

<span class="sd">    Returns:</span>
<span class="sd">        pandas.DataFrame: data gene matrix dataframe</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STORING DATA GENE MATRIX FILE&quot;</span><span class="p">)</span>
    <span class="n">data_gene_matrix_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;data_gene_matrix.txt&quot;</span><span class="p">)</span>
    <span class="c1"># Samples have already been removed</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">])</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">(</span>
        <span class="p">[</span><span class="n">data_gene_matrix</span><span class="p">,</span> <span class="n">clinicaldf</span><span class="p">[[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]]]</span>
    <span class="p">)</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">data_gene_matrix</span><span class="o">.</span><span class="n">rename</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">:</span> <span class="s2">&quot;mutations&quot;</span><span class="p">})</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">data_gene_matrix</span><span class="p">[</span><span class="n">data_gene_matrix</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">]</span>
    <span class="n">data_gene_matrix</span><span class="o">.</span><span class="n">drop_duplicates</span><span class="p">(</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="c1"># exclude wes assay_ids</span>
    <span class="n">wes_panel_mut</span> <span class="o">=</span> <span class="n">data_gene_matrix</span><span class="p">[</span><span class="s2">&quot;mutations&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">wes_seqassayids</span><span class="p">)</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">data_gene_matrix</span><span class="p">[</span><span class="o">~</span><span class="n">wes_panel_mut</span><span class="p">]</span>

    <span class="c1"># Add in CNA column into gene panel file</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">add_columns_to_data_gene_matrix</span><span class="p">(</span>
        <span class="n">data_gene_matrix</span><span class="o">=</span><span class="n">data_gene_matrix</span><span class="p">,</span> <span class="n">sample_list</span><span class="o">=</span><span class="n">cna_samples</span><span class="p">,</span> <span class="n">column_name</span><span class="o">=</span><span class="s2">&quot;cna&quot;</span>
    <span class="p">)</span>

    <span class="c1"># Add SV column into gene panel file</span>
    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">add_columns_to_data_gene_matrix</span><span class="p">(</span>
        <span class="n">data_gene_matrix</span><span class="o">=</span><span class="n">data_gene_matrix</span><span class="p">,</span> <span class="n">sample_list</span><span class="o">=</span><span class="n">sv_samples</span><span class="p">,</span> <span class="n">column_name</span><span class="o">=</span><span class="s2">&quot;sv&quot;</span>
    <span class="p">)</span>

    <span class="n">data_gene_matrix</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">data_gene_matrix_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">data_gene_matrix_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;data_gene_matrix.txt&quot;</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="k">return</span> <span class="n">data_gene_matrix</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.store_bed_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">store_bed_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_ids</span><span class="p">,</span> <span class="n">center_mappingdf</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="p">,</span> <span class="n">release_synid</span><span class="p">,</span> <span class="n">used</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.store_bed_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Store bed files, store the bed regions that had symbols remapped
Filters bed file by clinical dataframe seq assays</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version (ie. v6.1-consortium)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>beddf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Bed dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seq_assay_ids</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>All SEQ_ASSAY_IDs in the clinical file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>center_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Staging flag</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_synid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store release file</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1663</span>
<span class="normal">1664</span>
<span class="normal">1665</span>
<span class="normal">1666</span>
<span class="normal">1667</span>
<span class="normal">1668</span>
<span class="normal">1669</span>
<span class="normal">1670</span>
<span class="normal">1671</span>
<span class="normal">1672</span>
<span class="normal">1673</span>
<span class="normal">1674</span>
<span class="normal">1675</span>
<span class="normal">1676</span>
<span class="normal">1677</span>
<span class="normal">1678</span>
<span class="normal">1679</span>
<span class="normal">1680</span>
<span class="normal">1681</span>
<span class="normal">1682</span>
<span class="normal">1683</span>
<span class="normal">1684</span>
<span class="normal">1685</span>
<span class="normal">1686</span>
<span class="normal">1687</span>
<span class="normal">1688</span>
<span class="normal">1689</span>
<span class="normal">1690</span>
<span class="normal">1691</span>
<span class="normal">1692</span>
<span class="normal">1693</span>
<span class="normal">1694</span>
<span class="normal">1695</span>
<span class="normal">1696</span>
<span class="normal">1697</span>
<span class="normal">1698</span>
<span class="normal">1699</span>
<span class="normal">1700</span>
<span class="normal">1701</span>
<span class="normal">1702</span>
<span class="normal">1703</span>
<span class="normal">1704</span>
<span class="normal">1705</span>
<span class="normal">1706</span>
<span class="normal">1707</span>
<span class="normal">1708</span>
<span class="normal">1709</span>
<span class="normal">1710</span>
<span class="normal">1711</span>
<span class="normal">1712</span>
<span class="normal">1713</span>
<span class="normal">1714</span>
<span class="normal">1715</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">store_bed_files</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">beddf</span><span class="p">,</span>
    <span class="n">seq_assay_ids</span><span class="p">,</span>
    <span class="n">center_mappingdf</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="p">,</span>
    <span class="n">release_synid</span><span class="p">,</span>
    <span class="n">used</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Store bed files, store the bed regions that had symbols remapped</span>
<span class="sd">    Filters bed file by clinical dataframe seq assays</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version (ie. v6.1-consortium)</span>
<span class="sd">        beddf: Bed dataframe</span>
<span class="sd">        seq_assay_ids: All SEQ_ASSAY_IDs in the clinical file</span>
<span class="sd">        center_mappingdf: Center mapping dataframe</span>
<span class="sd">        current_release_staging: Staging flag</span>
<span class="sd">        release_synid: Synapse id to store release file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;STORING COMBINED BED FILE&quot;</span><span class="p">)</span>
    <span class="n">combined_bed_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="s2">&quot;genomic_information.txt&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">current_release_staging</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">seq_assay</span> <span class="ow">in</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">():</span>
            <span class="n">bed_seq_df</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">seq_assay</span><span class="p">]</span>
            <span class="n">center</span> <span class="o">=</span> <span class="n">seq_assay</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">bed_seq_df</span> <span class="o">=</span> <span class="n">bed_seq_df</span><span class="p">[</span><span class="n">bed_seq_df</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">bed_seq_df</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]]</span>
            <span class="c1"># There should always be a match here, because there should never</span>
            <span class="c1"># be a SEQ_ASSAY_ID that starts without the center name</span>
            <span class="c1"># If there is, check the bed db for SEQ_ASSAY_ID</span>
            <span class="n">center_ind</span> <span class="o">=</span> <span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;center&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">center</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">bed_seq_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                <span class="n">bed_seq_df</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">BED_DIFFS_SEQASSAY_PATH</span> <span class="o">%</span> <span class="n">seq_assay</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
                <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                    <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
                    <span class="n">filepath</span><span class="o">=</span><span class="n">BED_DIFFS_SEQASSAY_PATH</span> <span class="o">%</span> <span class="n">seq_assay</span><span class="p">,</span>
                    <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
                    <span class="n">parentid</span><span class="o">=</span><span class="n">center_mappingdf</span><span class="p">[</span><span class="s2">&quot;stagingSynId&quot;</span><span class="p">][</span><span class="n">center_ind</span><span class="p">][</span><span class="mi">0</span><span class="p">],</span>
                <span class="p">)</span>
    <span class="c1"># This clinicalDf is already filtered through most of the filters</span>
    <span class="n">beddf</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">seq_assay_ids</span><span class="p">)]</span>
    <span class="n">beddf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">combined_bed_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">filepath</span><span class="o">=</span><span class="n">combined_bed_path</span><span class="p">,</span>
        <span class="n">parentid</span><span class="o">=</span><span class="n">release_synid</span><span class="p">,</span>
        <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="n">name</span><span class="o">=</span><span class="s2">&quot;genomic_information.txt&quot;</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">used</span><span class="p">,</span>
    <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.stagingToCbio" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">stagingToCbio</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">processingDate</span><span class="p">,</span> <span class="n">genieVersion</span><span class="p">,</span> <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span> <span class="n">databaseSynIdMappingDf</span><span class="p">,</span> <span class="n">oncotree_url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">consortiumReleaseCutOff</span><span class="o">=</span><span class="mi">183</span><span class="p">,</span> <span class="n">current_release_staging</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">skipMutationsInCis</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">test</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.stagingToCbio" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Main function that takes the GENIE database and creates release files</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>processingDate</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Processing date in form of Apr-XXXX</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genieVersion</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version. Default is test.</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>CENTER_MAPPING_DF</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>center mapping dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>databaseSynIdMappingDf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Database to Synapse Id mapping</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>oncotree_url</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Oncotree link</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>None</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>consortiumReleaseCutOff</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Release cut off days</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>183</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>current_release_staging</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Is it staging. Default is False.</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>skipMutationsInCis</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Skip mutation in cis filter. Default is False.</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>test</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Testing parameter. Default is False.</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>False</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>list</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Gene panel entities</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1719</span>
<span class="normal">1720</span>
<span class="normal">1721</span>
<span class="normal">1722</span>
<span class="normal">1723</span>
<span class="normal">1724</span>
<span class="normal">1725</span>
<span class="normal">1726</span>
<span class="normal">1727</span>
<span class="normal">1728</span>
<span class="normal">1729</span>
<span class="normal">1730</span>
<span class="normal">1731</span>
<span class="normal">1732</span>
<span class="normal">1733</span>
<span class="normal">1734</span>
<span class="normal">1735</span>
<span class="normal">1736</span>
<span class="normal">1737</span>
<span class="normal">1738</span>
<span class="normal">1739</span>
<span class="normal">1740</span>
<span class="normal">1741</span>
<span class="normal">1742</span>
<span class="normal">1743</span>
<span class="normal">1744</span>
<span class="normal">1745</span>
<span class="normal">1746</span>
<span class="normal">1747</span>
<span class="normal">1748</span>
<span class="normal">1749</span>
<span class="normal">1750</span>
<span class="normal">1751</span>
<span class="normal">1752</span>
<span class="normal">1753</span>
<span class="normal">1754</span>
<span class="normal">1755</span>
<span class="normal">1756</span>
<span class="normal">1757</span>
<span class="normal">1758</span>
<span class="normal">1759</span>
<span class="normal">1760</span>
<span class="normal">1761</span>
<span class="normal">1762</span>
<span class="normal">1763</span>
<span class="normal">1764</span>
<span class="normal">1765</span>
<span class="normal">1766</span>
<span class="normal">1767</span>
<span class="normal">1768</span>
<span class="normal">1769</span>
<span class="normal">1770</span>
<span class="normal">1771</span>
<span class="normal">1772</span>
<span class="normal">1773</span>
<span class="normal">1774</span>
<span class="normal">1775</span>
<span class="normal">1776</span>
<span class="normal">1777</span>
<span class="normal">1778</span>
<span class="normal">1779</span>
<span class="normal">1780</span>
<span class="normal">1781</span>
<span class="normal">1782</span>
<span class="normal">1783</span>
<span class="normal">1784</span>
<span class="normal">1785</span>
<span class="normal">1786</span>
<span class="normal">1787</span>
<span class="normal">1788</span>
<span class="normal">1789</span>
<span class="normal">1790</span>
<span class="normal">1791</span>
<span class="normal">1792</span>
<span class="normal">1793</span>
<span class="normal">1794</span>
<span class="normal">1795</span>
<span class="normal">1796</span>
<span class="normal">1797</span>
<span class="normal">1798</span>
<span class="normal">1799</span>
<span class="normal">1800</span>
<span class="normal">1801</span>
<span class="normal">1802</span>
<span class="normal">1803</span>
<span class="normal">1804</span>
<span class="normal">1805</span>
<span class="normal">1806</span>
<span class="normal">1807</span>
<span class="normal">1808</span>
<span class="normal">1809</span>
<span class="normal">1810</span>
<span class="normal">1811</span>
<span class="normal">1812</span>
<span class="normal">1813</span>
<span class="normal">1814</span>
<span class="normal">1815</span>
<span class="normal">1816</span>
<span class="normal">1817</span>
<span class="normal">1818</span>
<span class="normal">1819</span>
<span class="normal">1820</span>
<span class="normal">1821</span>
<span class="normal">1822</span>
<span class="normal">1823</span>
<span class="normal">1824</span>
<span class="normal">1825</span>
<span class="normal">1826</span>
<span class="normal">1827</span>
<span class="normal">1828</span>
<span class="normal">1829</span>
<span class="normal">1830</span>
<span class="normal">1831</span>
<span class="normal">1832</span>
<span class="normal">1833</span>
<span class="normal">1834</span>
<span class="normal">1835</span>
<span class="normal">1836</span>
<span class="normal">1837</span>
<span class="normal">1838</span>
<span class="normal">1839</span>
<span class="normal">1840</span>
<span class="normal">1841</span>
<span class="normal">1842</span>
<span class="normal">1843</span>
<span class="normal">1844</span>
<span class="normal">1845</span>
<span class="normal">1846</span>
<span class="normal">1847</span>
<span class="normal">1848</span>
<span class="normal">1849</span>
<span class="normal">1850</span>
<span class="normal">1851</span>
<span class="normal">1852</span>
<span class="normal">1853</span>
<span class="normal">1854</span>
<span class="normal">1855</span>
<span class="normal">1856</span>
<span class="normal">1857</span>
<span class="normal">1858</span>
<span class="normal">1859</span>
<span class="normal">1860</span>
<span class="normal">1861</span>
<span class="normal">1862</span>
<span class="normal">1863</span>
<span class="normal">1864</span>
<span class="normal">1865</span>
<span class="normal">1866</span>
<span class="normal">1867</span>
<span class="normal">1868</span>
<span class="normal">1869</span>
<span class="normal">1870</span>
<span class="normal">1871</span>
<span class="normal">1872</span>
<span class="normal">1873</span>
<span class="normal">1874</span>
<span class="normal">1875</span>
<span class="normal">1876</span>
<span class="normal">1877</span>
<span class="normal">1878</span>
<span class="normal">1879</span>
<span class="normal">1880</span>
<span class="normal">1881</span>
<span class="normal">1882</span>
<span class="normal">1883</span>
<span class="normal">1884</span>
<span class="normal">1885</span>
<span class="normal">1886</span>
<span class="normal">1887</span>
<span class="normal">1888</span>
<span class="normal">1889</span>
<span class="normal">1890</span>
<span class="normal">1891</span>
<span class="normal">1892</span>
<span class="normal">1893</span>
<span class="normal">1894</span>
<span class="normal">1895</span>
<span class="normal">1896</span>
<span class="normal">1897</span>
<span class="normal">1898</span>
<span class="normal">1899</span>
<span class="normal">1900</span>
<span class="normal">1901</span>
<span class="normal">1902</span>
<span class="normal">1903</span>
<span class="normal">1904</span>
<span class="normal">1905</span>
<span class="normal">1906</span>
<span class="normal">1907</span>
<span class="normal">1908</span>
<span class="normal">1909</span>
<span class="normal">1910</span>
<span class="normal">1911</span>
<span class="normal">1912</span>
<span class="normal">1913</span>
<span class="normal">1914</span>
<span class="normal">1915</span>
<span class="normal">1916</span>
<span class="normal">1917</span>
<span class="normal">1918</span>
<span class="normal">1919</span>
<span class="normal">1920</span>
<span class="normal">1921</span>
<span class="normal">1922</span>
<span class="normal">1923</span>
<span class="normal">1924</span>
<span class="normal">1925</span>
<span class="normal">1926</span>
<span class="normal">1927</span>
<span class="normal">1928</span>
<span class="normal">1929</span>
<span class="normal">1930</span>
<span class="normal">1931</span>
<span class="normal">1932</span>
<span class="normal">1933</span>
<span class="normal">1934</span>
<span class="normal">1935</span>
<span class="normal">1936</span>
<span class="normal">1937</span>
<span class="normal">1938</span>
<span class="normal">1939</span>
<span class="normal">1940</span>
<span class="normal">1941</span>
<span class="normal">1942</span>
<span class="normal">1943</span>
<span class="normal">1944</span>
<span class="normal">1945</span>
<span class="normal">1946</span>
<span class="normal">1947</span>
<span class="normal">1948</span>
<span class="normal">1949</span>
<span class="normal">1950</span>
<span class="normal">1951</span>
<span class="normal">1952</span>
<span class="normal">1953</span>
<span class="normal">1954</span>
<span class="normal">1955</span>
<span class="normal">1956</span>
<span class="normal">1957</span>
<span class="normal">1958</span>
<span class="normal">1959</span>
<span class="normal">1960</span>
<span class="normal">1961</span>
<span class="normal">1962</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">stagingToCbio</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">processingDate</span><span class="p">,</span>
    <span class="n">genieVersion</span><span class="p">,</span>
    <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
    <span class="n">databaseSynIdMappingDf</span><span class="p">,</span>
    <span class="n">oncotree_url</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">consortiumReleaseCutOff</span><span class="o">=</span><span class="mi">183</span><span class="p">,</span>
    <span class="n">current_release_staging</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">skipMutationsInCis</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">test</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Main function that takes the GENIE database and creates release files</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        processingDate: Processing date in form of Apr-XXXX</span>
<span class="sd">        genieVersion: GENIE version. Default is test.</span>
<span class="sd">        CENTER_MAPPING_DF: center mapping dataframe</span>
<span class="sd">        databaseSynIdMappingDf: Database to Synapse Id mapping</span>
<span class="sd">        oncotree_url: Oncotree link</span>
<span class="sd">        consortiumReleaseCutOff: Release cut off days</span>
<span class="sd">        current_release_staging: Is it staging. Default is False.</span>
<span class="sd">        skipMutationsInCis: Skip mutation in cis filter. Default is False.</span>
<span class="sd">        test: Testing parameter. Default is False.</span>

<span class="sd">    Returns:</span>
<span class="sd">        list: Gene panel entities</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">):</span>
        <span class="n">os</span><span class="o">.</span><span class="n">mkdir</span><span class="p">(</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">)</span>
    <span class="n">consortiumReleaseSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;consortium&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">centerMafFileViewSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;centerMafView&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">patientSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;patient&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">sampleSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;sample&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">bedSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;bed&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">fileviewSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;fileview&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">segSynId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;seg&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">variant_filtering_synId</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;mutationsInCis&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">sv_synid</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span><span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;sv&quot;</span><span class="p">][</span>
        <span class="mi">0</span>
    <span class="p">]</span>
    <span class="n">clinical_tier_release_scope_synid</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;clinical_tier_release_scope&quot;</span>
    <span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Grab assay information</span>
    <span class="n">assay_info_ind</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;assayinfo&quot;</span>
    <span class="n">assay_info_synid</span> <span class="o">=</span> <span class="n">databaseSynIdMappingDf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span><span class="n">assay_info_ind</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>

    <span class="c1"># Using center mapping df to gate centers in release fileStage</span>
    <span class="n">center_query_str</span> <span class="o">=</span> <span class="s2">&quot;&#39;,&#39;&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">CENTER_MAPPING_DF</span><span class="o">.</span><span class="n">center</span><span class="p">)</span>

    <span class="n">patient_snapshot</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">create_snapshot_version</span><span class="p">(</span><span class="n">patientSynId</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">)</span>
    <span class="n">patient_used</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">patientSynId</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">patient_snapshot</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">patientDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;SELECT * FROM </span><span class="si">{</span><span class="n">patientSynId</span><span class="si">}</span><span class="s2"> where CENTER in (&#39;</span><span class="si">{</span><span class="n">center_query_str</span><span class="si">}</span><span class="s2">&#39;)&quot;</span>
    <span class="p">)</span>
    <span class="n">sample_snapshot</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">create_snapshot_version</span><span class="p">(</span><span class="n">sampleSynId</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">)</span>
    <span class="n">sample_used</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">sampleSynId</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">sample_snapshot</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">sampleDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;SELECT * FROM </span><span class="si">{</span><span class="n">sampleSynId</span><span class="si">}</span><span class="s2"> where CENTER in (&#39;</span><span class="si">{</span><span class="n">center_query_str</span><span class="si">}</span><span class="s2">&#39;)&quot;</span>
    <span class="p">)</span>
    <span class="n">bed_snapshot</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">create_snapshot_version</span><span class="p">(</span><span class="n">bedSynId</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="n">genieVersion</span><span class="p">)</span>
    <span class="n">bed_used</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">bedSynId</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">bed_snapshot</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="n">bedDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="s2">&quot;SELECT Chromosome,Start_Position,End_Position,Hugo_Symbol,ID,&quot;</span>
        <span class="s2">&quot;SEQ_ASSAY_ID,Feature_Type,includeInPanel,clinicalReported FROM&quot;</span>
        <span class="sa">f</span><span class="s2">&quot; </span><span class="si">{</span><span class="n">bedSynId</span><span class="si">}</span><span class="s2"> where CENTER in (&#39;</span><span class="si">{</span><span class="n">center_query_str</span><span class="si">}</span><span class="s2">&#39;)&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="c1"># Clinical release scope filter</span>
    <span class="c1"># If private -&gt; Don&#39;t release to public</span>
    <span class="n">clinicalReleaseScopeDf</span> <span class="o">=</span> <span class="n">extract</span><span class="o">.</span><span class="n">get_syntabledf</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;SELECT * FROM </span><span class="si">{</span><span class="n">clinical_tier_release_scope_synid</span><span class="si">}</span><span class="s2"> where releaseScope &lt;&gt; &#39;private&#39;&quot;</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">patientCols</span> <span class="o">=</span> <span class="n">clinicalReleaseScopeDf</span><span class="p">[</span><span class="s2">&quot;fieldName&quot;</span><span class="p">][</span>
        <span class="n">clinicalReleaseScopeDf</span><span class="p">[</span><span class="s2">&quot;level&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;patient&quot;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">sampleCols</span> <span class="o">=</span> <span class="n">clinicalReleaseScopeDf</span><span class="p">[</span><span class="s2">&quot;fieldName&quot;</span><span class="p">][</span>
        <span class="n">clinicalReleaseScopeDf</span><span class="p">[</span><span class="s2">&quot;level&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;sample&quot;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>

    <span class="c1"># Remove this when these columns are removed from both databases</span>
    <span class="k">if</span> <span class="n">sampleDf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;AGE_AT_SEQ_REPORT_NUMERICAL&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">sampleDf</span><span class="p">[</span><span class="s2">&quot;AGE_AT_SEQ_REPORT_NUMERICAL&quot;</span><span class="p">]</span>
    <span class="k">del</span> <span class="n">sampleDf</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span>
    <span class="c1"># Remove this when these columns are removed from both databases</span>
    <span class="k">if</span> <span class="n">patientDf</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="s2">&quot;BIRTH_YEAR_NUMERICAL&quot;</span><span class="p">)</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">del</span> <span class="n">patientDf</span><span class="p">[</span><span class="s2">&quot;BIRTH_YEAR_NUMERICAL&quot;</span><span class="p">]</span>
    <span class="c1"># del patientDf[&#39;BIRTH_YEAR_NUMERICAL&#39;]</span>

    <span class="n">totalSample</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;PATIENT_ID&quot;</span><span class="p">]</span>
    <span class="n">totalSample</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="n">sampleCols</span><span class="p">)</span>
    <span class="n">sampleCols</span> <span class="o">=</span> <span class="n">totalSample</span>
    <span class="c1"># Make sure to only grab samples that have patient information</span>
    <span class="n">sampleDf</span> <span class="o">=</span> <span class="n">sampleDf</span><span class="p">[</span><span class="n">sampleDf</span><span class="p">[</span><span class="s2">&quot;PATIENT_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">patientDf</span><span class="p">[</span><span class="s2">&quot;PATIENT_ID&quot;</span><span class="p">])]</span>
    <span class="n">clinicalDf</span> <span class="o">=</span> <span class="n">sampleDf</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">patientDf</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s2">&quot;PATIENT_ID&quot;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s2">&quot;outer&quot;</span><span class="p">)</span>
    <span class="c1"># Remove patients without any sample or patient ids</span>
    <span class="n">clinicalDf</span> <span class="o">=</span> <span class="n">clinicalDf</span><span class="p">[</span><span class="o">~</span><span class="n">clinicalDf</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
    <span class="n">clinicalDf</span> <span class="o">=</span> <span class="n">clinicalDf</span><span class="p">[</span><span class="o">~</span><span class="n">clinicalDf</span><span class="p">[</span><span class="s2">&quot;PATIENT_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>

    <span class="p">(</span>
        <span class="n">remove_mafInBed_variants</span><span class="p">,</span>
        <span class="n">removeForMergedConsortiumSamples</span><span class="p">,</span>
        <span class="n">removeForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">flagged_mutationInCis_variants</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">=</span> <span class="n">run_genie_filters</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">variant_filtering_synId</span><span class="p">,</span>
        <span class="n">clinicalDf</span><span class="p">,</span>
        <span class="n">bedDf</span><span class="p">,</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
        <span class="n">processingDate</span><span class="p">,</span>
        <span class="n">skipMutationsInCis</span><span class="p">,</span>
        <span class="n">consortiumReleaseCutOff</span><span class="p">,</span>
        <span class="n">test</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="o">=</span><span class="n">current_release_staging</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="p">(</span>
        <span class="n">clinicalDf</span><span class="p">,</span>
        <span class="n">keepForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">keepForMergedConsortiumSamples</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">=</span> <span class="n">store_clinical_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">clinicalDf</span><span class="p">,</span>
        <span class="n">oncotree_url</span><span class="p">,</span>
        <span class="n">sampleCols</span><span class="p">,</span>
        <span class="n">patientCols</span><span class="p">,</span>
        <span class="n">removeForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">removeForMergedConsortiumSamples</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="p">,</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
        <span class="n">databaseSynIdMappingDf</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="p">[</span><span class="n">sample_used</span><span class="p">,</span> <span class="n">patient_used</span><span class="p">],</span>
    <span class="p">)</span>

    <span class="k">assert</span> <span class="ow">not</span> <span class="n">clinicalDf</span><span class="p">[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">duplicated</span><span class="p">()</span><span class="o">.</span><span class="n">any</span><span class="p">()</span>

    <span class="n">store_maf_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">centerMafFileViewSynId</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">clinicalDf</span><span class="p">[[</span><span class="s2">&quot;SAMPLE_ID&quot;</span><span class="p">,</span> <span class="s2">&quot;CENTER&quot;</span><span class="p">]],</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
        <span class="n">keepForMergedConsortiumSamples</span><span class="p">,</span>
        <span class="n">keepForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">remove_mafInBed_variants</span><span class="p">,</span>
        <span class="n">flagged_mutationInCis_variants</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">cnaSamples</span> <span class="o">=</span> <span class="n">store_cna_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">centerMafFileViewSynId</span><span class="p">,</span>
        <span class="n">keepForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">keepForMergedConsortiumSamples</span><span class="p">,</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">wes_panelids</span> <span class="o">=</span> <span class="n">store_assay_info_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span> <span class="n">genieVersion</span><span class="p">,</span> <span class="n">assay_info_synid</span><span class="p">,</span> <span class="n">clinicalDf</span><span class="p">,</span> <span class="n">consortiumReleaseSynId</span>
    <span class="p">)</span>

    <span class="n">svSamples</span> <span class="o">=</span> <span class="n">store_sv_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">sv_synid</span><span class="p">,</span>
        <span class="n">keepForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">keepForMergedConsortiumSamples</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="p">,</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">data_gene_matrix</span> <span class="o">=</span> <span class="n">store_data_gene_matrix</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">clinicalDf</span><span class="p">,</span>
        <span class="n">cnaSamples</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">wes_panelids</span><span class="p">,</span>
        <span class="n">svSamples</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">genePanelEntities</span> <span class="o">=</span> <span class="n">store_gene_panel_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">fileviewSynId</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">data_gene_matrix</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">wes_panelids</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">store_seg_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">segSynId</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">keepForCenterConsortiumSamples</span><span class="p">,</span>
        <span class="n">keepForMergedConsortiumSamples</span><span class="p">,</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="n">store_bed_files</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span>
        <span class="n">genieVersion</span><span class="p">,</span>
        <span class="n">bedDf</span><span class="p">,</span>
        <span class="n">clinicalDf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">(),</span>
        <span class="n">CENTER_MAPPING_DF</span><span class="p">,</span>
        <span class="n">current_release_staging</span><span class="p">,</span>
        <span class="n">consortiumReleaseSynId</span><span class="p">,</span>
        <span class="n">used</span><span class="o">=</span><span class="n">bed_used</span><span class="p">,</span>
    <span class="p">)</span>

    <span class="k">return</span> <span class="n">genePanelEntities</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.revise_metadata_files" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">revise_metadata_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">consortiumid</span><span class="p">,</span> <span class="n">genie_version</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.revise_metadata_files" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Rewrite metadata files with the correct GENIE version</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>consortiumid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id of consortium release folder</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version, Default to None</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>None</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1966</span>
<span class="normal">1967</span>
<span class="normal">1968</span>
<span class="normal">1969</span>
<span class="normal">1970</span>
<span class="normal">1971</span>
<span class="normal">1972</span>
<span class="normal">1973</span>
<span class="normal">1974</span>
<span class="normal">1975</span>
<span class="normal">1976</span>
<span class="normal">1977</span>
<span class="normal">1978</span>
<span class="normal">1979</span>
<span class="normal">1980</span>
<span class="normal">1981</span>
<span class="normal">1982</span>
<span class="normal">1983</span>
<span class="normal">1984</span>
<span class="normal">1985</span>
<span class="normal">1986</span>
<span class="normal">1987</span>
<span class="normal">1988</span>
<span class="normal">1989</span>
<span class="normal">1990</span>
<span class="normal">1991</span>
<span class="normal">1992</span>
<span class="normal">1993</span>
<span class="normal">1994</span>
<span class="normal">1995</span>
<span class="normal">1996</span>
<span class="normal">1997</span>
<span class="normal">1998</span>
<span class="normal">1999</span>
<span class="normal">2000</span>
<span class="normal">2001</span>
<span class="normal">2002</span>
<span class="normal">2003</span>
<span class="normal">2004</span>
<span class="normal">2005</span>
<span class="normal">2006</span>
<span class="normal">2007</span>
<span class="normal">2008</span>
<span class="normal">2009</span>
<span class="normal">2010</span>
<span class="normal">2011</span>
<span class="normal">2012</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">revise_metadata_files</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">consortiumid</span><span class="p">,</span> <span class="n">genie_version</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Rewrite metadata files with the correct GENIE version</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        consortiumid: Synapse id of consortium release folder</span>
<span class="sd">        genie_version: GENIE version, Default to None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">release_files</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">getChildren</span><span class="p">(</span><span class="n">consortiumid</span><span class="p">)</span>
    <span class="n">meta_file_ents</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">syn</span><span class="o">.</span><span class="n">get</span><span class="p">(</span>
            <span class="n">i</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">],</span> <span class="n">downloadLocation</span><span class="o">=</span><span class="n">GENIE_RELEASE_DIR</span><span class="p">,</span> <span class="n">ifcollision</span><span class="o">=</span><span class="s2">&quot;overwrite.local&quot;</span>
        <span class="p">)</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">release_files</span>
        <span class="k">if</span> <span class="s2">&quot;meta&quot;</span> <span class="ow">in</span> <span class="n">i</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">and</span> <span class="n">i</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;meta_fusions.txt&quot;</span>
    <span class="p">]</span>

    <span class="k">for</span> <span class="n">meta_ent</span> <span class="ow">in</span> <span class="n">meta_file_ents</span><span class="p">:</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">meta_ent</span><span class="o">.</span><span class="n">path</span><span class="p">,</span> <span class="s2">&quot;r+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">meta</span><span class="p">:</span>
            <span class="n">meta_text</span> <span class="o">=</span> <span class="n">meta</span><span class="o">.</span><span class="n">read</span><span class="p">()</span>
            <span class="k">if</span> <span class="s2">&quot;meta_study&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">meta_ent</span><span class="o">.</span><span class="n">path</span><span class="p">:</span>
                <span class="n">version</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">version</span> <span class="o">=</span> <span class="n">re</span><span class="o">.</span><span class="n">search</span><span class="p">(</span><span class="s2">&quot;.+GENIE.+v(.+)&quot;</span><span class="p">,</span> <span class="n">meta_text</span><span class="p">)</span><span class="o">.</span><span class="n">group</span><span class="p">(</span><span class="mi">1</span><span class="p">)</span>
            <span class="c1"># Fix this line</span>
            <span class="n">genie_version</span> <span class="o">=</span> <span class="n">version</span> <span class="k">if</span> <span class="n">genie_version</span> <span class="ow">is</span> <span class="kc">None</span> <span class="k">else</span> <span class="n">genie_version</span>

            <span class="k">if</span> <span class="n">version</span> <span class="o">!=</span> <span class="n">genie_version</span><span class="p">:</span>
                <span class="n">meta_text</span> <span class="o">=</span> <span class="n">meta_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                    <span class="s2">&quot;GENIE Cohort v</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">),</span>
                    <span class="s2">&quot;GENIE Cohort v</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">genie_version</span><span class="p">),</span>
                <span class="p">)</span>

                <span class="n">meta_text</span> <span class="o">=</span> <span class="n">meta_text</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span>
                    <span class="s2">&quot;GENIE v</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">version</span><span class="p">),</span> <span class="s2">&quot;GENIE v</span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">genie_version</span><span class="p">)</span>
                <span class="p">)</span>

                <span class="n">meta</span><span class="o">.</span><span class="n">seek</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
                <span class="n">meta</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">meta_text</span><span class="p">)</span>
                <span class="n">meta</span><span class="o">.</span><span class="n">truncate</span><span class="p">()</span>
        <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
            <span class="n">syn</span><span class="o">=</span><span class="n">syn</span><span class="p">,</span>
            <span class="n">filepath</span><span class="o">=</span><span class="n">meta_ent</span><span class="o">.</span><span class="n">path</span><span class="p">,</span>
            <span class="n">parentid</span><span class="o">=</span><span class="n">consortiumid</span><span class="p">,</span>
            <span class="n">version_comment</span><span class="o">=</span><span class="n">genie_version</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.search_or_create_folder" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">search_or_create_folder</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">parentid</span><span class="p">,</span> <span class="n">folder_name</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.search_or_create_folder" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Searches for an existing Synapse Folder given a parent id
and creates the Synapse folder if it doesn't exist</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse connection</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="synapseclient.Synapse">Synapse</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parentid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse Id of a project or folder</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>folder_name</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Folde rname</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Synapse Folder id</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2016</span>
<span class="normal">2017</span>
<span class="normal">2018</span>
<span class="normal">2019</span>
<span class="normal">2020</span>
<span class="normal">2021</span>
<span class="normal">2022</span>
<span class="normal">2023</span>
<span class="normal">2024</span>
<span class="normal">2025</span>
<span class="normal">2026</span>
<span class="normal">2027</span>
<span class="normal">2028</span>
<span class="normal">2029</span>
<span class="normal">2030</span>
<span class="normal">2031</span>
<span class="normal">2032</span>
<span class="normal">2033</span>
<span class="normal">2034</span>
<span class="normal">2035</span>
<span class="normal">2036</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">search_or_create_folder</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">:</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Synapse</span><span class="p">,</span> <span class="n">parentid</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span> <span class="n">folder_name</span><span class="p">:</span> <span class="nb">str</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Searches for an existing Synapse Folder given a parent id</span>
<span class="sd">    and creates the Synapse folder if it doesn&#39;t exist</span>

<span class="sd">    Args:</span>
<span class="sd">        syn (synapseclient.Synapse): Synapse connection</span>
<span class="sd">        parentid (str): Synapse Id of a project or folder</span>
<span class="sd">        folder_name (str): Folde rname</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Synapse Folder id</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">folder_id</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">findEntityId</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">folder_name</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">parentid</span><span class="p">)</span>
    <span class="c1"># if case_lists doesn&#39;t exist</span>
    <span class="k">if</span> <span class="n">folder_id</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">folder_ent</span> <span class="o">=</span> <span class="n">synapseclient</span><span class="o">.</span><span class="n">Folder</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="n">folder_name</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">parentid</span><span class="p">)</span>
        <span class="n">folder_id</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">store</span><span class="p">(</span><span class="n">folder_ent</span><span class="p">)</span><span class="o">.</span><span class="n">id</span>
    <span class="k">return</span> <span class="n">folder_id</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie.database_to_staging.create_link_version" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_link_version</span><span class="p">(</span><span class="n">syn</span><span class="p">,</span> <span class="n">genie_version</span><span class="p">,</span> <span class="n">case_list_entities</span><span class="p">,</span> <span class="n">gene_panel_entities</span><span class="p">,</span> <span class="n">database_synid_mappingdf</span><span class="p">,</span> <span class="n">release_type</span><span class="o">=</span><span class="s1">&#39;consortium&#39;</span><span class="p">)</span></code>

<a href="#genie.database_to_staging.create_link_version" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create release links from the actual entity and version</p>
<p>TODO: Refactor to use fileviews</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>syn</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse object</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>genie_version</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE version number</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>case_list_entities</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Case list entities</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_panel_entities</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Gene panel entities</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>database_synid_mappingdf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>dataframe containing database to
                      synapse id mapping</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>release_type</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>'consortium' or 'public' release</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>&#39;consortium&#39;</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie/database_to_staging.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">2040</span>
<span class="normal">2041</span>
<span class="normal">2042</span>
<span class="normal">2043</span>
<span class="normal">2044</span>
<span class="normal">2045</span>
<span class="normal">2046</span>
<span class="normal">2047</span>
<span class="normal">2048</span>
<span class="normal">2049</span>
<span class="normal">2050</span>
<span class="normal">2051</span>
<span class="normal">2052</span>
<span class="normal">2053</span>
<span class="normal">2054</span>
<span class="normal">2055</span>
<span class="normal">2056</span>
<span class="normal">2057</span>
<span class="normal">2058</span>
<span class="normal">2059</span>
<span class="normal">2060</span>
<span class="normal">2061</span>
<span class="normal">2062</span>
<span class="normal">2063</span>
<span class="normal">2064</span>
<span class="normal">2065</span>
<span class="normal">2066</span>
<span class="normal">2067</span>
<span class="normal">2068</span>
<span class="normal">2069</span>
<span class="normal">2070</span>
<span class="normal">2071</span>
<span class="normal">2072</span>
<span class="normal">2073</span>
<span class="normal">2074</span>
<span class="normal">2075</span>
<span class="normal">2076</span>
<span class="normal">2077</span>
<span class="normal">2078</span>
<span class="normal">2079</span>
<span class="normal">2080</span>
<span class="normal">2081</span>
<span class="normal">2082</span>
<span class="normal">2083</span>
<span class="normal">2084</span>
<span class="normal">2085</span>
<span class="normal">2086</span>
<span class="normal">2087</span>
<span class="normal">2088</span>
<span class="normal">2089</span>
<span class="normal">2090</span>
<span class="normal">2091</span>
<span class="normal">2092</span>
<span class="normal">2093</span>
<span class="normal">2094</span>
<span class="normal">2095</span>
<span class="normal">2096</span>
<span class="normal">2097</span>
<span class="normal">2098</span>
<span class="normal">2099</span>
<span class="normal">2100</span>
<span class="normal">2101</span>
<span class="normal">2102</span>
<span class="normal">2103</span>
<span class="normal">2104</span>
<span class="normal">2105</span>
<span class="normal">2106</span>
<span class="normal">2107</span>
<span class="normal">2108</span>
<span class="normal">2109</span>
<span class="normal">2110</span>
<span class="normal">2111</span>
<span class="normal">2112</span>
<span class="normal">2113</span>
<span class="normal">2114</span>
<span class="normal">2115</span>
<span class="normal">2116</span>
<span class="normal">2117</span>
<span class="normal">2118</span>
<span class="normal">2119</span>
<span class="normal">2120</span>
<span class="normal">2121</span>
<span class="normal">2122</span>
<span class="normal">2123</span>
<span class="normal">2124</span>
<span class="normal">2125</span>
<span class="normal">2126</span>
<span class="normal">2127</span>
<span class="normal">2128</span>
<span class="normal">2129</span>
<span class="normal">2130</span>
<span class="normal">2131</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_link_version</span><span class="p">(</span>
    <span class="n">syn</span><span class="p">,</span>
    <span class="n">genie_version</span><span class="p">,</span>
    <span class="n">case_list_entities</span><span class="p">,</span>
    <span class="n">gene_panel_entities</span><span class="p">,</span>
    <span class="n">database_synid_mappingdf</span><span class="p">,</span>
    <span class="n">release_type</span><span class="o">=</span><span class="s2">&quot;consortium&quot;</span><span class="p">,</span>
<span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create release links from the actual entity and version</span>

<span class="sd">    TODO: Refactor to use fileviews</span>

<span class="sd">    Args:</span>
<span class="sd">        syn: Synapse object</span>
<span class="sd">        genie_version: GENIE version number</span>
<span class="sd">        case_list_entities: Case list entities</span>
<span class="sd">        gene_panel_entities: Gene panel entities</span>
<span class="sd">        database_synid_mappingdf: dataframe containing database to</span>
<span class="sd">                                  synapse id mapping</span>
<span class="sd">        release_type: &#39;consortium&#39; or &#39;public&#39; release</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Grab major release numbers (ie 1,2,3 ...)</span>
    <span class="n">major_release</span> <span class="o">=</span> <span class="n">genie_version</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">all_releases_synid</span> <span class="o">=</span> <span class="n">database_synid_mappingdf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">database_synid_mappingdf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;release&quot;</span>
    <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># Create major release folder</span>
    <span class="n">major_release_folder_synid</span> <span class="o">=</span> <span class="n">search_or_create_folder</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span> <span class="n">all_releases_synid</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;Release </span><span class="si">{</span><span class="n">major_release</span><span class="si">}</span><span class="s2">&quot;</span>
    <span class="p">)</span>
    <span class="c1"># If the major release folder didn&#39;t exist, go ahead and create the</span>
    <span class="c1"># release folder</span>
    <span class="n">release_folder_synid</span> <span class="o">=</span> <span class="n">search_or_create_folder</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span> <span class="n">major_release_folder_synid</span><span class="p">,</span> <span class="n">genie_version</span>
    <span class="p">)</span>
    <span class="c1"># Search or create case lists folder</span>
    <span class="n">caselist_folder_synid</span> <span class="o">=</span> <span class="n">search_or_create_folder</span><span class="p">(</span>
        <span class="n">syn</span><span class="p">,</span> <span class="n">release_folder_synid</span><span class="p">,</span> <span class="s2">&quot;case_lists&quot;</span>
    <span class="p">)</span>

    <span class="c1"># caselistId = findCaseListId(syn, release_folder_synid)</span>
    <span class="n">consortium_synid</span> <span class="o">=</span> <span class="n">database_synid_mappingdf</span><span class="p">[</span><span class="s2">&quot;Id&quot;</span><span class="p">][</span>
        <span class="n">database_synid_mappingdf</span><span class="p">[</span><span class="s2">&quot;Database&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">release_type</span>
    <span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">consortium_release_files</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">getChildren</span><span class="p">(</span><span class="n">consortium_synid</span><span class="p">)</span>

    <span class="k">for</span> <span class="n">release_file</span> <span class="ow">in</span> <span class="n">consortium_release_files</span><span class="p">:</span>
        <span class="n">not_folder</span> <span class="o">=</span> <span class="n">release_file</span><span class="p">[</span><span class="s2">&quot;type&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;org.sagebionetworks.repo.model.Folder&quot;</span>
        <span class="c1"># data_clinical.txt MUST be pulled in when doing consortium release</span>
        <span class="n">not_public</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">release_file</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;data_clinical.txt&quot;</span> <span class="ow">or</span> <span class="n">release_type</span> <span class="o">==</span> <span class="s2">&quot;consortium&quot;</span>
        <span class="p">)</span>
        <span class="n">is_gene_panel</span> <span class="o">=</span> <span class="n">release_file</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;data_gene_panel&quot;</span><span class="p">)</span>
        <span class="n">is_deprecated_file</span> <span class="o">=</span> <span class="n">release_file</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;data_fusions.txt&quot;</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">not_folder</span> <span class="ow">and</span> <span class="n">not_public</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">is_gene_panel</span> <span class="ow">and</span> <span class="ow">not</span> <span class="n">is_deprecated_file</span><span class="p">:</span>
            <span class="n">syn</span><span class="o">.</span><span class="n">store</span><span class="p">(</span>
                <span class="n">synapseclient</span><span class="o">.</span><span class="n">Link</span><span class="p">(</span>
                    <span class="n">release_file</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">],</span>
                    <span class="n">parent</span><span class="o">=</span><span class="n">release_folder_synid</span><span class="p">,</span>
                    <span class="n">targetVersion</span><span class="o">=</span><span class="n">release_file</span><span class="p">[</span><span class="s2">&quot;versionNumber&quot;</span><span class="p">],</span>
                <span class="p">)</span>
            <span class="p">)</span>

    <span class="n">release_files</span> <span class="o">=</span> <span class="n">syn</span><span class="o">.</span><span class="n">getChildren</span><span class="p">(</span><span class="n">release_folder_synid</span><span class="p">)</span>
    <span class="n">clinical_ent</span> <span class="o">=</span> <span class="p">[</span>
        <span class="n">ents</span><span class="p">[</span><span class="s2">&quot;id&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">ents</span> <span class="ow">in</span> <span class="n">release_files</span> <span class="k">if</span> <span class="n">ents</span><span class="p">[</span><span class="s2">&quot;name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;data_clinical.txt&quot;</span>
    <span class="p">]</span>
    <span class="k">if</span> <span class="n">clinical_ent</span><span class="p">:</span>
        <span class="c1"># Set private permission for the data_clinical.txt link</span>
        <span class="n">syn</span><span class="o">.</span><span class="n">setPermissions</span><span class="p">(</span><span class="n">clinical_ent</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">principalId</span><span class="o">=</span><span class="mi">3326313</span><span class="p">,</span> <span class="n">accessType</span><span class="o">=</span><span class="p">[])</span>

    <span class="k">for</span> <span class="n">ents</span> <span class="ow">in</span> <span class="n">case_list_entities</span><span class="p">:</span>
        <span class="n">syn</span><span class="o">.</span><span class="n">store</span><span class="p">(</span>
            <span class="n">synapseclient</span><span class="o">.</span><span class="n">Link</span><span class="p">(</span>
                <span class="n">ents</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">caselist_folder_synid</span><span class="p">,</span> <span class="n">targetVersion</span><span class="o">=</span><span class="n">ents</span><span class="o">.</span><span class="n">versionNumber</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="c1"># Store gene panels</span>
    <span class="k">for</span> <span class="n">ents</span> <span class="ow">in</span> <span class="n">gene_panel_entities</span><span class="p">:</span>
        <span class="n">syn</span><span class="o">.</span><span class="n">store</span><span class="p">(</span>
            <span class="n">synapseclient</span><span class="o">.</span><span class="n">Link</span><span class="p">(</span>
                <span class="n">ents</span><span class="o">.</span><span class="n">id</span><span class="p">,</span> <span class="n">parent</span><span class="o">=</span><span class="n">release_folder_synid</span><span class="p">,</span> <span class="n">targetVersion</span><span class="o">=</span><span class="n">ents</span><span class="o">.</span><span class="n">versionNumber</span>
            <span class="p">)</span>
        <span class="p">)</span>

    <span class="k">return</span> <span class="p">{</span>
        <span class="s2">&quot;release_folder&quot;</span><span class="p">:</span> <span class="n">release_folder_synid</span><span class="p">,</span>
        <span class="s2">&quot;caselist_folder&quot;</span><span class="p">:</span> <span class="n">caselist_folder_synid</span><span class="p">,</span>
    <span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tabs", "search.highlight", "search.share", "toc.integrate", "content.tabs.link", "navigation.indexes", "navigation.sections", "content.code.copy", "navigation.sections", "navigation.tabs", "navigation.top", "navigation.tracking", "navigation.instant", "header.autohide", "content.action.edit", "content.tabs.link", "content.tabs", "search.highlight", "search.share", "search.suggest", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>