
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="This repository documents code used to gather, QC, standardize, and analyze data uploaded by institutes participating in AACR's Project GENIE (Genomics, Evidence, Neoplasia, Information, Exchange).">
      
      
      
      
        <link rel="prev" href="../assay/">
      
      
        <link rel="next" href="../clinical/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.22">
    
    
      
        <title>Bed - AACR Project GENIE Repository Documentation</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.84d31ad4.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../assets/_mkdocstrings.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="green" data-md-color-accent="green">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#genie_registry.bed" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="AACR Project GENIE Repository Documentation" class="md-header__button md-logo" aria-label="AACR Project GENIE Repository Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            AACR Project GENIE Repository Documentation
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Bed
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="green" data-md-color-accent="green"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3zm3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95zm-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="green" data-md-color-accent="green"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5s-1.65.15-2.39.42zM3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29zm.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14zM20.65 7l-1.77 3.79a7.02 7.02 0 0 0-2.38-4.15zm-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29zM12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
          <a href="javascript:void(0)" class="md-search__icon md-icon" title="Share" aria-label="Share" data-clipboard data-clipboard-text="" data-md-component="search-share" tabindex="-1">
            
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M18 16.08c-.76 0-1.44.3-1.96.77L8.91 12.7c.05-.23.09-.46.09-.7s-.04-.47-.09-.7l7.05-4.11c.54.5 1.25.81 2.04.81a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3c0 .24.04.47.09.7L8.04 9.81C7.5 9.31 6.79 9 6 9a3 3 0 0 0-3 3 3 3 0 0 0 3 3c.79 0 1.5-.31 2.04-.81l7.12 4.15c-.05.21-.08.43-.08.66 0 1.61 1.31 2.91 2.92 2.91s2.92-1.3 2.92-2.91A2.92 2.92 0 0 0 18 16.08"/></svg>
          </a>
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/Sage-Bionetworks/Genie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    AACR Project GENIE
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../.." class="md-tabs__link">
        
  
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
  
    <li class="md-tabs__item">
      <a href="../../../getting_started/" class="md-tabs__link">
        
  
  
    
  
  Getting Started

      </a>
    </li>
  

      
        
  
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../tutorials/local_file_validation/" class="md-tabs__link">
          
  
  
    
  
  Tutorials

        </a>
      </li>
    
  

      
        
  
  
  
    
  
  
    
    
      
  
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../main_pipeline_commands/input_to_database/" class="md-tabs__link">
          
  
  
    
  
  Reference

        </a>
      </li>
    
  

    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


  

<nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="AACR Project GENIE Repository Documentation" class="md-nav__button md-logo" aria-label="AACR Project GENIE Repository Documentation" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    AACR Project GENIE Repository Documentation
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/Sage-Bionetworks/Genie" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 7.1.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2025 Fonticons, Inc.--><path d="M439.6 236.1 244 40.5c-5.4-5.5-12.8-8.5-20.4-8.5s-15 3-20.4 8.4L162.5 81l51.5 51.5c27.1-9.1 52.7 16.8 43.4 43.7l49.7 49.7c34.2-11.8 61.2 31 35.5 56.7-26.5 26.5-70.2-2.9-56-37.3L240.3 199v121.9c25.3 12.5 22.3 41.8 9.1 55-6.4 6.4-15.2 10.1-24.3 10.1s-17.8-3.6-24.3-10.1c-17.6-17.6-11.1-46.9 11.2-56v-123c-20.8-8.5-24.6-30.7-18.6-45L142.6 101 8.5 235.1C3 240.6 0 247.9 0 255.5s3 15 8.5 20.4l195.6 195.7c5.4 5.4 12.7 8.4 20.4 8.4s15-3 20.4-8.4l194.7-194.7c5.4-5.4 8.4-12.8 8.4-20.4s-3-15-8.4-20.4"/></svg>
  </div>
  <div class="md-source__repository">
    AACR Project GENIE
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../.." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../tutorials/local_file_validation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Local File Validation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Reference
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Reference
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_1" >
        
          
          <label class="md-nav__link" for="__nav_4_1" id="__nav_4_1_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Main Pipeline Commands
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_1">
            <span class="md-nav__icon md-icon"></span>
            Main Pipeline Commands
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/input_to_database/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Input to Database
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/database_to_staging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_commands/consortium_to_public/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Consortium to Public
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Main Pipeline Steps
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Main Pipeline Steps
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2_1" >
        
          
          <label class="md-nav__link" for="__nav_4_2_1" id="__nav_4_2_1_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Input to Database
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="3" aria-labelledby="__nav_4_2_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2_1">
            <span class="md-nav__icon md-icon"></span>
            Input to Database
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/input_to_database/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    input_to_database
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/validate.md" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    validate
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/write_invalid_reasons/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    write_invalid_reasons
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/process_mutation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process_mutation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/input_to_database/toRetract.py" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    to_retract
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/database_to_staging/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Database to Staging
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../main_pipeline_steps/consortium_to_public/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Consortium to Public
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_3" checked>
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Fileformats
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Fileformats
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../fileformat/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Fileformat
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../assay/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Assay Information
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Bed
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Bed
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#genie_registry.bed" class="md-nav__link">
    <span class="md-ellipsis">
      bed
    </span>
  </a>
  
    <nav class="md-nav" aria-label="bed">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed-attributes" class="md-nav__link">
    <span class="md-ellipsis">
      Attributes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Attributes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.LOGGER" class="md-nav__link">
    <span class="md-ellipsis">
      LOGGER
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed-classes" class="md-nav__link">
    <span class="md-ellipsis">
      Classes
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Classes">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.bed" class="md-nav__link">
    <span class="md-ellipsis">
      bed
    </span>
  </a>
  
    <nav class="md-nav" aria-label="bed">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.bed-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Functions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.bed.create_gene_panel" class="md-nav__link">
    <span class="md-ellipsis">
      create_gene_panel
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.bed.preprocess" class="md-nav__link">
    <span class="md-ellipsis">
      preprocess
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.bed.process_steps" class="md-nav__link">
    <span class="md-ellipsis">
      process_steps
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed-functions" class="md-nav__link">
    <span class="md-ellipsis">
      Functions
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Functions">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.create_gtf" class="md-nav__link">
    <span class="md-ellipsis">
      create_gtf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed._add_feature_type_tobeddf" class="md-nav__link">
    <span class="md-ellipsis">
      _add_feature_type_tobeddf
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.add_feature_type" class="md-nav__link">
    <span class="md-ellipsis">
      add_feature_type
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed._check_region_overlap" class="md-nav__link">
    <span class="md-ellipsis">
      _check_region_overlap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed._get_max_overlap_index" class="md-nav__link">
    <span class="md-ellipsis">
      _get_max_overlap_index
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed._map_position_within_boundary" class="md-nav__link">
    <span class="md-ellipsis">
      _map_position_within_boundary
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#genie_registry.bed.remap_symbols" class="md-nav__link">
    <span class="md-ellipsis">
      remap_symbols
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../clinical/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Clinical
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cna/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cna
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../maf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Maf
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../mutations_in_cis/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Mutations In Cis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../patient_retraction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Patient retraction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../sample_retraction/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Sample retraction
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Segmented
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../structural_variant/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Structural variant
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../vcf/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Vcf
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../workflow/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Workflow
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_4" >
        
          
          <label class="md-nav__link" for="__nav_4_4" id="__nav_4_4_label" tabindex="">
            
  
  
  <span class="md-ellipsis">
    Helper modules
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_4">
            <span class="md-nav__icon md-icon"></span>
            Helper modules
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/create_case_lists/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    create_case_lists
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/dashboard_table_updater/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    dashboard_table_updater
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/extract/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    extract
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/load/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    load
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/process_functions/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    process_functions
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../helper_modules/transform/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    transform
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
    <a href="https://github.com/Sage-Bionetworks/Genie/edit/master/docs/reference/fileformats/bed.md" title="Edit this page" class="md-content__button md-icon" rel="edit">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M10 20H6V4h7v5h5v3.1l2-2V8l-6-6H6c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h4zm10.2-7c.1 0 .3.1.4.2l1.3 1.3c.2.2.2.6 0 .8l-1 1-2.1-2.1 1-1c.1-.1.2-.2.4-.2m0 3.9L14.1 23H12v-2.1l6.1-6.1z"/></svg>
    </a>
  
  


  <h1>Bed</h1>

<div class="doc doc-object doc-module">



<h2 id="genie_registry.bed" class="doc doc-heading">
            <code>genie_registry.bed</code>


<a href="#genie_registry.bed" class="headerlink" title="Permanent link">&para;</a></h2>

    <div class="doc doc-contents first">

        <p>GENIE bed class and functions</p>










  <div class="doc doc-children">





<h3 id="genie_registry.bed-attributes">Attributes<a href="#genie_registry.bed-attributes" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-attribute">



<h4 id="genie_registry.bed.LOGGER" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">LOGGER</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="vm">__name__</span><span class="p">)</span></code>

  <span class="doc doc-labels">
      <small class="doc doc-label doc-label-module-attribute"><code>module-attribute</code></small>
  </span>

<a href="#genie_registry.bed.LOGGER" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

    </div>

</div>

<h3 id="genie_registry.bed-classes">Classes<a href="#genie_registry.bed-classes" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-class">



<h4 id="genie_registry.bed.bed" class="doc doc-heading">
            <code>bed</code>


<a href="#genie_registry.bed.bed" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">
            <p class="doc doc-class-bases">
              Bases: <code><a class="autorefs autorefs-internal" title="FileTypeFormat (genie.example_filetype_format.FileTypeFormat)" href="../fileformat/#genie.example_filetype_format.FileTypeFormat">FileTypeFormat</a></code></p>


        <p>GENIE bed format</p>








              <details class="quote">
                <summary>Source code in <code>genie_registry/bed.py</code></summary>
                <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">396</span>
<span class="normal">397</span>
<span class="normal">398</span>
<span class="normal">399</span>
<span class="normal">400</span>
<span class="normal">401</span>
<span class="normal">402</span>
<span class="normal">403</span>
<span class="normal">404</span>
<span class="normal">405</span>
<span class="normal">406</span>
<span class="normal">407</span>
<span class="normal">408</span>
<span class="normal">409</span>
<span class="normal">410</span>
<span class="normal">411</span>
<span class="normal">412</span>
<span class="normal">413</span>
<span class="normal">414</span>
<span class="normal">415</span>
<span class="normal">416</span>
<span class="normal">417</span>
<span class="normal">418</span>
<span class="normal">419</span>
<span class="normal">420</span>
<span class="normal">421</span>
<span class="normal">422</span>
<span class="normal">423</span>
<span class="normal">424</span>
<span class="normal">425</span>
<span class="normal">426</span>
<span class="normal">427</span>
<span class="normal">428</span>
<span class="normal">429</span>
<span class="normal">430</span>
<span class="normal">431</span>
<span class="normal">432</span>
<span class="normal">433</span>
<span class="normal">434</span>
<span class="normal">435</span>
<span class="normal">436</span>
<span class="normal">437</span>
<span class="normal">438</span>
<span class="normal">439</span>
<span class="normal">440</span>
<span class="normal">441</span>
<span class="normal">442</span>
<span class="normal">443</span>
<span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span>
<span class="normal">498</span>
<span class="normal">499</span>
<span class="normal">500</span>
<span class="normal">501</span>
<span class="normal">502</span>
<span class="normal">503</span>
<span class="normal">504</span>
<span class="normal">505</span>
<span class="normal">506</span>
<span class="normal">507</span>
<span class="normal">508</span>
<span class="normal">509</span>
<span class="normal">510</span>
<span class="normal">511</span>
<span class="normal">512</span>
<span class="normal">513</span>
<span class="normal">514</span>
<span class="normal">515</span>
<span class="normal">516</span>
<span class="normal">517</span>
<span class="normal">518</span>
<span class="normal">519</span>
<span class="normal">520</span>
<span class="normal">521</span>
<span class="normal">522</span>
<span class="normal">523</span>
<span class="normal">524</span>
<span class="normal">525</span>
<span class="normal">526</span>
<span class="normal">527</span>
<span class="normal">528</span>
<span class="normal">529</span>
<span class="normal">530</span>
<span class="normal">531</span>
<span class="normal">532</span>
<span class="normal">533</span>
<span class="normal">534</span>
<span class="normal">535</span>
<span class="normal">536</span>
<span class="normal">537</span>
<span class="normal">538</span>
<span class="normal">539</span>
<span class="normal">540</span>
<span class="normal">541</span>
<span class="normal">542</span>
<span class="normal">543</span>
<span class="normal">544</span>
<span class="normal">545</span>
<span class="normal">546</span>
<span class="normal">547</span>
<span class="normal">548</span>
<span class="normal">549</span>
<span class="normal">550</span>
<span class="normal">551</span>
<span class="normal">552</span>
<span class="normal">553</span>
<span class="normal">554</span>
<span class="normal">555</span>
<span class="normal">556</span>
<span class="normal">557</span>
<span class="normal">558</span>
<span class="normal">559</span>
<span class="normal">560</span>
<span class="normal">561</span>
<span class="normal">562</span>
<span class="normal">563</span>
<span class="normal">564</span>
<span class="normal">565</span>
<span class="normal">566</span>
<span class="normal">567</span>
<span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span>
<span class="normal">581</span>
<span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span>
<span class="normal">615</span>
<span class="normal">616</span>
<span class="normal">617</span>
<span class="normal">618</span>
<span class="normal">619</span>
<span class="normal">620</span>
<span class="normal">621</span>
<span class="normal">622</span>
<span class="normal">623</span>
<span class="normal">624</span>
<span class="normal">625</span>
<span class="normal">626</span>
<span class="normal">627</span>
<span class="normal">628</span>
<span class="normal">629</span>
<span class="normal">630</span>
<span class="normal">631</span>
<span class="normal">632</span>
<span class="normal">633</span>
<span class="normal">634</span>
<span class="normal">635</span>
<span class="normal">636</span>
<span class="normal">637</span>
<span class="normal">638</span>
<span class="normal">639</span>
<span class="normal">640</span>
<span class="normal">641</span>
<span class="normal">642</span>
<span class="normal">643</span>
<span class="normal">644</span>
<span class="normal">645</span>
<span class="normal">646</span>
<span class="normal">647</span>
<span class="normal">648</span>
<span class="normal">649</span>
<span class="normal">650</span>
<span class="normal">651</span>
<span class="normal">652</span>
<span class="normal">653</span>
<span class="normal">654</span>
<span class="normal">655</span>
<span class="normal">656</span>
<span class="normal">657</span>
<span class="normal">658</span>
<span class="normal">659</span>
<span class="normal">660</span>
<span class="normal">661</span>
<span class="normal">662</span>
<span class="normal">663</span>
<span class="normal">664</span>
<span class="normal">665</span>
<span class="normal">666</span>
<span class="normal">667</span>
<span class="normal">668</span>
<span class="normal">669</span>
<span class="normal">670</span>
<span class="normal">671</span>
<span class="normal">672</span>
<span class="normal">673</span>
<span class="normal">674</span>
<span class="normal">675</span>
<span class="normal">676</span>
<span class="normal">677</span>
<span class="normal">678</span>
<span class="normal">679</span>
<span class="normal">680</span>
<span class="normal">681</span>
<span class="normal">682</span>
<span class="normal">683</span>
<span class="normal">684</span>
<span class="normal">685</span>
<span class="normal">686</span>
<span class="normal">687</span>
<span class="normal">688</span>
<span class="normal">689</span>
<span class="normal">690</span>
<span class="normal">691</span>
<span class="normal">692</span>
<span class="normal">693</span>
<span class="normal">694</span>
<span class="normal">695</span>
<span class="normal">696</span>
<span class="normal">697</span>
<span class="normal">698</span>
<span class="normal">699</span>
<span class="normal">700</span>
<span class="normal">701</span>
<span class="normal">702</span>
<span class="normal">703</span>
<span class="normal">704</span>
<span class="normal">705</span>
<span class="normal">706</span>
<span class="normal">707</span>
<span class="normal">708</span>
<span class="normal">709</span>
<span class="normal">710</span>
<span class="normal">711</span>
<span class="normal">712</span>
<span class="normal">713</span>
<span class="normal">714</span>
<span class="normal">715</span>
<span class="normal">716</span>
<span class="normal">717</span>
<span class="normal">718</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">class</span><span class="w"> </span><span class="nc">bed</span><span class="p">(</span><span class="n">FileTypeFormat</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;GENIE bed format&quot;&quot;&quot;</span>

    <span class="n">_fileType</span> <span class="o">=</span> <span class="s2">&quot;bed&quot;</span>

    <span class="n">_process_kwargs</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;newPath&quot;</span><span class="p">,</span> <span class="s2">&quot;parentId&quot;</span><span class="p">,</span> <span class="s2">&quot;databaseSynId&quot;</span><span class="p">,</span> <span class="s2">&quot;seq_assay_id&quot;</span><span class="p">]</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_get_dataframe</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepathlist</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Bed files don&#39;t have a header</span>

<span class="sd">        Args:</span>
<span class="sd">            filePathList: List of files</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">filepath</span> <span class="o">=</span> <span class="n">filepathlist</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">beddf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">Exception</span><span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Can&#39;t read in your bed file. &quot;</span>
                <span class="s2">&quot;Please make sure the BED file is not binary and &quot;</span>
                <span class="s2">&quot;does not contain a comment/header line&quot;</span>
            <span class="p">)</span>
        <span class="n">first_value</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span>
        <span class="k">if</span> <span class="p">(</span>
            <span class="ow">not</span> <span class="n">first_value</span><span class="o">.</span><span class="n">isdigit</span><span class="p">()</span>
            <span class="ow">and</span> <span class="ow">not</span> <span class="n">first_value</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">)</span>
            <span class="ow">and</span> <span class="n">first_value</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;X&quot;</span><span class="p">,</span> <span class="s2">&quot;Y&quot;</span><span class="p">]</span>
        <span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span>
                <span class="s2">&quot;Please make sure your bed file does not &quot;</span>
                <span class="s2">&quot;contain a comment/header line&quot;</span>
            <span class="p">)</span>
        <span class="k">return</span> <span class="n">beddf</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validateFilename</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">filepath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validates filename</span>
<span class="sd">        CENTER-11.bed</span>

<span class="sd">        Args:</span>
<span class="sd">            filePath: Path to bedfile</span>
<span class="sd">        &quot;&quot;&quot;</span>

        <span class="k">assert</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span>
            <span class="s2">&quot;</span><span class="si">%s</span><span class="s2">-&quot;</span> <span class="o">%</span> <span class="bp">self</span><span class="o">.</span><span class="n">center</span>
        <span class="p">)</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">filepath</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">create_gene_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">parentid</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Create bed file and gene panel files from the bed file</span>

<span class="sd">        Args:</span>
<span class="sd">            beddf: bed dataframe</span>
<span class="sd">            seq_assay_id: GENIE SEQ_ASSAY_ID</span>
<span class="sd">            gene_panel_path: Gene panel folder path</span>
<span class="sd">            parentid: Synapse id of gene panel folder</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: configured bed dataframe</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">LOGGER</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;CREATING GENE PANEL&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">beddf</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">exonsdf</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Feature_Type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">]</span>
            <span class="c1"># Only include genes that should be included in the panels</span>
            <span class="n">include_exonsdf</span> <span class="o">=</span> <span class="n">exonsdf</span><span class="p">[</span><span class="n">exonsdf</span><span class="p">[</span><span class="s2">&quot;includeInPanel&quot;</span><span class="p">]]</span>
            <span class="c1"># Write gene panel</span>
            <span class="n">null_genes</span> <span class="o">=</span> <span class="n">include_exonsdf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span>
            <span class="n">unique_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">include_exonsdf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">][</span><span class="o">~</span><span class="n">null_genes</span><span class="p">])</span>
            <span class="n">gene_panel_text</span> <span class="o">=</span> <span class="p">(</span>
                <span class="s2">&quot;stable_id: </span><span class="si">{seq_assay_id}</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="s2">&quot;description: </span><span class="si">{seq_assay_id}</span><span class="s2">, &quot;</span>
                <span class="s2">&quot;Number of Genes - </span><span class="si">{num_genes}</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="s2">&quot;gene_list:</span><span class="se">\t</span><span class="si">{genelist}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                    <span class="n">seq_assay_id</span><span class="o">=</span><span class="n">seq_assay_id</span><span class="p">,</span>
                    <span class="n">num_genes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">),</span>
                    <span class="n">genelist</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">),</span>
                <span class="p">)</span>
            <span class="p">)</span>
            <span class="n">gene_panel_name</span> <span class="o">=</span> <span class="s2">&quot;data_gene_panel_&quot;</span> <span class="o">+</span> <span class="n">seq_assay_id</span> <span class="o">+</span> <span class="s2">&quot;.txt&quot;</span>

            <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">gene_panel_name</span><span class="p">),</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
                <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene_panel_text</span><span class="p">)</span>

            <span class="n">annotations</span> <span class="o">=</span> <span class="p">{</span>
                <span class="s2">&quot;fileFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;bed&quot;</span><span class="p">,</span>
                <span class="s2">&quot;dataSubType&quot;</span><span class="p">:</span> <span class="s2">&quot;metadata&quot;</span><span class="p">,</span>
                <span class="s2">&quot;cBioFileFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;genePanel&quot;</span><span class="p">,</span>
                <span class="s2">&quot;center&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">center</span><span class="p">,</span>
                <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="s2">&quot;Human&quot;</span><span class="p">,</span>
                <span class="s2">&quot;consortium&quot;</span><span class="p">:</span> <span class="s2">&quot;GENIE&quot;</span><span class="p">,</span>
                <span class="s2">&quot;dataType&quot;</span><span class="p">:</span> <span class="s2">&quot;genomicVariants&quot;</span><span class="p">,</span>
                <span class="s2">&quot;fundingAgency&quot;</span><span class="p">:</span> <span class="s2">&quot;AACR&quot;</span><span class="p">,</span>
                <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="s2">&quot;targetGeneSeq&quot;</span><span class="p">,</span>
                <span class="s2">&quot;fileStage&quot;</span><span class="p">:</span> <span class="s2">&quot;staging&quot;</span><span class="p">,</span>
            <span class="p">}</span>
            <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
                <span class="n">syn</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">syn</span><span class="p">,</span>
                <span class="n">filepath</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">gene_panel_name</span><span class="p">),</span>
                <span class="n">parentid</span><span class="o">=</span><span class="n">parentid</span><span class="p">,</span>
                <span class="n">annotations</span><span class="o">=</span><span class="n">annotations</span><span class="p">,</span>
            <span class="p">)</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_process</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">newpath</span><span class="p">,</span> <span class="n">parentid</span><span class="p">,</span> <span class="n">create_panel</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Process bed file, add feature type</span>

<span class="sd">        Args:</span>
<span class="sd">            gene: bed dataframe</span>
<span class="sd">            seq_assay_id: GENIE SEQ_ASSAY_ID</span>
<span class="sd">            newpath: new GENIE path</span>
<span class="sd">            parentid: Synapse id to store the gene panel</span>
<span class="sd">            create_panel: Create gene panel</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: Conigured bed dataframe</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seq_assay_id</span> <span class="o">=</span> <span class="n">seq_assay_id</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
        <span class="n">seq_assay_id</span> <span class="o">=</span> <span class="n">seq_assay_id</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>

        <span class="c1"># Add in 6th column which is the clinicalReported</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">beddf</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">5</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">])):</span>
                <span class="n">beddf</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">beddf</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">beddf</span><span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="n">beddf</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="mi">4</span><span class="p">,</span> <span class="mi">5</span><span class="p">]]</span>
        <span class="n">gene_panel_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span>
        <span class="c1"># Must be .astype(bool) because `1, 0 in [True, False]`</span>
        <span class="n">beddf</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">bool</span><span class="p">)</span>

        <span class="n">exon_gtf_path</span><span class="p">,</span> <span class="n">gene_gtf_path</span> <span class="o">=</span> <span class="n">create_gtf</span><span class="p">(</span><span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">)</span>
        <span class="n">LOGGER</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;REMAPPING </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">seq_assay_id</span><span class="p">))</span>
        <span class="c1"># bedname = seq_assay_id + &quot;.bed&quot;</span>
        <span class="n">beddf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;End_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span>
            <span class="s2">&quot;includeInPanel&quot;</span><span class="p">,</span>
            <span class="s2">&quot;clinicalReported&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="c1"># Validate gene symbols</span>
        <span class="c1"># Gene symbols can be split by ; and _ and : and .</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
            <span class="n">i</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;.&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>
        <span class="p">]</span>
        <span class="c1"># Replace all chr with blank</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;chr&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">]]</span>
        <span class="c1"># Change all start and end to int</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="nb">int</span><span class="p">)</span>

        <span class="n">gene_position_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">syn</span><span class="o">.</span><span class="n">tableQuery</span><span class="p">(</span><span class="s2">&quot;SELECT * FROM syn11806563&quot;</span><span class="p">)</span>
        <span class="n">gene_positiondf</span> <span class="o">=</span> <span class="n">gene_position_table</span><span class="o">.</span><span class="n">asDataFrame</span><span class="p">()</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>
        <span class="c1"># The apply function of a DataFrame is called twice on the first</span>
        <span class="c1"># row (known pandas behavior)</span>
        <span class="n">beddf</span> <span class="o">=</span> <span class="n">beddf</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">remap_symbols</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">seq_assay_id</span>
        <span class="n">temp_bed_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">,</span> <span class="s2">&quot;temp.bed&quot;</span><span class="p">)</span>
        <span class="n">beddf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">temp_bed_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">final_bed</span> <span class="o">=</span> <span class="n">add_feature_type</span><span class="p">(</span><span class="n">temp_bed_path</span><span class="p">,</span> <span class="n">exon_gtf_path</span><span class="p">,</span> <span class="n">gene_gtf_path</span><span class="p">)</span>
        <span class="n">final_bed</span><span class="p">[</span><span class="s2">&quot;CENTER&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">center</span>
        <span class="n">final_bed</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">final_bed</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">create_panel</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">create_gene_panel</span><span class="p">(</span><span class="n">final_bed</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">parentid</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">final_bed</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">preprocess</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">newpath</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Standardize and grab seq assay id from the bed file path</span>

<span class="sd">        Args:</span>
<span class="sd">            newpath: bed file path</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: GENIE seq assay id</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">seq_assay_id</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
        <span class="n">seq_assay_id</span> <span class="o">=</span> <span class="n">seq_assay_id</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
        <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;seq_assay_id&quot;</span><span class="p">:</span> <span class="n">seq_assay_id</span><span class="p">}</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">process_steps</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">beddf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
        <span class="n">newPath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">parentId</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">databaseSynId</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
        <span class="n">seq_assay_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Process bed file, update bed database, write bed file to path</span>

<span class="sd">        Args:</span>
<span class="sd">            beddf (pd.DataFrame): input bed data</span>
<span class="sd">            newPath (str): Path to new bed file</span>
<span class="sd">            parentId (str): Synapse id to store gene panel file</span>
<span class="sd">            databaseSynId (str): Synapse id of bed database</span>
<span class="sd">            seq_assay_id (str): GENIE seq assay id</span>

<span class="sd">        Returns:</span>
<span class="sd">            str: Path to new bed file</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">final_beddf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process</span><span class="p">(</span>
            <span class="n">beddf</span><span class="o">=</span><span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="o">=</span><span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">newpath</span><span class="o">=</span><span class="n">newPath</span><span class="p">,</span> <span class="n">parentid</span><span class="o">=</span><span class="n">parentId</span>
        <span class="p">)</span>
        <span class="n">load</span><span class="o">.</span><span class="n">update_table</span><span class="p">(</span>
            <span class="n">syn</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">syn</span><span class="p">,</span>
            <span class="n">databaseSynId</span><span class="o">=</span><span class="n">databaseSynId</span><span class="p">,</span>
            <span class="n">newData</span><span class="o">=</span><span class="n">final_beddf</span><span class="p">,</span>
            <span class="n">filterBy</span><span class="o">=</span><span class="n">seq_assay_id</span><span class="p">,</span>
            <span class="n">filterByColumn</span><span class="o">=</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">,</span>
            <span class="n">toDelete</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
        <span class="p">)</span>
        <span class="n">final_beddf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">newPath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">newPath</span>

    <span class="k">def</span><span class="w"> </span><span class="nf">_validate</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beddf</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Validate bed file</span>

<span class="sd">        Args:</span>
<span class="sd">            bed: Bed dataframe</span>

<span class="sd">        Returns:</span>
<span class="sd">            total_error: all the errors</span>
<span class="sd">            warning: all the warnings</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">total_error</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">warning</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">newcols</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;End_Position&quot;</span><span class="p">,</span>
            <span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span>
            <span class="s2">&quot;includeInPanel&quot;</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">beddf</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">newcols</span><span class="p">):</span>
            <span class="n">total_error</span> <span class="o">+=</span> <span class="p">(</span>
                <span class="s2">&quot;BED file: Must at least have five columns in this &quot;</span>
                <span class="s2">&quot;order: </span><span class="si">{}</span><span class="s2">. Make sure there are &quot;</span>
                <span class="s2">&quot;no headers.</span><span class="se">\n</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">newcols</span><span class="p">))</span>
            <span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">newcols</span><span class="o">.</span><span class="n">extend</span><span class="p">(</span><span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">beddf</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">newcols</span><span class="p">)))</span>
            <span class="n">beddf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">newcols</span>
            <span class="n">to_validate_symbol</span> <span class="o">=</span> <span class="kc">True</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">int</span><span class="p">))):</span>
                <span class="n">total_error</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="s2">&quot;BED file: &quot;</span>
                    <span class="s2">&quot;The Start_Position column must only be &quot;</span>
                    <span class="s2">&quot;integers. Make sure there are no headers.</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="n">to_validate_symbol</span> <span class="o">=</span> <span class="kc">False</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">isinstance</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="nb">int</span><span class="p">))):</span>
                <span class="n">total_error</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="s2">&quot;BED file: &quot;</span>
                    <span class="s2">&quot;The End_Position column must only be &quot;</span>
                    <span class="s2">&quot;integers. Make sure there are no headers.</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="p">)</span>
                <span class="n">to_validate_symbol</span> <span class="o">=</span> <span class="kc">False</span>

            <span class="n">LOGGER</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;VALIDATING GENE SYMBOLS&quot;</span><span class="p">)</span>
            <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()):</span>
                <span class="n">total_error</span> <span class="o">+=</span> <span class="s2">&quot;BED file: You cannot submit any null symbols.</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="n">beddf</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="o">~</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()]</span>
            <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span>
                <span class="nb">str</span><span class="p">(</span><span class="n">hugo</span><span class="p">)</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;;&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;:&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">for</span> <span class="n">hugo</span> <span class="ow">in</span> <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>
            <span class="p">]</span>
            <span class="k">if</span> <span class="p">(</span>
                <span class="nb">sum</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;+&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span>
                <span class="ow">or</span> <span class="nb">sum</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span>
            <span class="p">):</span>
                <span class="n">total_error</span> <span class="o">+=</span> <span class="p">(</span>
                    <span class="s2">&quot;BED file: Fourth column must be the &quot;</span>
                    <span class="s2">&quot;Hugo_Symbol column, not the strand column</span><span class="se">\n</span><span class="s2">&quot;</span>
                <span class="p">)</span>

            <span class="n">warn</span><span class="p">,</span> <span class="n">error</span> <span class="o">=</span> <span class="n">process_functions</span><span class="o">.</span><span class="n">check_col_and_values</span><span class="p">(</span>
                <span class="n">beddf</span><span class="p">,</span>
                <span class="s2">&quot;includeInPanel&quot;</span><span class="p">,</span>
                <span class="p">[</span><span class="kc">True</span><span class="p">,</span> <span class="kc">False</span><span class="p">],</span>
                <span class="n">filename</span><span class="o">=</span><span class="s2">&quot;BED file&quot;</span><span class="p">,</span>
                <span class="n">required</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
            <span class="p">)</span>
            <span class="n">warning</span> <span class="o">+=</span> <span class="n">warn</span>
            <span class="n">total_error</span> <span class="o">+=</span> <span class="n">error</span>

            <span class="k">if</span> <span class="n">to_validate_symbol</span><span class="p">:</span>
                <span class="n">gene_position_table</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">syn</span><span class="o">.</span><span class="n">tableQuery</span><span class="p">(</span><span class="s2">&quot;SELECT * FROM syn11806563&quot;</span><span class="p">)</span>
                <span class="n">gene_positiondf</span> <span class="o">=</span> <span class="n">gene_position_table</span><span class="o">.</span><span class="n">asDataFrame</span><span class="p">()</span>
                <span class="c1"># The apply function of a DataFrame is called twice on the first row (known</span>
                <span class="c1"># pandas behavior)</span>
                <span class="n">beddf</span> <span class="o">=</span> <span class="n">beddf</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">remap_symbols</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

                <span class="k">if</span> <span class="nb">any</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()):</span>
                    <span class="n">warning</span> <span class="o">+=</span> <span class="p">(</span>
                        <span class="s2">&quot;BED file: &quot;</span>
                        <span class="s2">&quot;Any gene names that can&#39;t be &quot;</span>
                        <span class="s2">&quot;remapped will be null.</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
                <span class="k">if</span> <span class="nb">all</span><span class="p">(</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()):</span>
                    <span class="n">total_error</span> <span class="o">+=</span> <span class="p">(</span>
                        <span class="s2">&quot;BED file: &quot;</span>
                        <span class="s2">&quot;You have no correct gene symbols. &quot;</span>
                        <span class="s2">&quot;Make sure your gene symbol column (4th &quot;</span>
                        <span class="s2">&quot;column) is formatted like so: SYMBOL&quot;</span>
                        <span class="s2">&quot;(;optionaltext).  Optional text can be &quot;</span>
                        <span class="s2">&quot;semi-colon separated.</span><span class="se">\n</span><span class="s2">&quot;</span>
                    <span class="p">)</span>
            <span class="c1"># validate on required chromosome col, chr is allowed</span>
            <span class="c1"># as we tackle it later in processing</span>
            <span class="n">error</span><span class="p">,</span> <span class="n">warn</span> <span class="o">=</span> <span class="n">validate</span><span class="o">.</span><span class="n">_validate_chromosome</span><span class="p">(</span>
                <span class="n">df</span><span class="o">=</span><span class="n">beddf</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span> <span class="n">fileformat</span><span class="o">=</span><span class="s2">&quot;BED file&quot;</span><span class="p">,</span> <span class="n">allow_chr</span><span class="o">=</span><span class="kc">True</span>
            <span class="p">)</span>
            <span class="n">total_error</span> <span class="o">+=</span> <span class="n">error</span>
            <span class="n">warning</span> <span class="o">+=</span> <span class="n">warn</span>

        <span class="k">return</span> <span class="p">(</span><span class="n">total_error</span><span class="p">,</span> <span class="n">warning</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
              </details>



  <div class="doc doc-children">








<h5 id="genie_registry.bed.bed-functions">Functions<a href="#genie_registry.bed.bed-functions" class="headerlink" title="Permanent link">&para;</a></h5>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.bed.bed.create_gene_panel" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_gene_panel</span><span class="p">(</span><span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">parentid</span><span class="p">)</span></code>

<a href="#genie_registry.bed.bed.create_gene_panel" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>Create bed file and gene panel files from the bed file</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>beddf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>bed dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seq_assay_id</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE SEQ_ASSAY_ID</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_panel_path</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Gene panel folder path</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parentid</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id of gene panel folder</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>pd.DataFrame: configured bed dataframe</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">444</span>
<span class="normal">445</span>
<span class="normal">446</span>
<span class="normal">447</span>
<span class="normal">448</span>
<span class="normal">449</span>
<span class="normal">450</span>
<span class="normal">451</span>
<span class="normal">452</span>
<span class="normal">453</span>
<span class="normal">454</span>
<span class="normal">455</span>
<span class="normal">456</span>
<span class="normal">457</span>
<span class="normal">458</span>
<span class="normal">459</span>
<span class="normal">460</span>
<span class="normal">461</span>
<span class="normal">462</span>
<span class="normal">463</span>
<span class="normal">464</span>
<span class="normal">465</span>
<span class="normal">466</span>
<span class="normal">467</span>
<span class="normal">468</span>
<span class="normal">469</span>
<span class="normal">470</span>
<span class="normal">471</span>
<span class="normal">472</span>
<span class="normal">473</span>
<span class="normal">474</span>
<span class="normal">475</span>
<span class="normal">476</span>
<span class="normal">477</span>
<span class="normal">478</span>
<span class="normal">479</span>
<span class="normal">480</span>
<span class="normal">481</span>
<span class="normal">482</span>
<span class="normal">483</span>
<span class="normal">484</span>
<span class="normal">485</span>
<span class="normal">486</span>
<span class="normal">487</span>
<span class="normal">488</span>
<span class="normal">489</span>
<span class="normal">490</span>
<span class="normal">491</span>
<span class="normal">492</span>
<span class="normal">493</span>
<span class="normal">494</span>
<span class="normal">495</span>
<span class="normal">496</span>
<span class="normal">497</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_gene_panel</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">parentid</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create bed file and gene panel files from the bed file</span>

<span class="sd">    Args:</span>
<span class="sd">        beddf: bed dataframe</span>
<span class="sd">        seq_assay_id: GENIE SEQ_ASSAY_ID</span>
<span class="sd">        gene_panel_path: Gene panel folder path</span>
<span class="sd">        parentid: Synapse id of gene panel folder</span>

<span class="sd">    Returns:</span>
<span class="sd">        pd.DataFrame: configured bed dataframe</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">LOGGER</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;CREATING GENE PANEL&quot;</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">beddf</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">exonsdf</span> <span class="o">=</span> <span class="n">beddf</span><span class="p">[</span><span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Feature_Type&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;exon&quot;</span><span class="p">]</span>
        <span class="c1"># Only include genes that should be included in the panels</span>
        <span class="n">include_exonsdf</span> <span class="o">=</span> <span class="n">exonsdf</span><span class="p">[</span><span class="n">exonsdf</span><span class="p">[</span><span class="s2">&quot;includeInPanel&quot;</span><span class="p">]]</span>
        <span class="c1"># Write gene panel</span>
        <span class="n">null_genes</span> <span class="o">=</span> <span class="n">include_exonsdf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">isnull</span><span class="p">()</span>
        <span class="n">unique_genes</span> <span class="o">=</span> <span class="nb">set</span><span class="p">(</span><span class="n">include_exonsdf</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">][</span><span class="o">~</span><span class="n">null_genes</span><span class="p">])</span>
        <span class="n">gene_panel_text</span> <span class="o">=</span> <span class="p">(</span>
            <span class="s2">&quot;stable_id: </span><span class="si">{seq_assay_id}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="s2">&quot;description: </span><span class="si">{seq_assay_id}</span><span class="s2">, &quot;</span>
            <span class="s2">&quot;Number of Genes - </span><span class="si">{num_genes}</span><span class="se">\n</span><span class="s2">&quot;</span>
            <span class="s2">&quot;gene_list:</span><span class="se">\t</span><span class="si">{genelist}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                <span class="n">seq_assay_id</span><span class="o">=</span><span class="n">seq_assay_id</span><span class="p">,</span>
                <span class="n">num_genes</span><span class="o">=</span><span class="nb">len</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">),</span>
                <span class="n">genelist</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">unique_genes</span><span class="p">),</span>
            <span class="p">)</span>
        <span class="p">)</span>
        <span class="n">gene_panel_name</span> <span class="o">=</span> <span class="s2">&quot;data_gene_panel_&quot;</span> <span class="o">+</span> <span class="n">seq_assay_id</span> <span class="o">+</span> <span class="s2">&quot;.txt&quot;</span>

        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">gene_panel_name</span><span class="p">),</span> <span class="s2">&quot;w+&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
            <span class="n">f</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene_panel_text</span><span class="p">)</span>

        <span class="n">annotations</span> <span class="o">=</span> <span class="p">{</span>
            <span class="s2">&quot;fileFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;bed&quot;</span><span class="p">,</span>
            <span class="s2">&quot;dataSubType&quot;</span><span class="p">:</span> <span class="s2">&quot;metadata&quot;</span><span class="p">,</span>
            <span class="s2">&quot;cBioFileFormat&quot;</span><span class="p">:</span> <span class="s2">&quot;genePanel&quot;</span><span class="p">,</span>
            <span class="s2">&quot;center&quot;</span><span class="p">:</span> <span class="bp">self</span><span class="o">.</span><span class="n">center</span><span class="p">,</span>
            <span class="s2">&quot;species&quot;</span><span class="p">:</span> <span class="s2">&quot;Human&quot;</span><span class="p">,</span>
            <span class="s2">&quot;consortium&quot;</span><span class="p">:</span> <span class="s2">&quot;GENIE&quot;</span><span class="p">,</span>
            <span class="s2">&quot;dataType&quot;</span><span class="p">:</span> <span class="s2">&quot;genomicVariants&quot;</span><span class="p">,</span>
            <span class="s2">&quot;fundingAgency&quot;</span><span class="p">:</span> <span class="s2">&quot;AACR&quot;</span><span class="p">,</span>
            <span class="s2">&quot;assay&quot;</span><span class="p">:</span> <span class="s2">&quot;targetGeneSeq&quot;</span><span class="p">,</span>
            <span class="s2">&quot;fileStage&quot;</span><span class="p">:</span> <span class="s2">&quot;staging&quot;</span><span class="p">,</span>
        <span class="p">}</span>
        <span class="n">load</span><span class="o">.</span><span class="n">store_file</span><span class="p">(</span>
            <span class="n">syn</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">syn</span><span class="p">,</span>
            <span class="n">filepath</span><span class="o">=</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_panel_path</span><span class="p">,</span> <span class="n">gene_panel_name</span><span class="p">),</span>
            <span class="n">parentid</span><span class="o">=</span><span class="n">parentid</span><span class="p">,</span>
            <span class="n">annotations</span><span class="o">=</span><span class="n">annotations</span><span class="p">,</span>
        <span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.bed.bed.preprocess" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">preprocess</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span></code>

<a href="#genie_registry.bed.bed.preprocess" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>Standardize and grab seq assay id from the bed file path</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>newpath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>bed file path</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>dict</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>GENIE seq assay id</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">568</span>
<span class="normal">569</span>
<span class="normal">570</span>
<span class="normal">571</span>
<span class="normal">572</span>
<span class="normal">573</span>
<span class="normal">574</span>
<span class="normal">575</span>
<span class="normal">576</span>
<span class="normal">577</span>
<span class="normal">578</span>
<span class="normal">579</span>
<span class="normal">580</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">preprocess</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">newpath</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Standardize and grab seq assay id from the bed file path</span>

<span class="sd">    Args:</span>
<span class="sd">        newpath: bed file path</span>

<span class="sd">    Returns:</span>
<span class="sd">        dict: GENIE seq assay id</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq_assay_id</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">newpath</span><span class="p">)</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;.bed&quot;</span><span class="p">,</span> <span class="s2">&quot;&quot;</span><span class="p">)</span>
    <span class="n">seq_assay_id</span> <span class="o">=</span> <span class="n">seq_assay_id</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">,</span> <span class="s2">&quot;-&quot;</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">{</span><span class="s2">&quot;seq_assay_id&quot;</span><span class="p">:</span> <span class="n">seq_assay_id</span><span class="p">}</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h6 id="genie_registry.bed.bed.process_steps" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">process_steps</span><span class="p">(</span><span class="n">beddf</span><span class="p">,</span> <span class="n">newPath</span><span class="p">,</span> <span class="n">parentId</span><span class="p">,</span> <span class="n">databaseSynId</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="p">)</span></code>

<a href="#genie_registry.bed.bed.process_steps" class="headerlink" title="Permanent link">&para;</a></h6>


    <div class="doc doc-contents ">

        <p>Process bed file, update bed database, write bed file to path</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>beddf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>input bed data</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="pandas.DataFrame">DataFrame</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>newPath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Path to new bed file</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>parentId</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id to store gene panel file</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>databaseSynId</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Synapse id of bed database</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>seq_assay_id</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>GENIE seq assay id</p>
              </div>
              <p>
                  <span class="doc-param-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>str</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Path to new bed file</p>
              </div>
                <p>
                  <span class="doc-returns-annotation">
                    <b>TYPE:</b>
                      <code><span title="str">str</span></code>
                  </span>
                </p>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">582</span>
<span class="normal">583</span>
<span class="normal">584</span>
<span class="normal">585</span>
<span class="normal">586</span>
<span class="normal">587</span>
<span class="normal">588</span>
<span class="normal">589</span>
<span class="normal">590</span>
<span class="normal">591</span>
<span class="normal">592</span>
<span class="normal">593</span>
<span class="normal">594</span>
<span class="normal">595</span>
<span class="normal">596</span>
<span class="normal">597</span>
<span class="normal">598</span>
<span class="normal">599</span>
<span class="normal">600</span>
<span class="normal">601</span>
<span class="normal">602</span>
<span class="normal">603</span>
<span class="normal">604</span>
<span class="normal">605</span>
<span class="normal">606</span>
<span class="normal">607</span>
<span class="normal">608</span>
<span class="normal">609</span>
<span class="normal">610</span>
<span class="normal">611</span>
<span class="normal">612</span>
<span class="normal">613</span>
<span class="normal">614</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">process_steps</span><span class="p">(</span>
    <span class="bp">self</span><span class="p">,</span>
    <span class="n">beddf</span><span class="p">:</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">,</span>
    <span class="n">newPath</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">parentId</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">databaseSynId</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
    <span class="n">seq_assay_id</span><span class="p">:</span> <span class="nb">str</span><span class="p">,</span>
<span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">str</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Process bed file, update bed database, write bed file to path</span>

<span class="sd">    Args:</span>
<span class="sd">        beddf (pd.DataFrame): input bed data</span>
<span class="sd">        newPath (str): Path to new bed file</span>
<span class="sd">        parentId (str): Synapse id to store gene panel file</span>
<span class="sd">        databaseSynId (str): Synapse id of bed database</span>
<span class="sd">        seq_assay_id (str): GENIE seq assay id</span>

<span class="sd">    Returns:</span>
<span class="sd">        str: Path to new bed file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">final_beddf</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_process</span><span class="p">(</span>
        <span class="n">beddf</span><span class="o">=</span><span class="n">beddf</span><span class="p">,</span> <span class="n">seq_assay_id</span><span class="o">=</span><span class="n">seq_assay_id</span><span class="p">,</span> <span class="n">newpath</span><span class="o">=</span><span class="n">newPath</span><span class="p">,</span> <span class="n">parentid</span><span class="o">=</span><span class="n">parentId</span>
    <span class="p">)</span>
    <span class="n">load</span><span class="o">.</span><span class="n">update_table</span><span class="p">(</span>
        <span class="n">syn</span><span class="o">=</span><span class="bp">self</span><span class="o">.</span><span class="n">syn</span><span class="p">,</span>
        <span class="n">databaseSynId</span><span class="o">=</span><span class="n">databaseSynId</span><span class="p">,</span>
        <span class="n">newData</span><span class="o">=</span><span class="n">final_beddf</span><span class="p">,</span>
        <span class="n">filterBy</span><span class="o">=</span><span class="n">seq_assay_id</span><span class="p">,</span>
        <span class="n">filterByColumn</span><span class="o">=</span><span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">,</span>
        <span class="n">toDelete</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="p">)</span>
    <span class="n">final_beddf</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">newPath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">newPath</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>
<h3 id="genie_registry.bed-functions">Functions<a href="#genie_registry.bed-functions" class="headerlink" title="Permanent link">&para;</a></h3>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed.create_gtf" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">create_gtf</span><span class="p">(</span><span class="n">dirname</span><span class="p">)</span></code>

<a href="#genie_registry.bed.create_gtf" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Create exon.gtf and gene.gtf from GRCh37 gtf</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>dirname</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Directory where these files should live</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>exon_gtf_path</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>exon GTF</p>
              </div>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_gtf_path</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>gene GTF</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">61</span>
<span class="normal">62</span>
<span class="normal">63</span>
<span class="normal">64</span>
<span class="normal">65</span>
<span class="normal">66</span>
<span class="normal">67</span>
<span class="normal">68</span>
<span class="normal">69</span>
<span class="normal">70</span>
<span class="normal">71</span>
<span class="normal">72</span>
<span class="normal">73</span>
<span class="normal">74</span>
<span class="normal">75</span>
<span class="normal">76</span>
<span class="normal">77</span>
<span class="normal">78</span>
<span class="normal">79</span>
<span class="normal">80</span>
<span class="normal">81</span>
<span class="normal">82</span>
<span class="normal">83</span>
<span class="normal">84</span>
<span class="normal">85</span>
<span class="normal">86</span>
<span class="normal">87</span>
<span class="normal">88</span>
<span class="normal">89</span>
<span class="normal">90</span>
<span class="normal">91</span>
<span class="normal">92</span>
<span class="normal">93</span>
<span class="normal">94</span>
<span class="normal">95</span>
<span class="normal">96</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">create_gtf</span><span class="p">(</span><span class="n">dirname</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Create exon.gtf and gene.gtf from GRCh37 gtf</span>

<span class="sd">    Args:</span>
<span class="sd">        dirname: Directory where these files should live</span>

<span class="sd">    Returns:</span>
<span class="sd">        exon_gtf_path: exon GTF</span>
<span class="sd">        gene_gtf_path: gene GTF</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">exon_gtf_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="s2">&quot;exon.gtf&quot;</span><span class="p">)</span>
    <span class="n">gene_gtf_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="s2">&quot;gene.gtf&quot;</span><span class="p">)</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">exon_gtf_path</span><span class="p">)</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="n">gene_gtf_path</span><span class="p">):</span>
        <span class="n">download_cmd</span> <span class="o">=</span> <span class="p">[</span>
            <span class="s2">&quot;wget&quot;</span><span class="p">,</span>
            <span class="s2">&quot;http://ftp.ensembl.org/pub/release-75/gtf/homo_sapiens/Homo_sapiens.GRCh37.75.gtf.gz&quot;</span><span class="p">,</span>
            <span class="s2">&quot;-P&quot;</span><span class="p">,</span>
            <span class="n">dirname</span><span class="p">,</span>
        <span class="p">]</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="n">download_cmd</span><span class="p">)</span>
        <span class="n">gtfgz_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="s2">&quot;Homo_sapiens.GRCh37.75.gtf.gz&quot;</span><span class="p">)</span>
        <span class="n">gunzip_cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;gunzip&quot;</span><span class="p">,</span> <span class="s2">&quot;-f&quot;</span><span class="p">,</span> <span class="n">gtfgz_path</span><span class="p">]</span>
        <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="n">gunzip_cmd</span><span class="p">)</span>
        <span class="n">gtf_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">dirname</span><span class="p">,</span> <span class="s2">&quot;Homo_sapiens.GRCh37.75.gtf&quot;</span><span class="p">)</span>

        <span class="n">exon_awk_cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;awk&quot;</span><span class="p">,</span> <span class="s1">&#39;$3 == &quot;exon&quot; </span><span class="si">{print}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">gtf_path</span><span class="p">]</span>
        <span class="n">exon_gtf</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">exon_awk_cmd</span><span class="p">,</span> <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">exon_gtf_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gtf_file</span><span class="p">:</span>
            <span class="n">gtf_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">exon_gtf</span><span class="p">)</span>
        <span class="n">gene_awk_cmd</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;awk&quot;</span><span class="p">,</span> <span class="s1">&#39;$3 == &quot;gene&quot; </span><span class="si">{print}</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">gtf_path</span><span class="p">]</span>
        <span class="n">gene_gtf</span> <span class="o">=</span> <span class="n">subprocess</span><span class="o">.</span><span class="n">check_output</span><span class="p">(</span><span class="n">gene_awk_cmd</span><span class="p">,</span> <span class="n">universal_newlines</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="n">gene_gtf_path</span><span class="p">,</span> <span class="s2">&quot;w&quot;</span><span class="p">)</span> <span class="k">as</span> <span class="n">gtf_file</span><span class="p">:</span>
            <span class="n">gtf_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">gene_gtf</span><span class="p">)</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">exon_gtf_path</span><span class="p">,</span> <span class="n">gene_gtf_path</span><span class="p">)</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed._add_feature_type_tobeddf" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_add_feature_type_tobeddf</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">featuretype</span><span class="p">)</span></code>

<a href="#genie_registry.bed._add_feature_type_tobeddf" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Add Feature_Type to dataframe</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>filepath</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>path to bed</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>featuretype</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>exon, intron, or intergenic</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>df</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>empty dataframe or dataframe with appended feature type</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 99</span>
<span class="normal">100</span>
<span class="normal">101</span>
<span class="normal">102</span>
<span class="normal">103</span>
<span class="normal">104</span>
<span class="normal">105</span>
<span class="normal">106</span>
<span class="normal">107</span>
<span class="normal">108</span>
<span class="normal">109</span>
<span class="normal">110</span>
<span class="normal">111</span>
<span class="normal">112</span>
<span class="normal">113</span>
<span class="normal">114</span>
<span class="normal">115</span>
<span class="normal">116</span>
<span class="normal">117</span>
<span class="normal">118</span>
<span class="normal">119</span>
<span class="normal">120</span>
<span class="normal">121</span>
<span class="normal">122</span>
<span class="normal">123</span>
<span class="normal">124</span>
<span class="normal">125</span>
<span class="normal">126</span>
<span class="normal">127</span>
<span class="normal">128</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_add_feature_type_tobeddf</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">featuretype</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add Feature_Type to dataframe</span>

<span class="sd">    Args:</span>
<span class="sd">        filepath: path to bed</span>
<span class="sd">        featuretype: exon, intron, or intergenic</span>

<span class="sd">    Returns:</span>
<span class="sd">        df: empty dataframe or dataframe with appended feature type</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">bed_columns</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;Chromosome&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Start_Position&quot;</span><span class="p">,</span>
        <span class="s2">&quot;End_Position&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">,</span>
        <span class="s2">&quot;includeInPanel&quot;</span><span class="p">,</span>
        <span class="s2">&quot;clinicalReported&quot;</span><span class="p">,</span>
        <span class="s2">&quot;ID&quot;</span><span class="p">,</span>
        <span class="s2">&quot;SEQ_ASSAY_ID&quot;</span><span class="p">,</span>
        <span class="s2">&quot;Feature_Type&quot;</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="c1"># No need to add anything if the dataframe is empty</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">stat</span><span class="p">(</span><span class="n">filepath</span><span class="p">)</span><span class="o">.</span><span class="n">st_size</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">beddf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">filepath</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">)</span>
        <span class="n">beddf</span><span class="p">[</span><span class="s2">&quot;Feature_Type&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">featuretype</span>
        <span class="n">beddf</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="n">bed_columns</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">beddf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">columns</span><span class="o">=</span><span class="n">bed_columns</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">beddf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed.add_feature_type" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">add_feature_type</span><span class="p">(</span><span class="n">temp_bed_path</span><span class="p">,</span> <span class="n">exon_gtf_path</span><span class="p">,</span> <span class="n">gene_gtf_path</span><span class="p">)</span></code>

<a href="#genie_registry.bed.add_feature_type" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Add Feature_Type to bed file (exon, intron, intergenic)</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>temp_bed_path</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>BED file without feature type</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>exon_gtf_path</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>exon gtf</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_gtf_path</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>gene gtf</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>genie_combined_path</code>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Path to final bed file</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">131</span>
<span class="normal">132</span>
<span class="normal">133</span>
<span class="normal">134</span>
<span class="normal">135</span>
<span class="normal">136</span>
<span class="normal">137</span>
<span class="normal">138</span>
<span class="normal">139</span>
<span class="normal">140</span>
<span class="normal">141</span>
<span class="normal">142</span>
<span class="normal">143</span>
<span class="normal">144</span>
<span class="normal">145</span>
<span class="normal">146</span>
<span class="normal">147</span>
<span class="normal">148</span>
<span class="normal">149</span>
<span class="normal">150</span>
<span class="normal">151</span>
<span class="normal">152</span>
<span class="normal">153</span>
<span class="normal">154</span>
<span class="normal">155</span>
<span class="normal">156</span>
<span class="normal">157</span>
<span class="normal">158</span>
<span class="normal">159</span>
<span class="normal">160</span>
<span class="normal">161</span>
<span class="normal">162</span>
<span class="normal">163</span>
<span class="normal">164</span>
<span class="normal">165</span>
<span class="normal">166</span>
<span class="normal">167</span>
<span class="normal">168</span>
<span class="normal">169</span>
<span class="normal">170</span>
<span class="normal">171</span>
<span class="normal">172</span>
<span class="normal">173</span>
<span class="normal">174</span>
<span class="normal">175</span>
<span class="normal">176</span>
<span class="normal">177</span>
<span class="normal">178</span>
<span class="normal">179</span>
<span class="normal">180</span>
<span class="normal">181</span>
<span class="normal">182</span>
<span class="normal">183</span>
<span class="normal">184</span>
<span class="normal">185</span>
<span class="normal">186</span>
<span class="normal">187</span>
<span class="normal">188</span>
<span class="normal">189</span>
<span class="normal">190</span>
<span class="normal">191</span>
<span class="normal">192</span>
<span class="normal">193</span>
<span class="normal">194</span>
<span class="normal">195</span>
<span class="normal">196</span>
<span class="normal">197</span>
<span class="normal">198</span>
<span class="normal">199</span>
<span class="normal">200</span>
<span class="normal">201</span>
<span class="normal">202</span>
<span class="normal">203</span>
<span class="normal">204</span>
<span class="normal">205</span>
<span class="normal">206</span>
<span class="normal">207</span>
<span class="normal">208</span>
<span class="normal">209</span>
<span class="normal">210</span>
<span class="normal">211</span>
<span class="normal">212</span>
<span class="normal">213</span>
<span class="normal">214</span>
<span class="normal">215</span>
<span class="normal">216</span>
<span class="normal">217</span>
<span class="normal">218</span>
<span class="normal">219</span>
<span class="normal">220</span>
<span class="normal">221</span>
<span class="normal">222</span>
<span class="normal">223</span>
<span class="normal">224</span>
<span class="normal">225</span>
<span class="normal">226</span>
<span class="normal">227</span>
<span class="normal">228</span>
<span class="normal">229</span>
<span class="normal">230</span>
<span class="normal">231</span>
<span class="normal">232</span>
<span class="normal">233</span>
<span class="normal">234</span>
<span class="normal">235</span>
<span class="normal">236</span>
<span class="normal">237</span>
<span class="normal">238</span>
<span class="normal">239</span>
<span class="normal">240</span>
<span class="normal">241</span>
<span class="normal">242</span>
<span class="normal">243</span>
<span class="normal">244</span>
<span class="normal">245</span>
<span class="normal">246</span>
<span class="normal">247</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">add_feature_type</span><span class="p">(</span><span class="n">temp_bed_path</span><span class="p">,</span> <span class="n">exon_gtf_path</span><span class="p">,</span> <span class="n">gene_gtf_path</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Add Feature_Type to bed file (exon, intron, intergenic)</span>

<span class="sd">    Args:</span>
<span class="sd">        temp_bed_path: BED file without feature type</span>
<span class="sd">        exon_gtf_path: exon gtf</span>
<span class="sd">        gene_gtf_path: gene gtf</span>

<span class="sd">    Returns:</span>
<span class="sd">        genie_combined_path: Path to final bed file</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genie_exon_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">,</span> <span class="s2">&quot;genie_exons.bed&quot;</span><span class="p">)</span>
    <span class="n">genie_intron_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">,</span> <span class="s2">&quot;genie_introns.bed&quot;</span><span class="p">)</span>
    <span class="n">genie_intergenic_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">,</span> <span class="s2">&quot;genie_intergenic.bed&quot;</span>
    <span class="p">)</span>
    <span class="n">intron_intergenic_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span>
        <span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">,</span> <span class="s2">&quot;intron_intergenic.bed&quot;</span>
    <span class="p">)</span>
    <span class="n">gene_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">process_functions</span><span class="o">.</span><span class="n">SCRIPT_DIR</span><span class="p">,</span> <span class="s2">&quot;gene.bed&quot;</span><span class="p">)</span>
    <span class="c1"># GET EXON REGIONS</span>
    <span class="c1"># Get intersection between true exon regions and submitted bed regions</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;bedtools&quot;</span><span class="p">,</span>
        <span class="s2">&quot;intersect&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-a&quot;</span><span class="p">,</span>
        <span class="n">temp_bed_path</span><span class="p">,</span>
        <span class="s2">&quot;-b&quot;</span><span class="p">,</span>
        <span class="n">exon_gtf_path</span><span class="p">,</span>
        <span class="s2">&quot;-wa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sort&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uniq&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">genie_exon_path</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">),</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># get intergenic/intron regions</span>
    <span class="c1"># Get opposite of intersection between true exon regions and submitted</span>
    <span class="c1"># bed regions</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;bedtools&quot;</span><span class="p">,</span>
        <span class="s2">&quot;intersect&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-a&quot;</span><span class="p">,</span>
        <span class="n">temp_bed_path</span><span class="p">,</span>
        <span class="s2">&quot;-b&quot;</span><span class="p">,</span>
        <span class="n">exon_gtf_path</span><span class="p">,</span>
        <span class="s2">&quot;-wa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-v&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sort&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uniq&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">intron_intergenic_path</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">),</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># get gene regions</span>
    <span class="c1"># Get intersection between true gene regions and submitted bed regions</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;bedtools&quot;</span><span class="p">,</span>
        <span class="s2">&quot;intersect&quot;</span><span class="p">,</span>
        <span class="s2">&quot;-a&quot;</span><span class="p">,</span>
        <span class="n">temp_bed_path</span><span class="p">,</span>
        <span class="s2">&quot;-b&quot;</span><span class="p">,</span>
        <span class="n">gene_gtf_path</span><span class="p">,</span>
        <span class="s2">&quot;-wa&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sort&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;uniq&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">gene_path</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">),</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># GET INTRON REGSIONS</span>
    <span class="c1"># Difference between the gene regions and exon regions will give</span>
    <span class="c1"># intron regions</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;diff&quot;</span><span class="p">,</span>
        <span class="n">gene_path</span><span class="p">,</span>
        <span class="n">genie_exon_path</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;grep&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&#39;&lt;&#39;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&#39;s/&lt; //&#39;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">genie_intron_path</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">),</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
    <span class="c1"># GET INTERGENIC REGIONS</span>
    <span class="c1"># Difference between the intron/intergenic and intron regions will</span>
    <span class="c1"># give intergenic regions</span>
    <span class="n">command</span> <span class="o">=</span> <span class="p">[</span>
        <span class="s2">&quot;diff&quot;</span><span class="p">,</span>
        <span class="n">intron_intergenic_path</span><span class="p">,</span>
        <span class="n">genie_intron_path</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;grep&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&#39;&lt;&#39;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;|&quot;</span><span class="p">,</span>
        <span class="s2">&quot;sed&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&#39;s/&lt; //&#39;&quot;</span><span class="p">,</span>
        <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span>
        <span class="n">genie_intergenic_path</span><span class="p">,</span>
    <span class="p">]</span>
    <span class="n">subprocess</span><span class="o">.</span><span class="n">check_call</span><span class="p">(</span><span class="s2">&quot; &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">command</span><span class="p">),</span> <span class="n">shell</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">genie_exondf</span> <span class="o">=</span> <span class="n">_add_feature_type_tobeddf</span><span class="p">(</span><span class="n">genie_exon_path</span><span class="p">,</span> <span class="s2">&quot;exon&quot;</span><span class="p">)</span>
    <span class="n">genie_introndf</span> <span class="o">=</span> <span class="n">_add_feature_type_tobeddf</span><span class="p">(</span><span class="n">genie_intron_path</span><span class="p">,</span> <span class="s2">&quot;intron&quot;</span><span class="p">)</span>
    <span class="n">genie_intergenicdf</span> <span class="o">=</span> <span class="n">_add_feature_type_tobeddf</span><span class="p">(</span><span class="n">genie_intergenic_path</span><span class="p">,</span> <span class="s2">&quot;intergenic&quot;</span><span class="p">)</span>
    <span class="n">genie_combineddf</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">genie_exondf</span><span class="p">,</span> <span class="n">genie_introndf</span><span class="p">,</span> <span class="n">genie_intergenicdf</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">genie_combineddf</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed._check_region_overlap" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_check_region_overlap</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">)</span></code>

<a href="#genie_registry.bed._check_region_overlap" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Check if the submitted bed symbol + region overlaps with
the actual gene's positions</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>row</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>row in bed file (genomic region)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_positiondf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Reference gene position dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<details class="return" open>
  <summary>Return</summary>
  <p>True if the region does overlap.</p>
</details>

            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">250</span>
<span class="normal">251</span>
<span class="normal">252</span>
<span class="normal">253</span>
<span class="normal">254</span>
<span class="normal">255</span>
<span class="normal">256</span>
<span class="normal">257</span>
<span class="normal">258</span>
<span class="normal">259</span>
<span class="normal">260</span>
<span class="normal">261</span>
<span class="normal">262</span>
<span class="normal">263</span>
<span class="normal">264</span>
<span class="normal">265</span>
<span class="normal">266</span>
<span class="normal">267</span>
<span class="normal">268</span>
<span class="normal">269</span>
<span class="normal">270</span>
<span class="normal">271</span>
<span class="normal">272</span>
<span class="normal">273</span>
<span class="normal">274</span>
<span class="normal">275</span>
<span class="normal">276</span>
<span class="normal">277</span>
<span class="normal">278</span>
<span class="normal">279</span>
<span class="normal">280</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_check_region_overlap</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Check if the submitted bed symbol + region overlaps with</span>
<span class="sd">    the actual gene&#39;s positions</span>

<span class="sd">    Args:</span>
<span class="sd">        row: row in bed file (genomic region)</span>
<span class="sd">        gene_positiondf: Reference gene position dataframe</span>

<span class="sd">    Return:</span>
<span class="sd">        True if the region does overlap.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">matching_symbol_ind</span> <span class="o">=</span> <span class="n">gene_positiondf</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span>
    <span class="n">match_with_genedb</span> <span class="o">=</span> <span class="n">gene_positiondf</span><span class="p">[</span><span class="n">matching_symbol_ind</span><span class="p">]</span>

    <span class="k">if</span> <span class="ow">not</span> <span class="n">match_with_genedb</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="c1"># We are assuming that there is only one matching gene, but</span>
        <span class="c1"># there are actually duplicated gene symbols</span>
        <span class="n">start_position</span> <span class="o">=</span> <span class="n">match_with_genedb</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">end_position</span> <span class="o">=</span> <span class="n">match_with_genedb</span><span class="p">[</span><span class="s2">&quot;end_position&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
        <span class="c1"># Submitted bed start position in a gene</span>
        <span class="n">start_in_gene</span> <span class="o">=</span> <span class="n">start_position</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">end_position</span>
        <span class="c1"># Submitted bed end position in a gene</span>
        <span class="n">end_in_gene</span> <span class="o">=</span> <span class="n">start_position</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">end_position</span>
        <span class="c1"># Submitted bed region surrounds a gene</span>
        <span class="n">gene_in_region</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">end_position</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span>
            <span class="ow">and</span> <span class="n">start_position</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span>
        <span class="p">)</span>
        <span class="k">return</span> <span class="n">start_in_gene</span> <span class="ow">or</span> <span class="n">end_in_gene</span> <span class="ow">or</span> <span class="n">gene_in_region</span>
    <span class="k">return</span> <span class="kc">False</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed._get_max_overlap_index" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_get_max_overlap_index</span><span class="p">(</span><span class="n">overlap</span><span class="p">,</span> <span class="n">bed_length</span><span class="p">,</span> <span class="n">boundary</span><span class="p">)</span></code>

<a href="#genie_registry.bed._get_max_overlap_index" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Calculate the ratio of overlap between the submitted bed
region and gene position dataframe and return the index of
the max overlapping region</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>overlap</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Possible overlapping region</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>bed_length</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Length of submitted region</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>boundary</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>specified ratio overlap</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">RETURNS</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
            </td>
            <td class="doc-returns-details">
              <div class="doc-md-description">
                <p>Index of regions with maximum overlap or None</p>
              </div>
            </td>
          </tr>
      </tbody>
    </table>


            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">283</span>
<span class="normal">284</span>
<span class="normal">285</span>
<span class="normal">286</span>
<span class="normal">287</span>
<span class="normal">288</span>
<span class="normal">289</span>
<span class="normal">290</span>
<span class="normal">291</span>
<span class="normal">292</span>
<span class="normal">293</span>
<span class="normal">294</span>
<span class="normal">295</span>
<span class="normal">296</span>
<span class="normal">297</span>
<span class="normal">298</span>
<span class="normal">299</span>
<span class="normal">300</span>
<span class="normal">301</span>
<span class="normal">302</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_get_max_overlap_index</span><span class="p">(</span><span class="n">overlap</span><span class="p">,</span> <span class="n">bed_length</span><span class="p">,</span> <span class="n">boundary</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Calculate the ratio of overlap between the submitted bed</span>
<span class="sd">    region and gene position dataframe and return the index of</span>
<span class="sd">    the max overlapping region</span>

<span class="sd">    Args:</span>
<span class="sd">        overlap: Possible overlapping region</span>
<span class="sd">        bed_length: Length of submitted region</span>
<span class="sd">        boundary: specified ratio overlap</span>

<span class="sd">    Returns:</span>
<span class="sd">        Index of regions with maximum overlap or None</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">ratio_overlap</span> <span class="o">=</span> <span class="n">overlap</span> <span class="o">/</span> <span class="n">bed_length</span>
    <span class="n">ratio_overlap</span> <span class="o">=</span> <span class="n">ratio_overlap</span><span class="p">[</span><span class="n">ratio_overlap</span> <span class="o">&gt;</span> <span class="n">boundary</span><span class="p">]</span>
    <span class="n">ratio_overlap</span> <span class="o">=</span> <span class="n">ratio_overlap</span><span class="p">[</span><span class="n">ratio_overlap</span> <span class="o">&lt;=</span> <span class="mi">1</span><span class="p">]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">ratio_overlap</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">ratio_overlap</span><span class="o">.</span><span class="n">idxmax</span><span class="p">()</span>
    <span class="k">return</span> <span class="kc">None</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed._map_position_within_boundary" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">_map_position_within_boundary</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">positiondf</span><span class="p">,</span> <span class="n">boundary</span><span class="o">=</span><span class="mf">0.9</span><span class="p">)</span></code>

<a href="#genie_registry.bed._map_position_within_boundary" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Map positions and checks if posision is contained within the
specified percentage boundary</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>row</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Row in bed file (genomic region)</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>positiondf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Reference bed position dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>boundary</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Percent boundary defined</p>
              </div>
              <p>
                  <span class="doc-param-default">
                    <b>DEFAULT:</b>
                      <code>0.9</code>
                  </span>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<details class="return" open>
  <summary>Return</summary>
  <p>pd.Series: mapped position</p>
</details>

            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">305</span>
<span class="normal">306</span>
<span class="normal">307</span>
<span class="normal">308</span>
<span class="normal">309</span>
<span class="normal">310</span>
<span class="normal">311</span>
<span class="normal">312</span>
<span class="normal">313</span>
<span class="normal">314</span>
<span class="normal">315</span>
<span class="normal">316</span>
<span class="normal">317</span>
<span class="normal">318</span>
<span class="normal">319</span>
<span class="normal">320</span>
<span class="normal">321</span>
<span class="normal">322</span>
<span class="normal">323</span>
<span class="normal">324</span>
<span class="normal">325</span>
<span class="normal">326</span>
<span class="normal">327</span>
<span class="normal">328</span>
<span class="normal">329</span>
<span class="normal">330</span>
<span class="normal">331</span>
<span class="normal">332</span>
<span class="normal">333</span>
<span class="normal">334</span>
<span class="normal">335</span>
<span class="normal">336</span>
<span class="normal">337</span>
<span class="normal">338</span>
<span class="normal">339</span>
<span class="normal">340</span>
<span class="normal">341</span>
<span class="normal">342</span>
<span class="normal">343</span>
<span class="normal">344</span>
<span class="normal">345</span>
<span class="normal">346</span>
<span class="normal">347</span>
<span class="normal">348</span>
<span class="normal">349</span>
<span class="normal">350</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">_map_position_within_boundary</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">positiondf</span><span class="p">,</span> <span class="n">boundary</span><span class="o">=</span><span class="mf">0.9</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Map positions and checks if posision is contained within the</span>
<span class="sd">    specified percentage boundary</span>

<span class="sd">    Args:</span>
<span class="sd">        row: Row in bed file (genomic region)</span>
<span class="sd">        positiondf: Reference bed position dataframe</span>
<span class="sd">        boundary: Percent boundary defined</span>

<span class="sd">    Return:</span>
<span class="sd">        pd.Series: mapped position</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># First filter just based on whether or not region is within the gene</span>
    <span class="c1"># position dataframe</span>
    <span class="n">match_chr</span> <span class="o">=</span> <span class="n">positiondf</span><span class="p">[</span><span class="s2">&quot;chromosome_name&quot;</span><span class="p">]</span> <span class="o">==</span> <span class="nb">str</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Chromosome&quot;</span><span class="p">])</span>
    <span class="n">chrom_rows</span> <span class="o">=</span> <span class="n">positiondf</span><span class="p">[</span><span class="n">match_chr</span><span class="p">]</span>
    <span class="n">match_start</span> <span class="o">=</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span>
    <span class="n">start_rows</span> <span class="o">=</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="n">match_start</span><span class="p">]</span>
    <span class="n">end_rows</span> <span class="o">=</span> <span class="n">start_rows</span><span class="p">[</span><span class="n">start_rows</span><span class="p">[</span><span class="s2">&quot;end_position&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]]</span>

    <span class="n">bed_length</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span>
    <span class="c1"># as long as the strand is within the boundary % defined</span>
    <span class="c1"># Start goes over start boundary, but end is contained in position</span>
    <span class="k">if</span> <span class="n">end_rows</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;end_position&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">overlap</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span>
            <span class="c1"># difference =  difference * -1.0</span>
            <span class="n">max_overlap</span> <span class="o">=</span> <span class="n">_get_max_overlap_index</span><span class="p">(</span><span class="n">overlap</span><span class="p">,</span> <span class="n">bed_length</span><span class="p">,</span> <span class="n">boundary</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">max_overlap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">end_rows</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">end_rows</span><span class="p">,</span> <span class="n">chrom_rows</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">max_overlap</span><span class="p">]]])</span>
        <span class="c1"># End goes over end boundary, but start is contained in position</span>
        <span class="k">if</span> <span class="nb">sum</span><span class="p">(</span><span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">])</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
            <span class="n">overlap</span> <span class="o">=</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;end_position&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]</span>
            <span class="n">max_overlap</span> <span class="o">=</span> <span class="n">_get_max_overlap_index</span><span class="p">(</span><span class="n">overlap</span><span class="p">,</span> <span class="n">bed_length</span><span class="p">,</span> <span class="n">boundary</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">max_overlap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">end_rows</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">end_rows</span><span class="p">,</span> <span class="n">chrom_rows</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">max_overlap</span><span class="p">]]])</span>
        <span class="c1"># Start and end go over position boundary</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Start_Position&quot;</span><span class="p">]]</span>
        <span class="n">check</span> <span class="o">=</span> <span class="n">check</span><span class="p">[</span><span class="n">check</span><span class="p">[</span><span class="s2">&quot;end_position&quot;</span><span class="p">]</span> <span class="o">&lt;=</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;End_Position&quot;</span><span class="p">]]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">check</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">overlap</span> <span class="o">=</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;end_position&quot;</span><span class="p">]</span> <span class="o">-</span> <span class="n">chrom_rows</span><span class="p">[</span><span class="s2">&quot;start_position&quot;</span><span class="p">]</span>
            <span class="n">max_overlap</span> <span class="o">=</span> <span class="n">_get_max_overlap_index</span><span class="p">(</span><span class="n">overlap</span><span class="p">,</span> <span class="n">bed_length</span><span class="p">,</span> <span class="n">boundary</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">max_overlap</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="n">end_rows</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">end_rows</span><span class="p">,</span> <span class="n">chrom_rows</span><span class="o">.</span><span class="n">loc</span><span class="p">[[</span><span class="n">max_overlap</span><span class="p">]]])</span>
    <span class="k">return</span> <span class="n">end_rows</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>

<div class="doc doc-object doc-function">


<h4 id="genie_registry.bed.remap_symbols" class="doc doc-heading">
            <code class="highlight language-python"><span class="n">remap_symbols</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">)</span></code>

<a href="#genie_registry.bed.remap_symbols" class="headerlink" title="Permanent link">&para;</a></h4>


    <div class="doc doc-contents ">

        <p>Remap hugo symbols if there is no overlap between submitted bed region and
gene positions.</p>


<table>
      <thead>
        <tr>
          <th><span class="doc-section-title">PARAMETER</span></th>
          <th><span>DESCRIPTION</span></th>
        </tr>
      </thead>
      <tbody>
          <tr class="doc-section-item">
            <td>
                <code>row</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>start and end position</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
          <tr class="doc-section-item">
            <td>
                <code>gene_positiondf</code>
            </td>
            <td class="doc-param-details">
              <div class="doc-md-description">
                <p>Actual gene position dataframe</p>
              </div>
              <p>
              </p>
            </td>
          </tr>
      </tbody>
    </table>


<details class="return" open>
  <summary>Return</summary>
  <p>bool or Series: if the gene passed in need to be remapped or
                the remapped gene</p>
</details>

            <details class="quote">
              <summary>Source code in <code>genie_registry/bed.py</code></summary>
              <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">353</span>
<span class="normal">354</span>
<span class="normal">355</span>
<span class="normal">356</span>
<span class="normal">357</span>
<span class="normal">358</span>
<span class="normal">359</span>
<span class="normal">360</span>
<span class="normal">361</span>
<span class="normal">362</span>
<span class="normal">363</span>
<span class="normal">364</span>
<span class="normal">365</span>
<span class="normal">366</span>
<span class="normal">367</span>
<span class="normal">368</span>
<span class="normal">369</span>
<span class="normal">370</span>
<span class="normal">371</span>
<span class="normal">372</span>
<span class="normal">373</span>
<span class="normal">374</span>
<span class="normal">375</span>
<span class="normal">376</span>
<span class="normal">377</span>
<span class="normal">378</span>
<span class="normal">379</span>
<span class="normal">380</span>
<span class="normal">381</span>
<span class="normal">382</span>
<span class="normal">383</span>
<span class="normal">384</span>
<span class="normal">385</span>
<span class="normal">386</span>
<span class="normal">387</span>
<span class="normal">388</span>
<span class="normal">389</span>
<span class="normal">390</span>
<span class="normal">391</span>
<span class="normal">392</span>
<span class="normal">393</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="k">def</span><span class="w"> </span><span class="nf">remap_symbols</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Remap hugo symbols if there is no overlap between submitted bed region and</span>
<span class="sd">    gene positions.</span>

<span class="sd">    Args:</span>
<span class="sd">        row: start and end position</span>
<span class="sd">        gene_positiondf: Actual gene position dataframe</span>

<span class="sd">    Return:</span>
<span class="sd">        bool or Series: if the gene passed in need to be remapped or</span>
<span class="sd">                        the remapped gene</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">region_overlap</span> <span class="o">=</span> <span class="n">_check_region_overlap</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">region_overlap</span><span class="p">:</span>
        <span class="n">overlap_positions</span> <span class="o">=</span> <span class="n">_map_position_within_boundary</span><span class="p">(</span><span class="n">row</span><span class="p">,</span> <span class="n">gene_positiondf</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">overlap_positions</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">LOGGER</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> cannot be remapped. &quot;</span>
                <span class="s2">&quot;These rows will have an empty gene symbol&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">])</span>
            <span class="p">)</span>
            <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">overlap_positions</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">symbol_list</span> <span class="o">=</span> <span class="n">overlap_positions</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">symbol_list</span><span class="p">:</span>
                <span class="c1"># if &quot;MLL4&quot;, then the HUGO symbol should be KMT2D and KMT2B</span>
                <span class="n">LOGGER</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span>
                    <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> can be mapped to different symbols: </span><span class="si">{}</span><span class="s2">. &quot;</span>
                    <span class="s2">&quot;Please correct or it will be removed.&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span>
                        <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">],</span> <span class="s2">&quot;, &quot;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">symbol_list</span><span class="p">)</span>
                    <span class="p">)</span>
                <span class="p">)</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="s2">&quot;nan&quot;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">symbol</span> <span class="o">=</span> <span class="n">overlap_positions</span><span class="p">[</span><span class="s2">&quot;hgnc_symbol&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">!=</span> <span class="n">symbol</span><span class="p">:</span>
                <span class="n">LOGGER</span><span class="o">.</span><span class="n">info</span><span class="p">(</span>
                    <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> will be remapped to </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">],</span> <span class="n">symbol</span><span class="p">)</span>
                <span class="p">)</span>
                <span class="n">row</span><span class="p">[</span><span class="s2">&quot;Hugo_Symbol&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">symbol</span>
    <span class="k">return</span> <span class="n">row</span>
</code></pre></div></td></tr></table></div>
            </details>
    </div>

</div>



  </div>

    </div>

</div>












                
              </article>
            </div>
          
          
  <script>var tabs=__md_get("__tabs");if(Array.isArray(tabs))e:for(var set of document.querySelectorAll(".tabbed-set")){var labels=set.querySelector(".tabbed-labels");for(var tab of tabs)for(var label of labels.getElementsByTagName("label"))if(label.innerText.trim()===tab){var input=document.getElementById(label.htmlFor);input.checked=!0;continue e}}</script>

<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.instant", "navigation.tabs", "search.highlight", "search.share", "toc.integrate", "content.tabs.link", "navigation.indexes", "navigation.sections", "content.code.copy", "navigation.sections", "navigation.tabs", "navigation.top", "navigation.tracking", "navigation.instant", "header.autohide", "content.action.edit", "content.tabs.link", "content.tabs", "search.highlight", "search.share", "search.suggest", "toc.integrate"], "search": "../../../assets/javascripts/workers/search.973d3a69.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f55a23d4.min.js"></script>
      
        <script src="../../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>